{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "DnQGl_D4786T"
   },
   "source": [
    "DRKG\n",
    "\n",
    "Adapted from: https://github.com/gnn4dr/DRKG/blob/master/drkg_with_dgl/loading_drkg_in_dgl.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "PV4N5XSIKfTa"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os \n",
    "import torch_geometric.transforms as T\n",
    "import sys\n",
    "sys.path.append('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SIMP_LLM.DRKG_loading   import  get_triplets, read_tsv,filter_drkg,map_drkg_relationships,filter_interaction_subset,print_head\n",
    "from SIMP_LLM.DRKG_translate import  load_lookups\n",
    "from SIMP_LLM.DRKG_entity_processing import get_unique_entities, get_entity_lookup, convert_entitynames, flip_headtail\n",
    "from SIMP_LLM.raredisease_loading import get_orphan_data # SP 05/24/23 added"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "VsLJKwgRxArZ"
   },
   "source": [
    "# 1) Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Triplets:\n",
      "\n",
      "[['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2157'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5264'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2158'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::3309'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::28912'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::811'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2159'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::821'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5627'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5624']]\n",
      "\n",
      " ../data/drkg.tsv  Dataframe:\n",
      "\n",
      "+----+------------+--------------------------------+-------------+\n",
      "|    | 0          | 1                              | 2           |\n",
      "|----+------------+--------------------------------+-------------|\n",
      "|  0 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2157  |\n",
      "|  1 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::5264  |\n",
      "|  2 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2158  |\n",
      "|  3 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::3309  |\n",
      "|  4 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::28912 |\n",
      "+----+------------+--------------------------------+-------------+\n",
      "\n",
      " ../data/relation_glossary.tsv  Dataframe:\n",
      "\n",
      "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
      "|    | Relation-name                              | Data-source   | Connected entity-types   | Interaction-type      | Description                                                                                                                                                                     | Reference for the description        |\n",
      "|----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------|\n",
      "|  0 | DGIDB::ACTIVATOR::Gene:Compound            | DGIDB         | Compound:Gene            | activation            | An activator interaction is when a drug activates a biological response from a target, although the mechanism by which it does so may not be understood.                        | http://www.dgidb.org/getting_started |\n",
      "|  1 | DGIDB::AGONIST::Gene:Compound              | DGIDB         | Compound:Gene            | agonism               | An agonist interaction occurs when a drug binds to a target receptor and activates the receptor to produce a biological response.                                               | http://www.dgidb.org/getting_started |\n",
      "|  2 | DGIDB::ALLOSTERIC MODULATOR::Gene:Compound | DGIDB         | Compound:Gene            | allosteric modulation | An allosteric modulator interaction occurs when drugs exert their effects on their protein targets via a different binding site than the natural (orthosteric) ligand site.     | http://www.dgidb.org/getting_started |\n",
      "|  3 | DGIDB::ANTAGONIST::Gene:Compound           | DGIDB         | Compound:Gene            | antagonism            | An antagonist interaction occurs when a drug blocks or dampens agonist-mediated responses rather than provoking a biological response itself upon binding to a target receptor. | http://www.dgidb.org/getting_started |\n",
      "|  4 | DGIDB::ANTIBODY::Gene:Compound             | DGIDB         | Compound:Gene            | antibody              | An antibody interaction occurs when an antibody drug specifically binds the target molecule.                                                                                    | http://www.dgidb.org/getting_started |\n",
      "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
      "Number of Rows Before Filtering: 107\n",
      "Number of Rows After Filtering: 10\n",
      "\n",
      "\n",
      "Filtered:\n",
      "24    DRUGBANK::treats::Compound:Disease\n",
      "35             GNBR::T::Compound:Disease\n",
      "36             GNBR::C::Compound:Disease\n",
      "37            GNBR::Sa::Compound:Disease\n",
      "38            GNBR::Pa::Compound:Disease\n",
      "39            GNBR::Mp::Compound:Disease\n",
      "40            GNBR::Pr::Compound:Disease\n",
      "41             GNBR::J::Compound:Disease\n",
      "68       Hetionet::CtD::Compound:Disease\n",
      "73       Hetionet::CpD::Compound:Disease\n",
      "dtype: object\n",
      "\n",
      "Relationships Mapped:\n",
      "0                    Compound treats the disease\n",
      "1    treatment/therapy (including investigatory)\n",
      "2            inhibits cell growth (esp. cancers)\n",
      "3                      side effect/adverse event\n",
      "4                            alleviates, reduces\n",
      "5            biomarkers (of disease progression)\n",
      "6                           prevents, suppresses\n",
      "7                   role in disease pathogenesis\n",
      "8                                      treatment\n",
      "9                                     palliation\n",
      "Name: Interaction-type, dtype: object\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n",
      "|         | 0                 | 1                                  | 2                     |\n",
      "|---------+-------------------+------------------------------------+-----------------------|\n",
      "| 1518268 | Compound::DB00003 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D003550 |\n",
      "| 1518269 | Compound::DB00004 | DRUGBANK::treats::Compound:Disease | Disease::MESH:C063419 |\n",
      "| 1518270 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D013167 |\n",
      "| 1518271 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D006086 |\n",
      "| 1518272 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D017497 |\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n"
     ]
    }
   ],
   "source": [
    "### 1) Read: This section reads DRKG and a glossary (used to map entities from codes to words)\n",
    "DATA_DIR           = os.path.join(\"../data\")\n",
    "verbose            =  True \n",
    "triplets,drkg_df   =  get_triplets(drkg_file = os.path.join(DATA_DIR  ,'drkg.tsv'),             verbose=verbose)  # Read triplets (head,relationship,tail)\n",
    "relation_glossary  =  read_tsv(relation_file = os.path.join(DATA_DIR  ,'relation_glossary.tsv'),verbose=verbose)  # Read relationship mapping  \n",
    "\n",
    "\n",
    "### 2) Filter & Map Interactions: This section returns a list of interactions e.g. DRUGBANK::treats::Compound:Disease )\n",
    "# 2.1: First  we filter the interactions to only Compound-Disease\n",
    "# 2.2: Then   we map the codes -> text  (this will be use to further filter interactions based on text) e.g.  Hetionet::CpD::Compound:Disease -> palliation\n",
    "# 2.3: We use natural text to fitler  interactions based on terms such as \"treat\" (but we return the orignal interaction name )\n",
    "\n",
    "\n",
    "\n",
    "# modularize this in create_dataframe\n",
    "drkg_rx_dx_relations        = filter_drkg(data_frame = drkg_df ,  filter_column = 1 ,  filter_term = r'.*?Compound:Disease', verbose = verbose) # 2.1 Filter only Compound-Disease Interactions\n",
    "drkg_rx_dx_relations_mapped = map_drkg_relationships(drkg_rx_dx_relations,relation_glossary,verbose=verbose)                                    # 2.2 Map codes to text \n",
    "\n",
    "### 2.3 Filter Drug interactions Interaction types to only include: treat inhibit or alleviate interactions  ###\n",
    "drkg_rx_dx_relation_subset =  filter_interaction_subset(df                  = drkg_rx_dx_relations_mapped,\n",
    "                                                        filter_colunm_name = 'Interaction-type' ,\n",
    "                                                        regex_string       =  'treat|inhibit|alleviate',\n",
    "                                                        return_colunm_name =  'Relation-name')\n",
    "\n",
    "# 3) Use Filter Interactions to get Gilter DRKG \n",
    "drkg_df_filtered = drkg_df[drkg_df[1].isin(drkg_rx_dx_relation_subset)] # 3.1 Filter DRKG  to only  Compund-Disease \n",
    "print_head(df=drkg_df_filtered)\n",
    "\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "rx_dx_triplets   = drkg_df_filtered.values.tolist()                     # 3.2 Convert filtered DRKG to list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "KYBnEGL4Nixb",
    "outputId": "5bc6b100-d023-43d7-cf3c-488dc73b6475"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " ../data/hetionet-v1.0-nodes.tsv  Dataframe:\n",
      "\n",
      "+----+-------------------------+---------------------------+---------+\n",
      "|    | id                      | name                      | kind    |\n",
      "|----+-------------------------+---------------------------+---------|\n",
      "|  0 | Anatomy::UBERON:0000002 | uterine cervix            | Anatomy |\n",
      "|  1 | Anatomy::UBERON:0000004 | nose                      | Anatomy |\n",
      "|  2 | Anatomy::UBERON:0000006 | islet of Langerhans       | Anatomy |\n",
      "|  3 | Anatomy::UBERON:0000007 | pituitary gland           | Anatomy |\n",
      "|  4 | Anatomy::UBERON:0000010 | peripheral nervous system | Anatomy |\n",
      "+----+-------------------------+---------------------------+---------+\n",
      "\n",
      " Sample of Hetionet Data Types (Before processing):\n",
      "\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "|       | id                               | name                                      | kind                |\n",
      "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
      "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
      "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
      "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
      "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
      "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
      "| 14863 | Gene::1                          | A1BG                                      | Gene                |\n",
      "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
      "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
      "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
      "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
      "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "\n",
      " Sample of Hetionet Data Types (After processing):\n",
      "\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "|       | id                               | name                                      | kind                |\n",
      "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
      "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
      "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
      "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
      "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
      "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
      "| 14863 | Gene::1                          | A1BG gene                                 | Gene                |\n",
      "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
      "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
      "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
      "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
      "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "\n",
      " Gene ID Dataframe (After processing):\n",
      "\n",
      "+----+----------+--------------------------------------------+\n",
      "|    | GeneID   | description                                |\n",
      "|----+----------+--------------------------------------------|\n",
      "|  0 | Gene::1  | alpha-1-B glycoprotein (A1BG)              |\n",
      "|  1 | Gene::2  | alpha-2-macroglobulin (A2M)                |\n",
      "|  2 | Gene::3  | alpha-2-macroglobulin pseudogene 1 (A2MP1) |\n",
      "|  3 | Gene::9  | N-acetyltransferase 1 (NAT1)               |\n",
      "|  4 | Gene::10 | N-acetyltransferase 2 (NAT2)               |\n",
      "+----+----------+--------------------------------------------+\n",
      "\n",
      " DrugBank Dataframe (Before processing):\n",
      "\n",
      "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
      "|    | DrugBank ID   | Accession Numbers    | Common name         | CAS         | UNII       | Synonyms                                                                                                                                                              |   Standard InChI Key |\n",
      "|----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------|\n",
      "|  0 | DB00001       | BIOD00024 | BTD00024 | Lepirudin           | 138068-37-8 | Y43GF64R34 | [Leu1, Thr2]-63-desulfohirudin | Desulfatohirudin | Hirudin variant-1 | Lepirudin | Lepirudin recombinant | R-hirudin                                                 |                  nan |\n",
      "|  1 | DB00002       | BIOD00071 | BTD00071 | Cetuximab           | 205923-56-4 | PQX0D8J21J | Cetuximab | Cétuximab | Cetuximabum                                                                                                                                   |                  nan |\n",
      "|  2 | DB00003       | BIOD00001 | BTD00001 | Dornase alfa        | 143831-71-4 | 953A26OA1Y | Deoxyribonuclease (human clone 18-1 protein moiety) | Dornasa alfa | Dornase alfa | Dornase alfa, recombinant | Dornase alpha | Recombinant deoxyribonuclease (DNAse) |                  nan |\n",
      "|  3 | DB00004       | BIOD00084 | BTD00084 | Denileukin diftitox | 173146-27-5 | 25E79B5CTM | Denileukin | Denileukin diftitox | Interleukin-2/diptheria toxin fusion protein                                                                                       |                  nan |\n",
      "|  4 | DB00005       | BIOD00052 | BTD00052 | Etanercept          | 185243-69-0 | OP401G7OJC | Etanercept | etanercept-szzs | etanercept-ykro | Recombinant human TNF | rhu TNFR:Fc | rhu-TNFR:Fc | TNFR-Immunoadhesin                                               |                  nan |\n",
      "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
      "\n",
      " DrugBank Dataframe (After processing):\n",
      "\n",
      "+----+-------------------+---------------------+\n",
      "|    | DrugBank ID       | Common name         |\n",
      "|----+-------------------+---------------------|\n",
      "|  0 | Compound::DB00001 | Lepirudin           |\n",
      "|  1 | Compound::DB00002 | Cetuximab           |\n",
      "|  2 | Compound::DB00003 | Dornase alfa        |\n",
      "|  3 | Compound::DB00004 | Denileukin diftitox |\n",
      "|  4 | Compound::DB00005 | Etanercept          |\n",
      "+----+-------------------+---------------------+\n",
      "\n",
      " ../data/mimTitles.txt  Dataframe (Before processing):\n",
      "\n",
      "+----+----------------------+----------------------------------------+\n",
      "|    | MIM Number           | name                                   |\n",
      "|----+----------------------+----------------------------------------|\n",
      "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
      "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
      "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
      "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
      "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
      "+----+----------------------+----------------------------------------+\n",
      "\n",
      " ../data/mimTitles.txt  Dataframe (After processing):\n",
      "\n",
      "+----+----------------------+----------------------------------------+\n",
      "|    | MIM Number           | name                                   |\n",
      "|----+----------------------+----------------------------------------|\n",
      "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
      "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
      "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
      "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
      "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
      "+----+----------------------+----------------------------------------+\n",
      "\n",
      " MeSH Dictionary:\n",
      "\n",
      "+---------------+------------------------+\n",
      "|               | 0                      |\n",
      "|---------------+------------------------|\n",
      "| MESH::D000001 | Calcimycin             |\n",
      "| MESH::D000002 | Temefos                |\n",
      "| MESH::D000003 | Abattoirs              |\n",
      "| MESH::D000004 | Abbreviations as Topic |\n",
      "| MESH::D000005 | Abdomen                |\n",
      "+---------------+------------------------+\n"
     ]
    }
   ],
   "source": [
    "# 4) Load Data frames for translation\n",
    "hetionet_df, gene_df, drugbank_df, omim_df, mesh_dict, chebi_df, chembl_df = load_lookups(data_path=DATA_DIR,verbose=verbose)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Long-form orphan disease data (before processing):\n",
      "\n",
      "+----+-----------+------------------------------------------------------------------------+\n",
      "|    | cols      | data                                                                   |\n",
      "|----+-----------+------------------------------------------------------------------------|\n",
      "|  0 | Orphacode | 166024                                                                 |\n",
      "|  1 | Name      | Multiple epiphyseal dysplasia, Al-Gazali type                          |\n",
      "|  2 | Synonym   | Multiple epiphyseal dysplasia-macrocephaly-distinctive facies syndrome |\n",
      "|  3 | Source    | ICD-10                                                                 |\n",
      "|  4 | Reference | Q77.3                                                                  |\n",
      "+----+-----------+------------------------------------------------------------------------+\n",
      "\n",
      " Long-form orphan disease data (after processing):\n",
      "\n",
      "+----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------+\n",
      "|    | cols                    | data                                                                   |   disease_id | code_source   | code   |\n",
      "|----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------|\n",
      "|  0 | Orphacode               | 166024                                                                 |       166024 |               |        |\n",
      "|  1 | Name                    | Multiple epiphyseal dysplasia, Al-Gazali type                          |       166024 |               |        |\n",
      "|  2 | Synonym                 | Multiple epiphyseal dysplasia-macrocephaly-distinctive facies syndrome |       166024 |               |        |\n",
      "|  4 | Reference               | Q77.3                                                                  |       166024 | ICD-10        | Q77.3  |\n",
      "|  6 | Disordermappingrelation | NTBT (ORPHA code's Narrower Term maps to a Broader Term)               |       166024 | ICD-10        | Q77.3  |\n",
      "+----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------+\n",
      "\n",
      " Orphan disease name/summary data:\n",
      "\n",
      "+--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+\n",
      "|   disease_id |   Orphacode | Name                                                   | Synonym                                                    | Definition                                                                                                                                                                                                                                                                                                                                                                                             |   Info |\n",
      "|--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------|\n",
      "|           10 |          10 | 48,XXYY syndrome                                       | nan                                                        | A rare sex chromosome number anomaly disorder characterized, genetically, by the presence of an extra X and Y chromosome in males and, clinically, by tall stature, dysfunctional testes associated with infertility and insufficient testosterone production, cognitive, affective and social functioning impairments, global developmental delay, and an increased risk of congenital malformations. |    nan |\n",
      "|          100 |         100 | Ataxia-telangiectasia                                  | Louis-Bar syndrome                                         | A rare disorder characterized by the association of severe combined immunodeficiency (affecting mainly the humoral immune response) with progressive cerebellar ataxia. It is characterized by neurological signs, telangiectasia, increased susceptibility to infections and a higher risk of cancer.                                                                                                 |    nan |\n",
      "|         1000 |        1000 | Ocular albinism with late-onset sensorineural deafness | Ocular albinism with late-onset sensorineural hearing loss | Ocular albinism with late-onset sensorineural deafness is a rare, X-linked inherited subtype of ocular albinism characterized by severe visual impairment, translucent pale-blue irises, a reduction in the retinal pigment and moderately severe deafness with onset ranging from adolescence to fourth or fifth decade of life.                                                                      |    nan |\n",
      "|       100000 |      100000 | Reticular perineurioma                                 | nan                                                        | nan                                                                                                                                                                                                                                                                                                                                                                                                    |    nan |\n",
      "|       100001 |      100001 | Sclerosing perineurioma                                | nan                                                        | nan                                                                                                                                                                                                                                                                                                                                                                                                    |    nan |\n",
      "+--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+\n",
      "\n",
      " Orphan disease codes:\n",
      "\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+\n",
      "|    |   Orphacode | Name                                          | code_source   | code     | Disordermappingrelation                                  | Disordermappingicdrelation                          | Disordermappingvalidationstatus   |\n",
      "|----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------|\n",
      "|  0 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | ICD-10        | Q77.3    | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         |\n",
      "|  1 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | OMIM          | 607131   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  2 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | UMLS          | C1846722 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  3 |          58 | Alexander disease                             | OMIM          | 203450   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  4 |          58 | Alexander disease                             | MeSH          | D038261  | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+\n"
     ]
    }
   ],
   "source": [
    "# Load orphan disease names and codes (28 Nov 2022 version) # SP 05/24/23\n",
    "orphan_names, orphan_codes = get_orphan_data(os.path.join(DATA_DIR, 'en_product1-Orphadata.xml'), verbose=verbose)\n",
    "\n",
    "# Get orphan disease MeSH codes\n",
    "orphan_codes_mesh = orphan_codes[orphan_codes['code_source']=='MeSH'].copy()\n",
    "orphan_codes_mesh['id'] = 'MESH::'+orphan_codes_mesh['code']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HoImejcxpFYP",
    "outputId": "904e49c4-8889-4f2d-fcce-26dc99331b86"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+------------------------------+--------------------------------+------------------------------------------------------+\n",
      "|    | 0                            | 1                              | 2                                                    |\n",
      "|----+------------------------------+--------------------------------+------------------------------------------------------|\n",
      "|  0 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | coagulation factor VIII (F8)                         |\n",
      "|  1 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | phytanoyl-CoA 2-hydroxylase (PHYH)                   |\n",
      "|  2 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | coagulation factor IX (F9)                           |\n",
      "|  3 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | heat shock protein family A (Hsp70) member 5 (HSPA5) |\n",
      "|  4 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | immunoglobulin kappa variable 3-20 (IGKV3-20)        |\n",
      "+----+------------------------------+--------------------------------+------------------------------------------------------+\n",
      "Number of unique DRKG entities:  97238\n",
      "Number of translated entities:  64384\n",
      "Number of untranslated entities:  32854\n",
      "Percentage of entities translated:  66.2 %\n",
      "Total DRKG relationships:  5874261\n",
      "Translated DRKG relationships:  5707119\n",
      "Percentage of relationships fully translated:  97.2 %\n"
     ]
    }
   ],
   "source": [
    "# Make dictionaries for codes\n",
    "code_df   = pd.concat([hetionet_df[['name', 'id']], \n",
    "                       gene_df.rename(columns = {\"description\":\"name\", \"GeneID\":\"id\"}),\n",
    "                       drugbank_df.rename(columns = {\"Common name\":\"name\", \"DrugBank ID\":\"id\"}),\n",
    "                       omim_df.rename(columns = {\"MIM Number\":\"id\"}),\n",
    "                       chebi_df.rename(columns = {\"NAME\":\"name\", \"CHEBI_ACCESSION\":\"id\"}),\n",
    "                       chembl_df.rename(columns = {\"pref_name\":\"name\", \"chembl_id\":\"id\"}),\n",
    "                       orphan_codes_mesh.rename(columns = {\"Name\":\"name\"}) # SP 05/24/23 added orphan disease MeSH terms\n",
    "                       ], ignore_index=True, axis=0).drop_duplicates() \n",
    "code_dict = pd.Series(code_df['name'].values, index=code_df['id']).to_dict() | mesh_dict # Convert node df to dict and merge with MeSH dictionary\n",
    "\n",
    "# Get unique DRKG entities\n",
    "drkg_entities = get_unique_entities(drkg_df, [0,2])\n",
    "\n",
    "# Create and use convert_entitynames function\n",
    "drkg_entity_df, drkg_unmatched = get_entity_lookup(drkg_entities, code_dict)\n",
    "\n",
    "# Create final node dictionary\n",
    "node_dict = pd.Series(drkg_entity_df['name'].values, index=drkg_entity_df['drkg_id']).to_dict() \n",
    "\n",
    "# Initialize translated DRKG and manually clean heads/tails for one case where they were flipped\n",
    "drkg_translated    = drkg_df.copy()\n",
    "drkg_translated = flip_headtail(drkg_translated, 'Gene:Compound')\n",
    "\n",
    "# Map DRKG to translated entity names\n",
    "drkg_translated = convert_entitynames(drkg_translated, 0, node_dict)\n",
    "drkg_translated = convert_entitynames(drkg_translated, 2, node_dict)\n",
    "drkg_translated = drkg_translated.dropna()\n",
    "print_head(drkg_translated) \n",
    "\n",
    "# Summarize percentage translated\n",
    "print(\"Number of unique DRKG entities: \", len(drkg_entities)) # should be 97238\n",
    "print(\"Number of translated entities: \", drkg_entity_df.shape[0])\n",
    "print(\"Number of untranslated entities: \", drkg_unmatched.shape[0])\n",
    "pct_entity_translated = drkg_entity_df.shape[0]/len(drkg_entities)\n",
    "print('Percentage of entities translated: ', round(pct_entity_translated*100,1), '%')\n",
    "\n",
    "print('Total DRKG relationships: ', drkg_df.shape[0])\n",
    "print('Translated DRKG relationships: ', drkg_translated.shape[0])\n",
    "pct_translated = drkg_translated.shape[0]/drkg_df.shape[0]\n",
    "print('Percentage of relationships fully translated: ', round(pct_translated*100,1), '%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>Data-source</th>\n",
       "      <th>Connected entity-types</th>\n",
       "      <th>Interaction-type</th>\n",
       "      <th>Description</th>\n",
       "      <th>Reference for the description</th>\n",
       "      <th>head_entity</th>\n",
       "      <th>tail_entity</th>\n",
       "      <th>relation_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DGIDB::ACTIVATOR::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>activation</td>\n",
       "      <td>An activator interaction is when a drug activa...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DGIDB::AGONIST::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>agonism</td>\n",
       "      <td>An agonist interaction occurs when a drug bind...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DGIDB::ALLOSTERIC MODULATOR::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>allosteric modulation</td>\n",
       "      <td>An allosteric modulator interaction occurs whe...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>allosteric modulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DGIDB::ANTAGONIST::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>antagonism</td>\n",
       "      <td>An antagonist interaction occurs when a drug b...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>antagonism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DGIDB::ANTIBODY::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>antibody</td>\n",
       "      <td>An antibody interaction occurs when an antibod...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>antibody</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>bioarx::Covid2_acc_host_gene::Disease:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Disease:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Interactions between 27 viral  proteins,  and ...</td>\n",
       "      <td></td>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>bioarx::DrugHumGen:Compound:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>bioarx::DrugVirGen:Compound:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>bioarx::HumGenHumGen:Gene:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Gene:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Protein-protein interaction</td>\n",
       "      <td></td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>bioarx::VirGenHumGen:Gene:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Gene:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Protein-protein interaction</td>\n",
       "      <td></td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        drkg_id   Data-source   \n",
       "0               DGIDB::ACTIVATOR::Gene:Compound         DGIDB  \\\n",
       "1                 DGIDB::AGONIST::Gene:Compound         DGIDB   \n",
       "2    DGIDB::ALLOSTERIC MODULATOR::Gene:Compound         DGIDB   \n",
       "3              DGIDB::ANTAGONIST::Gene:Compound         DGIDB   \n",
       "4                DGIDB::ANTIBODY::Gene:Compound         DGIDB   \n",
       "..                                          ...           ...   \n",
       "102  bioarx::Covid2_acc_host_gene::Disease:Gene  BIBLIOGRAPHY   \n",
       "103            bioarx::DrugHumGen:Compound:Gene  BIBLIOGRAPHY   \n",
       "104            bioarx::DrugVirGen:Compound:Gene  BIBLIOGRAPHY   \n",
       "105              bioarx::HumGenHumGen:Gene:Gene  BIBLIOGRAPHY   \n",
       "106              bioarx::VirGenHumGen:Gene:Gene  BIBLIOGRAPHY   \n",
       "\n",
       "    Connected entity-types       Interaction-type   \n",
       "0            Compound:Gene             activation  \\\n",
       "1            Compound:Gene                agonism   \n",
       "2            Compound:Gene  allosteric modulation   \n",
       "3            Compound:Gene             antagonism   \n",
       "4            Compound:Gene               antibody   \n",
       "..                     ...                    ...   \n",
       "102           Disease:Gene            interaction   \n",
       "103          Compound:Gene            interaction   \n",
       "104          Compound:Gene            interaction   \n",
       "105              Gene:Gene            interaction   \n",
       "106              Gene:Gene            interaction   \n",
       "\n",
       "                                           Description   \n",
       "0    An activator interaction is when a drug activa...  \\\n",
       "1    An agonist interaction occurs when a drug bind...   \n",
       "2    An allosteric modulator interaction occurs whe...   \n",
       "3    An antagonist interaction occurs when a drug b...   \n",
       "4    An antibody interaction occurs when an antibod...   \n",
       "..                                                 ...   \n",
       "102  Interactions between 27 viral  proteins,  and ...   \n",
       "103                                                      \n",
       "104                                                      \n",
       "105                        Protein-protein interaction   \n",
       "106                        Protein-protein interaction   \n",
       "\n",
       "            Reference for the description head_entity tail_entity   \n",
       "0    http://www.dgidb.org/getting_started    Compound        Gene  \\\n",
       "1    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "2    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "3    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "4    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "..                                    ...         ...         ...   \n",
       "102                                           Disease        Gene   \n",
       "103                                          Compound        Gene   \n",
       "104                                          Compound        Gene   \n",
       "105                                              Gene        Gene   \n",
       "106                                              Gene        Gene   \n",
       "\n",
       "             relation_name  \n",
       "0               activation  \n",
       "1               activation  \n",
       "2    allosteric modulation  \n",
       "3               antagonism  \n",
       "4                 antibody  \n",
       "..                     ...  \n",
       "102            interaction  \n",
       "103            interaction  \n",
       "104            interaction  \n",
       "105            interaction  \n",
       "106            interaction  \n",
       "\n",
       "[107 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "# Update relation glossary \n",
    "relation_df = relation_glossary.copy().rename(columns={'Relation-name':'drkg_id'})\n",
    "relation_df[['head_entity','tail_entity']] = relation_df['drkg_id'].str.split('::', expand=True)[2].str.split(':', expand=True) # Set head and tail nodes\n",
    "\n",
    "# Manually fix head and tail nodes for DGIDB relations, which reverse compound-gene interactions\n",
    "relation_df.loc[relation_df['drkg_id'].str.contains('Gene:Compound'),'head_entity'] = 'Compound'\n",
    "relation_df.loc[relation_df['drkg_id'].str.contains('Gene:Compound'),'tail_entity'] = 'Gene'\n",
    "\n",
    "# Fix bioarx entries without the second \"::\" delimiter\n",
    "bioarx_ht = relation_df['drkg_id'].str.split(':', expand=True)[[3,4]]\n",
    "relation_df['head_entity'] = np.where(relation_df['head_entity'].isna(), bioarx_ht[3], relation_df['head_entity'])\n",
    "relation_df['tail_entity'] = np.where(relation_df['tail_entity'].isna(), bioarx_ht[4], relation_df['tail_entity'])\n",
    "\n",
    "# Add mapped relation group labels\n",
    "relation_groups = [['activation', 'agonism', 'agonism, activation', 'activates, stimulates'],\n",
    "    ['antagonism', 'blocking', 'antagonism, blocking'],\n",
    "    ['binding', 'binding, ligand (esp. receptors)'],\n",
    "    ['blocking', 'channel blocking'],\n",
    "    ['inhibition', 'inhibits cell growth (esp. cancers)', 'inhibits'],\n",
    "    ['enzyme', 'enzyme activity'],\n",
    "    ['upregulation', 'increases expression/production'],\n",
    "    ['downregulation', 'decreases expression/production'],\n",
    "    ['Compound treats the disease', 'treatment/therapy (including investigatory)', 'treatment']]\n",
    "\n",
    "relation_df['relation_name'] = relation_df['Interaction-type']\n",
    "\n",
    "for grp in relation_groups:\n",
    "    relation_df_subset = relation_df[relation_df['Interaction-type'].isin(grp)].copy()\n",
    "    for entities in relation_df_subset['Connected entity-types'].unique():\n",
    "        subgrp = relation_df_subset[relation_df_subset['Connected entity-types'] == entities]['Interaction-type'].unique()\n",
    "        relation_df.loc[(relation_df_subset['Connected entity-types'] == entities) & (relation_df['Interaction-type'].isin(subgrp)), 'relation_name'] = subgrp[0]\n",
    "\n",
    "# Remove special characters from relation names\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace(',|/', ' or ', regex=True)\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace('esp.','especially')\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace('\\(|\\)|-|\\.', '', regex=True)\n",
    "\n",
    "# Check if any relationshp names still have non alpha numeric values except space\n",
    "error_relation_names = relation_df['relation_name'][relation_df['relation_name'].str.replace(' ', '').str.contains(r\"[^a-zA-Z0-9]+\", regex=True)].drop_duplicates()\n",
    "if len(error_relation_names):\n",
    "    print('Warning: The following relation names contain special characters, which can interfere with PyG/GraphSage')\n",
    "    print(error_relation_names)\n",
    "\n",
    "relation_df\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Identify rare diseases in DRKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SIMP_LLM.raredisease_loading import get_drkg_entity_ontologies, read_and_process_doid, create_orphanet_regex, merge_regex, find_drkg_rarediseases, check_raredisease_multiple_codes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download and process data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DRKG code counts:\n",
      "   matched ontology_name  count\n",
      "0        0          MESH   1002\n",
      "1        1          DOID    127\n",
      "2        1          MESH   4284\n",
      "3        1          OMIM     78\n",
      "4        1      UMLS CUI   5701\n",
      "\n",
      "Orphanet code types:  ['ICD-10' 'OMIM' 'UMLS' 'MeSH' 'ICD-11' 'GARD' 'MedDRA']\n",
      "\n",
      " DOID Dataframe (After processing):\n",
      "\n",
      "+-----+--------------+--------------------------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------------------------------------------------------------------------------------------------------------+----------------------+------------------------------------------------+---------------------------------------------+\n",
      "|     | id           | Preferred Label          | Synonyms                                       | Definitions                                                                                                                                                                     |   CUI | database_cross_reference                                                                                         | has_alternative_id   | has_exact_synonym                              | Parents                                     |\n",
      "|-----+--------------+--------------------------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------------------------------------------------------------------------------------------------------------+----------------------+------------------------------------------------+---------------------------------------------|\n",
      "| 572 | DOID:0001816 | angiosarcoma             | hemangiosarcoma                                | A vascular cancer that derives_from the cells that line the walls of blood vessels or lymphatic vessels.                                                                        |   nan | SNOMEDCT_US_2022_09_01:39000009|NCI:C3088|UMLS_CUI:C0018923|MESH:D006394|ICDO:9120/3|NCI:C9275|UMLS_CUI:C0854893 | DOID:4508|DOID:267   | hemangiosarcoma                                | http://purl.obolibrary.org/obo/DOID_175     |\n",
      "| 573 | DOID:0002116 | pterygium                | surfer's eye                                   | A corneal disease that is characterized by a triangular tissue growth located_in cornea of the eye that is the result of collagen degeneration and fibrovascular proliferation. |   nan | UMLS_CUI:C0033999                                                                                                | nan                  | surfer's eye                                   | http://purl.obolibrary.org/obo/DOID_10124   |\n",
      "| 574 | DOID:0014667 | disease of metabolism    | metabolic disease                              | A disease that involving errors in metabolic processes of building or degradation of molecules.                                                                                 |   nan | ICD10CM:E88.9|ICD9CM:277.9|NCI:C3235|SNOMEDCT_US_2022_09_01:75934005|MESH:D008659|UMLS_CUI:C0025517              | nan                  | metabolic disease                              | http://purl.obolibrary.org/obo/DOID_4       |\n",
      "| 576 | DOID:0040002 | aspirin allergy          | ASA allergy|acetylsalicylic acid allergy       | A drug allergy that has_allergic_trigger acetylsalicylic acid.                                                                                                                  |   nan | UMLS_CUI:C0004058|SNOMEDCT_US_2022_09_01:293586001                                                               | nan                  | ASA allergy|acetylsalicylic acid allergy       | http://purl.obolibrary.org/obo/DOID_0060500 |\n",
      "| 577 | DOID:0040003 | benzylpenicillin allergy | penicillin G allergy|benzyl penicillin allergy | A beta-lactam allergy that has_allergic_trigger benzylpenicillin.                                                                                                               |   nan | SNOMEDCT_US_2022_09_01:294499007|UMLS_CUI:C0571411                                                               | nan                  | penicillin G allergy|benzyl penicillin allergy | http://purl.obolibrary.org/obo/DOID_0060519 |\n",
      "+-----+--------------+--------------------------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------------------------------------------------------------------------------------------------------------+----------------------+------------------------------------------------+---------------------------------------------+\n",
      "\n",
      " Orphanet Dataframe with regex (After processing):\n",
      "\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+-------------------------------------------+\n",
      "|    |   Orphacode | Name                                          | code_source   | code     | Disordermappingrelation                                  | Disordermappingicdrelation                          | Disordermappingvalidationstatus   | regex                                     |\n",
      "|----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+-------------------------------------------|\n",
      "|  0 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | ICD-10        | Q77\\.3   | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         | (?:^|\\|)ICD10(?:(?!\\|).)*:Q77.3.*         |\n",
      "|  1 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | OMIM          | 607131   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | (?:^|\\|)OMIM(?:(?!\\|).)*:607131(?:$|\\|)   |\n",
      "|  2 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | UMLS          | C1846722 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | (?:^|\\|)UMLS(?:(?!\\|).)*:C1846722(?:$|\\|) |\n",
      "|  3 |          58 | Alexander disease                             | OMIM          | 203450   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | (?:^|\\|)OMIM(?:(?!\\|).)*:203450(?:$|\\|)   |\n",
      "|  4 |          58 | Alexander disease                             | MeSH          | D038261  | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | (?:^|\\|)MESH(?:(?!\\|).)*:D038261(?:$|\\|)  |\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+-------------------------------------------+\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/selinapi/SIMPLLM/Debug/../SIMP_LLM/raredisease_loading.py:204: DtypeWarning: Columns (8,16,17,18,26,31,32,33,47,52,53,55,56,58,62,63,65,69,71,73,77,83,84,92,104,108) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_raw = pd.read_csv(relation_file)\n"
     ]
    }
   ],
   "source": [
    "# All drkg entities \n",
    "drkg_all_entities = pd.concat([drkg_entity_df, drkg_unmatched], ignore_index=True, axis=0) \n",
    "\n",
    " # Check disease code ontologies in DRKG - extra code in DKRG called DOID - and Orphanet\n",
    "if verbose:\n",
    "    condition_list = ['Disease', 'Symptom', 'Side Effect']\n",
    "    drkg_ontology_counts = get_drkg_entity_ontologies(drkg_all_entities, condition_list)\n",
    "    print('DRKG code counts:')\n",
    "    print(drkg_ontology_counts)\n",
    "    print('\\nOrphanet code types: ', orphan_codes['code_source'].unique())\n",
    "\n",
    "# Download and process DOID (disease ontology) codes\n",
    "doid_df  =  read_and_process_doid(relation_file = os.path.join(DATA_DIR,'DOID.csv'), verbose=verbose) \n",
    "\n",
    "# Create regex for match Orphanet mapped codes to cross reference codes in DOID\n",
    "orphan_codes_match = create_orphanet_regex(orphan_codes, verbose=verbose)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code and regex matching between Orphacodes and DRKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " DRKG-Orphacode matches:\n",
      "\n",
      "+----+-----------------------+----------------+------------------------------------------------------------------------+---------------+-----------------+-----------------+---------+-----------+-------------+------------------------------------------------------------------------+---------------+----------------------------------------------------+------------------------------+-----------------------------------+---------------------+--------------+----------+------+----------+---------+\n",
      "|    | drkg_id               | drkg_dict_id   | name                                                                   | entity_type   | ontology_code   | ontology_name   | code    |   matched |   Orphacode | Name                                                                   | code_source   | Disordermappingrelation                            |   Disordermappingicdrelation | Disordermappingvalidationstatus   | code_source_upper   | match_type   |   code_x |   id |   code_y |   key_0 |\n",
      "|----+-----------------------+----------------+------------------------------------------------------------------------+---------------+-----------------+-----------------+---------+-----------+-------------+------------------------------------------------------------------------+---------------+----------------------------------------------------+------------------------------+-----------------------------------+---------------------+--------------+----------+------+----------+---------|\n",
      "|  0 | Disease::MESH:C531609 | MESH::C531609  | Pseudopelade of Brocq                                                  | Disease       | MESH:C531609    | MESH            | C531609 |         1 |         129 | Pseudopelade of Brocq                                                  | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "|  1 | Disease::MESH:C531616 | MESH::C531616  | AL amyloidosis                                                         | Disease       | MESH:C531616    | MESH            | C531616 |         1 |       85443 | AL amyloidosis                                                         | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "|  2 | Disease::MESH:C531629 | MESH::C531629  | Balantidiasis                                                          | Disease       | MESH:C531629    | MESH            | C531629 |         1 |        1223 | Balantidiasis                                                          | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "|  3 | Disease::MESH:C531816 | MESH::C531816  | Severe combined immunodeficiency due to adenosine deaminase deficiency | Disease       | MESH:C531816    | MESH            | C531816 |         1 |         277 | Severe combined immunodeficiency due to adenosine deaminase deficiency | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "|  4 | Disease::MESH:C535276 | MESH::C535276  | Primary angiitis of the central nervous system                         | Disease       | MESH:C535276    | MESH            | C535276 |         1 |      140989 | Primary angiitis of the central nervous system                         | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "+----+-----------------------+----------------+------------------------------------------------------------------------+---------------+-----------------+-----------------+---------+-----------+-------------+------------------------------------------------------------------------+---------------+----------------------------------------------------+------------------------------+-----------------------------------+---------------------+--------------+----------+------+----------+---------+\n",
      "                                        count\n",
      "match_type   entity_type ontology_name       \n",
      "DOID regex   Disease     DOID              21\n",
      "Disease name Disease     DOID               7\n",
      "                         MESH              54\n",
      "             Side Effect UMLS CUI         134\n",
      "             Symptom     MESH               3\n",
      "MeSH/OMIM    Disease     MESH            1033\n",
      "                         OMIM              67\n",
      "             Symptom     MESH              41\n",
      "\n",
      "Unique matched rare disease DRKG IDs:  1321\n",
      "Unique matched Orphacodes:  1197\n",
      "DRKG IDs associated with multiple Orphacodes:  31\n",
      "Orphacodes associated with multiple DRKG IDs:  44\n",
      "+-----+-----------------------+--------------------------------------+--------------+----------------+\n",
      "|     | drkg_id               | name                                 | match_type   |   ct_orphacode |\n",
      "|-----+-----------------------+--------------------------------------+--------------+----------------|\n",
      "|   1 | Disease::DOID:10021   | duodenum cancer                      | DOID regex   |              2 |\n",
      "|   2 | Disease::DOID:10153   | ileum cancer                         | DOID regex   |              3 |\n",
      "|  10 | Disease::DOID:14268   | sclerosing cholangitis               | DOID regex   |              4 |\n",
      "| 153 | Disease::MESH:C536385 | Focal facial dermal dysplasia type I | MeSH/OMIM    |              2 |\n",
      "| 184 | Disease::MESH:C536611 | Brooke-Spiegler syndrome             | MeSH/OMIM    |              2 |\n",
      "+-----+-----------------------+--------------------------------------+--------------+----------------+\n",
      "+----+-------------+----------------------------------+--------------+--------------+\n",
      "|    |   Orphacode | name                             | match_type   |   ct_drkg_id |\n",
      "|----+-------------+----------------------------------+--------------+--------------|\n",
      "|  1 |         100 | Ataxia Telangiectasia            | MeSH/OMIM    |            2 |\n",
      "|  4 |      100067 | Waterhouse-Friderichsen Syndrome | MeSH/OMIM    |            2 |\n",
      "| 21 |        1041 | Hydrops Fetalis                  | MeSH/OMIM    |            2 |\n",
      "| 43 |        1183 | Opsoclonus-Myoclonus Syndrome    | MeSH/OMIM    |            2 |\n",
      "| 49 |        1223 | Balantidiasis                    | MeSH/OMIM    |            2 |\n",
      "+----+-------------+----------------------------------+--------------+--------------+\n",
      "Number of unmatched Orphacodes:  7132\n",
      "+----+-------------+---------------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+---------------------+\n",
      "|    |   Orphacode | Name                                              | code_source   | code     | Disordermappingrelation                                  | Disordermappingicdrelation                          | Disordermappingvalidationstatus   | code_source_upper   |\n",
      "|----+-------------+---------------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+---------------------|\n",
      "|  0 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type     | ICD-10        | Q77.3    | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         | ICD-10              |\n",
      "|  1 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type     | OMIM          | 607131   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | OMIM                |\n",
      "|  2 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type     | UMLS          | C1846722 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | UMLS                |\n",
      "|  9 |      166032 | Multiple epiphyseal dysplasia, with miniepiphyses | ICD-10        | Q77.3    | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         | ICD-10              |\n",
      "| 10 |      166032 | Multiple epiphyseal dysplasia, with miniepiphyses | UMLS          | C1836307 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | UMLS                |\n",
      "+----+-------------+---------------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+---------------------+\n",
      "Rare disease heads in untranslated DRKG:  3610\n",
      "+-------+-----------------------+-------------------------------------------------+------------+\n",
      "|       | 0                     | 1                                               | 2          |\n",
      "|-------+-----------------------+-------------------------------------------------+------------|\n",
      "| 84697 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::6502 |\n",
      "| 84698 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::3840 |\n",
      "| 84699 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::8575 |\n",
      "| 84700 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::928  |\n",
      "| 84701 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::7113 |\n",
      "+-------+-----------------------+-------------------------------------------------+------------+\n",
      "\n",
      " Rare disease tails in untranslated DRKG:  31854\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n",
      "|         | 0                 | 1                                  | 2                     |\n",
      "|---------+-------------------+------------------------------------+-----------------------|\n",
      "| 1518268 | Compound::DB00003 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D003550 |\n",
      "| 1518271 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D006086 |\n",
      "| 1518273 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D001171 |\n",
      "| 1518276 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D017511 |\n",
      "| 1518278 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D013262 |\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n",
      "+-------+----------------------------------------------+-------------------------------------------------+------------------------------------------------------------------------+\n",
      "|       | 0                                            | 1                                               | 2                                                                      |\n",
      "|-------+----------------------------------------------+-------------------------------------------------+------------------------------------------------------------------------|\n",
      "| 84697 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | S-phase kinase associated protein 2 (SKP2)                             |\n",
      "| 84698 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | karyopherin subunit alpha 4 (KPNA4)                                    |\n",
      "| 84699 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | protein activator of interferon induced protein kinase EIF2AK2 (PRKRA) |\n",
      "| 84700 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | CD9 molecule (CD9)                                                     |\n",
      "| 84701 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | transmembrane serine protease 2 (TMPRSS2)                              |\n",
      "+-------+----------------------------------------------+-------------------------------------------------+------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "# Match Orphacodes and DOIDs - Takes ~3min to run\n",
    "orphacode_doid_regex = merge_regex(orphan_codes_match, 'regex', doid_df, 'database_cross_reference')\n",
    "\n",
    "# Match rare diseases in DRKG\n",
    "matched_rarediseases = find_drkg_rarediseases(drkg_all_entities, orphan_codes, orphacode_doid_regex, verbose=verbose)\n",
    "\n",
    "# Check for rare diseases with multiple codes\n",
    "if verbose:\n",
    "    multiple_orphacode, multiple_drkg = check_raredisease_multiple_codes(matched_rarediseases)\n",
    "    print_head(multiple_orphacode)\n",
    "    print_head(multiple_drkg)\n",
    "\n",
    "# Get non-matched rare diseases (could still be in DRKG but under different name or code - need to check that embeddings are separate)\n",
    "unmatched_rarediseases = orphan_codes[~orphan_codes['Orphacode'].isin(matched_rarediseases['Orphacode'])]\n",
    "if verbose:\n",
    "    print('Number of unmatched Orphacodes: ', len(unmatched_rarediseases['Orphacode'].unique()))\n",
    "    print_head(unmatched_rarediseases)\n",
    "\n",
    "# Get all DRKG relationships involving matched rare diseases\n",
    "raredisease_heads = drkg_df[drkg_df[0].isin(matched_rarediseases['drkg_id'])]\n",
    "raredisease_tails = drkg_df[drkg_df[2].isin(matched_rarediseases['drkg_id'])]\n",
    "\n",
    "if verbose:\n",
    "    print('Rare disease heads in untranslated DRKG: ', raredisease_heads.shape[0])\n",
    "    print_head(raredisease_heads)\n",
    "    print('\\n Rare disease tails in untranslated DRKG: ',raredisease_tails.shape[0])\n",
    "    print_head(raredisease_tails)\n",
    "\n",
    "# Get DRKG rare disease entries\n",
    "raredisease_index = raredisease_heads.index.tolist() + raredisease_tails.index.tolist()\n",
    "drkg_raredisease = drkg_translated.loc[drkg_translated.index.isin(raredisease_index)]\n",
    "if verbose:\n",
    "    print_head(drkg_raredisease)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>84697</th>\n",
       "      <td>Middle East Respiratory Syndrome Coronavirus</td>\n",
       "      <td>bioarx::Coronavirus_ass_host_gene::Disease:Gene</td>\n",
       "      <td>S-phase kinase associated protein 2 (SKP2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84698</th>\n",
       "      <td>Middle East Respiratory Syndrome Coronavirus</td>\n",
       "      <td>bioarx::Coronavirus_ass_host_gene::Disease:Gene</td>\n",
       "      <td>karyopherin subunit alpha 4 (KPNA4)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84699</th>\n",
       "      <td>Middle East Respiratory Syndrome Coronavirus</td>\n",
       "      <td>bioarx::Coronavirus_ass_host_gene::Disease:Gene</td>\n",
       "      <td>protein activator of interferon induced protei...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84700</th>\n",
       "      <td>Middle East Respiratory Syndrome Coronavirus</td>\n",
       "      <td>bioarx::Coronavirus_ass_host_gene::Disease:Gene</td>\n",
       "      <td>CD9 molecule (CD9)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84701</th>\n",
       "      <td>Middle East Respiratory Syndrome Coronavirus</td>\n",
       "      <td>bioarx::Coronavirus_ass_host_gene::Disease:Gene</td>\n",
       "      <td>transmembrane serine protease 2 (TMPRSS2)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594805</th>\n",
       "      <td>Azithromycin</td>\n",
       "      <td>Hetionet::CcSE::Compound:Side Effect</td>\n",
       "      <td>Stevens-Johnson syndrome</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594863</th>\n",
       "      <td>Amifostine</td>\n",
       "      <td>Hetionet::CcSE::Compound:Side Effect</td>\n",
       "      <td>Toxic epidermal necrolysis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594923</th>\n",
       "      <td>Quinine</td>\n",
       "      <td>Hetionet::CcSE::Compound:Side Effect</td>\n",
       "      <td>Fixed drug eruption</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594932</th>\n",
       "      <td>Olmesartan</td>\n",
       "      <td>Hetionet::CcSE::Compound:Side Effect</td>\n",
       "      <td>Systemic lupus erythematosus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3594978</th>\n",
       "      <td>Tobramycin</td>\n",
       "      <td>Hetionet::CcSE::Compound:Side Effect</td>\n",
       "      <td>Endophthalmitis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29217 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0   \n",
       "84697    Middle East Respiratory Syndrome Coronavirus  \\\n",
       "84698    Middle East Respiratory Syndrome Coronavirus   \n",
       "84699    Middle East Respiratory Syndrome Coronavirus   \n",
       "84700    Middle East Respiratory Syndrome Coronavirus   \n",
       "84701    Middle East Respiratory Syndrome Coronavirus   \n",
       "...                                               ...   \n",
       "3594805                                  Azithromycin   \n",
       "3594863                                    Amifostine   \n",
       "3594923                                       Quinine   \n",
       "3594932                                    Olmesartan   \n",
       "3594978                                    Tobramycin   \n",
       "\n",
       "                                                       1   \n",
       "84697    bioarx::Coronavirus_ass_host_gene::Disease:Gene  \\\n",
       "84698    bioarx::Coronavirus_ass_host_gene::Disease:Gene   \n",
       "84699    bioarx::Coronavirus_ass_host_gene::Disease:Gene   \n",
       "84700    bioarx::Coronavirus_ass_host_gene::Disease:Gene   \n",
       "84701    bioarx::Coronavirus_ass_host_gene::Disease:Gene   \n",
       "...                                                  ...   \n",
       "3594805             Hetionet::CcSE::Compound:Side Effect   \n",
       "3594863             Hetionet::CcSE::Compound:Side Effect   \n",
       "3594923             Hetionet::CcSE::Compound:Side Effect   \n",
       "3594932             Hetionet::CcSE::Compound:Side Effect   \n",
       "3594978             Hetionet::CcSE::Compound:Side Effect   \n",
       "\n",
       "                                                         2  \n",
       "84697           S-phase kinase associated protein 2 (SKP2)  \n",
       "84698                  karyopherin subunit alpha 4 (KPNA4)  \n",
       "84699    protein activator of interferon induced protei...  \n",
       "84700                                   CD9 molecule (CD9)  \n",
       "84701            transmembrane serine protease 2 (TMPRSS2)  \n",
       "...                                                    ...  \n",
       "3594805                           Stevens-Johnson syndrome  \n",
       "3594863                         Toxic epidermal necrolysis  \n",
       "3594923                                Fixed drug eruption  \n",
       "3594932                       Systemic lupus erythematosus  \n",
       "3594978                                    Endophthalmitis  \n",
       "\n",
       "[29217 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drkg_raredisease"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Rare disease tests"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>drkg_dict_id</th>\n",
       "      <th>name</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_code</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>code</th>\n",
       "      <th>matched</th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>Name</th>\n",
       "      <th>code_source</th>\n",
       "      <th>Disordermappingrelation</th>\n",
       "      <th>Disordermappingicdrelation</th>\n",
       "      <th>Disordermappingvalidationstatus</th>\n",
       "      <th>code_source_upper</th>\n",
       "      <th>match_type</th>\n",
       "      <th>code_x</th>\n",
       "      <th>id</th>\n",
       "      <th>code_y</th>\n",
       "      <th>key_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1067</th>\n",
       "      <td>Disease::MESH:D065207</td>\n",
       "      <td>MESH::D065207</td>\n",
       "      <td>Middle East Respiratory Syndrome Coronavirus</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MESH:D065207</td>\n",
       "      <td>MESH</td>\n",
       "      <td>D065207</td>\n",
       "      <td>1</td>\n",
       "      <td>576074</td>\n",
       "      <td>Middle East respiratory syndrome</td>\n",
       "      <td>MeSH</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>MESH</td>\n",
       "      <td>MeSH/OMIM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    drkg_id   drkg_dict_id   \n",
       "1067  Disease::MESH:D065207  MESH::D065207  \\\n",
       "\n",
       "                                              name entity_type ontology_code   \n",
       "1067  Middle East Respiratory Syndrome Coronavirus     Disease  MESH:D065207  \\\n",
       "\n",
       "     ontology_name     code  matched Orphacode   \n",
       "1067          MESH  D065207        1    576074  \\\n",
       "\n",
       "                                  Name code_source   \n",
       "1067  Middle East respiratory syndrome        MeSH  \\\n",
       "\n",
       "                                Disordermappingrelation   \n",
       "1067  E (Exact mapping: the two concepts are equival...  \\\n",
       "\n",
       "     Disordermappingicdrelation Disordermappingvalidationstatus   \n",
       "1067                        NaN                       Validated  \\\n",
       "\n",
       "     code_source_upper match_type code_x   id code_y key_0  \n",
       "1067              MESH  MeSH/OMIM    NaN  NaN    NaN   NaN  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# checks for rare disease matches in drkg\n",
    "matched_rarediseases[matched_rarediseases['drkg_id']=='Disease::MESH:D065207']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 246,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Preferred Label</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Definitions</th>\n",
       "      <th>CUI</th>\n",
       "      <th>database_cross_reference</th>\n",
       "      <th>has_alternative_id</th>\n",
       "      <th>has_exact_synonym</th>\n",
       "      <th>Parents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>DOID:0001816</td>\n",
       "      <td>angiosarcoma</td>\n",
       "      <td>hemangiosarcoma</td>\n",
       "      <td>A vascular cancer that derives_from the cells ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNOMEDCT_US_2022_09_01:39000009|NCI:C3088|UMLS...</td>\n",
       "      <td>DOID:4508|DOID:267</td>\n",
       "      <td>hemangiosarcoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>DOID:0002116</td>\n",
       "      <td>pterygium</td>\n",
       "      <td>surfer's eye</td>\n",
       "      <td>A corneal disease that is characterized by a t...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UMLS_CUI:C0033999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surfer's eye</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_10124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>DOID:0014667</td>\n",
       "      <td>disease of metabolism</td>\n",
       "      <td>metabolic disease</td>\n",
       "      <td>A disease that involving errors in metabolic p...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD10CM:E88.9|ICD9CM:277.9|NCI:C3235|SNOMEDCT_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolic disease</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>DOID:0040002</td>\n",
       "      <td>aspirin allergy</td>\n",
       "      <td>ASA allergy|acetylsalicylic acid allergy</td>\n",
       "      <td>A drug allergy that has_allergic_trigger acety...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UMLS_CUI:C0004058|SNOMEDCT_US_2022_09_01:29358...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASA allergy|acetylsalicylic acid allergy</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_0060500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>DOID:0040003</td>\n",
       "      <td>benzylpenicillin allergy</td>\n",
       "      <td>penicillin G allergy|benzyl penicillin allergy</td>\n",
       "      <td>A beta-lactam allergy that has_allergic_trigge...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNOMEDCT_US_2022_09_01:294499007|UMLS_CUI:C057...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>penicillin G allergy|benzyl penicillin allergy</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_0060519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17114</th>\n",
       "      <td>SYMP:0000818</td>\n",
       "      <td>localized superficial mass</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD9CM_2005:782.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0000488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17115</th>\n",
       "      <td>SYMP:0000819</td>\n",
       "      <td>localized superficial swelling</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD9CM_2005:782.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0000488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17116</th>\n",
       "      <td>SYMP:0000820</td>\n",
       "      <td>mass in chest</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD9CM_2005:786.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0000514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17148</th>\n",
       "      <td>SYMP:0000852</td>\n",
       "      <td>anuria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A urinary system symptom that is characterized...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UMLS_ICD9CM_2005_AUI:A0243854|UMLS_CUI:C002896...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0000486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17304</th>\n",
       "      <td>SYMP:0020061</td>\n",
       "      <td>outlet dysfunction constipation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A constipation that is characterized by diffic...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UMLS_CUI:C0949134|ICD10CM:K59.02|ICD9CM:564.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0019180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10761 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id                  Preferred Label   \n",
       "572    DOID:0001816                     angiosarcoma  \\\n",
       "573    DOID:0002116                        pterygium   \n",
       "574    DOID:0014667            disease of metabolism   \n",
       "576    DOID:0040002                  aspirin allergy   \n",
       "577    DOID:0040003         benzylpenicillin allergy   \n",
       "...             ...                              ...   \n",
       "17114  SYMP:0000818       localized superficial mass   \n",
       "17115  SYMP:0000819   localized superficial swelling   \n",
       "17116  SYMP:0000820                    mass in chest   \n",
       "17148  SYMP:0000852                           anuria   \n",
       "17304  SYMP:0020061  outlet dysfunction constipation   \n",
       "\n",
       "                                             Synonyms   \n",
       "572                                   hemangiosarcoma  \\\n",
       "573                                      surfer's eye   \n",
       "574                                 metabolic disease   \n",
       "576          ASA allergy|acetylsalicylic acid allergy   \n",
       "577    penicillin G allergy|benzyl penicillin allergy   \n",
       "...                                               ...   \n",
       "17114                                             NaN   \n",
       "17115                                             NaN   \n",
       "17116                                             NaN   \n",
       "17148                                             NaN   \n",
       "17304                                             NaN   \n",
       "\n",
       "                                             Definitions  CUI   \n",
       "572    A vascular cancer that derives_from the cells ...  NaN  \\\n",
       "573    A corneal disease that is characterized by a t...  NaN   \n",
       "574    A disease that involving errors in metabolic p...  NaN   \n",
       "576    A drug allergy that has_allergic_trigger acety...  NaN   \n",
       "577    A beta-lactam allergy that has_allergic_trigge...  NaN   \n",
       "...                                                  ...  ...   \n",
       "17114                                                NaN  NaN   \n",
       "17115                                                NaN  NaN   \n",
       "17116                                                NaN  NaN   \n",
       "17148  A urinary system symptom that is characterized...  NaN   \n",
       "17304  A constipation that is characterized by diffic...  NaN   \n",
       "\n",
       "                                database_cross_reference  has_alternative_id   \n",
       "572    SNOMEDCT_US_2022_09_01:39000009|NCI:C3088|UMLS...  DOID:4508|DOID:267  \\\n",
       "573                                    UMLS_CUI:C0033999                 NaN   \n",
       "574    ICD10CM:E88.9|ICD9CM:277.9|NCI:C3235|SNOMEDCT_...                 NaN   \n",
       "576    UMLS_CUI:C0004058|SNOMEDCT_US_2022_09_01:29358...                 NaN   \n",
       "577    SNOMEDCT_US_2022_09_01:294499007|UMLS_CUI:C057...                 NaN   \n",
       "...                                                  ...                 ...   \n",
       "17114                                  ICD9CM_2005:782.2                 NaN   \n",
       "17115                                  ICD9CM_2005:782.2                 NaN   \n",
       "17116                                  ICD9CM_2005:786.6                 NaN   \n",
       "17148  UMLS_ICD9CM_2005_AUI:A0243854|UMLS_CUI:C002896...                 NaN   \n",
       "17304     UMLS_CUI:C0949134|ICD10CM:K59.02|ICD9CM:564.02                 NaN   \n",
       "\n",
       "                                    has_exact_synonym   \n",
       "572                                   hemangiosarcoma  \\\n",
       "573                                      surfer's eye   \n",
       "574                                 metabolic disease   \n",
       "576          ASA allergy|acetylsalicylic acid allergy   \n",
       "577    penicillin G allergy|benzyl penicillin allergy   \n",
       "...                                               ...   \n",
       "17114                                             NaN   \n",
       "17115                                             NaN   \n",
       "17116                                             NaN   \n",
       "17148                                             NaN   \n",
       "17304                                             NaN   \n",
       "\n",
       "                                           Parents  \n",
       "572        http://purl.obolibrary.org/obo/DOID_175  \n",
       "573      http://purl.obolibrary.org/obo/DOID_10124  \n",
       "574          http://purl.obolibrary.org/obo/DOID_4  \n",
       "576    http://purl.obolibrary.org/obo/DOID_0060500  \n",
       "577    http://purl.obolibrary.org/obo/DOID_0060519  \n",
       "...                                            ...  \n",
       "17114  http://purl.obolibrary.org/obo/SYMP_0000488  \n",
       "17115  http://purl.obolibrary.org/obo/SYMP_0000488  \n",
       "17116  http://purl.obolibrary.org/obo/SYMP_0000514  \n",
       "17148  http://purl.obolibrary.org/obo/SYMP_0000486  \n",
       "17304  http://purl.obolibrary.org/obo/SYMP_0019180  \n",
       "\n",
       "[10761 rows x 9 columns]"
      ]
     },
     "execution_count": 246,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# All drkg entities (and remove 'CUI' form 'UMLS CUI' in dataframe for merging later)\n",
    "# drkg_all_entities['ontology_name'] = np.where(drkg_all_entities['ontology_name'].str.contains('UMLS'), 'UMLS', drkg_all_entities['ontology_name'])\n",
    "\n",
    "# doid_df_raw[doid_df_raw['id'] =='CHEBI:102166'].iloc[:, 100:120]\n",
    "# doid_df[doid_df['label'] != doid_df['Preferred Label']]\n",
    "doid_df[doid_df['database_cross_reference'].isna()==False] # different code types\n",
    "# doid_df[doid_df['has_alternative_id'].isna()==False] # different values of same ID type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gb/zy8l_k3d1s1d1dfxqb6h6_1r0000gn/T/ipykernel_10810/4016115131.py:2: DtypeWarning: Columns (8,16,17,18,26,31,32,33,47,52,53,55,56,58,62,63,65,69,71,73,77,83,84,92,104,108) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  doid_df_raw  =  pd.read_csv(os.path.join(DATA_DIR,'DOID.csv'))  # Read relationship mapping\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>Preferred Label</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>Definitions</th>\n",
       "      <th>CUI</th>\n",
       "      <th>database_cross_reference</th>\n",
       "      <th>has_alternative_id</th>\n",
       "      <th>has_exact_synonym</th>\n",
       "      <th>Parents</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>572</th>\n",
       "      <td>DOID:0001816</td>\n",
       "      <td>angiosarcoma</td>\n",
       "      <td>hemangiosarcoma</td>\n",
       "      <td>A vascular cancer that derives_from the cells ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNOMEDCT_US_2022_09_01:39000009|NCI:C3088|UMLS...</td>\n",
       "      <td>DOID:4508|DOID:267</td>\n",
       "      <td>hemangiosarcoma</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>573</th>\n",
       "      <td>DOID:0002116</td>\n",
       "      <td>pterygium</td>\n",
       "      <td>surfer's eye</td>\n",
       "      <td>A corneal disease that is characterized by a t...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UMLS_CUI:C0033999</td>\n",
       "      <td>NaN</td>\n",
       "      <td>surfer's eye</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_10124</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>DOID:0014667</td>\n",
       "      <td>disease of metabolism</td>\n",
       "      <td>metabolic disease</td>\n",
       "      <td>A disease that involving errors in metabolic p...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD10CM:E88.9|ICD9CM:277.9|NCI:C3235|SNOMEDCT_...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>metabolic disease</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>576</th>\n",
       "      <td>DOID:0040002</td>\n",
       "      <td>aspirin allergy</td>\n",
       "      <td>ASA allergy|acetylsalicylic acid allergy</td>\n",
       "      <td>A drug allergy that has_allergic_trigger acety...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UMLS_CUI:C0004058|SNOMEDCT_US_2022_09_01:29358...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ASA allergy|acetylsalicylic acid allergy</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_0060500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>577</th>\n",
       "      <td>DOID:0040003</td>\n",
       "      <td>benzylpenicillin allergy</td>\n",
       "      <td>penicillin G allergy|benzyl penicillin allergy</td>\n",
       "      <td>A beta-lactam allergy that has_allergic_trigge...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SNOMEDCT_US_2022_09_01:294499007|UMLS_CUI:C057...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>penicillin G allergy|benzyl penicillin allergy</td>\n",
       "      <td>http://purl.obolibrary.org/obo/DOID_0060519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17114</th>\n",
       "      <td>SYMP:0000818</td>\n",
       "      <td>localized superficial mass</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD9CM_2005:782.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0000488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17115</th>\n",
       "      <td>SYMP:0000819</td>\n",
       "      <td>localized superficial swelling</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD9CM_2005:782.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0000488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17116</th>\n",
       "      <td>SYMP:0000820</td>\n",
       "      <td>mass in chest</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICD9CM_2005:786.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0000514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17148</th>\n",
       "      <td>SYMP:0000852</td>\n",
       "      <td>anuria</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A urinary system symptom that is characterized...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UMLS_ICD9CM_2005_AUI:A0243854|UMLS_CUI:C002896...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0000486</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17304</th>\n",
       "      <td>SYMP:0020061</td>\n",
       "      <td>outlet dysfunction constipation</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A constipation that is characterized by diffic...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UMLS_CUI:C0949134|ICD10CM:K59.02|ICD9CM:564.02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>http://purl.obolibrary.org/obo/SYMP_0019180</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10568 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 id                  Preferred Label   \n",
       "572    DOID:0001816                     angiosarcoma  \\\n",
       "573    DOID:0002116                        pterygium   \n",
       "574    DOID:0014667            disease of metabolism   \n",
       "576    DOID:0040002                  aspirin allergy   \n",
       "577    DOID:0040003         benzylpenicillin allergy   \n",
       "...             ...                              ...   \n",
       "17114  SYMP:0000818       localized superficial mass   \n",
       "17115  SYMP:0000819   localized superficial swelling   \n",
       "17116  SYMP:0000820                    mass in chest   \n",
       "17148  SYMP:0000852                           anuria   \n",
       "17304  SYMP:0020061  outlet dysfunction constipation   \n",
       "\n",
       "                                             Synonyms   \n",
       "572                                   hemangiosarcoma  \\\n",
       "573                                      surfer's eye   \n",
       "574                                 metabolic disease   \n",
       "576          ASA allergy|acetylsalicylic acid allergy   \n",
       "577    penicillin G allergy|benzyl penicillin allergy   \n",
       "...                                               ...   \n",
       "17114                                             NaN   \n",
       "17115                                             NaN   \n",
       "17116                                             NaN   \n",
       "17148                                             NaN   \n",
       "17304                                             NaN   \n",
       "\n",
       "                                             Definitions  CUI   \n",
       "572    A vascular cancer that derives_from the cells ...  NaN  \\\n",
       "573    A corneal disease that is characterized by a t...  NaN   \n",
       "574    A disease that involving errors in metabolic p...  NaN   \n",
       "576    A drug allergy that has_allergic_trigger acety...  NaN   \n",
       "577    A beta-lactam allergy that has_allergic_trigge...  NaN   \n",
       "...                                                  ...  ...   \n",
       "17114                                                NaN  NaN   \n",
       "17115                                                NaN  NaN   \n",
       "17116                                                NaN  NaN   \n",
       "17148  A urinary system symptom that is characterized...  NaN   \n",
       "17304  A constipation that is characterized by diffic...  NaN   \n",
       "\n",
       "                                database_cross_reference  has_alternative_id   \n",
       "572    SNOMEDCT_US_2022_09_01:39000009|NCI:C3088|UMLS...  DOID:4508|DOID:267  \\\n",
       "573                                    UMLS_CUI:C0033999                 NaN   \n",
       "574    ICD10CM:E88.9|ICD9CM:277.9|NCI:C3235|SNOMEDCT_...                 NaN   \n",
       "576    UMLS_CUI:C0004058|SNOMEDCT_US_2022_09_01:29358...                 NaN   \n",
       "577    SNOMEDCT_US_2022_09_01:294499007|UMLS_CUI:C057...                 NaN   \n",
       "...                                                  ...                 ...   \n",
       "17114                                  ICD9CM_2005:782.2                 NaN   \n",
       "17115                                  ICD9CM_2005:782.2                 NaN   \n",
       "17116                                  ICD9CM_2005:786.6                 NaN   \n",
       "17148  UMLS_ICD9CM_2005_AUI:A0243854|UMLS_CUI:C002896...                 NaN   \n",
       "17304     UMLS_CUI:C0949134|ICD10CM:K59.02|ICD9CM:564.02                 NaN   \n",
       "\n",
       "                                    has_exact_synonym   \n",
       "572                                   hemangiosarcoma  \\\n",
       "573                                      surfer's eye   \n",
       "574                                 metabolic disease   \n",
       "576          ASA allergy|acetylsalicylic acid allergy   \n",
       "577    penicillin G allergy|benzyl penicillin allergy   \n",
       "...                                               ...   \n",
       "17114                                             NaN   \n",
       "17115                                             NaN   \n",
       "17116                                             NaN   \n",
       "17148                                             NaN   \n",
       "17304                                             NaN   \n",
       "\n",
       "                                           Parents  \n",
       "572        http://purl.obolibrary.org/obo/DOID_175  \n",
       "573      http://purl.obolibrary.org/obo/DOID_10124  \n",
       "574          http://purl.obolibrary.org/obo/DOID_4  \n",
       "576    http://purl.obolibrary.org/obo/DOID_0060500  \n",
       "577    http://purl.obolibrary.org/obo/DOID_0060519  \n",
       "...                                            ...  \n",
       "17114  http://purl.obolibrary.org/obo/SYMP_0000488  \n",
       "17115  http://purl.obolibrary.org/obo/SYMP_0000488  \n",
       "17116  http://purl.obolibrary.org/obo/SYMP_0000514  \n",
       "17148  http://purl.obolibrary.org/obo/SYMP_0000486  \n",
       "17304  http://purl.obolibrary.org/obo/SYMP_0019180  \n",
       "\n",
       "[10568 rows x 9 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# CHECK filter DOID to relevant codes\n",
    "# doid_df_raw  =  pd.read_csv(os.path.join(DATA_DIR,'DOID.csv'))  # Read relationship mapping  \n",
    "\n",
    "# doid_vars = ['id', 'Preferred Label', 'Synonyms', 'Definitions', 'CUI', 'database_cross_reference', 'has_alternative_id', 'has_exact_synonym', 'Parents']\n",
    "\n",
    "# doid_df = doid_df_raw[doid_vars]\n",
    "# # doid_df\n",
    "\n",
    "# doid_othercodes_df = doid_df[doid_df['database_cross_reference'].str.upper().str.contains(r'UMLS|ICD|OMIM|MESH', na=False)]\n",
    "# doid_othercodes_df\n",
    "# doid_othercodes_df[doid_othercodes_df['database_cross_reference'].str.contains('MESH')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 332,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     True\n",
       "1     True\n",
       "2    False\n",
       "3    False\n",
       "4    False\n",
       "5    False\n",
       "Name: 0, dtype: bool"
      ]
     },
     "execution_count": 332,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Test regex strings\n",
    "test_regex = pd.DataFrame([\"ICD10CM:E88.9|ICD9CM:277.9|NCI:C3235|UMLS_CUI:C0033\", \n",
    "                           \"abc|ICD10CM:E88.9|ICD9CM:277.9|UMLS_CUI:C0033|NCI:C3235\", \n",
    "                           \"ICD10CM:E8889|ICD9CM:277.9|NCI:C3235|UMLS_CUI:C0033asdf\", \n",
    "                           \"ICD10CM:xE88.9|ICD9CM:277.9|NCI:C3235|UMLS_CUI:xyczC0033\",\n",
    "                           \"bICD10CM:E88.9|ICD9CM:277.9|NCI:C3235|asdfUMLS_CUI:C0033\",\n",
    "                           \"ICD10CM:xyz|UMLS:E88.9|ICD9CM:277.9|NCI:C3235|UMLS_CUI:zdfsj|xya:C0033\"])\n",
    "test_regex[0].str.contains('(?:^|\\|)ICD10(?:(?!\\|).)*:E88\\.9.*', regex=True) # For ICD codes, need to make sure to escape middle period\n",
    "\n",
    "# Expected results for rows (starting index = 1)\n",
    "# 1 and 2 true, matches and ontology is at start (see 5)\n",
    "# 3 false, For ICD codes, need to make sure to escape middle period, \n",
    "# 4 false, have : in front of all codes\n",
    "# 5 false, ontology can't start with another value\n",
    "# 6 prevent | between code and value\n",
    "\n",
    "# Below: For non-ICD, prevent additional characters after end (ICD ok)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 333,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     True\n",
       "1     True\n",
       "2    False\n",
       "3    False\n",
       "4    False\n",
       "5    False\n",
       "Name: 0, dtype: bool"
      ]
     },
     "execution_count": 333,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_regex[0].str.contains('(?:^|\\|)UMLS(?:(?!\\|).)*:C0033(?:$|\\|)', regex=True) # For non-ICD, prevent additional characters after end (ICD ok)\n",
    "\n",
    "# Expected results for rows (starting index = 1)\n",
    "# 1 and 2 true, matches and ontology is at start (see 5)\n",
    "# 3 false, for non-ICD, prevent additional characters after end (ICD ok)\n",
    "# 4 false, have : in front of all codes\n",
    "# 5 false, ontology can't start with another value\n",
    "# 6 prevent | between code and value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>Name</th>\n",
       "      <th>code_source</th>\n",
       "      <th>code</th>\n",
       "      <th>Disordermappingrelation</th>\n",
       "      <th>Disordermappingicdrelation</th>\n",
       "      <th>Disordermappingvalidationstatus</th>\n",
       "      <th>code_source_upper</th>\n",
       "      <th>regex</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>166024</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>Q77\\.3</td>\n",
       "      <td>NTBT (ORPHA code's Narrower Term maps to a Bro...</td>\n",
       "      <td>Attributed (The ICD code is attributed by Orph...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>(?:^|\\|)ICD10(?:(?!\\|).)*:Q77.3.*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>166024</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>OMIM</td>\n",
       "      <td>607131</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>OMIM</td>\n",
       "      <td>(?:^|\\|)OMIM(?:(?!\\|).)*:607131(?:$|\\|)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>166024</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C1846722</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>(?:^|\\|)UMLS(?:(?!\\|).)*:C1846722(?:$|\\|)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>58</td>\n",
       "      <td>Alexander disease</td>\n",
       "      <td>OMIM</td>\n",
       "      <td>203450</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>OMIM</td>\n",
       "      <td>(?:^|\\|)OMIM(?:(?!\\|).)*:203450(?:$|\\|)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>58</td>\n",
       "      <td>Alexander disease</td>\n",
       "      <td>MeSH</td>\n",
       "      <td>D038261</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>MESH</td>\n",
       "      <td>(?:^|\\|)MESH(?:(?!\\|).)*:D038261(?:$|\\|)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30794</th>\n",
       "      <td>620368</td>\n",
       "      <td>EGF-related primary hypomagnesemia with intell...</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C5681825</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>(?:^|\\|)UMLS(?:(?!\\|).)*:C5681825(?:$|\\|)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30795</th>\n",
       "      <td>617910</td>\n",
       "      <td>Conjunctival malignant melanoma</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C0346360</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>(?:^|\\|)UMLS(?:(?!\\|).)*:C0346360(?:$|\\|)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30796</th>\n",
       "      <td>619948</td>\n",
       "      <td>Early-onset autoimmunity-autoinflammation-immu...</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C5680416</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>(?:^|\\|)UMLS(?:(?!\\|).)*:C5680416(?:$|\\|)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30797</th>\n",
       "      <td>619360</td>\n",
       "      <td>NON RARE IN EUROPE: Isolated hereditary persis...</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>D56\\.4</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>Specific code (The ORPHA code has its own code...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>(?:^|\\|)ICD10(?:(?!\\|).)*:D56.4.*</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30798</th>\n",
       "      <td>619233</td>\n",
       "      <td>Hereditary persistence of fetal hemoglobin-int...</td>\n",
       "      <td>OMIM</td>\n",
       "      <td>617101</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>OMIM</td>\n",
       "      <td>(?:^|\\|)OMIM(?:(?!\\|).)*:617101(?:$|\\|)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30799 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orphacode                                               Name   \n",
       "0        166024      Multiple epiphyseal dysplasia, Al-Gazali type  \\\n",
       "1        166024      Multiple epiphyseal dysplasia, Al-Gazali type   \n",
       "2        166024      Multiple epiphyseal dysplasia, Al-Gazali type   \n",
       "3            58                                  Alexander disease   \n",
       "4            58                                  Alexander disease   \n",
       "...         ...                                                ...   \n",
       "30794    620368  EGF-related primary hypomagnesemia with intell...   \n",
       "30795    617910                    Conjunctival malignant melanoma   \n",
       "30796    619948  Early-onset autoimmunity-autoinflammation-immu...   \n",
       "30797    619360  NON RARE IN EUROPE: Isolated hereditary persis...   \n",
       "30798    619233  Hereditary persistence of fetal hemoglobin-int...   \n",
       "\n",
       "      code_source      code   \n",
       "0          ICD-10    Q77\\.3  \\\n",
       "1            OMIM    607131   \n",
       "2            UMLS  C1846722   \n",
       "3            OMIM    203450   \n",
       "4            MeSH   D038261   \n",
       "...           ...       ...   \n",
       "30794        UMLS  C5681825   \n",
       "30795        UMLS  C0346360   \n",
       "30796        UMLS  C5680416   \n",
       "30797      ICD-10    D56\\.4   \n",
       "30798        OMIM    617101   \n",
       "\n",
       "                                 Disordermappingrelation   \n",
       "0      NTBT (ORPHA code's Narrower Term maps to a Bro...  \\\n",
       "1      E (Exact mapping: the two concepts are equival...   \n",
       "2      E (Exact mapping: the two concepts are equival...   \n",
       "3      E (Exact mapping: the two concepts are equival...   \n",
       "4      E (Exact mapping: the two concepts are equival...   \n",
       "...                                                  ...   \n",
       "30794  E (Exact mapping: the two concepts are equival...   \n",
       "30795  E (Exact mapping: the two concepts are equival...   \n",
       "30796  E (Exact mapping: the two concepts are equival...   \n",
       "30797  E (Exact mapping: the two concepts are equival...   \n",
       "30798  E (Exact mapping: the two concepts are equival...   \n",
       "\n",
       "                              Disordermappingicdrelation   \n",
       "0      Attributed (The ICD code is attributed by Orph...  \\\n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "30794                                                NaN   \n",
       "30795                                                NaN   \n",
       "30796                                                NaN   \n",
       "30797  Specific code (The ORPHA code has its own code...   \n",
       "30798                                                NaN   \n",
       "\n",
       "      Disordermappingvalidationstatus code_source_upper   \n",
       "0                           Validated            ICD-10  \\\n",
       "1                           Validated              OMIM   \n",
       "2                           Validated              UMLS   \n",
       "3                           Validated              OMIM   \n",
       "4                           Validated              MESH   \n",
       "...                               ...               ...   \n",
       "30794                       Validated              UMLS   \n",
       "30795                       Validated              UMLS   \n",
       "30796                       Validated              UMLS   \n",
       "30797                       Validated            ICD-10   \n",
       "30798                       Validated              OMIM   \n",
       "\n",
       "                                           regex  \n",
       "0              (?:^|\\|)ICD10(?:(?!\\|).)*:Q77.3.*  \n",
       "1        (?:^|\\|)OMIM(?:(?!\\|).)*:607131(?:$|\\|)  \n",
       "2      (?:^|\\|)UMLS(?:(?!\\|).)*:C1846722(?:$|\\|)  \n",
       "3        (?:^|\\|)OMIM(?:(?!\\|).)*:203450(?:$|\\|)  \n",
       "4       (?:^|\\|)MESH(?:(?!\\|).)*:D038261(?:$|\\|)  \n",
       "...                                          ...  \n",
       "30794  (?:^|\\|)UMLS(?:(?!\\|).)*:C5681825(?:$|\\|)  \n",
       "30795  (?:^|\\|)UMLS(?:(?!\\|).)*:C0346360(?:$|\\|)  \n",
       "30796  (?:^|\\|)UMLS(?:(?!\\|).)*:C5680416(?:$|\\|)  \n",
       "30797          (?:^|\\|)ICD10(?:(?!\\|).)*:D56.4.*  \n",
       "30798    (?:^|\\|)OMIM(?:(?!\\|).)*:617101(?:$|\\|)  \n",
       "\n",
       "[30799 rows x 9 columns]"
      ]
     },
     "execution_count": 335,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Create regex strings for orphan disease codes\n",
    "\n",
    "# regex format: \n",
    "# ICD:   '(?:^|\\|)ICD10(?:(?!\\|).)*:E88\\.9.*'\n",
    "# Other: '(?:^|\\|)UMLS(?:(?!\\|).)*:C0033(?:$|\\|)'\n",
    "orphan_codes_match = orphan_codes.copy()\n",
    "orphan_codes_match['code'] = np.where(orphan_codes_match['code_source'].str.startswith('ICD'), \n",
    "                                       orphan_codes_match['code'].str.replace('.','\\.'),\n",
    "                                       orphan_codes_match['code'])\n",
    "orphan_codes_match['regex'] = '(?:^|\\|)'+orphan_codes['code_source'].str.upper()+'(?:(?!\\|).)*:'+orphan_codes['code']\n",
    "orphan_codes_match['regex'] = orphan_codes_match['regex'].str.replace('-', '', regex=True)\n",
    "orphan_codes_match['regex'] = np.where(orphan_codes_match['code_source'].str.startswith('ICD'), \n",
    "                                       orphan_codes_match['regex']+'.*',\n",
    "                                       orphan_codes_match['regex']+'(?:$|\\|)')\n",
    "\n",
    "# .isin(doid_df['database_cross_reference'].str.upper())\n",
    "orphan_codes_match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 336,
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge by regex\n",
    "# source: https://stackoverflow.com/questions/62521616/can-i-perform-a-left-join-merge-between-two-dataframes-using-regular-expressions\n",
    "import re\n",
    "def merge_regex(regex_df, regex_col, search_df, search_col):\n",
    "    idx = [(i,j) for i,r in enumerate(regex_df[regex_col]) for j,v in enumerate(search_df[search_col].astype(str)) if re.match(r,v)]\n",
    "    regex_df_idx, search_df_idx = zip(*idx)\n",
    "    t = regex_df.iloc[list(regex_df_idx),0].reset_index(drop=True)\n",
    "    t1 = search_df.iloc[list(search_df_idx),0].reset_index(drop=True)\n",
    "    return pd.concat([t,t1],axis=1)\n",
    "\n",
    "test = merge_regex(orphan_codes_match, 'regex', doid_df, 'database_cross_reference')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 341,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>58</td>\n",
       "      <td>DOID:4252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>58</td>\n",
       "      <td>DOID:1926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>93</td>\n",
       "      <td>DOID:0050461</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>166035</td>\n",
       "      <td>DOID:0110137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>166035</td>\n",
       "      <td>DOID:0110139</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21508</th>\n",
       "      <td>99792</td>\n",
       "      <td>DOID:0110065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21509</th>\n",
       "      <td>99792</td>\n",
       "      <td>DOID:701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21510</th>\n",
       "      <td>99772</td>\n",
       "      <td>DOID:0110214</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21511</th>\n",
       "      <td>99777</td>\n",
       "      <td>DOID:0050602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21512</th>\n",
       "      <td>617910</td>\n",
       "      <td>DOID:1751</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>21513 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orphacode            id\n",
       "0            58     DOID:4252\n",
       "1            58     DOID:1926\n",
       "2            93  DOID:0050461\n",
       "3        166035  DOID:0110137\n",
       "4        166035  DOID:0110139\n",
       "...         ...           ...\n",
       "21508     99792  DOID:0110065\n",
       "21509     99792      DOID:701\n",
       "21510     99772  DOID:0110214\n",
       "21511     99777  DOID:0050602\n",
       "21512    617910     DOID:1751\n",
       "\n",
       "[21513 rows x 2 columns]"
      ]
     },
     "execution_count": 341,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>drkg_dict_id</th>\n",
       "      <th>name</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_code</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>code</th>\n",
       "      <th>matched</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>58268</th>\n",
       "      <td>Side Effect::C0000727</td>\n",
       "      <td>Side Effect::C0000727</td>\n",
       "      <td>Acute abdomen</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C0000727</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C0000727</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58269</th>\n",
       "      <td>Side Effect::C0000729</td>\n",
       "      <td>Side Effect::C0000729</td>\n",
       "      <td>Abdominal cramps</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C0000729</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C0000729</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58270</th>\n",
       "      <td>Side Effect::C0000731</td>\n",
       "      <td>Side Effect::C0000731</td>\n",
       "      <td>Abdominal distension</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C0000731</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C0000731</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58271</th>\n",
       "      <td>Side Effect::C0000735</td>\n",
       "      <td>Side Effect::C0000735</td>\n",
       "      <td>Abdominal neoplasm</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C0000735</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C0000735</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58272</th>\n",
       "      <td>Side Effect::C0000737</td>\n",
       "      <td>Side Effect::C0000737</td>\n",
       "      <td>Abdominal pain</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C0000737</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C0000737</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63964</th>\n",
       "      <td>Side Effect::C3665624</td>\n",
       "      <td>Side Effect::C3665624</td>\n",
       "      <td>Blood calcium decreased</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C3665624</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C3665624</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63965</th>\n",
       "      <td>Side Effect::C3665770</td>\n",
       "      <td>Side Effect::C3665770</td>\n",
       "      <td>Acquired lipoatrophic diabetes</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C3665770</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C3665770</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63966</th>\n",
       "      <td>Side Effect::C3665818</td>\n",
       "      <td>Side Effect::C3665818</td>\n",
       "      <td>Medication residue present</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C3665818</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C3665818</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63967</th>\n",
       "      <td>Side Effect::C3665888</td>\n",
       "      <td>Side Effect::C3665888</td>\n",
       "      <td>Reproductive toxicity</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C3665888</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C3665888</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63968</th>\n",
       "      <td>Side Effect::C3665900</td>\n",
       "      <td>Side Effect::C3665900</td>\n",
       "      <td>Intraspinal bleeding</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C3665900</td>\n",
       "      <td>UMLS CUI</td>\n",
       "      <td>C3665900</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5701 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     drkg_id           drkg_dict_id   \n",
       "58268  Side Effect::C0000727  Side Effect::C0000727  \\\n",
       "58269  Side Effect::C0000729  Side Effect::C0000729   \n",
       "58270  Side Effect::C0000731  Side Effect::C0000731   \n",
       "58271  Side Effect::C0000735  Side Effect::C0000735   \n",
       "58272  Side Effect::C0000737  Side Effect::C0000737   \n",
       "...                      ...                    ...   \n",
       "63964  Side Effect::C3665624  Side Effect::C3665624   \n",
       "63965  Side Effect::C3665770  Side Effect::C3665770   \n",
       "63966  Side Effect::C3665818  Side Effect::C3665818   \n",
       "63967  Side Effect::C3665888  Side Effect::C3665888   \n",
       "63968  Side Effect::C3665900  Side Effect::C3665900   \n",
       "\n",
       "                                 name  entity_type ontology_code   \n",
       "58268                   Acute abdomen  Side Effect      C0000727  \\\n",
       "58269                Abdominal cramps  Side Effect      C0000729   \n",
       "58270            Abdominal distension  Side Effect      C0000731   \n",
       "58271              Abdominal neoplasm  Side Effect      C0000735   \n",
       "58272                  Abdominal pain  Side Effect      C0000737   \n",
       "...                               ...          ...           ...   \n",
       "63964         Blood calcium decreased  Side Effect      C3665624   \n",
       "63965  Acquired lipoatrophic diabetes  Side Effect      C3665770   \n",
       "63966      Medication residue present  Side Effect      C3665818   \n",
       "63967           Reproductive toxicity  Side Effect      C3665888   \n",
       "63968            Intraspinal bleeding  Side Effect      C3665900   \n",
       "\n",
       "      ontology_name      code  matched  \n",
       "58268      UMLS CUI  C0000727        1  \n",
       "58269      UMLS CUI  C0000729        1  \n",
       "58270      UMLS CUI  C0000731        1  \n",
       "58271      UMLS CUI  C0000735        1  \n",
       "58272      UMLS CUI  C0000737        1  \n",
       "...             ...       ...      ...  \n",
       "63964      UMLS CUI  C3665624        1  \n",
       "63965      UMLS CUI  C3665770        1  \n",
       "63966      UMLS CUI  C3665818        1  \n",
       "63967      UMLS CUI  C3665888        1  \n",
       "63968      UMLS CUI  C3665900        1  \n",
       "\n",
       "[5701 rows x 8 columns]"
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drkg_all_entities[drkg_all_entities['ontology_name'].str.contains('UMLS', na=False)]\n",
    "# drkg_all_entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>ct</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>match_type</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">1</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Disease</th>\n",
       "      <th>MESH</th>\n",
       "      <td>1033</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OMIM</th>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Side Effect</th>\n",
       "      <th>UMLS</th>\n",
       "      <td>407</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symptom</th>\n",
       "      <th>MESH</th>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <th>Disease</th>\n",
       "      <th>DOID</th>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">3</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Disease</th>\n",
       "      <th>DOID</th>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MESH</th>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Side Effect</th>\n",
       "      <th>UMLS</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Symptom</th>\n",
       "      <th>MESH</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        ct\n",
       "match_type entity_type ontology_name      \n",
       "1          Disease     MESH           1033\n",
       "                       OMIM             67\n",
       "           Side Effect UMLS            407\n",
       "           Symptom     MESH             41\n",
       "2          Disease     DOID             21\n",
       "3          Disease     DOID              7\n",
       "                       MESH             54\n",
       "           Side Effect UMLS              6\n",
       "           Symptom     MESH              3"
      ]
     },
     "execution_count": 342,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NEW\n",
    "doid_orphan_codes = test.merge(orphan_codes, how='left', on='Orphacode').drop_duplicates()\n",
    "\n",
    "drkg_all_entities['ontology_name'] = np.where(drkg_all_entities['ontology_name'].str.contains('UMLS'), 'UMLS', drkg_all_entities['ontology_name'])\n",
    "\n",
    "# Get matches on MeSH and OMIM codes\n",
    "orphan_codes['code_source_upper'] = orphan_codes['code_source'].str.upper()\n",
    "match_try1 = drkg_all_entities.merge(orphan_codes, how='inner', left_on=['ontology_name', 'code'], right_on=['code_source_upper', 'code'])\n",
    "match_try1['match_type'] = 1\n",
    "\n",
    "match_try2 = drkg_all_entities.merge(doid_orphan_codes, how='inner', left_on='ontology_code', right_on='id')\n",
    "match_try2['match_type'] = 2\n",
    "\n",
    "# Match by name\n",
    "match_try3 = drkg_all_entities.merge(orphan_codes, how='inner', left_on=drkg_all_entities['name'].str.upper(), right_on=orphan_codes['Name'].str.upper())\n",
    "match_try3['match_type'] = 3\n",
    "\n",
    "# Stack matched entities\n",
    "matched_rarediseases = pd.concat([match_try1, match_try2, match_try3], ignore_index=True, axis=0).drop_duplicates(subset=['drkg_id', 'Orphacode']) # keeps first entry of duplicates\n",
    "matched_rarediseases.drop_duplicates(subset='drkg_id').drop(columns=['Disordermappingicdrelation', 'Disordermappingvalidationstatus']) # Show just unique entities in DRKG\n",
    "matched_rarediseases.groupby(by=['match_type', 'entity_type', 'ontology_name']).agg(\n",
    "    ct = ('drkg_id', 'count')\n",
    ")\n",
    "# matched_rarediseases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 348,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>drkg_dict_id</th>\n",
       "      <th>name</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_code</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>code</th>\n",
       "      <th>matched</th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>Name</th>\n",
       "      <th>code_source</th>\n",
       "      <th>Disordermappingrelation</th>\n",
       "      <th>Disordermappingicdrelation</th>\n",
       "      <th>Disordermappingvalidationstatus</th>\n",
       "      <th>code_source_upper</th>\n",
       "      <th>match_type</th>\n",
       "      <th>code_x</th>\n",
       "      <th>id</th>\n",
       "      <th>code_y</th>\n",
       "      <th>key_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1711</th>\n",
       "      <td>Disease::DOID:0050156</td>\n",
       "      <td>Disease::DOID:0050156</td>\n",
       "      <td>idiopathic pulmonary fibrosis</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:0050156</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2032</td>\n",
       "      <td>Idiopathic pulmonary fibrosis</td>\n",
       "      <td>MedDRA</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>MEDDRA</td>\n",
       "      <td>3</td>\n",
       "      <td>0050156</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10021240</td>\n",
       "      <td>IDIOPATHIC PULMONARY FIBROSIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1865</th>\n",
       "      <td>Disease::DOID:12930</td>\n",
       "      <td>Disease::DOID:12930</td>\n",
       "      <td>dilated cardiomyopathy</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:12930</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>217604</td>\n",
       "      <td>Dilated cardiomyopathy</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>Specific code (The ORPHA code has its own code...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>3</td>\n",
       "      <td>12930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>I42.0</td>\n",
       "      <td>DILATED CARDIOMYOPATHY</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1870</th>\n",
       "      <td>Disease::DOID:13378</td>\n",
       "      <td>Disease::DOID:13378</td>\n",
       "      <td>Kawasaki disease</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:13378</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>2331</td>\n",
       "      <td>Kawasaki disease</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>Specific code (The ORPHA code has its own code...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>3</td>\n",
       "      <td>13378</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4A44.5</td>\n",
       "      <td>KAWASAKI DISEASE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876</th>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>sclerosing cholangitis</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>447771</td>\n",
       "      <td>Sclerosing cholangitis</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>3</td>\n",
       "      <td>14268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C0008313</td>\n",
       "      <td>SCLEROSING CHOLANGITIS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1877</th>\n",
       "      <td>Disease::DOID:2394</td>\n",
       "      <td>Disease::DOID:2394</td>\n",
       "      <td>ovarian cancer</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:2394</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>213500</td>\n",
       "      <td>Ovarian cancer</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>3</td>\n",
       "      <td>2394</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C1140680</td>\n",
       "      <td>OVARIAN CANCER</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8061</th>\n",
       "      <td>Side Effect::C0221242</td>\n",
       "      <td>Side Effect::C0221242</td>\n",
       "      <td>Fixed drug eruption</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C0221242</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>293812</td>\n",
       "      <td>Fixed drug eruption</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>3</td>\n",
       "      <td>C0221242</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C0877391</td>\n",
       "      <td>FIXED DRUG ERUPTION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8104</th>\n",
       "      <td>Side Effect::C0334634</td>\n",
       "      <td>Side Effect::C0334634</td>\n",
       "      <td>Mantle cell lymphoma</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C0334634</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>52416</td>\n",
       "      <td>Mantle cell lymphoma</td>\n",
       "      <td>GARD</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>GARD</td>\n",
       "      <td>3</td>\n",
       "      <td>C0334634</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6969</td>\n",
       "      <td>MANTLE CELL LYMPHOMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8110</th>\n",
       "      <td>Side Effect::C0334660</td>\n",
       "      <td>Side Effect::C0334660</td>\n",
       "      <td>Intravascular large B-cell lymphoma</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C0334660</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>98839</td>\n",
       "      <td>Intravascular large B-cell lymphoma</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>Specific code (The ORPHA code has its own code...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>3</td>\n",
       "      <td>C0334660</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2A81.1</td>\n",
       "      <td>INTRAVASCULAR LARGE B-CELL LYMPHOMA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8187</th>\n",
       "      <td>Side Effect::C2242577</td>\n",
       "      <td>Side Effect::C2242577</td>\n",
       "      <td>Oromandibular dystonia</td>\n",
       "      <td>Side Effect</td>\n",
       "      <td>C2242577</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>93958</td>\n",
       "      <td>Oromandibular dystonia</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>3</td>\n",
       "      <td>C2242577</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C0393607</td>\n",
       "      <td>OROMANDIBULAR DYSTONIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8223</th>\n",
       "      <td>Symptom::D019569</td>\n",
       "      <td>Symptom::D019569</td>\n",
       "      <td>Hemifacial Spasm</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>D019569</td>\n",
       "      <td>MESH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>221083</td>\n",
       "      <td>Hemifacial spasm</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>Specific code (The ORPHA code has its own code...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>3</td>\n",
       "      <td>D019569</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8B88.2</td>\n",
       "      <td>HEMIFACIAL SPASM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>73 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    drkg_id           drkg_dict_id   \n",
       "1711  Disease::DOID:0050156  Disease::DOID:0050156  \\\n",
       "1865    Disease::DOID:12930    Disease::DOID:12930   \n",
       "1870    Disease::DOID:13378    Disease::DOID:13378   \n",
       "1876    Disease::DOID:14268    Disease::DOID:14268   \n",
       "1877     Disease::DOID:2394     Disease::DOID:2394   \n",
       "...                     ...                    ...   \n",
       "8061  Side Effect::C0221242  Side Effect::C0221242   \n",
       "8104  Side Effect::C0334634  Side Effect::C0334634   \n",
       "8110  Side Effect::C0334660  Side Effect::C0334660   \n",
       "8187  Side Effect::C2242577  Side Effect::C2242577   \n",
       "8223       Symptom::D019569       Symptom::D019569   \n",
       "\n",
       "                                     name  entity_type ontology_code   \n",
       "1711        idiopathic pulmonary fibrosis      Disease  DOID:0050156  \\\n",
       "1865               dilated cardiomyopathy      Disease    DOID:12930   \n",
       "1870                     Kawasaki disease      Disease    DOID:13378   \n",
       "1876               sclerosing cholangitis      Disease    DOID:14268   \n",
       "1877                       ovarian cancer      Disease     DOID:2394   \n",
       "...                                   ...          ...           ...   \n",
       "8061                  Fixed drug eruption  Side Effect      C0221242   \n",
       "8104                 Mantle cell lymphoma  Side Effect      C0334634   \n",
       "8110  Intravascular large B-cell lymphoma  Side Effect      C0334660   \n",
       "8187               Oromandibular dystonia  Side Effect      C2242577   \n",
       "8223                     Hemifacial Spasm      Symptom       D019569   \n",
       "\n",
       "     ontology_name code  matched Orphacode   \n",
       "1711          DOID  NaN        1      2032  \\\n",
       "1865          DOID  NaN        1    217604   \n",
       "1870          DOID  NaN        1      2331   \n",
       "1876          DOID  NaN        1    447771   \n",
       "1877          DOID  NaN        1    213500   \n",
       "...            ...  ...      ...       ...   \n",
       "8061          UMLS  NaN        1    293812   \n",
       "8104          UMLS  NaN        1     52416   \n",
       "8110          UMLS  NaN        1     98839   \n",
       "8187          UMLS  NaN        1     93958   \n",
       "8223          MESH  NaN        1    221083   \n",
       "\n",
       "                                     Name code_source   \n",
       "1711        Idiopathic pulmonary fibrosis      MedDRA  \\\n",
       "1865               Dilated cardiomyopathy      ICD-10   \n",
       "1870                     Kawasaki disease      ICD-11   \n",
       "1876               Sclerosing cholangitis        UMLS   \n",
       "1877                       Ovarian cancer        UMLS   \n",
       "...                                   ...         ...   \n",
       "8061                  Fixed drug eruption        UMLS   \n",
       "8104                 Mantle cell lymphoma        GARD   \n",
       "8110  Intravascular large B-cell lymphoma      ICD-11   \n",
       "8187               Oromandibular dystonia        UMLS   \n",
       "8223                     Hemifacial spasm      ICD-11   \n",
       "\n",
       "                                Disordermappingrelation   \n",
       "1711  E (Exact mapping: the two concepts are equival...  \\\n",
       "1865  E (Exact mapping: the two concepts are equival...   \n",
       "1870  E (Exact mapping: the two concepts are equival...   \n",
       "1876  E (Exact mapping: the two concepts are equival...   \n",
       "1877  E (Exact mapping: the two concepts are equival...   \n",
       "...                                                 ...   \n",
       "8061  E (Exact mapping: the two concepts are equival...   \n",
       "8104  E (Exact mapping: the two concepts are equival...   \n",
       "8110  E (Exact mapping: the two concepts are equival...   \n",
       "8187  E (Exact mapping: the two concepts are equival...   \n",
       "8223  E (Exact mapping: the two concepts are equival...   \n",
       "\n",
       "                             Disordermappingicdrelation   \n",
       "1711                                                NaN  \\\n",
       "1865  Specific code (The ORPHA code has its own code...   \n",
       "1870  Specific code (The ORPHA code has its own code...   \n",
       "1876                                                NaN   \n",
       "1877                                                NaN   \n",
       "...                                                 ...   \n",
       "8061                                                NaN   \n",
       "8104                                                NaN   \n",
       "8110  Specific code (The ORPHA code has its own code...   \n",
       "8187                                                NaN   \n",
       "8223  Specific code (The ORPHA code has its own code...   \n",
       "\n",
       "     Disordermappingvalidationstatus code_source_upper  match_type    code_x   \n",
       "1711                       Validated            MEDDRA           3   0050156  \\\n",
       "1865                       Validated            ICD-10           3     12930   \n",
       "1870                       Validated            ICD-11           3     13378   \n",
       "1876                       Validated              UMLS           3     14268   \n",
       "1877                       Validated              UMLS           3      2394   \n",
       "...                              ...               ...         ...       ...   \n",
       "8061                       Validated              UMLS           3  C0221242   \n",
       "8104                       Validated              GARD           3  C0334634   \n",
       "8110                       Validated            ICD-11           3  C0334660   \n",
       "8187                       Validated              UMLS           3  C2242577   \n",
       "8223                       Validated            ICD-11           3   D019569   \n",
       "\n",
       "       id    code_y                                key_0  \n",
       "1711  NaN  10021240        IDIOPATHIC PULMONARY FIBROSIS  \n",
       "1865  NaN     I42.0               DILATED CARDIOMYOPATHY  \n",
       "1870  NaN    4A44.5                     KAWASAKI DISEASE  \n",
       "1876  NaN  C0008313               SCLEROSING CHOLANGITIS  \n",
       "1877  NaN  C1140680                       OVARIAN CANCER  \n",
       "...   ...       ...                                  ...  \n",
       "8061  NaN  C0877391                  FIXED DRUG ERUPTION  \n",
       "8104  NaN      6969                 MANTLE CELL LYMPHOMA  \n",
       "8110  NaN    2A81.1  INTRAVASCULAR LARGE B-CELL LYMPHOMA  \n",
       "8187  NaN  C0393607               OROMANDIBULAR DYSTONIA  \n",
       "8223  NaN    8B88.2                     HEMIFACIAL SPASM  \n",
       "\n",
       "[73 rows x 20 columns]"
      ]
     },
     "execution_count": 348,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check those matched by name only\n",
    "matched_rarediseases[matched_rarediseases['match_type']==3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 346,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>name</th>\n",
       "      <th>match_type</th>\n",
       "      <th>ct_orphacode</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Disease::DOID:10021</td>\n",
       "      <td>duodenum cancer</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Disease::DOID:10153</td>\n",
       "      <td>ileum cancer</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>sclerosing cholangitis</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153</th>\n",
       "      <td>Disease::MESH:C536385</td>\n",
       "      <td>Focal facial dermal dysplasia type I</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>184</th>\n",
       "      <td>Disease::MESH:C536611</td>\n",
       "      <td>Brooke-Spiegler syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>301</th>\n",
       "      <td>Disease::MESH:C537771</td>\n",
       "      <td>Anorectal malformation</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>Disease::MESH:D000168</td>\n",
       "      <td>Acrocephalosyndactylia</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>390</th>\n",
       "      <td>Disease::MESH:D000699</td>\n",
       "      <td>Pain Insensitivity, Congenital</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>429</th>\n",
       "      <td>Disease::MESH:D002292</td>\n",
       "      <td>Carcinoma, Renal Cell</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>499</th>\n",
       "      <td>Disease::MESH:D005359</td>\n",
       "      <td>Fibrous Dysplasia, Polyostotic</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>590</th>\n",
       "      <td>Disease::MESH:D008228</td>\n",
       "      <td>Lymphoma, Non-Hodgkin</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>643</th>\n",
       "      <td>Disease::MESH:D009456</td>\n",
       "      <td>Neurofibromatosis 1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>649</th>\n",
       "      <td>Disease::MESH:D009477</td>\n",
       "      <td>Hereditary Sensory and Autonomic Neuropathies</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>Disease::MESH:D011666</td>\n",
       "      <td>Pulmonary Valve Stenosis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>768</th>\n",
       "      <td>Disease::MESH:D014884</td>\n",
       "      <td>Waterhouse-Friderichsen Syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>788</th>\n",
       "      <td>Disease::MESH:D015432</td>\n",
       "      <td>Glomerulonephritis, Membranoproliferative</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>791</th>\n",
       "      <td>Disease::MESH:D015456</td>\n",
       "      <td>Leukemia, Biphenotypic, Acute</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>895</th>\n",
       "      <td>Disease::MESH:D018901</td>\n",
       "      <td>Peroxisomal Disorders</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>932</th>\n",
       "      <td>Disease::MESH:D020389</td>\n",
       "      <td>Muscular Dystrophy, Emery-Dreifuss</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>Disease::MESH:D020967</td>\n",
       "      <td>Myotonic Disorders</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>944</th>\n",
       "      <td>Disease::MESH:D024741</td>\n",
       "      <td>Cardiomyopathy, Hypertrophic, Familial</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>984</th>\n",
       "      <td>Disease::MESH:D052456</td>\n",
       "      <td>Hypoalphalipoproteinemias</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1090</th>\n",
       "      <td>Disease::OMIM:102530</td>\n",
       "      <td>SPERMATOGENIC FAILURE 6</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>Disease::OMIM:155600</td>\n",
       "      <td>MELANOMA, CUTANEOUS MALIGNANT, SUSCEPTIBILITY ...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1109</th>\n",
       "      <td>Disease::OMIM:162000</td>\n",
       "      <td>TUBULOINTERSTITIAL KIDNEY DISEASE, AUTOSOMAL D...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1110</th>\n",
       "      <td>Disease::OMIM:165800</td>\n",
       "      <td>SHORT STATURE AND ADVANCED BONE AGE, WITH OR W...</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1126</th>\n",
       "      <td>Disease::OMIM:277000</td>\n",
       "      <td>MAYER-ROKITANSKY-KUSTER-HAUSER SYNDROME</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1129</th>\n",
       "      <td>Disease::OMIM:300049</td>\n",
       "      <td>PERIVENTRICULAR NODULAR HETEROTOPIA 1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1130</th>\n",
       "      <td>Disease::OMIM:301050</td>\n",
       "      <td>ALPORT SYNDROME 1, X-LINKED</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1143</th>\n",
       "      <td>Disease::OMIM:613091</td>\n",
       "      <td>SHORT-RIB THORACIC DYSPLASIA 3 WITH OR WITHOUT...</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1151</th>\n",
       "      <td>Side Effect::C0001403</td>\n",
       "      <td>Addison's disease</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>Side Effect::C0002893</td>\n",
       "      <td>Refractory anemia</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1162</th>\n",
       "      <td>Side Effect::C0002894</td>\n",
       "      <td>Refractory anaemia with an excess of blasts</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1174</th>\n",
       "      <td>Side Effect::C0005779</td>\n",
       "      <td>Coagulopathy</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1180</th>\n",
       "      <td>Side Effect::C0006413</td>\n",
       "      <td>Burkitt's lymphoma</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1192</th>\n",
       "      <td>Side Effect::C0008924</td>\n",
       "      <td>Cleft lip</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1195</th>\n",
       "      <td>Side Effect::C0010278</td>\n",
       "      <td>Craniosynostosis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1198</th>\n",
       "      <td>Side Effect::C0010481</td>\n",
       "      <td>Cushing's syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1204</th>\n",
       "      <td>Side Effect::C0011860</td>\n",
       "      <td>Type 2 diabetes mellitus</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1212</th>\n",
       "      <td>Side Effect::C0014518</td>\n",
       "      <td>Toxic epidermal necrolysis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1291</th>\n",
       "      <td>Side Effect::C0026272</td>\n",
       "      <td>Mixed connective tissue disease</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1318</th>\n",
       "      <td>Side Effect::C0031036</td>\n",
       "      <td>Polyarteritis nodosa</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1365</th>\n",
       "      <td>Side Effect::C0041408</td>\n",
       "      <td>Turner's syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1390</th>\n",
       "      <td>Side Effect::C0085786</td>\n",
       "      <td>Hamman-Rich syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1399</th>\n",
       "      <td>Side Effect::C0152171</td>\n",
       "      <td>Primary pulmonary hypertension</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1425</th>\n",
       "      <td>Side Effect::C0206762</td>\n",
       "      <td>Limb malformation</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1563</th>\n",
       "      <td>Symptom::D014884</td>\n",
       "      <td>Waterhouse-Friderichsen Syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    drkg_id   \n",
       "1       Disease::DOID:10021  \\\n",
       "2       Disease::DOID:10153   \n",
       "10      Disease::DOID:14268   \n",
       "153   Disease::MESH:C536385   \n",
       "184   Disease::MESH:C536611   \n",
       "301   Disease::MESH:C537771   \n",
       "375   Disease::MESH:D000168   \n",
       "390   Disease::MESH:D000699   \n",
       "429   Disease::MESH:D002292   \n",
       "499   Disease::MESH:D005359   \n",
       "590   Disease::MESH:D008228   \n",
       "643   Disease::MESH:D009456   \n",
       "649   Disease::MESH:D009477   \n",
       "699   Disease::MESH:D011666   \n",
       "768   Disease::MESH:D014884   \n",
       "788   Disease::MESH:D015432   \n",
       "791   Disease::MESH:D015456   \n",
       "895   Disease::MESH:D018901   \n",
       "932   Disease::MESH:D020389   \n",
       "940   Disease::MESH:D020967   \n",
       "944   Disease::MESH:D024741   \n",
       "984   Disease::MESH:D052456   \n",
       "1090   Disease::OMIM:102530   \n",
       "1105   Disease::OMIM:155600   \n",
       "1109   Disease::OMIM:162000   \n",
       "1110   Disease::OMIM:165800   \n",
       "1126   Disease::OMIM:277000   \n",
       "1129   Disease::OMIM:300049   \n",
       "1130   Disease::OMIM:301050   \n",
       "1143   Disease::OMIM:613091   \n",
       "1151  Side Effect::C0001403   \n",
       "1161  Side Effect::C0002893   \n",
       "1162  Side Effect::C0002894   \n",
       "1174  Side Effect::C0005779   \n",
       "1180  Side Effect::C0006413   \n",
       "1192  Side Effect::C0008924   \n",
       "1195  Side Effect::C0010278   \n",
       "1198  Side Effect::C0010481   \n",
       "1204  Side Effect::C0011860   \n",
       "1212  Side Effect::C0014518   \n",
       "1291  Side Effect::C0026272   \n",
       "1318  Side Effect::C0031036   \n",
       "1365  Side Effect::C0041408   \n",
       "1390  Side Effect::C0085786   \n",
       "1399  Side Effect::C0152171   \n",
       "1425  Side Effect::C0206762   \n",
       "1563       Symptom::D014884   \n",
       "\n",
       "                                                   name  match_type   \n",
       "1                                       duodenum cancer           2  \\\n",
       "2                                          ileum cancer           2   \n",
       "10                               sclerosing cholangitis           2   \n",
       "153                Focal facial dermal dysplasia type I           1   \n",
       "184                            Brooke-Spiegler syndrome           1   \n",
       "301                              Anorectal malformation           1   \n",
       "375                              Acrocephalosyndactylia           1   \n",
       "390                      Pain Insensitivity, Congenital           1   \n",
       "429                               Carcinoma, Renal Cell           1   \n",
       "499                      Fibrous Dysplasia, Polyostotic           1   \n",
       "590                               Lymphoma, Non-Hodgkin           1   \n",
       "643                                 Neurofibromatosis 1           1   \n",
       "649       Hereditary Sensory and Autonomic Neuropathies           1   \n",
       "699                            Pulmonary Valve Stenosis           1   \n",
       "768                    Waterhouse-Friderichsen Syndrome           1   \n",
       "788           Glomerulonephritis, Membranoproliferative           1   \n",
       "791                       Leukemia, Biphenotypic, Acute           1   \n",
       "895                               Peroxisomal Disorders           1   \n",
       "932                  Muscular Dystrophy, Emery-Dreifuss           1   \n",
       "940                                  Myotonic Disorders           1   \n",
       "944              Cardiomyopathy, Hypertrophic, Familial           1   \n",
       "984                           Hypoalphalipoproteinemias           1   \n",
       "1090                            SPERMATOGENIC FAILURE 6           1   \n",
       "1105  MELANOMA, CUTANEOUS MALIGNANT, SUSCEPTIBILITY ...           1   \n",
       "1109  TUBULOINTERSTITIAL KIDNEY DISEASE, AUTOSOMAL D...           1   \n",
       "1110  SHORT STATURE AND ADVANCED BONE AGE, WITH OR W...           1   \n",
       "1126            MAYER-ROKITANSKY-KUSTER-HAUSER SYNDROME           1   \n",
       "1129              PERIVENTRICULAR NODULAR HETEROTOPIA 1           1   \n",
       "1130                        ALPORT SYNDROME 1, X-LINKED           1   \n",
       "1143  SHORT-RIB THORACIC DYSPLASIA 3 WITH OR WITHOUT...           1   \n",
       "1151                                  Addison's disease           1   \n",
       "1161                                  Refractory anemia           1   \n",
       "1162        Refractory anaemia with an excess of blasts           1   \n",
       "1174                                       Coagulopathy           1   \n",
       "1180                                 Burkitt's lymphoma           1   \n",
       "1192                                          Cleft lip           1   \n",
       "1195                                   Craniosynostosis           1   \n",
       "1198                                 Cushing's syndrome           1   \n",
       "1204                           Type 2 diabetes mellitus           1   \n",
       "1212                         Toxic epidermal necrolysis           1   \n",
       "1291                    Mixed connective tissue disease           1   \n",
       "1318                               Polyarteritis nodosa           1   \n",
       "1365                                  Turner's syndrome           1   \n",
       "1390                               Hamman-Rich syndrome           1   \n",
       "1399                     Primary pulmonary hypertension           1   \n",
       "1425                                  Limb malformation           1   \n",
       "1563                   Waterhouse-Friderichsen Syndrome           1   \n",
       "\n",
       "      ct_orphacode  \n",
       "1                2  \n",
       "2                3  \n",
       "10               4  \n",
       "153              2  \n",
       "184              2  \n",
       "301              2  \n",
       "375              2  \n",
       "390              2  \n",
       "429              2  \n",
       "499              2  \n",
       "590              2  \n",
       "643              2  \n",
       "649              2  \n",
       "699              2  \n",
       "768              2  \n",
       "788              2  \n",
       "791              2  \n",
       "895              3  \n",
       "932              2  \n",
       "940              2  \n",
       "944              2  \n",
       "984              2  \n",
       "1090             2  \n",
       "1105             2  \n",
       "1109             2  \n",
       "1110             2  \n",
       "1126             2  \n",
       "1129             2  \n",
       "1130             2  \n",
       "1143             4  \n",
       "1151             2  \n",
       "1161             2  \n",
       "1162             2  \n",
       "1174             2  \n",
       "1180             2  \n",
       "1192             2  \n",
       "1195             2  \n",
       "1198             2  \n",
       "1204             2  \n",
       "1212             2  \n",
       "1291             2  \n",
       "1318             2  \n",
       "1365             2  \n",
       "1390             2  \n",
       "1399             2  \n",
       "1425             2  \n",
       "1563             2  "
      ]
     },
     "execution_count": 346,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# DRKG entities with multiple orphanet codes?\n",
    "ct_orphacode = matched_rarediseases.groupby(by=['drkg_id','name', 'match_type']).agg(\n",
    "    ct_orphacode = ('Orphacode', 'count')\n",
    ").reset_index()\n",
    "ct_orphacode[ct_orphacode['ct_orphacode']>1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 344,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>drkg_dict_id</th>\n",
       "      <th>name</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_code</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>code</th>\n",
       "      <th>matched</th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>Name</th>\n",
       "      <th>code_source</th>\n",
       "      <th>Disordermappingrelation</th>\n",
       "      <th>Disordermappingicdrelation</th>\n",
       "      <th>Disordermappingvalidationstatus</th>\n",
       "      <th>code_source_upper</th>\n",
       "      <th>match_type</th>\n",
       "      <th>code_x</th>\n",
       "      <th>id</th>\n",
       "      <th>code_y</th>\n",
       "      <th>key_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1654</th>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>sclerosing cholangitis</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>171</td>\n",
       "      <td>Primary sclerosing cholangitis</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>Specific code (The ORPHA code has its own code...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>2</td>\n",
       "      <td>14268</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>DB96.2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1662</th>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>sclerosing cholangitis</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>447774</td>\n",
       "      <td>Secondary sclerosing cholangitis</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>NTBT (ORPHA code's Narrower Term maps to a Bro...</td>\n",
       "      <td>Index term (The ORPHA code is listed in the IC...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-11</td>\n",
       "      <td>2</td>\n",
       "      <td>14268</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>DC13</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1664</th>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>sclerosing cholangitis</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>447764</td>\n",
       "      <td>IgG4-related sclerosing cholangitis</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>NTBT (ORPHA code's Narrower Term maps to a Bro...</td>\n",
       "      <td>Attributed (The ICD code is attributed by Orph...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>2</td>\n",
       "      <td>14268</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>K83.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1665</th>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>sclerosing cholangitis</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>480556</td>\n",
       "      <td>Isolated neonatal sclerosing cholangitis</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>NTBT (ORPHA code's Narrower Term maps to a Bro...</td>\n",
       "      <td>Attributed (The ICD code is attributed by Orph...</td>\n",
       "      <td>Not yet validated</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>2</td>\n",
       "      <td>14268</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>K83.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1876</th>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>Disease::DOID:14268</td>\n",
       "      <td>sclerosing cholangitis</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:14268</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>447771</td>\n",
       "      <td>Sclerosing cholangitis</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>3</td>\n",
       "      <td>14268</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C0008313</td>\n",
       "      <td>SCLEROSING CHOLANGITIS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  drkg_id         drkg_dict_id                    name   \n",
       "1654  Disease::DOID:14268  Disease::DOID:14268  sclerosing cholangitis  \\\n",
       "1662  Disease::DOID:14268  Disease::DOID:14268  sclerosing cholangitis   \n",
       "1664  Disease::DOID:14268  Disease::DOID:14268  sclerosing cholangitis   \n",
       "1665  Disease::DOID:14268  Disease::DOID:14268  sclerosing cholangitis   \n",
       "1876  Disease::DOID:14268  Disease::DOID:14268  sclerosing cholangitis   \n",
       "\n",
       "     entity_type ontology_code ontology_name code  matched Orphacode   \n",
       "1654     Disease    DOID:14268          DOID  NaN        1       171  \\\n",
       "1662     Disease    DOID:14268          DOID  NaN        1    447774   \n",
       "1664     Disease    DOID:14268          DOID  NaN        1    447764   \n",
       "1665     Disease    DOID:14268          DOID  NaN        1    480556   \n",
       "1876     Disease    DOID:14268          DOID  NaN        1    447771   \n",
       "\n",
       "                                          Name code_source   \n",
       "1654            Primary sclerosing cholangitis      ICD-11  \\\n",
       "1662          Secondary sclerosing cholangitis      ICD-11   \n",
       "1664       IgG4-related sclerosing cholangitis      ICD-10   \n",
       "1665  Isolated neonatal sclerosing cholangitis      ICD-10   \n",
       "1876                    Sclerosing cholangitis        UMLS   \n",
       "\n",
       "                                Disordermappingrelation   \n",
       "1654  E (Exact mapping: the two concepts are equival...  \\\n",
       "1662  NTBT (ORPHA code's Narrower Term maps to a Bro...   \n",
       "1664  NTBT (ORPHA code's Narrower Term maps to a Bro...   \n",
       "1665  NTBT (ORPHA code's Narrower Term maps to a Bro...   \n",
       "1876  E (Exact mapping: the two concepts are equival...   \n",
       "\n",
       "                             Disordermappingicdrelation   \n",
       "1654  Specific code (The ORPHA code has its own code...  \\\n",
       "1662  Index term (The ORPHA code is listed in the IC...   \n",
       "1664  Attributed (The ICD code is attributed by Orph...   \n",
       "1665  Attributed (The ICD code is attributed by Orph...   \n",
       "1876                                                NaN   \n",
       "\n",
       "     Disordermappingvalidationstatus code_source_upper  match_type code_x   \n",
       "1654                       Validated            ICD-11           2  14268  \\\n",
       "1662                       Validated            ICD-11           2  14268   \n",
       "1664                       Validated            ICD-10           2  14268   \n",
       "1665               Not yet validated            ICD-10           2  14268   \n",
       "1876                       Validated              UMLS           3  14268   \n",
       "\n",
       "              id    code_y                   key_0  \n",
       "1654  DOID:14268    DB96.2                     NaN  \n",
       "1662  DOID:14268      DC13                     NaN  \n",
       "1664  DOID:14268     K83.0                     NaN  \n",
       "1665  DOID:14268     K83.0                     NaN  \n",
       "1876         NaN  C0008313  SCLEROSING CHOLANGITIS  "
      ]
     },
     "execution_count": 344,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_rarediseases[matched_rarediseases['drkg_id']=='Disease::DOID:14268']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 347,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>name</th>\n",
       "      <th>match_type</th>\n",
       "      <th>ct_drkg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>100</td>\n",
       "      <td>Ataxia Telangiectasia</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100067</td>\n",
       "      <td>Waterhouse-Friderichsen Syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1041</td>\n",
       "      <td>Hydrops Fetalis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>1163</td>\n",
       "      <td>Aspergillosis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>1183</td>\n",
       "      <td>Opsoclonus-Myoclonus Syndrome</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>963</td>\n",
       "      <td>Acromegaly</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1437</th>\n",
       "      <td>97275</td>\n",
       "      <td>Encephalitis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1438</th>\n",
       "      <td>97279</td>\n",
       "      <td>Insulinoma</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1459</th>\n",
       "      <td>98292</td>\n",
       "      <td>Mastocytosis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1488</th>\n",
       "      <td>98715</td>\n",
       "      <td>Uveitis</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>79 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Orphacode                              name  match_type  ct_drkg_id\n",
       "1          100             Ataxia Telangiectasia           1           2\n",
       "4       100067  Waterhouse-Friderichsen Syndrome           1           2\n",
       "36        1041                   Hydrops Fetalis           1           2\n",
       "56        1163                     Aspergillosis           1           2\n",
       "61        1183     Opsoclonus-Myoclonus Syndrome           1           2\n",
       "...        ...                               ...         ...         ...\n",
       "1431       963                        Acromegaly           1           2\n",
       "1437     97275                      Encephalitis           1           2\n",
       "1438     97279                        Insulinoma           1           2\n",
       "1459     98292                      Mastocytosis           1           2\n",
       "1488     98715                           Uveitis           1           2\n",
       "\n",
       "[79 rows x 4 columns]"
      ]
     },
     "execution_count": 347,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Orphacodes with multiple DRKG entities?\n",
    "ct_drkg_id = matched_rarediseases.groupby(by=['Orphacode','name', 'match_type']).agg(\n",
    "    ct_drkg_id = ('drkg_id', 'count')\n",
    ").reset_index()\n",
    "ct_drkg_id[ct_drkg_id['ct_drkg_id']>1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>drkg_dict_id</th>\n",
       "      <th>name</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_code</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>code</th>\n",
       "      <th>matched</th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>Name</th>\n",
       "      <th>code_source</th>\n",
       "      <th>Disordermappingrelation</th>\n",
       "      <th>Disordermappingicdrelation</th>\n",
       "      <th>Disordermappingvalidationstatus</th>\n",
       "      <th>code_source_upper</th>\n",
       "      <th>key_0</th>\n",
       "      <th>code_x</th>\n",
       "      <th>code_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>371</th>\n",
       "      <td>Disease::MESH:D000699</td>\n",
       "      <td>MESH::D000699</td>\n",
       "      <td>Pain Insensitivity, Congenital</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MESH:D000699</td>\n",
       "      <td>MESH</td>\n",
       "      <td>D000699</td>\n",
       "      <td>1.0</td>\n",
       "      <td>88642</td>\n",
       "      <td>Congenital insensitivity to pain-anosmia-neuro...</td>\n",
       "      <td>MeSH</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>MESH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>636</th>\n",
       "      <td>Disease::MESH:D009477</td>\n",
       "      <td>MESH::D009477</td>\n",
       "      <td>Hereditary Sensory and Autonomic Neuropathies</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MESH:D009477</td>\n",
       "      <td>MESH</td>\n",
       "      <td>D009477</td>\n",
       "      <td>1.0</td>\n",
       "      <td>88642</td>\n",
       "      <td>Congenital insensitivity to pain-anosmia-neuro...</td>\n",
       "      <td>MeSH</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>MESH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   drkg_id   drkg_dict_id   \n",
       "371  Disease::MESH:D000699  MESH::D000699  \\\n",
       "636  Disease::MESH:D009477  MESH::D009477   \n",
       "\n",
       "                                              name entity_type ontology_code   \n",
       "371                 Pain Insensitivity, Congenital     Disease  MESH:D000699  \\\n",
       "636  Hereditary Sensory and Autonomic Neuropathies     Disease  MESH:D009477   \n",
       "\n",
       "    ontology_name     code  matched Orphacode   \n",
       "371          MESH  D000699      1.0     88642  \\\n",
       "636          MESH  D009477      1.0     88642   \n",
       "\n",
       "                                                  Name code_source   \n",
       "371  Congenital insensitivity to pain-anosmia-neuro...        MeSH  \\\n",
       "636  Congenital insensitivity to pain-anosmia-neuro...        MeSH   \n",
       "\n",
       "                               Disordermappingrelation   \n",
       "371  E (Exact mapping: the two concepts are equival...  \\\n",
       "636  E (Exact mapping: the two concepts are equival...   \n",
       "\n",
       "    Disordermappingicdrelation Disordermappingvalidationstatus   \n",
       "371                        NaN                       Validated  \\\n",
       "636                        NaN                       Validated   \n",
       "\n",
       "    code_source_upper key_0 code_x code_y  \n",
       "371              MESH   NaN    NaN    NaN  \n",
       "636              MESH   NaN    NaN    NaN  "
      ]
     },
     "execution_count": 232,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_rarediseases[matched_rarediseases['Orphacode']=='88642']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>drkg_dict_id</th>\n",
       "      <th>name</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_code</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>code</th>\n",
       "      <th>matched</th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>Name</th>\n",
       "      <th>code_source</th>\n",
       "      <th>Disordermappingrelation</th>\n",
       "      <th>Disordermappingicdrelation</th>\n",
       "      <th>Disordermappingvalidationstatus</th>\n",
       "      <th>code_source_upper</th>\n",
       "      <th>key_0</th>\n",
       "      <th>code_x</th>\n",
       "      <th>code_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>988</th>\n",
       "      <td>Disease::MESH:D053578</td>\n",
       "      <td>MESH::D053578</td>\n",
       "      <td>Opsoclonus-Myoclonus Syndrome</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MESH:D053578</td>\n",
       "      <td>MESH</td>\n",
       "      <td>D053578</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1183</td>\n",
       "      <td>Opsoclonus-myoclonus syndrome</td>\n",
       "      <td>MeSH</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>MESH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>989</th>\n",
       "      <td>Symptom::D053578</td>\n",
       "      <td>Symptom::D053578</td>\n",
       "      <td>Opsoclonus-Myoclonus Syndrome</td>\n",
       "      <td>Symptom</td>\n",
       "      <td>D053578</td>\n",
       "      <td>MESH</td>\n",
       "      <td>D053578</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1183</td>\n",
       "      <td>Opsoclonus-myoclonus syndrome</td>\n",
       "      <td>MeSH</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>MESH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   drkg_id      drkg_dict_id                           name   \n",
       "988  Disease::MESH:D053578     MESH::D053578  Opsoclonus-Myoclonus Syndrome  \\\n",
       "989       Symptom::D053578  Symptom::D053578  Opsoclonus-Myoclonus Syndrome   \n",
       "\n",
       "    entity_type ontology_code ontology_name     code  matched Orphacode   \n",
       "988     Disease  MESH:D053578          MESH  D053578      1.0      1183  \\\n",
       "989     Symptom       D053578          MESH  D053578      1.0      1183   \n",
       "\n",
       "                              Name code_source   \n",
       "988  Opsoclonus-myoclonus syndrome        MeSH  \\\n",
       "989  Opsoclonus-myoclonus syndrome        MeSH   \n",
       "\n",
       "                               Disordermappingrelation   \n",
       "988  E (Exact mapping: the two concepts are equival...  \\\n",
       "989  E (Exact mapping: the two concepts are equival...   \n",
       "\n",
       "    Disordermappingicdrelation Disordermappingvalidationstatus   \n",
       "988                        NaN                       Validated  \\\n",
       "989                        NaN                       Validated   \n",
       "\n",
       "    code_source_upper key_0 code_x code_y  \n",
       "988              MESH   NaN    NaN    NaN  \n",
       "989              MESH   NaN    NaN    NaN  "
      ]
     },
     "execution_count": 233,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_rarediseases[matched_rarediseases['Orphacode']=='1183']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>Name</th>\n",
       "      <th>code_source</th>\n",
       "      <th>code</th>\n",
       "      <th>Disordermappingrelation</th>\n",
       "      <th>Disordermappingicdrelation</th>\n",
       "      <th>Disordermappingvalidationstatus</th>\n",
       "      <th>code_source_upper</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>166024</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>Q77.3</td>\n",
       "      <td>NTBT (ORPHA code's Narrower Term maps to a Bro...</td>\n",
       "      <td>Attributed (The ICD code is attributed by Orph...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>166024</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>OMIM</td>\n",
       "      <td>607131</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>OMIM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>166024</td>\n",
       "      <td>Multiple epiphyseal dysplasia, Al-Gazali type</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C1846722</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>166032</td>\n",
       "      <td>Multiple epiphyseal dysplasia, with miniepiphyses</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>Q77.3</td>\n",
       "      <td>NTBT (ORPHA code's Narrower Term maps to a Bro...</td>\n",
       "      <td>Attributed (The ICD code is attributed by Orph...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>166032</td>\n",
       "      <td>Multiple epiphyseal dysplasia, with miniepiphyses</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C1836307</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30794</th>\n",
       "      <td>620368</td>\n",
       "      <td>EGF-related primary hypomagnesemia with intell...</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C5681825</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30795</th>\n",
       "      <td>617910</td>\n",
       "      <td>Conjunctival malignant melanoma</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C0346360</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30796</th>\n",
       "      <td>619948</td>\n",
       "      <td>Early-onset autoimmunity-autoinflammation-immu...</td>\n",
       "      <td>UMLS</td>\n",
       "      <td>C5680416</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>UMLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30797</th>\n",
       "      <td>619360</td>\n",
       "      <td>NON RARE IN EUROPE: Isolated hereditary persis...</td>\n",
       "      <td>ICD-10</td>\n",
       "      <td>D56.4</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>Specific code (The ORPHA code has its own code...</td>\n",
       "      <td>Validated</td>\n",
       "      <td>ICD-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30798</th>\n",
       "      <td>619233</td>\n",
       "      <td>Hereditary persistence of fetal hemoglobin-int...</td>\n",
       "      <td>OMIM</td>\n",
       "      <td>617101</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>OMIM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22807 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Orphacode                                               Name   \n",
       "0        166024      Multiple epiphyseal dysplasia, Al-Gazali type  \\\n",
       "1        166024      Multiple epiphyseal dysplasia, Al-Gazali type   \n",
       "2        166024      Multiple epiphyseal dysplasia, Al-Gazali type   \n",
       "9        166032  Multiple epiphyseal dysplasia, with miniepiphyses   \n",
       "10       166032  Multiple epiphyseal dysplasia, with miniepiphyses   \n",
       "...         ...                                                ...   \n",
       "30794    620368  EGF-related primary hypomagnesemia with intell...   \n",
       "30795    617910                    Conjunctival malignant melanoma   \n",
       "30796    619948  Early-onset autoimmunity-autoinflammation-immu...   \n",
       "30797    619360  NON RARE IN EUROPE: Isolated hereditary persis...   \n",
       "30798    619233  Hereditary persistence of fetal hemoglobin-int...   \n",
       "\n",
       "      code_source      code   \n",
       "0          ICD-10     Q77.3  \\\n",
       "1            OMIM    607131   \n",
       "2            UMLS  C1846722   \n",
       "9          ICD-10     Q77.3   \n",
       "10           UMLS  C1836307   \n",
       "...           ...       ...   \n",
       "30794        UMLS  C5681825   \n",
       "30795        UMLS  C0346360   \n",
       "30796        UMLS  C5680416   \n",
       "30797      ICD-10     D56.4   \n",
       "30798        OMIM    617101   \n",
       "\n",
       "                                 Disordermappingrelation   \n",
       "0      NTBT (ORPHA code's Narrower Term maps to a Bro...  \\\n",
       "1      E (Exact mapping: the two concepts are equival...   \n",
       "2      E (Exact mapping: the two concepts are equival...   \n",
       "9      NTBT (ORPHA code's Narrower Term maps to a Bro...   \n",
       "10     E (Exact mapping: the two concepts are equival...   \n",
       "...                                                  ...   \n",
       "30794  E (Exact mapping: the two concepts are equival...   \n",
       "30795  E (Exact mapping: the two concepts are equival...   \n",
       "30796  E (Exact mapping: the two concepts are equival...   \n",
       "30797  E (Exact mapping: the two concepts are equival...   \n",
       "30798  E (Exact mapping: the two concepts are equival...   \n",
       "\n",
       "                              Disordermappingicdrelation   \n",
       "0      Attributed (The ICD code is attributed by Orph...  \\\n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "9      Attributed (The ICD code is attributed by Orph...   \n",
       "10                                                   NaN   \n",
       "...                                                  ...   \n",
       "30794                                                NaN   \n",
       "30795                                                NaN   \n",
       "30796                                                NaN   \n",
       "30797  Specific code (The ORPHA code has its own code...   \n",
       "30798                                                NaN   \n",
       "\n",
       "      Disordermappingvalidationstatus code_source_upper  \n",
       "0                           Validated            ICD-10  \n",
       "1                           Validated              OMIM  \n",
       "2                           Validated              UMLS  \n",
       "9                           Validated            ICD-10  \n",
       "10                          Validated              UMLS  \n",
       "...                               ...               ...  \n",
       "30794                       Validated              UMLS  \n",
       "30795                       Validated              UMLS  \n",
       "30796                       Validated              UMLS  \n",
       "30797                       Validated            ICD-10  \n",
       "30798                       Validated              OMIM  \n",
       "\n",
       "[22807 rows x 8 columns]"
      ]
     },
     "execution_count": 244,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Non-matched rare diseases? (could still be in DRKG but under different name or code - need to check that embeddings are separate)\n",
    "\n",
    "unmatched_rarediseases = orphan_codes[~orphan_codes['Orphacode'].isin(matched_rarediseases['Orphacode'])]\n",
    "unmatched_rarediseases"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "oTAJOqDVc_Hg"
   },
   "source": [
    "# 3) BioLinkBERT embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import HeteroData\n",
    "from SIMP_LLM.llm_encode import EntityEncoder\n",
    "# from SIMP_LLM.dataloader_mappings import create_mapping, create_edges, embed_entities, embed_edges\n",
    "from DEBUG_dataloader_mappings import create_mapping, create_edges, embed_entities, embed_edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set variables and load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "id": "QkZKp6CJiw6f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+------------------------------------+---------------+--------------------------+---------------------------------------------+------------------------------------------------------------------------------------------+---------------------------------+---------------+---------------+------------------------------------+\n",
      "|    | drkg_id                            | Data-source   | Connected entity-types   | Interaction-type                            | Description                                                                              | Reference for the description   | head_entity   | tail_entity   | relation_name                      |\n",
      "|----+------------------------------------+---------------+--------------------------+---------------------------------------------+------------------------------------------------------------------------------------------+---------------------------------+---------------+---------------+------------------------------------|\n",
      "| 18 | DRUGBANK::treats::Compound:Disease | DRUGBANK      | Compound:Disease         | Compound treats the disease                 | We extract  the  associated  conditions  related  to  compound entities in the drugbank. |                                 | Compound      | Disease       | Compound treats the disease        |\n",
      "| 25 | GNBR::E+::Compound:Gene            | GNBR          | Compound:Gene            | increases expression/production             |                                                                                          |                                 | Compound      | Gene          | increases expression or production |\n",
      "| 26 | GNBR::E+::Gene:Gene                | GNBR          | Gene:Gene                | increases expression/production             |                                                                                          |                                 | Gene          | Gene          | increases expression or production |\n",
      "| 37 | GNBR::Md::Gene:Disease             | GNBR          | Disease:Gene             | biomarkers (diagnostic)                     |                                                                                          |                                 | Gene          | Disease       | biomarkers diagnostic              |\n",
      "| 46 | GNBR::T::Compound:Disease          | GNBR          | Compound:Disease         | treatment/therapy (including investigatory) |                                                                                          |                                 | Compound      | Disease       | Compound treats the disease        |\n",
      "+----+------------------------------------+---------------+--------------------------+---------------------------------------------+------------------------------------------------------------------------------------------+---------------------------------+---------------+---------------+------------------------------------+\n",
      "+----+-----------------------------+------------------------------------+-----------------------------------------------+\n",
      "|    | 0                           | 1                                  | 2                                             |\n",
      "|----+-----------------------------+------------------------------------+-----------------------------------------------|\n",
      "|  0 | Dornase alfa                | DRUGBANK::treats::Compound:Disease | Cystic Fibrosis                               |\n",
      "|  1 | Etanercept                  | DRUGBANK::treats::Compound:Disease | Spondylitis, Ankylosing                       |\n",
      "|  2 | Etanercept                  | DRUGBANK::treats::Compound:Disease | Graft vs Host Disease                         |\n",
      "|  3 | cysteine                    | GNBR::E+::Compound:Gene            | lipoprotein lipase (LPL)                      |\n",
      "|  4 | 5,6,7,8-tetrahydrobiopterin | GNBR::E+::Compound:Gene            | prostaglandin-endoperoxide synthase 2 (PTGS2) |\n",
      "+----+-----------------------------+------------------------------------+-----------------------------------------------+\n",
      "+-------+------------------------+------------------------+-----------------------------+---------------+-----------------+-----------------+--------------+\n",
      "|       | drkg_id                | drkg_dict_id           | name                        | entity_type   | ontology_code   | ontology_name   | code         |\n",
      "|-------+------------------------+------------------------+-----------------------------+---------------+-----------------+-----------------+--------------|\n",
      "| 17241 | Compound::CHEBI:15356  | Compound::CHEBI:15356  | cysteine                    | Compound      | CHEBI:15356     | CHEBI           | 15356        |\n",
      "| 17242 | Compound::CHEBI:15372  | Compound::CHEBI:15372  | 5,6,7,8-tetrahydrobiopterin | Compound      | CHEBI:15372     | CHEBI           | 15372        |\n",
      "| 17243 | Compound::CHEBI:15414  | Compound::CHEBI:15414  | S-adenosyl-L-methionine     | Compound      | CHEBI:15414     | CHEBI           | 15414        |\n",
      "| 20247 | Compound::CHEMBL224120 | Compound::CHEMBL224120 | S-adenosyl-L-methionine     | Compound      | CHEMBL224120    | CHEMBL          | CHEMBL224120 |\n",
      "| 21962 | Compound::DB00003      | Compound::DB00003      | Dornase alfa                | Compound      | DB00003         | drugbank        | DB00003      |\n",
      "+-------+------------------------+------------------------+-----------------------------+---------------+-----------------+-----------------+--------------+\n"
     ]
    }
   ],
   "source": [
    "## Set variables\n",
    "device   = \"cpu\"\n",
    "Encoder  = EntityEncoder(device = device )\n",
    "run_full_sample = 0\n",
    "\n",
    "if run_full_sample:\n",
    "    # Run full DRKG\n",
    "    entity_df = drkg_entity_df.copy()\n",
    "    hrt_data = drkg_translated.copy()\n",
    "    relation_lookup = relation_df.copy()\n",
    "else:\n",
    "    # Create relationship subset for testing\n",
    "    test_list = ['Compound treats the disease', 'increases expression or production', 'biomarkers diagnostic']\n",
    "\n",
    "    # Create test sample of DRKG relationships filtering to these relations (for full sample: delete and use drkg_entity_df)\n",
    "    test_relation_df = relation_df[relation_df['relation_name'].isin(test_list)].copy()\n",
    "    print_head(test_relation_df)\n",
    "\n",
    "    # Create test sample of DRKG relationships filtering to these relations (for full sample: delete and use drkg_entity_df)\n",
    "    test_hrt_df = drkg_translated[drkg_translated[1].isin(test_relation_df['drkg_id'])]\n",
    "    test_hrt_df = test_hrt_df.groupby(1).head(3).reset_index(drop=True)\n",
    "    test_unique_entities = get_unique_entities(test_hrt_df, columns=[0,2])\n",
    "    test_entity_df = drkg_entity_df[drkg_entity_df['name'].isin(test_unique_entities)]\n",
    "    print_head(test_hrt_df)\n",
    "    print_head(test_entity_df)\n",
    "\n",
    "    entity_df = test_entity_df.copy()\n",
    "    hrt_data = test_hrt_df.copy()\n",
    "    relation_lookup = relation_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relation_name</th>\n",
       "      <th>new_relation_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>drug-drug interaction</td>\n",
       "      <td>drugdrug interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Compound belongs to Anatomical Therapeutic Che...</td>\n",
       "      <td>Compound belongs to Anatomical Therapeutic Che...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>binding, ligand (esp. receptors)</td>\n",
       "      <td>binding or  ligand especially receptors</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>inhibits cell growth (esp. cancers)</td>\n",
       "      <td>inhibits cell growth especially cancers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>increases expression/production</td>\n",
       "      <td>increases expression or production</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>decreases expression/production</td>\n",
       "      <td>decreases expression or production</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>affects expression/production (neutral)</td>\n",
       "      <td>affects expression or production neutral</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>metabolism, pharmacokinetics</td>\n",
       "      <td>metabolism or  pharmacokinetics</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>biomarkers (diagnostic)</td>\n",
       "      <td>biomarkers diagnostic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>biomarkers (of disease progression)</td>\n",
       "      <td>biomarkers of disease progression</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>transport, channels</td>\n",
       "      <td>transport or  channels</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>alleviates, reduces</td>\n",
       "      <td>alleviates or  reduces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>prevents, suppresses</td>\n",
       "      <td>prevents or  suppresses</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>side effect/adverse event</td>\n",
       "      <td>side effect or adverse event</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>activates, stimulates</td>\n",
       "      <td>activates or  stimulates</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>post-translational modification</td>\n",
       "      <td>posttranslational modification</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        relation_name   \n",
       "14                              drug-drug interaction  \\\n",
       "17  Compound belongs to Anatomical Therapeutic Che...   \n",
       "22                   binding, ligand (esp. receptors)   \n",
       "23                inhibits cell growth (esp. cancers)   \n",
       "25                    increases expression/production   \n",
       "27                    decreases expression/production   \n",
       "28            affects expression/production (neutral)   \n",
       "35                       metabolism, pharmacokinetics   \n",
       "37                            biomarkers (diagnostic)   \n",
       "38                biomarkers (of disease progression)   \n",
       "40                                transport, channels   \n",
       "41                                alleviates, reduces   \n",
       "42                               prevents, suppresses   \n",
       "45                          side effect/adverse event   \n",
       "50                              activates, stimulates   \n",
       "99                    post-translational modification   \n",
       "\n",
       "                                    new_relation_name  \n",
       "14                               drugdrug interaction  \n",
       "17  Compound belongs to Anatomical Therapeutic Che...  \n",
       "22            binding or  ligand especially receptors  \n",
       "23            inhibits cell growth especially cancers  \n",
       "25                 increases expression or production  \n",
       "27                 decreases expression or production  \n",
       "28           affects expression or production neutral  \n",
       "35                    metabolism or  pharmacokinetics  \n",
       "37                              biomarkers diagnostic  \n",
       "38                  biomarkers of disease progression  \n",
       "40                             transport or  channels  \n",
       "41                             alleviates or  reduces  \n",
       "42                            prevents or  suppresses  \n",
       "45                       side effect or adverse event  \n",
       "50                           activates or  stimulates  \n",
       "99                     posttranslational modification  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for special characters in relation name\n",
    "relation_lookup[relation_lookup['relation_name'].str.contains('/')]\n",
    "relation_df['relation_name'][relation_df['relation_name'].str.replace(' ', '').str.contains(r\"[^a-zA-Z0-9]+\", regex=True)].drop_duplicates()\n",
    "\n",
    "# Replace and check\n",
    "relation_df_test = relation_df.copy()\n",
    "relation_df_test['relation_name'] = relation_df_test['relation_name'].str.replace(',|/', ' or ', regex=True)\n",
    "relation_df_test['relation_name'] = relation_df_test['relation_name'].str.replace('esp.','especially')\n",
    "relation_df_test['relation_name'] = relation_df_test['relation_name'].str.replace('\\(|\\)|-|\\.', '', regex=True)\n",
    "relation_df_test['relation_name'][relation_df_test['relation_name'].str.replace(' ', '').str.contains(r\"[^a-zA-Z0-9]+\", regex=True)].drop_duplicates()\n",
    "\n",
    "# Check differences\n",
    "relation_df_test = relation_df.copy()\n",
    "relation_df_test['new_relation_name'] = relation_df_test['relation_name']\n",
    "relation_df_test['new_relation_name'] = relation_df_test['new_relation_name'].str.replace(',|/', ' or ', regex=True)\n",
    "relation_df_test['new_relation_name'] = relation_df_test['new_relation_name'].str.replace('esp.','especially')\n",
    "relation_df_test['new_relation_name'] = relation_df_test['new_relation_name'].str.replace('\\(|\\)|-|\\.', '', regex=True)\n",
    "relation_df_test[['relation_name', 'new_relation_name']][relation_df_test['relation_name'] != relation_df_test['new_relation_name']].drop_duplicates() #.loc[17, 'new_relation_name']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Compound belongs to Anatomical Therapeutic Chemical ATC code'"
      ]
     },
     "execution_count": 134,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "relation_df.loc[17,'relation_name']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "ywHcEVMiAdOv"
   },
   "source": [
    "### Build HeteroData Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HeteroData(\n",
      "  \u001b[1mCompound\u001b[0m={ x=[12, 768] },\n",
      "  \u001b[1mDisease\u001b[0m={ x=[11, 768] },\n",
      "  \u001b[1mGene\u001b[0m={ x=[12, 768] },\n",
      "  \u001b[1m(Compound, Compound treats the disease, Disease)\u001b[0m={\n",
      "    edge_index=[2, 9],\n",
      "    edge_label=[9, 768]\n",
      "  },\n",
      "  \u001b[1m(Compound, increases expression or production, Gene)\u001b[0m={\n",
      "    edge_index=[2, 6],\n",
      "    edge_label=[6, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, increases expression or production, Gene)\u001b[0m={\n",
      "    edge_index=[2, 3],\n",
      "    edge_label=[3, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, biomarkers diagnostic, Disease)\u001b[0m={\n",
      "    edge_index=[2, 3],\n",
      "    edge_label=[3, 768]\n",
      "  }\n",
      ")\n",
      "Unique Compounds: 12 \t Matrix shape: torch.Size([12, 768])\n",
      "Unique Diseases: 11 \t Matrix shape: torch.Size([11, 768])\n",
      "Unique Genes: 12 \t Matrix shape: torch.Size([12, 768])\n"
     ]
    }
   ],
   "source": [
    "# Initialize heterograph object\n",
    "data = HeteroData()\n",
    "\n",
    "# Embed entities, add to graph, and save embedding mapping dictionary of dictionaries\n",
    "mapping_dict = embed_entities(entity_df, data, Encoder, device) \n",
    "\n",
    "# Embed relationships, add to graph, and save relation embeddings/mapping dictionary\n",
    "relation_X, relation_mapping = embed_edges(hrt_data, relation_lookup, data, mapping_dict, Encoder, device, data_path=DATA_DIR)\n",
    "\n",
    "# Print summary\n",
    "#data = T.ToUndirected()(data)\n",
    "\n",
    "print(data)\n",
    "for ent_type in entity_df['entity_type'].unique():\n",
    "    print(f\"Unique {ent_type}s: {len(mapping_dict[ent_type])} \\t Matrix shape: {data[ent_type].x.shape }\")\n",
    "    # print(mapping_dict[ent_type]) # Prints whole dictionary so delete/uncomment if using all entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HeteroData(\n",
       "  \u001b[1mCompound\u001b[0m={ x=[12, 768] },\n",
       "  \u001b[1mDisease\u001b[0m={ x=[11, 768] },\n",
       "  \u001b[1mGene\u001b[0m={ x=[12, 768] },\n",
       "  \u001b[1m(Compound, Compound treats the disease, Disease)\u001b[0m={\n",
       "    edge_index=[2, 9],\n",
       "    edge_label=[9, 768]\n",
       "  },\n",
       "  \u001b[1m(Compound, increases expression or production, Gene)\u001b[0m={\n",
       "    edge_index=[2, 6],\n",
       "    edge_label=[6, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, increases expression or production, Gene)\u001b[0m={\n",
       "    edge_index=[2, 6],\n",
       "    edge_label=[6, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, biomarkers diagnostic, Disease)\u001b[0m={\n",
       "    edge_index=[2, 3],\n",
       "    edge_label=[3, 768]\n",
       "  },\n",
       "  \u001b[1m(Disease, rev_Compound treats the disease, Compound)\u001b[0m={\n",
       "    edge_index=[2, 9],\n",
       "    edge_label=[9, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, rev_increases expression or production, Compound)\u001b[0m={\n",
       "    edge_index=[2, 6],\n",
       "    edge_label=[6, 768]\n",
       "  },\n",
       "  \u001b[1m(Disease, rev_biomarkers diagnostic, Gene)\u001b[0m={\n",
       "    edge_index=[2, 3],\n",
       "    edge_label=[3, 768]\n",
       "  }\n",
       ")"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "data1 = T.ToUndirected()(data)\n",
    "data1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[ 7,  6,  3,  8,  1,  4],\n",
       "        [ 3,  9,  2, 10,  7,  6]])"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[('Compound', 'increases expression/production', 'Gene')].edge_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([2, 3])\n",
      "HeteroData(\n",
      "  \u001b[1mCompound\u001b[0m={ x=[13, 768] },\n",
      "  \u001b[1mDisease\u001b[0m={ x=[3, 768] },\n",
      "  \u001b[1m(Compound, treats, Disease)\u001b[0m={ edge_index=[2, 3] },\n",
      "  \u001b[1m(Disease, rev_treats, Compound)\u001b[0m={ edge_index=[2, 3] }\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "# Experimental: Ignore ##\n",
    "data2 = HeteroData()\n",
    "data2['Compound'].x  = data[\"Compound\"][\"x\"]\n",
    "data2['Disease'].x   = data[\"Disease\"][\"x\"]\n",
    "\n",
    "ctd                  = data[(\"Compound\", \"Compound treats the disease\", \"Disease\")][\"edge_index\"] #.to_sparse()\n",
    "data2['Compound', 'treats', 'Disease'].edge_index = ctd\n",
    "target_label                                     = data2['Compound', 'treats', 'Disease'].edge_index\n",
    "#data['Compound', 'treats', 'Disease'].edge_label =  torch.ones(target_label.shape[1],)\n",
    "data2 = T.ToUndirected()(data2)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print(target_label.shape)\n",
    "\n",
    "print(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+--------------+-----------------------------+-------------------------+\n",
      "|    | 0            | 1                           | 2                       |\n",
      "|----+--------------+-----------------------------+-------------------------|\n",
      "|  0 | Dornase alfa | Compound treats the disease | Cystic Fibrosis         |\n",
      "|  1 | Etanercept   | Compound treats the disease | Spondylitis, Ankylosing |\n",
      "|  2 | Etanercept   | Compound treats the disease | Graft vs Host Disease   |\n",
      "|  3 | Tetrandrine  | Compound treats the disease | Dermatitis, Atopic      |\n",
      "|  4 | Tetrandrine  | Compound treats the disease | Q Fever                 |\n",
      "+----+--------------+-----------------------------+-------------------------+\n"
     ]
    }
   ],
   "source": [
    "# SP create new test sample\n",
    "run_test_sample3 = 1\n",
    "if  run_test_sample3:  \n",
    "    test_relation_df = relation_df[relation_df['relation_name']=='Compound treats the disease']\n",
    "\n",
    "    # Create test sample of DRKG relationships filtering to these relations (for full sample: delete and use drkg_entity_df)\n",
    "    test_hrt_df = drkg_translated[drkg_translated[1].isin(test_relation_df['drkg_id'])]\n",
    "    test_hrt_df = test_hrt_df.groupby(1).head(3).reset_index(drop=True)\n",
    "    test_hrt_df[1] = 'Compound treats the disease'\n",
    "    print_head(test_hrt_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Drugs:   6 Matrix shape: torch.Size([6, 768])\n",
      "Unique Disases: 8 Matrix shape: torch.Size([8, 768])\n",
      "HeteroData(\n",
      "  \u001b[1mcompounds\u001b[0m={ x=[6, 768] },\n",
      "  \u001b[1mdisease\u001b[0m={ x=[8, 768] },\n",
      "  \u001b[1m(compounds, treats, disease)\u001b[0m={ edge_index=[2, 9] },\n",
      "  \u001b[1m(disease, rev_treats, compounds)\u001b[0m={ edge_index=[2, 9] }\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "#### Graph that works\n",
    "rx_dx_subset = test_hrt_df[:10].copy()\n",
    "\n",
    "\n",
    "### DX RX Relationship ###\n",
    "rx_X,rx_mapping = create_mapping(rx_dx_subset[0].to_list(),encoder= Encoder ,device=device) # Maps drugs to indices\n",
    "dx_X,dx_mapping = create_mapping(rx_dx_subset[2].to_list(),encoder= Encoder ,device=device) # Maps drugs to indices\n",
    "## As of now this only encodes 'Compound treats the disease', but  the idea is that this is used to encode every entity \n",
    "relationship_X,relationship_mapping = create_mapping(rx_dx_subset[1].to_list(),encoder= Encoder ,device=device)  \n",
    "\n",
    "print(f\"Unique Drugs:   {len(rx_mapping)} Matrix shape: {rx_X.shape}\")\n",
    "print(f\"Unique Disases: {len(dx_mapping)} Matrix shape: {dx_X.shape }\")\n",
    "relationship_feature = relationship_X[relationship_mapping['Compound treats the disease'],:].reshape(1,-1)\n",
    "\n",
    "data3 = HeteroData()\n",
    "data3['compounds'].x = rx_X\n",
    "data3['disease'].x   = dx_X\n",
    "\n",
    "#data['compounds2'].x = rx_X\n",
    "#data['disease2'].x   = dx_X\n",
    "#print(data)\n",
    "\n",
    "Edge_index,edge_attribute = create_edges(df             =  rx_dx_subset,\n",
    "                                          src_index_col  = 0, \n",
    "                                          src_mapping    = rx_mapping , \n",
    "                                          dst_index_col  = 2, \n",
    "                                          dst_mapping    = dx_mapping ,\n",
    "                                          edge_attr      = relationship_feature)\n",
    "\n",
    "data3['compounds', 'treats', 'disease'].edge_index = Edge_index\n",
    "data3 = T.ToUndirected()(data3)\n",
    "\n",
    "\n",
    "print(data3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[0, 2, 2, 5, 5, 5, 3, 1, 4],\n",
       "        [0, 7, 1, 2, 6, 4, 3, 5, 5]])"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "data3['compounds', 'treats', 'disease'].edge_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[ 1, 10, 10],\n",
       "        [ 0,  1,  2]])"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['Compound', 'Compound treats the disease', 'Disease'].edge_index\n",
    "data2['Compound', 'treats', 'Disease'].edge_index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "99     post-translational modification\n",
       "100                           reaction\n",
       "Name: relation_name, dtype: object"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "relation_df['relation_name'].drop_duplicates()[70:90]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GRAPH SAGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch_geometric as pyg\n",
    "import torch_geometric\n",
    "from torch_geometric.nn import SAGEConv, to_hetero\n",
    "from   torch.utils.data      import Dataset, DataLoader\n",
    "from   torch_geometric.data  import Data\n",
    "from   torch_geometric.utils import negative_sampling\n",
    "\n",
    "from torch_geometric.nn import SAGEConv, to_hetero\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class GNNStack(torch.nn.Module):\n",
    "    def __init__(self, input_dim:int, hidden_dim:int, output_dim:int, layers:int, dropout:float=0.3, return_embedding=False):\n",
    "        \"\"\"\n",
    "            A stack of GraphSAGE Module \n",
    "            input_dim        <int>:   Input dimension\n",
    "            hidden_dim       <int>:   Hidden dimension\n",
    "            output_dim       <int>:   Output dimension\n",
    "            layers           <int>:   Number of layers\n",
    "            dropout          <float>: Dropout rate\n",
    "            return_embedding <bool>:  Whether to return the return_embeddingedding of the input graph\n",
    "        \"\"\"\n",
    "        \n",
    "        super(GNNStack, self).__init__()\n",
    "        graphSage_conv               = pyg.nn.SAGEConv\n",
    "        self.dropout                 = dropout\n",
    "        self.layers                  = layers\n",
    "        self.return_embedding        = return_embedding\n",
    "\n",
    "        ### Initalize the layers ###\n",
    "        self.convs                   = nn.ModuleList()                      # ModuleList to hold the layers\n",
    "        for l in range(self.layers):\n",
    "            if l == 0:\n",
    "                ### First layer  maps from input_dim to hidden_dim ###\n",
    "                self.convs.append(graphSage_conv(input_dim, hidden_dim))\n",
    "            else:\n",
    "                ### All other layers map from hidden_dim to hidden_dim ###\n",
    "                self.convs.append(graphSage_conv(hidden_dim, hidden_dim))\n",
    "\n",
    "        # post-message-passing processing MLP\n",
    "        self.post_mp = nn.Sequential(\n",
    "                                     nn.Linear(hidden_dim, hidden_dim), \n",
    "                                     nn.Dropout(self.dropout),\n",
    "                                     nn.Linear(hidden_dim, output_dim))\n",
    "\n",
    "    def forward(self, x, edge_index):\n",
    "        for i in range(self.layers):\n",
    "            x = self.convs[i](x, edge_index)\n",
    "            x = F.relu(x)\n",
    "            x = F.dropout(x, p=self.dropout, training=self.training)\n",
    "\n",
    "        x = self.post_mp(x)\n",
    "\n",
    "        # Return final layer of return_embeddingeddings if specified\n",
    "        if self.return_embedding:\n",
    "            return x\n",
    "\n",
    "        # Else return class probabilities\n",
    "        return F.log_softmax(x, dim=1)\n",
    "\n",
    "    def loss(self, pred, label):\n",
    "        return F.nll_loss(pred, label)\n",
    "    \n",
    "\n",
    "\n",
    "class LinkPredictorMLP(nn.Module):\n",
    "    def __init__(self, in_channels:int, hidden_channels:int, out_channels:int, n_layers:int,dropout_probabilty:float=0.3):\n",
    "        \"\"\"\n",
    "        Args:\n",
    "            in_channels (int):     Number of input features.\n",
    "            hidden_channels (int): Number of hidden features.\n",
    "            out_channels (int):    Number of output features.\n",
    "            n_layers (int):        Number of MLP layers.\n",
    "            dropout (float):       Dropout probability.\n",
    "            \"\"\"\n",
    "        super(LinkPredictorMLP, self).__init__()\n",
    "        self.dropout_probabilty    = dropout_probabilty  # dropout probability\n",
    "        self.mlp_layers            = nn.ModuleList()     # ModuleList: is a list of modules\n",
    "        self.non_linearity         = F.relu              # non-linearity\n",
    "        \n",
    "        for i in range(n_layers - 1):                                 \n",
    "            if i == 0:\n",
    "                self.mlp_layers.append(nn.Linear(in_channels, hidden_channels))          # input layer (in_channels, hidden_channels)\n",
    "            else:\n",
    "                self.mlp_layers.append(nn.Linear(hidden_channels, hidden_channels))      # hidden layers (hidden_channels, hidden_channels)\n",
    "\n",
    "        self.mlp_layers.append(nn.Linear(hidden_channels, out_channels))                 # output layer (hidden_channels, out_channels)\n",
    "\n",
    "\n",
    "    def reset_parameters(self):\n",
    "        for mlp_layer in self.mlp_layers:\n",
    "            mlp_layer.reset_parameters()\n",
    "\n",
    "    def forward(self, x_i, x_j):\n",
    "        x = x_i * x_j                                                     # element-wise multiplication\n",
    "        for mlp_layer in self.mlp_layers[:-1]:                            # iterate over all layers except the last one\n",
    "            x = mlp_layer(x)                                              # apply linear transformation\n",
    "            x = self.non_linearity(x)                                     # Apply non linear activation function\n",
    "            x = F.dropout(x, p=self.dropout_probabilty,training=self.training)      # Apply dropout\n",
    "        x = self.mlp_layers[-1](x)                                        # apply linear transformation to the last layer\n",
    "        x = torch.sigmoid(x)                                              # apply sigmoid activation function to get the probability\n",
    "        return x\n",
    "    \n",
    "### We will use This function to save our best model during trainnig ###\n",
    "def save_torch_model(model,epoch,PATH:str,optimizer):\n",
    "    print(f\"Saving Model in Path {PATH}\")\n",
    "    torch.save({'epoch': epoch,\n",
    "                'model_state_dict': model.state_dict(),\n",
    "                'optimizer':optimizer,      \n",
    "                }, PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GraphModule(\n",
      "  (convs): ModuleList(\n",
      "    (0): ModuleDict(\n",
      "      (Compound__Compound_treats_the_disease__Disease): SAGEConv(768, 524, aggr=mean)\n",
      "      (Compound__increases_expression_or_production__Gene): SAGEConv(768, 524, aggr=mean)\n",
      "      (Gene__increases_expression_or_production__Gene): SAGEConv(768, 524, aggr=mean)\n",
      "      (Gene__biomarkers_diagnostic__Disease): SAGEConv(768, 524, aggr=mean)\n",
      "      (Disease__rev_Compound_treats_the_disease__Compound): SAGEConv(768, 524, aggr=mean)\n",
      "      (Gene__rev_increases_expression_or_production__Compound): SAGEConv(768, 524, aggr=mean)\n",
      "      (Disease__rev_biomarkers_diagnostic__Gene): SAGEConv(768, 524, aggr=mean)\n",
      "    )\n",
      "    (1-2): 2 x ModuleDict(\n",
      "      (Compound__Compound_treats_the_disease__Disease): SAGEConv(524, 524, aggr=mean)\n",
      "      (Compound__increases_expression_or_production__Gene): SAGEConv(524, 524, aggr=mean)\n",
      "      (Gene__increases_expression_or_production__Gene): SAGEConv(524, 524, aggr=mean)\n",
      "      (Gene__biomarkers_diagnostic__Disease): SAGEConv(524, 524, aggr=mean)\n",
      "      (Disease__rev_Compound_treats_the_disease__Compound): SAGEConv(524, 524, aggr=mean)\n",
      "      (Gene__rev_increases_expression_or_production__Compound): SAGEConv(524, 524, aggr=mean)\n",
      "      (Disease__rev_biomarkers_diagnostic__Gene): SAGEConv(524, 524, aggr=mean)\n",
      "    )\n",
      "  )\n",
      "  (post_mp): ModuleList(\n",
      "    (0): ModuleDict(\n",
      "      (Compound): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Disease): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Gene): Linear(in_features=524, out_features=524, bias=True)\n",
      "    )\n",
      "    (1): ModuleDict(\n",
      "      (Compound): Dropout(p=0.7, inplace=False)\n",
      "      (Disease): Dropout(p=0.7, inplace=False)\n",
      "      (Gene): Dropout(p=0.7, inplace=False)\n",
      "    )\n",
      "    (2): ModuleDict(\n",
      "      (Compound): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Disease): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Gene): Linear(in_features=524, out_features=524, bias=True)\n",
      "    )\n",
      "  )\n",
      ")\n",
      "\n",
      "\n",
      "\n",
      "def forward(self, x, edge_index):\n",
      "    x_dict = torch_geometric_nn_to_hetero_transformer_get_dict(x);  x = None\n",
      "    x__Compound = x_dict.get('Compound', None)\n",
      "    x__Disease = x_dict.get('Disease', None)\n",
      "    x__Gene = x_dict.get('Gene', None);  x_dict = None\n",
      "    edge_index_dict = torch_geometric_nn_to_hetero_transformer_get_dict(edge_index);  edge_index = None\n",
      "    edge_index__Compound__Compound_treats_the_disease__Disease = edge_index_dict.get(('Compound', 'Compound treats the disease', 'Disease'), None)\n",
      "    edge_index__Compound__increases_expression_or_production__Gene = edge_index_dict.get(('Compound', 'increases expression or production', 'Gene'), None)\n",
      "    edge_index__Gene__increases_expression_or_production__Gene = edge_index_dict.get(('Gene', 'increases expression or production', 'Gene'), None)\n",
      "    edge_index__Gene__biomarkers_diagnostic__Disease = edge_index_dict.get(('Gene', 'biomarkers diagnostic', 'Disease'), None)\n",
      "    edge_index__Disease__rev_Compound_treats_the_disease__Compound = edge_index_dict.get(('Disease', 'rev_Compound treats the disease', 'Compound'), None)\n",
      "    edge_index__Gene__rev_increases_expression_or_production__Compound = edge_index_dict.get(('Gene', 'rev_increases expression or production', 'Compound'), None)\n",
      "    edge_index__Disease__rev_biomarkers_diagnostic__Gene = edge_index_dict.get(('Disease', 'rev_biomarkers diagnostic', 'Gene'), None);  edge_index_dict = None\n",
      "    convs_0__Disease1 = getattr(self.convs, \"0\").Compound__Compound_treats_the_disease__Disease((x__Compound, x__Disease), edge_index__Compound__Compound_treats_the_disease__Disease)\n",
      "    convs_0__Gene1 = getattr(self.convs, \"0\").Compound__increases_expression_or_production__Gene((x__Compound, x__Gene), edge_index__Compound__increases_expression_or_production__Gene)\n",
      "    convs_0__Gene2 = getattr(self.convs, \"0\").Gene__increases_expression_or_production__Gene(x__Gene, edge_index__Gene__increases_expression_or_production__Gene)\n",
      "    convs_0__Disease2 = getattr(self.convs, \"0\").Gene__biomarkers_diagnostic__Disease((x__Gene, x__Disease), edge_index__Gene__biomarkers_diagnostic__Disease)\n",
      "    convs_0__Compound1 = getattr(self.convs, \"0\").Disease__rev_Compound_treats_the_disease__Compound((x__Disease, x__Compound), edge_index__Disease__rev_Compound_treats_the_disease__Compound)\n",
      "    convs_0__Compound2 = getattr(self.convs, \"0\").Gene__rev_increases_expression_or_production__Compound((x__Gene, x__Compound), edge_index__Gene__rev_increases_expression_or_production__Compound);  x__Compound = None\n",
      "    convs_0__Gene3 = getattr(self.convs, \"0\").Disease__rev_biomarkers_diagnostic__Gene((x__Disease, x__Gene), edge_index__Disease__rev_biomarkers_diagnostic__Gene);  x__Disease = x__Gene = None\n",
      "    convs_0__Disease = torch.add(convs_0__Disease1, convs_0__Disease2);  convs_0__Disease1 = convs_0__Disease2 = None\n",
      "    convs_0__Gene_1 = torch.add(convs_0__Gene1, convs_0__Gene2);  convs_0__Gene1 = convs_0__Gene2 = None\n",
      "    convs_0__Gene = torch.add(convs_0__Gene3, convs_0__Gene_1);  convs_0__Gene3 = convs_0__Gene_1 = None\n",
      "    convs_0__Compound = torch.add(convs_0__Compound1, convs_0__Compound2);  convs_0__Compound1 = convs_0__Compound2 = None\n",
      "    relu__Compound = torch.nn.functional.relu(convs_0__Compound, inplace = False);  convs_0__Compound = None\n",
      "    relu__Disease = torch.nn.functional.relu(convs_0__Disease, inplace = False);  convs_0__Disease = None\n",
      "    relu__Gene = torch.nn.functional.relu(convs_0__Gene, inplace = False);  convs_0__Gene = None\n",
      "    dropout__Compound = torch.nn.functional.dropout(relu__Compound, p = 0.7, training = True, inplace = False);  relu__Compound = None\n",
      "    dropout__Disease = torch.nn.functional.dropout(relu__Disease, p = 0.7, training = True, inplace = False);  relu__Disease = None\n",
      "    dropout__Gene = torch.nn.functional.dropout(relu__Gene, p = 0.7, training = True, inplace = False);  relu__Gene = None\n",
      "    convs_1__Disease1 = getattr(self.convs, \"1\").Compound__Compound_treats_the_disease__Disease((dropout__Compound, dropout__Disease), edge_index__Compound__Compound_treats_the_disease__Disease)\n",
      "    convs_1__Gene1 = getattr(self.convs, \"1\").Compound__increases_expression_or_production__Gene((dropout__Compound, dropout__Gene), edge_index__Compound__increases_expression_or_production__Gene)\n",
      "    convs_1__Gene2 = getattr(self.convs, \"1\").Gene__increases_expression_or_production__Gene(dropout__Gene, edge_index__Gene__increases_expression_or_production__Gene)\n",
      "    convs_1__Disease2 = getattr(self.convs, \"1\").Gene__biomarkers_diagnostic__Disease((dropout__Gene, dropout__Disease), edge_index__Gene__biomarkers_diagnostic__Disease)\n",
      "    convs_1__Compound1 = getattr(self.convs, \"1\").Disease__rev_Compound_treats_the_disease__Compound((dropout__Disease, dropout__Compound), edge_index__Disease__rev_Compound_treats_the_disease__Compound)\n",
      "    convs_1__Compound2 = getattr(self.convs, \"1\").Gene__rev_increases_expression_or_production__Compound((dropout__Gene, dropout__Compound), edge_index__Gene__rev_increases_expression_or_production__Compound);  dropout__Compound = None\n",
      "    convs_1__Gene3 = getattr(self.convs, \"1\").Disease__rev_biomarkers_diagnostic__Gene((dropout__Disease, dropout__Gene), edge_index__Disease__rev_biomarkers_diagnostic__Gene);  dropout__Disease = dropout__Gene = None\n",
      "    convs_1__Disease = torch.add(convs_1__Disease1, convs_1__Disease2);  convs_1__Disease1 = convs_1__Disease2 = None\n",
      "    convs_1__Gene_1 = torch.add(convs_1__Gene1, convs_1__Gene2);  convs_1__Gene1 = convs_1__Gene2 = None\n",
      "    convs_1__Gene = torch.add(convs_1__Gene3, convs_1__Gene_1);  convs_1__Gene3 = convs_1__Gene_1 = None\n",
      "    convs_1__Compound = torch.add(convs_1__Compound1, convs_1__Compound2);  convs_1__Compound1 = convs_1__Compound2 = None\n",
      "    relu_1__Compound = torch.nn.functional.relu(convs_1__Compound, inplace = False);  convs_1__Compound = None\n",
      "    relu_1__Disease = torch.nn.functional.relu(convs_1__Disease, inplace = False);  convs_1__Disease = None\n",
      "    relu_1__Gene = torch.nn.functional.relu(convs_1__Gene, inplace = False);  convs_1__Gene = None\n",
      "    dropout_1__Compound = torch.nn.functional.dropout(relu_1__Compound, p = 0.7, training = True, inplace = False);  relu_1__Compound = None\n",
      "    dropout_1__Disease = torch.nn.functional.dropout(relu_1__Disease, p = 0.7, training = True, inplace = False);  relu_1__Disease = None\n",
      "    dropout_1__Gene = torch.nn.functional.dropout(relu_1__Gene, p = 0.7, training = True, inplace = False);  relu_1__Gene = None\n",
      "    convs_2__Disease1 = getattr(self.convs, \"2\").Compound__Compound_treats_the_disease__Disease((dropout_1__Compound, dropout_1__Disease), edge_index__Compound__Compound_treats_the_disease__Disease);  edge_index__Compound__Compound_treats_the_disease__Disease = None\n",
      "    convs_2__Gene1 = getattr(self.convs, \"2\").Compound__increases_expression_or_production__Gene((dropout_1__Compound, dropout_1__Gene), edge_index__Compound__increases_expression_or_production__Gene);  edge_index__Compound__increases_expression_or_production__Gene = None\n",
      "    convs_2__Gene2 = getattr(self.convs, \"2\").Gene__increases_expression_or_production__Gene(dropout_1__Gene, edge_index__Gene__increases_expression_or_production__Gene);  edge_index__Gene__increases_expression_or_production__Gene = None\n",
      "    convs_2__Disease2 = getattr(self.convs, \"2\").Gene__biomarkers_diagnostic__Disease((dropout_1__Gene, dropout_1__Disease), edge_index__Gene__biomarkers_diagnostic__Disease);  edge_index__Gene__biomarkers_diagnostic__Disease = None\n",
      "    convs_2__Compound1 = getattr(self.convs, \"2\").Disease__rev_Compound_treats_the_disease__Compound((dropout_1__Disease, dropout_1__Compound), edge_index__Disease__rev_Compound_treats_the_disease__Compound);  edge_index__Disease__rev_Compound_treats_the_disease__Compound = None\n",
      "    convs_2__Compound2 = getattr(self.convs, \"2\").Gene__rev_increases_expression_or_production__Compound((dropout_1__Gene, dropout_1__Compound), edge_index__Gene__rev_increases_expression_or_production__Compound);  dropout_1__Compound = edge_index__Gene__rev_increases_expression_or_production__Compound = None\n",
      "    convs_2__Gene3 = getattr(self.convs, \"2\").Disease__rev_biomarkers_diagnostic__Gene((dropout_1__Disease, dropout_1__Gene), edge_index__Disease__rev_biomarkers_diagnostic__Gene);  dropout_1__Disease = dropout_1__Gene = edge_index__Disease__rev_biomarkers_diagnostic__Gene = None\n",
      "    convs_2__Disease = torch.add(convs_2__Disease1, convs_2__Disease2);  convs_2__Disease1 = convs_2__Disease2 = None\n",
      "    convs_2__Gene_1 = torch.add(convs_2__Gene1, convs_2__Gene2);  convs_2__Gene1 = convs_2__Gene2 = None\n",
      "    convs_2__Gene = torch.add(convs_2__Gene3, convs_2__Gene_1);  convs_2__Gene3 = convs_2__Gene_1 = None\n",
      "    convs_2__Compound = torch.add(convs_2__Compound1, convs_2__Compound2);  convs_2__Compound1 = convs_2__Compound2 = None\n",
      "    relu_2__Compound = torch.nn.functional.relu(convs_2__Compound, inplace = False);  convs_2__Compound = None\n",
      "    relu_2__Disease = torch.nn.functional.relu(convs_2__Disease, inplace = False);  convs_2__Disease = None\n",
      "    relu_2__Gene = torch.nn.functional.relu(convs_2__Gene, inplace = False);  convs_2__Gene = None\n",
      "    dropout_2__Compound = torch.nn.functional.dropout(relu_2__Compound, p = 0.7, training = True, inplace = False);  relu_2__Compound = None\n",
      "    dropout_2__Disease = torch.nn.functional.dropout(relu_2__Disease, p = 0.7, training = True, inplace = False);  relu_2__Disease = None\n",
      "    dropout_2__Gene = torch.nn.functional.dropout(relu_2__Gene, p = 0.7, training = True, inplace = False);  relu_2__Gene = None\n",
      "    post_mp_0__Compound = getattr(self.post_mp, \"0\").Compound(dropout_2__Compound);  dropout_2__Compound = None\n",
      "    post_mp_0__Disease = getattr(self.post_mp, \"0\").Disease(dropout_2__Disease);  dropout_2__Disease = None\n",
      "    post_mp_0__Gene = getattr(self.post_mp, \"0\").Gene(dropout_2__Gene);  dropout_2__Gene = None\n",
      "    post_mp_1__Compound = getattr(self.post_mp, \"1\").Compound(post_mp_0__Compound);  post_mp_0__Compound = None\n",
      "    post_mp_1__Disease = getattr(self.post_mp, \"1\").Disease(post_mp_0__Disease);  post_mp_0__Disease = None\n",
      "    post_mp_1__Gene = getattr(self.post_mp, \"1\").Gene(post_mp_0__Gene);  post_mp_0__Gene = None\n",
      "    post_mp_2__Compound = getattr(self.post_mp, \"2\").Compound(post_mp_1__Compound);  post_mp_1__Compound = None\n",
      "    post_mp_2__Disease = getattr(self.post_mp, \"2\").Disease(post_mp_1__Disease);  post_mp_1__Disease = None\n",
      "    post_mp_2__Gene = getattr(self.post_mp, \"2\").Gene(post_mp_1__Gene);  post_mp_1__Gene = None\n",
      "    return {'Compound': post_mp_2__Compound, 'Disease': post_mp_2__Disease, 'Gene': post_mp_2__Gene}\n",
      "    \n",
      "# To see more debug info, please use `graph_module.print_readable()`\n",
      "LinkPredictorMLP(\n",
      "  (mlp_layers): ModuleList(\n",
      "    (0-1): 2 x Linear(in_features=524, out_features=524, bias=True)\n",
      "    (2): Linear(in_features=524, out_features=1, bias=True)\n",
      "  )\n",
      ")\n",
      "Models Loaded to cpu\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch_geometric/nn/to_hetero_transformer.py:165: UserWarning: The type 'Compound treats the disease' contains invalid characters which may lead to unexpected behavior. To avoid any issues, ensure that your types only contain letters, numbers and underscores.\n",
      "  warnings.warn(\n",
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch_geometric/nn/to_hetero_transformer.py:165: UserWarning: The type 'increases expression or production' contains invalid characters which may lead to unexpected behavior. To avoid any issues, ensure that your types only contain letters, numbers and underscores.\n",
      "  warnings.warn(\n",
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch_geometric/nn/to_hetero_transformer.py:165: UserWarning: The type 'biomarkers diagnostic' contains invalid characters which may lead to unexpected behavior. To avoid any issues, ensure that your types only contain letters, numbers and underscores.\n",
      "  warnings.warn(\n",
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch_geometric/nn/to_hetero_transformer.py:165: UserWarning: The type 'rev_Compound treats the disease' contains invalid characters which may lead to unexpected behavior. To avoid any issues, ensure that your types only contain letters, numbers and underscores.\n",
      "  warnings.warn(\n",
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch_geometric/nn/to_hetero_transformer.py:165: UserWarning: The type 'rev_increases expression or production' contains invalid characters which may lead to unexpected behavior. To avoid any issues, ensure that your types only contain letters, numbers and underscores.\n",
      "  warnings.warn(\n",
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch_geometric/nn/to_hetero_transformer.py:165: UserWarning: The type 'rev_biomarkers diagnostic' contains invalid characters which may lead to unexpected behavior. To avoid any issues, ensure that your types only contain letters, numbers and underscores.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "epochs        = 500\n",
    "hidden_dim    = 524      # 256 \n",
    "dropout       = 0.7\n",
    "num_layers    = 3\n",
    "learning_rate = 1e-4\n",
    "node_emb_dim  = 768\n",
    "\n",
    "test_data = data1\n",
    "\n",
    "HomoGNN         = GNNStack(node_emb_dim, hidden_dim, hidden_dim, num_layers, dropout, return_embedding=True).to(device) # the graph neural network that takes all the node embeddings as inputs to message pass and agregate\n",
    "HeteroGNN       = to_hetero(HomoGNN   , test_data.metadata(), aggr='sum')\n",
    "link_predictor  = LinkPredictorMLP(hidden_dim, hidden_dim, 1, num_layers , dropout).to(device) # the MLP that takes embeddings of a pair of nodes and predicts the existence of an edge between them\n",
    "#optimizer      = torch.optim.AdamW(list(model.parameters()) + list(link_predictor.parameters() ), lr=learning_rate, weight_decay=1e-4)\n",
    "optimizer       = torch.optim.Adam(list(HeteroGNN.parameters()) + list(link_predictor.parameters() ), lr=learning_rate)\n",
    "\n",
    "print(HeteroGNN )\n",
    "print(link_predictor)\n",
    "print(f\"Models Loaded to {device}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HeteroData(\n",
      "  \u001b[1mCompound\u001b[0m={ x=[12, 768] },\n",
      "  \u001b[1mDisease\u001b[0m={ x=[11, 768] },\n",
      "  \u001b[1mGene\u001b[0m={ x=[12, 768] },\n",
      "  \u001b[1m(Compound, Compound treats the disease, Disease)\u001b[0m={\n",
      "    edge_index=[2, 9],\n",
      "    edge_label=[9, 768]\n",
      "  },\n",
      "  \u001b[1m(Compound, increases expression/production, Gene)\u001b[0m={\n",
      "    edge_index=[2, 6],\n",
      "    edge_label=[6, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, increases expression/production, Gene)\u001b[0m={\n",
      "    edge_index=[2, 6],\n",
      "    edge_label=[6, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, biomarkers (diagnostic), Disease)\u001b[0m={\n",
      "    edge_index=[2, 3],\n",
      "    edge_label=[3, 768]\n",
      "  },\n",
      "  \u001b[1m(Disease, rev_Compound treats the disease, Compound)\u001b[0m={\n",
      "    edge_index=[2, 9],\n",
      "    edge_label=[9, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, rev_increases expression/production, Compound)\u001b[0m={\n",
      "    edge_index=[2, 6],\n",
      "    edge_label=[6, 768]\n",
      "  },\n",
      "  \u001b[1m(Disease, rev_biomarkers (diagnostic), Gene)\u001b[0m={\n",
      "    edge_index=[2, 3],\n",
      "    edge_label=[3, 768]\n",
      "  }\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "print(test_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(test_data.x_dict.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('Compound',\n",
       "  'activation',\n",
       "  'Gene'): tensor([[11,  0,  4,  6,  7,  5,  2,  2,  9],\n",
       "         [16, 12, 14,  3, 23,  5,  1, 22,  6]]),\n",
       " ('Compound',\n",
       "  'inhibition',\n",
       "  'Gene'): tensor([[ 1, 10,  3],\n",
       "         [ 7, 21, 19]]),\n",
       " ('Compound',\n",
       "  'Compound treats the disease',\n",
       "  'Disease'): tensor([[12,  8,  8],\n",
       "         [ 1,  0,  2]]),\n",
       " ('Gene',\n",
       "  'activates, stimulates',\n",
       "  'Gene'): tensor([[ 2,  4,  8, 11, 15, 17],\n",
       "         [11, 17, 15,  2,  8,  4]]),\n",
       " ('Gene',\n",
       "  'inhibition',\n",
       "  'Gene'): tensor([[ 0,  9, 10, 13, 18, 20],\n",
       "         [20, 18, 13, 10,  9,  0]]),\n",
       " ('Gene',\n",
       "  'rev_activation',\n",
       "  'Compound'): tensor([[16, 12, 14,  3, 23,  5,  1, 22,  6],\n",
       "         [11,  0,  4,  6,  7,  5,  2,  2,  9]]),\n",
       " ('Gene',\n",
       "  'rev_inhibition',\n",
       "  'Compound'): tensor([[ 7, 21, 19],\n",
       "         [ 1, 10,  3]]),\n",
       " ('Disease',\n",
       "  'rev_Compound treats the disease',\n",
       "  'Compound'): tensor([[ 1,  0,  2],\n",
       "         [12,  8,  8]])}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data.edge_index_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('compounds',\n",
       "  'treats',\n",
       "  'disease'): tensor([[5, 2, 2, 3, 3, 3, 0, 4, 1],\n",
       "         [5, 0, 2, 4, 6, 7, 3, 1, 1]]),\n",
       " ('disease',\n",
       "  'rev_treats',\n",
       "  'compounds'): tensor([[5, 0, 2, 4, 6, 7, 3, 1, 1],\n",
       "         [5, 2, 2, 3, 3, 3, 0, 4, 1]])}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data3.edge_index_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 252,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_emb   = HeteroGNN(test_data.x_dict, test_data.edge_index_dict)\n",
    "# edge_index = test_data['compounds', 'treats', 'disease'].edge_index \n",
    "# pos_pred    = link_predictor(node_emb[\"compounds\"][edge_index[0]], node_emb[\"disease\"][edge_index[0]])   # (B, )\n",
    "edge_index = test_data['Compound', 'Compound treats the disease', 'Disease'].edge_index \n",
    "pos_pred    = link_predictor(node_emb[\"Compound\"][edge_index[0]], node_emb[\"Disease\"][edge_index[1]])   # (B, )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 253,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "tensor([[0.4970],\n",
       "        [0.4745],\n",
       "        [0.4823],\n",
       "        [0.4923],\n",
       "        [0.4851],\n",
       "        [0.4945],\n",
       "        [0.4230],\n",
       "        [0.4087],\n",
       "        [0.4884]], grad_fn=<SigmoidBackward0>)"
      ]
     },
     "execution_count": 253,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pos_pred"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 254,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HeteroData(\n",
       "  \u001b[1mCompound\u001b[0m={ x=[12, 768] },\n",
       "  \u001b[1mDisease\u001b[0m={ x=[11, 768] },\n",
       "  \u001b[1mGene\u001b[0m={ x=[12, 768] },\n",
       "  \u001b[1m(Compound, Compound treats the disease, Disease)\u001b[0m={\n",
       "    edge_index=[2, 9],\n",
       "    edge_label=[9, 768]\n",
       "  },\n",
       "  \u001b[1m(Compound, increases expression or production, Gene)\u001b[0m={\n",
       "    edge_index=[2, 6],\n",
       "    edge_label=[6, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, increases expression or production, Gene)\u001b[0m={\n",
       "    edge_index=[2, 6],\n",
       "    edge_label=[6, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, biomarkers diagnostic, Disease)\u001b[0m={\n",
       "    edge_index=[2, 3],\n",
       "    edge_label=[3, 768]\n",
       "  },\n",
       "  \u001b[1m(Disease, rev_Compound treats the disease, Compound)\u001b[0m={\n",
       "    edge_index=[2, 9],\n",
       "    edge_label=[9, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, rev_increases expression or production, Compound)\u001b[0m={\n",
       "    edge_index=[2, 6],\n",
       "    edge_label=[6, 768]\n",
       "  },\n",
       "  \u001b[1m(Disease, rev_biomarkers diagnostic, Gene)\u001b[0m={\n",
       "    edge_index=[2, 3],\n",
       "    edge_label=[3, 768]\n",
       "  }\n",
       ")"
      ]
     },
     "execution_count": 254,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "2d42bd2c408a4bd3bdebc2865ea42483": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": "hidden",
      "width": null
     }
    },
    "6e79954f9c514f099b336e3a11ad635e": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_6f2e0878218c4b38974a74a673b70751",
      "placeholder": "​",
      "style": "IPY_MODEL_98ccadd9e4a34d8282282a332fad6e0b",
      "value": "Downloading chembl_32_sqlite.tar.gz: 100%"
     }
    },
    "6f2e0878218c4b38974a74a673b70751": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "98ccadd9e4a34d8282282a332fad6e0b": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "a95662c4dc0949f4af40e8331cde378d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "info",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_eec53708d80f41e8b402060c9a1352b1",
      "max": 1,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_f87f682db17a4e6b837b9c14c85ec36f",
      "value": 1
     }
    },
    "b0797e6674144b1b92d2ff70918b917c": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "b355679609b749d698029d4f39d45503": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_6e79954f9c514f099b336e3a11ad635e",
       "IPY_MODEL_a95662c4dc0949f4af40e8331cde378d",
       "IPY_MODEL_d89d003254ce42f9a44572c7e26940ec"
      ],
      "layout": "IPY_MODEL_2d42bd2c408a4bd3bdebc2865ea42483"
     }
    },
    "d0a9a73a5fe14afb8b15a2addf4f080d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "d89d003254ce42f9a44572c7e26940ec": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_b0797e6674144b1b92d2ff70918b917c",
      "placeholder": "​",
      "style": "IPY_MODEL_d0a9a73a5fe14afb8b15a2addf4f080d",
      "value": " 4.27G/4.27G [03:38&lt;00:00, 21.7MB/s]"
     }
    },
    "eec53708d80f41e8b402060c9a1352b1": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": "20px"
     }
    },
    "f87f682db17a4e6b837b9c14c85ec36f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
