{
  "cells": [
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {
        "id": "DnQGl_D4786T"
      },
      "source": [
        "DRKG\n",
        "\n",
        "Adapted from: https://github.com/gnn4dr/DRKG/blob/master/drkg_with_dgl/loading_drkg_in_dgl.ipynb"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "id": "PV4N5XSIKfTa"
      },
      "outputs": [],
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import os "
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [],
      "source": [
        "from SIMP_LLM.DRKG_loading   import  get_triplets, read_tsv,filter_drkg,map_drkg_relationships,filter_interaction_subset,print_head\n",
        "from SIMP_LLM.DRKG_translate import  load_lookups\n",
        "from SIMP_LLM.DRKG_entity_processing import get_unique_entities, get_entity_lookup, convert_entitynames, flip_headtail"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {
        "id": "VsLJKwgRxArZ"
      },
      "source": [
        "# 1) Load Data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            " Triplets:\n",
            "\n",
            "[['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2157'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5264'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2158'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::3309'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::28912'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::811'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2159'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::821'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5627'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5624']]\n",
            "\n",
            " data/drkg.tsv  Dataframe:\n",
            "\n",
            "+----+------------+--------------------------------+-------------+\n",
            "|    | 0          | 1                              | 2           |\n",
            "|----+------------+--------------------------------+-------------|\n",
            "|  0 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2157  |\n",
            "|  1 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::5264  |\n",
            "|  2 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2158  |\n",
            "|  3 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::3309  |\n",
            "|  4 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::28912 |\n",
            "+----+------------+--------------------------------+-------------+\n",
            "\n",
            " data/relation_glossary.tsv  Dataframe:\n",
            "\n",
            "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
            "|    | Relation-name                              | Data-source   | Connected entity-types   | Interaction-type      | Description                                                                                                                                                                     | Reference for the description        |\n",
            "|----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------|\n",
            "|  0 | DGIDB::ACTIVATOR::Gene:Compound            | DGIDB         | Compound:Gene            | activation            | An activator interaction is when a drug activates a biological response from a target, although the mechanism by which it does so may not be understood.                        | http://www.dgidb.org/getting_started |\n",
            "|  1 | DGIDB::AGONIST::Gene:Compound              | DGIDB         | Compound:Gene            | agonism               | An agonist interaction occurs when a drug binds to a target receptor and activates the receptor to produce a biological response.                                               | http://www.dgidb.org/getting_started |\n",
            "|  2 | DGIDB::ALLOSTERIC MODULATOR::Gene:Compound | DGIDB         | Compound:Gene            | allosteric modulation | An allosteric modulator interaction occurs when drugs exert their effects on their protein targets via a different binding site than the natural (orthosteric) ligand site.     | http://www.dgidb.org/getting_started |\n",
            "|  3 | DGIDB::ANTAGONIST::Gene:Compound           | DGIDB         | Compound:Gene            | antagonism            | An antagonist interaction occurs when a drug blocks or dampens agonist-mediated responses rather than provoking a biological response itself upon binding to a target receptor. | http://www.dgidb.org/getting_started |\n",
            "|  4 | DGIDB::ANTIBODY::Gene:Compound             | DGIDB         | Compound:Gene            | antibody              | An antibody interaction occurs when an antibody drug specifically binds the target molecule.                                                                                    | http://www.dgidb.org/getting_started |\n",
            "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
            "Number of Rows Before Filtering: 107\n",
            "Number of Rows After Filtering: 10\n",
            "\n",
            "\n",
            "Filtered:\n",
            "24    DRUGBANK::treats::Compound:Disease\n",
            "35             GNBR::T::Compound:Disease\n",
            "36             GNBR::C::Compound:Disease\n",
            "37            GNBR::Sa::Compound:Disease\n",
            "38            GNBR::Pa::Compound:Disease\n",
            "39            GNBR::Mp::Compound:Disease\n",
            "40            GNBR::Pr::Compound:Disease\n",
            "41             GNBR::J::Compound:Disease\n",
            "68       Hetionet::CtD::Compound:Disease\n",
            "73       Hetionet::CpD::Compound:Disease\n",
            "dtype: object\n",
            "\n",
            "Relationships Mapped:\n",
            "0                    Compound treats the disease\n",
            "1    treatment/therapy (including investigatory)\n",
            "2            inhibits cell growth (esp. cancers)\n",
            "3                      side effect/adverse event\n",
            "4                            alleviates, reduces\n",
            "5            biomarkers (of disease progression)\n",
            "6                           prevents, suppresses\n",
            "7                   role in disease pathogenesis\n",
            "8                                      treatment\n",
            "9                                     palliation\n",
            "Name: Interaction-type, dtype: object\n",
            "+---------+-------------------+------------------------------------+-----------------------+\n",
            "|         | 0                 | 1                                  | 2                     |\n",
            "|---------+-------------------+------------------------------------+-----------------------|\n",
            "| 1518268 | Compound::DB00003 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D003550 |\n",
            "| 1518269 | Compound::DB00004 | DRUGBANK::treats::Compound:Disease | Disease::MESH:C063419 |\n",
            "| 1518270 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D013167 |\n",
            "| 1518271 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D006086 |\n",
            "| 1518272 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D017497 |\n",
            "+---------+-------------------+------------------------------------+-----------------------+\n"
          ]
        }
      ],
      "source": [
        "### 1) Read: This section reads DRKG and a glossary (used to map entities from codes to words)\n",
        "DATA_DIR           = os.path.join(\"data\")\n",
        "verbose            =  True \n",
        "triplets,drkg_df   =  get_triplets(drkg_file = os.path.join(DATA_DIR  ,'drkg.tsv'),             verbose=verbose)  # Read triplets (head,relationship,tail)\n",
        "relation_glossary  =  read_tsv(relation_file = os.path.join(DATA_DIR  ,'relation_glossary.tsv'),verbose=verbose)  # Read relationship mapping  \n",
        "\n",
        "\n",
        "### 2) Filter & Map Interactions: This section returns a list of interactions e.g. DRUGBANK::treats::Compound:Disease )\n",
        "# 2.1: First  we filter the interactions to only Compound-Disease\n",
        "# 2.2: Then   we map the codes -> text  (this will be use to further filter interactions based on text) e.g.  Hetionet::CpD::Compound:Disease -> palliation\n",
        "# 2.3: We use natural text to fitler  interactions based on terms such as \"treat\" (but we return the orignal interaction name )\n",
        "\n",
        "\n",
        "\n",
        "# modularize this in create_dataframe\n",
        "drkg_rx_dx_relations        = filter_drkg(data_frame = drkg_df ,  filter_column = 1 ,  filter_term = r'.*?Compound:Disease', verbose = verbose) # 2.1 Filter only Compound-Disease Interactions\n",
        "drkg_rx_dx_relations_mapped = map_drkg_relationships(drkg_rx_dx_relations,relation_glossary,verbose=verbose)                                    # 2.2 Map codes to text \n",
        "\n",
        "### 2.3 Filter Drug interactions Interaction types to only include: treat inhibit or alleviate interactions  ###\n",
        "drkg_rx_dx_relation_subset =  filter_interaction_subset(df                  = drkg_rx_dx_relations_mapped,\n",
        "                                                        filter_colunm_name = 'Interaction-type' ,\n",
        "                                                        regex_string       =  'treat|inhibit|alleviate',\n",
        "                                                        return_colunm_name =  'Relation-name')\n",
        "\n",
        "# 3) Use Filter Interactions to get Gilter DRKG \n",
        "drkg_df_filtered = drkg_df[drkg_df[1].isin(drkg_rx_dx_relation_subset)] # 3.1 Filter DRKG  to only  Compund-Disease \n",
        "print_head(df=drkg_df_filtered)\n",
        "\n",
        "\n",
        "\n",
        "###\n",
        "\n",
        "rx_dx_triplets   = drkg_df_filtered.values.tolist()                     # 3.2 Convert filtered DRKG to list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KYBnEGL4Nixb",
        "outputId": "5bc6b100-d023-43d7-cf3c-488dc73b6475"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "\n",
            " data/hetionet-v1.0-nodes.tsv  Dataframe:\n",
            "\n",
            "+----+-------------------------+---------------------------+---------+\n",
            "|    | id                      | name                      | kind    |\n",
            "|----+-------------------------+---------------------------+---------|\n",
            "|  0 | Anatomy::UBERON:0000002 | uterine cervix            | Anatomy |\n",
            "|  1 | Anatomy::UBERON:0000004 | nose                      | Anatomy |\n",
            "|  2 | Anatomy::UBERON:0000006 | islet of Langerhans       | Anatomy |\n",
            "|  3 | Anatomy::UBERON:0000007 | pituitary gland           | Anatomy |\n",
            "|  4 | Anatomy::UBERON:0000010 | peripheral nervous system | Anatomy |\n",
            "+----+-------------------------+---------------------------+---------+\n",
            "\n",
            " Sample of Hetionet Data Types (Before processing):\n",
            "\n",
            "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
            "|       | id                               | name                                      | kind                |\n",
            "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
            "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
            "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
            "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
            "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
            "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
            "| 14863 | Gene::1                          | A1BG                                      | Gene                |\n",
            "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
            "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
            "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
            "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
            "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
            "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
            "\n",
            " Sample of Hetionet Data Types (After processing):\n",
            "\n",
            "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
            "|       | id                               | name                                      | kind                |\n",
            "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
            "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
            "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
            "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
            "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
            "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
            "| 14863 | Gene::1                          | A1BG gene                                 | Gene                |\n",
            "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
            "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
            "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
            "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
            "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
            "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
            "\n",
            " Gene ID Dataframe (After processing):\n",
            "\n",
            "+----+----------+--------------------------------------------+\n",
            "|    | GeneID   | description                                |\n",
            "|----+----------+--------------------------------------------|\n",
            "|  0 | Gene::1  | alpha-1-B glycoprotein (A1BG)              |\n",
            "|  1 | Gene::2  | alpha-2-macroglobulin (A2M)                |\n",
            "|  2 | Gene::3  | alpha-2-macroglobulin pseudogene 1 (A2MP1) |\n",
            "|  3 | Gene::9  | N-acetyltransferase 1 (NAT1)               |\n",
            "|  4 | Gene::10 | N-acetyltransferase 2 (NAT2)               |\n",
            "+----+----------+--------------------------------------------+\n",
            "\n",
            " DrugBank Dataframe (Before processing):\n",
            "\n",
            "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
            "|    | DrugBank ID   | Accession Numbers    | Common name         | CAS         | UNII       | Synonyms                                                                                                                                                              |   Standard InChI Key |\n",
            "|----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------|\n",
            "|  0 | DB00001       | BIOD00024 | BTD00024 | Lepirudin           | 138068-37-8 | Y43GF64R34 | [Leu1, Thr2]-63-desulfohirudin | Desulfatohirudin | Hirudin variant-1 | Lepirudin | Lepirudin recombinant | R-hirudin                                                 |                  nan |\n",
            "|  1 | DB00002       | BIOD00071 | BTD00071 | Cetuximab           | 205923-56-4 | PQX0D8J21J | Cetuximab | Cétuximab | Cetuximabum                                                                                                                                   |                  nan |\n",
            "|  2 | DB00003       | BIOD00001 | BTD00001 | Dornase alfa        | 143831-71-4 | 953A26OA1Y | Deoxyribonuclease (human clone 18-1 protein moiety) | Dornasa alfa | Dornase alfa | Dornase alfa, recombinant | Dornase alpha | Recombinant deoxyribonuclease (DNAse) |                  nan |\n",
            "|  3 | DB00004       | BIOD00084 | BTD00084 | Denileukin diftitox | 173146-27-5 | 25E79B5CTM | Denileukin | Denileukin diftitox | Interleukin-2/diptheria toxin fusion protein                                                                                       |                  nan |\n",
            "|  4 | DB00005       | BIOD00052 | BTD00052 | Etanercept          | 185243-69-0 | OP401G7OJC | Etanercept | etanercept-szzs | etanercept-ykro | Recombinant human TNF | rhu TNFR:Fc | rhu-TNFR:Fc | TNFR-Immunoadhesin                                               |                  nan |\n",
            "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
            "\n",
            " DrugBank Dataframe (After processing):\n",
            "\n",
            "+----+-------------------+---------------------+\n",
            "|    | DrugBank ID       | Common name         |\n",
            "|----+-------------------+---------------------|\n",
            "|  0 | Compound::DB00001 | Lepirudin           |\n",
            "|  1 | Compound::DB00002 | Cetuximab           |\n",
            "|  2 | Compound::DB00003 | Dornase alfa        |\n",
            "|  3 | Compound::DB00004 | Denileukin diftitox |\n",
            "|  4 | Compound::DB00005 | Etanercept          |\n",
            "+----+-------------------+---------------------+\n",
            "\n",
            " data/mimTitles.txt  Dataframe (Before processing):\n",
            "\n",
            "+----+----------------------+----------------------------------------+\n",
            "|    | MIM Number           | name                                   |\n",
            "|----+----------------------+----------------------------------------|\n",
            "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
            "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
            "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
            "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
            "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
            "+----+----------------------+----------------------------------------+\n",
            "\n",
            " data/mimTitles.txt  Dataframe (After processing):\n",
            "\n",
            "+----+----------------------+----------------------------------------+\n",
            "|    | MIM Number           | name                                   |\n",
            "|----+----------------------+----------------------------------------|\n",
            "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
            "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
            "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
            "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
            "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
            "+----+----------------------+----------------------------------------+\n",
            "\n",
            " MeSH Dictionary:\n",
            "\n",
            "+---------------+------------------------+\n",
            "|               | 0                      |\n",
            "|---------------+------------------------|\n",
            "| MESH::D000001 | Calcimycin             |\n",
            "| MESH::D000002 | Temefos                |\n",
            "| MESH::D000003 | Abattoirs              |\n",
            "| MESH::D000004 | Abbreviations as Topic |\n",
            "| MESH::D000005 | Abdomen                |\n",
            "+---------------+------------------------+\n"
          ]
        }
      ],
      "source": [
        "# 4) Load Data frames for translation\n",
        "hetionet_df, gene_df, drugbank_df, omim_df, mesh_dict, chebi_df, chembl_df = load_lookups(data_path=DATA_DIR,verbose=verbose)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "HoImejcxpFYP",
        "outputId": "904e49c4-8889-4f2d-fcce-26dc99331b86"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "+----+------------------------------+--------------------------------+------------------------------------------------------+\n",
            "|    | 0                            | 1                              | 2                                                    |\n",
            "|----+------------------------------+--------------------------------+------------------------------------------------------|\n",
            "|  0 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | coagulation factor VIII (F8)                         |\n",
            "|  1 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | phytanoyl-CoA 2-hydroxylase (PHYH)                   |\n",
            "|  2 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | coagulation factor IX (F9)                           |\n",
            "|  3 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | heat shock protein family A (Hsp70) member 5 (HSPA5) |\n",
            "|  4 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | immunoglobulin kappa variable 3-20 (IGKV3-20)        |\n",
            "+----+------------------------------+--------------------------------+------------------------------------------------------+\n",
            "Number of unique DRKG entities:  97238\n",
            "Number of translated entities:  64033\n",
            "Number of untranslated entities:  33205\n",
            "Percentage of entities translated:  65.9 %\n",
            "Total DRKG relationships:  5874261\n",
            "Translated DRKG relationships:  5705613\n",
            "Percentage of relationships fully translated:  97.1 %\n"
          ]
        }
      ],
      "source": [
        "# Make dictionaries for codes\n",
        "code_df   = pd.concat([hetionet_df[['name', 'id']], \n",
        "                       gene_df.rename(columns = {\"description\":\"name\", \"GeneID\":\"id\"}),\n",
        "                       drugbank_df.rename(columns = {\"Common name\":\"name\", \"DrugBank ID\":\"id\"}),\n",
        "                       omim_df.rename(columns = {\"MIM Number\":\"id\"}),\n",
        "                       chebi_df.rename(columns = {\"NAME\":\"name\", \"CHEBI_ACCESSION\":\"id\"}),\n",
        "                       chembl_df.rename(columns = {\"pref_name\":\"name\", \"chembl_id\":\"id\"})\n",
        "                       ], ignore_index=True, axis=0).drop_duplicates() \n",
        "code_dict = pd.Series(code_df['name'].values, index=code_df['id']).to_dict() | mesh_dict # Convert node df to dict and merge with MeSH dictionary\n",
        "\n",
        "# Get unique DRKG entities\n",
        "drkg_entities = get_unique_entities(drkg_df, [0,2])\n",
        "\n",
        "# Create and use convert_entitynames function\n",
        "drkg_entity_df, drkg_unmatched = get_entity_lookup(drkg_entities, code_dict)\n",
        "\n",
        "# Create final node dictionary\n",
        "node_dict = pd.Series(drkg_entity_df['name'].values, index=drkg_entity_df['drkg_id']).to_dict() \n",
        "\n",
        "# Initialize translated DRKG and manually clean heads/tails for one case where they were flipped\n",
        "drkg_translated    = drkg_df.copy()\n",
        "drkg_translated = flip_headtail(drkg_translated, 'Gene:Compound')\n",
        "\n",
        "# Map DRKG to translated entity names\n",
        "drkg_translated = convert_entitynames(drkg_translated, 0, node_dict)\n",
        "drkg_translated = convert_entitynames(drkg_translated, 2, node_dict)\n",
        "drkg_translated = drkg_translated.dropna()\n",
        "print_head(drkg_translated) \n",
        "\n",
        "# Summarize percentage translated\n",
        "print(\"Number of unique DRKG entities: \", len(drkg_entities)) # should be 97238\n",
        "print(\"Number of translated entities: \", drkg_entity_df.shape[0])\n",
        "print(\"Number of untranslated entities: \", drkg_unmatched.shape[0])\n",
        "pct_entity_translated = drkg_entity_df.shape[0]/len(drkg_entities)\n",
        "print('Percentage of entities translated: ', round(pct_entity_translated*100,1), '%')\n",
        "\n",
        "print('Total DRKG relationships: ', drkg_df.shape[0])\n",
        "print('Translated DRKG relationships: ', drkg_translated.shape[0])\n",
        "pct_translated = drkg_translated.shape[0]/drkg_df.shape[0]\n",
        "print('Percentage of relationships fully translated: ', round(pct_translated*100,1), '%')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 27,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>drkg_id</th>\n",
              "      <th>Data-source</th>\n",
              "      <th>Connected entity-types</th>\n",
              "      <th>Interaction-type</th>\n",
              "      <th>Description</th>\n",
              "      <th>Reference for the description</th>\n",
              "      <th>head_entity</th>\n",
              "      <th>tail_entity</th>\n",
              "      <th>relation_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>DGIDB::ACTIVATOR::Gene:Compound</td>\n",
              "      <td>DGIDB</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>activation</td>\n",
              "      <td>An activator interaction is when a drug activa...</td>\n",
              "      <td>http://www.dgidb.org/getting_started</td>\n",
              "      <td>Compound</td>\n",
              "      <td>Gene</td>\n",
              "      <td>activation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>DGIDB::AGONIST::Gene:Compound</td>\n",
              "      <td>DGIDB</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>agonism</td>\n",
              "      <td>An agonist interaction occurs when a drug bind...</td>\n",
              "      <td>http://www.dgidb.org/getting_started</td>\n",
              "      <td>Compound</td>\n",
              "      <td>Gene</td>\n",
              "      <td>activation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>DGIDB::ALLOSTERIC MODULATOR::Gene:Compound</td>\n",
              "      <td>DGIDB</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>allosteric modulation</td>\n",
              "      <td>An allosteric modulator interaction occurs whe...</td>\n",
              "      <td>http://www.dgidb.org/getting_started</td>\n",
              "      <td>Compound</td>\n",
              "      <td>Gene</td>\n",
              "      <td>allosteric modulation</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>DGIDB::ANTAGONIST::Gene:Compound</td>\n",
              "      <td>DGIDB</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>antagonism</td>\n",
              "      <td>An antagonist interaction occurs when a drug b...</td>\n",
              "      <td>http://www.dgidb.org/getting_started</td>\n",
              "      <td>Compound</td>\n",
              "      <td>Gene</td>\n",
              "      <td>antagonism</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>DGIDB::ANTIBODY::Gene:Compound</td>\n",
              "      <td>DGIDB</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>antibody</td>\n",
              "      <td>An antibody interaction occurs when an antibod...</td>\n",
              "      <td>http://www.dgidb.org/getting_started</td>\n",
              "      <td>Compound</td>\n",
              "      <td>Gene</td>\n",
              "      <td>antibody</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>102</th>\n",
              "      <td>bioarx::Covid2_acc_host_gene::Disease:Gene</td>\n",
              "      <td>BIBLIOGRAPHY</td>\n",
              "      <td>Disease:Gene</td>\n",
              "      <td>interaction</td>\n",
              "      <td>Interactions between 27 viral  proteins,  and ...</td>\n",
              "      <td></td>\n",
              "      <td>Disease</td>\n",
              "      <td>Gene</td>\n",
              "      <td>interaction</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>103</th>\n",
              "      <td>bioarx::DrugHumGen:Compound:Gene</td>\n",
              "      <td>BIBLIOGRAPHY</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>interaction</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>interaction</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>104</th>\n",
              "      <td>bioarx::DrugVirGen:Compound:Gene</td>\n",
              "      <td>BIBLIOGRAPHY</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>interaction</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>interaction</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>105</th>\n",
              "      <td>bioarx::HumGenHumGen:Gene:Gene</td>\n",
              "      <td>BIBLIOGRAPHY</td>\n",
              "      <td>Gene:Gene</td>\n",
              "      <td>interaction</td>\n",
              "      <td>Protein-protein interaction</td>\n",
              "      <td></td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>interaction</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>106</th>\n",
              "      <td>bioarx::VirGenHumGen:Gene:Gene</td>\n",
              "      <td>BIBLIOGRAPHY</td>\n",
              "      <td>Gene:Gene</td>\n",
              "      <td>interaction</td>\n",
              "      <td>Protein-protein interaction</td>\n",
              "      <td></td>\n",
              "      <td>None</td>\n",
              "      <td>None</td>\n",
              "      <td>interaction</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>107 rows × 9 columns</p>\n",
              "</div>"
            ],
            "text/plain": [
              "                                        drkg_id   Data-source   \n",
              "0               DGIDB::ACTIVATOR::Gene:Compound         DGIDB  \\\n",
              "1                 DGIDB::AGONIST::Gene:Compound         DGIDB   \n",
              "2    DGIDB::ALLOSTERIC MODULATOR::Gene:Compound         DGIDB   \n",
              "3              DGIDB::ANTAGONIST::Gene:Compound         DGIDB   \n",
              "4                DGIDB::ANTIBODY::Gene:Compound         DGIDB   \n",
              "..                                          ...           ...   \n",
              "102  bioarx::Covid2_acc_host_gene::Disease:Gene  BIBLIOGRAPHY   \n",
              "103            bioarx::DrugHumGen:Compound:Gene  BIBLIOGRAPHY   \n",
              "104            bioarx::DrugVirGen:Compound:Gene  BIBLIOGRAPHY   \n",
              "105              bioarx::HumGenHumGen:Gene:Gene  BIBLIOGRAPHY   \n",
              "106              bioarx::VirGenHumGen:Gene:Gene  BIBLIOGRAPHY   \n",
              "\n",
              "    Connected entity-types       Interaction-type   \n",
              "0            Compound:Gene             activation  \\\n",
              "1            Compound:Gene                agonism   \n",
              "2            Compound:Gene  allosteric modulation   \n",
              "3            Compound:Gene             antagonism   \n",
              "4            Compound:Gene               antibody   \n",
              "..                     ...                    ...   \n",
              "102           Disease:Gene            interaction   \n",
              "103          Compound:Gene            interaction   \n",
              "104          Compound:Gene            interaction   \n",
              "105              Gene:Gene            interaction   \n",
              "106              Gene:Gene            interaction   \n",
              "\n",
              "                                           Description   \n",
              "0    An activator interaction is when a drug activa...  \\\n",
              "1    An agonist interaction occurs when a drug bind...   \n",
              "2    An allosteric modulator interaction occurs whe...   \n",
              "3    An antagonist interaction occurs when a drug b...   \n",
              "4    An antibody interaction occurs when an antibod...   \n",
              "..                                                 ...   \n",
              "102  Interactions between 27 viral  proteins,  and ...   \n",
              "103                                                      \n",
              "104                                                      \n",
              "105                        Protein-protein interaction   \n",
              "106                        Protein-protein interaction   \n",
              "\n",
              "            Reference for the description head_entity tail_entity   \n",
              "0    http://www.dgidb.org/getting_started    Compound        Gene  \\\n",
              "1    http://www.dgidb.org/getting_started    Compound        Gene   \n",
              "2    http://www.dgidb.org/getting_started    Compound        Gene   \n",
              "3    http://www.dgidb.org/getting_started    Compound        Gene   \n",
              "4    http://www.dgidb.org/getting_started    Compound        Gene   \n",
              "..                                    ...         ...         ...   \n",
              "102                                           Disease        Gene   \n",
              "103                                              None        None   \n",
              "104                                              None        None   \n",
              "105                                              None        None   \n",
              "106                                              None        None   \n",
              "\n",
              "             relation_name  \n",
              "0               activation  \n",
              "1               activation  \n",
              "2    allosteric modulation  \n",
              "3               antagonism  \n",
              "4                 antibody  \n",
              "..                     ...  \n",
              "102            interaction  \n",
              "103            interaction  \n",
              "104            interaction  \n",
              "105            interaction  \n",
              "106            interaction  \n",
              "\n",
              "[107 rows x 9 columns]"
            ]
          },
          "execution_count": 27,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Update relation glossary \n",
        "relation_df = relation_glossary.copy().rename(columns={'Relation-name':'drkg_id'})\n",
        "relation_df[['head_entity','tail_entity']] = relation_df['drkg_id'].str.split('::', expand=True)[2].str.split(':', expand=True) # Set head and tail nodes\n",
        "\n",
        "# Manually fix head and tail nodes for DGIDB relations, which reverse compound-gene interactions\n",
        "relation_df.loc[relation_df['drkg_id'].str.contains('Gene:Compound'),'head_entity'] = 'Compound'\n",
        "relation_df.loc[relation_df['drkg_id'].str.contains('Gene:Compound'),'tail_entity'] = 'Gene'\n",
        "\n",
        "# Add mapped relation group labels\n",
        "relation_groups = [['activation', 'agonism', 'agonism, activation', 'activates, stimulates'],\n",
        "    ['antagonism', 'blocking', 'antagonism, blocking'],\n",
        "    ['binding', 'binding, ligand (esp. receptors)'],\n",
        "    ['blocking', 'channel blocking'],\n",
        "    ['inhibition', 'inhibits cell growth (esp. cancers)', 'inhibits'],\n",
        "    ['enzyme', 'enzyme activity'],\n",
        "    ['upregulation', 'increases expression/production'],\n",
        "    ['downregulation', 'decreases expression/production'],\n",
        "    ['Compound treats the disease', 'treatment/therapy (including investigatory)', 'treatment']]\n",
        "\n",
        "relation_df['relation_name'] = relation_df['Interaction-type']\n",
        "\n",
        "for grp in relation_groups:\n",
        "    relation_df_subset = relation_df[relation_df['Interaction-type'].isin(grp)].copy()\n",
        "    for entities in relation_df_subset['Connected entity-types'].unique():\n",
        "        subgrp = relation_df_subset[relation_df_subset['Connected entity-types'] == entities]['Interaction-type'].unique()\n",
        "        relation_df.loc[(relation_df_subset['Connected entity-types'] == entities) & (relation_df['Interaction-type'].isin(subgrp)), 'relation_name'] = subgrp[0]\n",
        "\n",
        "relation_df\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 31,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>drkg_id</th>\n",
              "      <th>Data-source</th>\n",
              "      <th>Connected entity-types</th>\n",
              "      <th>Interaction-type</th>\n",
              "      <th>Description</th>\n",
              "      <th>Reference for the description</th>\n",
              "      <th>head_entity</th>\n",
              "      <th>tail_entity</th>\n",
              "      <th>relation_name</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>DGIDB::INHIBITOR::Gene:Compound</td>\n",
              "      <td>DGIDB</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>inhibition</td>\n",
              "      <td>In inhibitor interactions, the drug binds to a...</td>\n",
              "      <td>http://www.dgidb.org/getting_started</td>\n",
              "      <td>Compound</td>\n",
              "      <td>Gene</td>\n",
              "      <td>inhibition</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>39</th>\n",
              "      <td>GNBR::N::Compound:Gene</td>\n",
              "      <td>GNBR</td>\n",
              "      <td>Compound:Gene</td>\n",
              "      <td>inhibits</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>Compound</td>\n",
              "      <td>Gene</td>\n",
              "      <td>inhibition</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>97</th>\n",
              "      <td>STRING::INHIBITION::Gene:Gene</td>\n",
              "      <td>STRING</td>\n",
              "      <td>Gene:Gene</td>\n",
              "      <td>inhibition</td>\n",
              "      <td></td>\n",
              "      <td></td>\n",
              "      <td>Gene</td>\n",
              "      <td>Gene</td>\n",
              "      <td>inhibition</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                            drkg_id Data-source Connected entity-types   \n",
              "8   DGIDB::INHIBITOR::Gene:Compound       DGIDB          Compound:Gene  \\\n",
              "39           GNBR::N::Compound:Gene        GNBR          Compound:Gene   \n",
              "97    STRING::INHIBITION::Gene:Gene      STRING              Gene:Gene   \n",
              "\n",
              "   Interaction-type                                        Description   \n",
              "8        inhibition  In inhibitor interactions, the drug binds to a...  \\\n",
              "39         inhibits                                                      \n",
              "97       inhibition                                                      \n",
              "\n",
              "           Reference for the description head_entity tail_entity relation_name  \n",
              "8   http://www.dgidb.org/getting_started    Compound        Gene    inhibition  \n",
              "39                                          Compound        Gene    inhibition  \n",
              "97                                              Gene        Gene    inhibition  "
            ]
          },
          "execution_count": 31,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Test specific cases\n",
        "test = 'inhibition'\n",
        "relation_df[relation_df['relation_name']==test]\n",
        "# relation_df[relation_df['Connected entity-types']=='Compound:Disease']"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 419
        },
        "id": "IcZZ_cEn-hIF",
        "outputId": "83b65c42-cf6f-41e2-9f30-f1c3618d957a"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>0</th>\n",
              "      <th>1</th>\n",
              "      <th>2</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1518268</th>\n",
              "      <td>Dornase alfa</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Cystic Fibrosis</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518270</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Spondylitis, Ankylosing</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518271</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Graft vs Host Disease</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518272</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Hidradenitis Suppurativa</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518273</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Arthritis, Juvenile</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518274</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Psoriasis</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518275</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Arthritis, Psoriatic</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518276</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Pyoderma Gangrenosum</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518277</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Arthritis, Rheumatoid</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1518278</th>\n",
              "      <td>Etanercept</td>\n",
              "      <td>DRUGBANK::treats::Compound:Disease</td>\n",
              "      <td>Stevens-Johnson Syndrome</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                    0                                   1   \n",
              "1518268  Dornase alfa  DRUGBANK::treats::Compound:Disease  \\\n",
              "1518270    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "1518271    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "1518272    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "1518273    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "1518274    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "1518275    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "1518276    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "1518277    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "1518278    Etanercept  DRUGBANK::treats::Compound:Disease   \n",
              "\n",
              "                                2  \n",
              "1518268           Cystic Fibrosis  \n",
              "1518270   Spondylitis, Ankylosing  \n",
              "1518271     Graft vs Host Disease  \n",
              "1518272  Hidradenitis Suppurativa  \n",
              "1518273       Arthritis, Juvenile  \n",
              "1518274                 Psoriasis  \n",
              "1518275      Arthritis, Psoriatic  \n",
              "1518276      Pyoderma Gangrenosum  \n",
              "1518277     Arthritis, Rheumatoid  \n",
              "1518278  Stevens-Johnson Syndrome  "
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Filter DRKG in natural language to drug-treats-disease relationships\n",
        "# rx_dx        = df_med[df_med.iloc[:,1] ==   'Compound treats the disease']\n",
        "rx_dx        =  drkg_translated[drkg_translated[1].isin(drkg_rx_dx_relation_subset)]\n",
        "rx_dx_subset =  rx_dx[0:10]\n",
        "rx_dx_subset"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {
        "id": "oTAJOqDVc_Hg"
      },
      "source": [
        "# 3) BioLinkBERT embedding"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 10,
      "metadata": {},
      "outputs": [],
      "source": [
        "from torch_geometric.data import HeteroData\n",
        "from SIMP_LLM.llm_encode import EntityEncoder\n",
        "from SIMP_LLM.dataloader_mappings import create_mapping, create_edges, embed_entities, embed_edges"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "### Set variables and load data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 37,
      "metadata": {
        "id": "QkZKp6CJiw6f"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "+----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-------------------------+\n",
            "|    | drkg_id                            | Data-source   | Connected entity-types   | Interaction-type            | Description                                                                                                                                                                                            | Reference for the description        | head_entity   | tail_entity   | relation_name           |\n",
            "|----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-------------------------|\n",
            "|  0 | DGIDB::ACTIVATOR::Gene:Compound    | DGIDB         | Compound:Gene            | activation                  | An activator interaction is when a drug activates a biological response from a target, although the mechanism by which it does so may not be understood.                                               | http://www.dgidb.org/getting_started | Compound      | Gene          | Compound activates gene |\n",
            "|  1 | DGIDB::AGONIST::Gene:Compound      | DGIDB         | Compound:Gene            | agonism                     | An agonist interaction occurs when a drug binds to a target receptor and activates the receptor to produce a biological response.                                                                      | http://www.dgidb.org/getting_started | Compound      | Gene          | Compound activates gene |\n",
            "|  8 | DGIDB::INHIBITOR::Gene:Compound    | DGIDB         | Compound:Gene            | inhibition                  | In inhibitor interactions, the drug binds to a target and decreases its expression or activity. Most interactions of this class are enzyme inhibitors, which bind an enzyme to reduce enzyme activity. | http://www.dgidb.org/getting_started | Compound      | Gene          | Inhibition              |\n",
            "| 18 | DRUGBANK::treats::Compound:Disease | DRUGBANK      | Compound:Disease         | Compound treats the disease | We extract  the  associated  conditions  related  to  compound entities in the drugbank.                                                                                                               |                                      | Compound      | Disease       | Compound treats disease |\n",
            "| 19 | GNBR::A+::Compound:Gene            | GNBR          | Compound:Gene            | agonism, activation         |                                                                                                                                                                                                        |                                      | Compound      | Gene          | Compound activates gene |\n",
            "+----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-------------------------+\n",
            "+----+-------------+---------------------------------+---------------------------------------------------------+\n",
            "|    | 0           | 1                               | 2                                                       |\n",
            "|----+-------------+---------------------------------+---------------------------------------------------------|\n",
            "|  0 | SNS-032     | DGIDB::INHIBITOR::Gene:Compound | cyclin dependent kinase 7 (CDK7)                        |\n",
            "|  1 | Quizartinib | DGIDB::INHIBITOR::Gene:Compound | platelet derived growth factor receptor beta (PDGFRB)   |\n",
            "|  2 | Sufentanil  | DGIDB::AGONIST::Gene:Compound   | opioid receptor mu 1 (OPRM1)                            |\n",
            "|  3 | PF-04691502 | DGIDB::INHIBITOR::Gene:Compound | phosphoinositide-3-kinase regulatory subunit 2 (PIK3R2) |\n",
            "|  4 | Ramelteon   | DGIDB::AGONIST::Gene:Compound   | melatonin receptor 1A (MTNR1A)                          |\n",
            "+----+-------------+---------------------------------+---------------------------------------------------------+\n",
            "+-------+-------------------+-------------------+----------------+---------------+-----------------+-----------------+---------+\n",
            "|       | drkg_id           | drkg_dict_id      | name           | entity_type   | ontology_code   | ontology_name   | code    |\n",
            "|-------+-------------------+-------------------+----------------+---------------+-----------------+-----------------+---------|\n",
            "| 21962 | Compound::DB00003 | Compound::DB00003 | Dornase alfa   | Compound      | DB00003         | drugbank        | DB00003 |\n",
            "| 21964 | Compound::DB00005 | Compound::DB00005 | Etanercept     | Compound      | DB00005         | drugbank        | DB00005 |\n",
            "| 22319 | Compound::DB00371 | Compound::DB00371 | Meprobamate    | Compound      | DB00371         | drugbank        | DB00371 |\n",
            "| 22653 | Compound::DB00708 | Compound::DB00708 | Sufentanil     | Compound      | DB00708         | drugbank        | DB00708 |\n",
            "| 22688 | Compound::DB00743 | Compound::DB00743 | Gadobenic acid | Compound      | DB00743         | drugbank        | DB00743 |\n",
            "+-------+-------------------+-------------------+----------------+---------------+-----------------+-----------------+---------+\n"
          ]
        }
      ],
      "source": [
        "## Example of loading data without anything to encode\n",
        "device   = \"cpu\"\n",
        "Encoder  = EntityEncoder(device = device )\n",
        "\n",
        "\n",
        "# Create relationship subset for testing\n",
        "test_relation_df = relation_df[relation_df['Connected entity-types'].isin(['Compound:Gene', 'Disease:Gene', 'Compound:Disease', 'Gene:Gene'])].copy()\n",
        "test_relation_df['relation_name'] = None\n",
        "\n",
        "activation_list = ['activation', 'agonism', 'agonism, activation'] \n",
        "treat_list = ['Compound treats the disease', 'treats']\n",
        "gene_drug_list = ['inhibition']\n",
        "\n",
        "test_relation_df['relation_name'][test_relation_df['Interaction-type'].isin(activation_list)] = 'Compound activates gene'\n",
        "test_relation_df['relation_name'][test_relation_df['Interaction-type'].isin(treat_list)] = 'Compound treats disease'\n",
        "test_relation_df['relation_name'][test_relation_df['Interaction-type'].isin(gene_drug_list)] = 'Inhibition'\n",
        "test_relation_df = test_relation_df[~test_relation_df['relation_name'].isna()]\n",
        "print_head(test_relation_df)\n",
        "\n",
        "# Create test sample of DRKG relationships filtering to these relations (for full sample: delete and use drkg_entity_df)\n",
        "test_hrt_df = drkg_translated[drkg_translated[1].isin(test_relation_df['drkg_id'])]\n",
        "test_hrt_df = test_hrt_df.groupby(1).head(3).reset_index(drop=True)\n",
        "test_unique_entities = get_unique_entities(test_hrt_df, columns=[0,2])\n",
        "test_entity_df = drkg_entity_df[drkg_entity_df['name'].isin(test_unique_entities)]\n",
        "print_head(test_hrt_df)\n",
        "print_head(test_entity_df)\n",
        "\n",
        "\n",
        "entity_df = test_entity_df.copy() # (for full sample: replace test_entity_df with drkg_entity_df)\n",
        "hrt_data = test_hrt_df.copy()  # (for full sample: replace test_hrt_df with drkg_translated)\n",
        "relation_lookup = relation_df.copy() # (for full sample: replace test_relation_df with the updated relation_glossary with relation_name)"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {
        "id": "ywHcEVMiAdOv"
      },
      "source": [
        "### Build HeteroData Object"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 38,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "HeteroData(\n",
            "  \u001b[1mCompound\u001b[0m={ x=[13, 768] },\n",
            "  \u001b[1mDisease\u001b[0m={ x=[3, 768] },\n",
            "  \u001b[1mGene\u001b[0m={ x=[24, 768] },\n",
            "  \u001b[1m(Compound, activation, Gene)\u001b[0m={\n",
            "    edge_index=[2, 9],\n",
            "    edge_label=[9, 768]\n",
            "  },\n",
            "  \u001b[1m(Compound, inhibition, Gene)\u001b[0m={\n",
            "    edge_index=[2, 3],\n",
            "    edge_label=[3, 768]\n",
            "  },\n",
            "  \u001b[1m(Compound, Compound treats the disease, Disease)\u001b[0m={\n",
            "    edge_index=[2, 3],\n",
            "    edge_label=[3, 768]\n",
            "  },\n",
            "  \u001b[1m(Gene, activates, stimulates, Gene)\u001b[0m={\n",
            "    edge_index=[2, 3],\n",
            "    edge_label=[3, 768]\n",
            "  },\n",
            "  \u001b[1m(Gene, inhibition, Gene)\u001b[0m={\n",
            "    edge_index=[2, 3],\n",
            "    edge_label=[3, 768]\n",
            "  }\n",
            ")\n",
            "Unique Compounds: 13 \t Matrix shape: torch.Size([13, 768])\n",
            "Unique Diseases: 3 \t Matrix shape: torch.Size([3, 768])\n",
            "Unique Genes: 24 \t Matrix shape: torch.Size([24, 768])\n"
          ]
        }
      ],
      "source": [
        "# Initialize heterograph object\n",
        "graph_obj = HeteroData()\n",
        "\n",
        "# Embed entities, add to graph, and save embedding mapping dictionary of dictionaries\n",
        "mapping_dict = embed_entities(entity_df, graph_obj, Encoder, device) \n",
        "\n",
        "# Embed relationships, add to graph, and save relation embeddings/mapping dictionary\n",
        "relation_X, relation_mapping = embed_edges(hrt_data, relation_lookup, graph_obj, mapping_dict, Encoder, device)\n",
        "\n",
        "# Print summary\n",
        "print(graph_obj)\n",
        "for ent_type in entity_df['entity_type'].unique():\n",
        "    print(f\"Unique {ent_type}s: {len(mapping_dict[ent_type])} \\t Matrix shape: {graph_obj[ent_type].x.shape }\")\n",
        "    # print(mapping_dict[ent_type]) # Prints whole dictionary so delete/uncomment if using all entities"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "\n",
        "### Old example: DX RX Relationship ###"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "\n",
        "# Create embeddings\n",
        "rx_X,rx_mapping = create_mapping(rx_dx_subset[0].to_list(),encoder= Encoder ,device=device) # Maps drugs to indices\n",
        "dx_X,dx_mapping = create_mapping(rx_dx_subset[2].to_list(),encoder= Encoder ,device=device) # Maps drugs to indices\n",
        "## As of now this only encodes 'Compound treats the disease', but  the idea is that this is used to encode every entity \n",
        "relationship_X,relationship_mapping = create_mapping(rx_dx_subset[1].to_list(),encoder= Encoder ,device=device)  \n",
        "\n",
        "print(f\"Unique Drugs:   {len(rx_mapping)} Matrix shape: {rx_X.shape}\")\n",
        "print(f\"Unique Disases: {len(dx_mapping)} Matrix shape: {dx_X.shape }\")\n",
        "relationship_feature = relationship_X[relationship_mapping['Compound treats the disease'],:].reshape(1,-1)\n",
        "\n",
        "# Build heterograph object\n",
        "### EXAMPLE: Create PyG Hetero Graph:\n",
        "data = HeteroData()\n",
        "data['compounds'].x = rx_X\n",
        "data['disease'].x   = dx_X\n",
        "print(data)\n",
        "\n",
        "Edge_index,edge_attribute = create_edges(df             =  rx_dx_subset,\n",
        "                                          src_index_col  = 0, \n",
        "                                          src_mapping    = rx_mapping , \n",
        "                                          dst_index_col  = 2, \n",
        "                                          dst_mapping    = dx_mapping ,\n",
        "                                          edge_attr      = relationship_feature)\n",
        "\n",
        "data['compounds', 'Compound treats the disease', 'disease'].edge_index = Edge_index\n",
        "data['compounds', 'Compound treats the disease', 'disease'].edge_label = edge_attribute "
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "gpuClass": "standard",
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.3"
    },
    "widgets": {
      "application/vnd.jupyter.widget-state+json": {
        "2d42bd2c408a4bd3bdebc2865ea42483": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": "hidden",
            "width": null
          }
        },
        "6e79954f9c514f099b336e3a11ad635e": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_6f2e0878218c4b38974a74a673b70751",
            "placeholder": "​",
            "style": "IPY_MODEL_98ccadd9e4a34d8282282a332fad6e0b",
            "value": "Downloading chembl_32_sqlite.tar.gz: 100%"
          }
        },
        "6f2e0878218c4b38974a74a673b70751": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "98ccadd9e4a34d8282282a332fad6e0b": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "a95662c4dc0949f4af40e8331cde378d": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "FloatProgressModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "FloatProgressModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "ProgressView",
            "bar_style": "info",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_eec53708d80f41e8b402060c9a1352b1",
            "max": 1,
            "min": 0,
            "orientation": "horizontal",
            "style": "IPY_MODEL_f87f682db17a4e6b837b9c14c85ec36f",
            "value": 1
          }
        },
        "b0797e6674144b1b92d2ff70918b917c": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": null
          }
        },
        "b355679609b749d698029d4f39d45503": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HBoxModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HBoxModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HBoxView",
            "box_style": "",
            "children": [
              "IPY_MODEL_6e79954f9c514f099b336e3a11ad635e",
              "IPY_MODEL_a95662c4dc0949f4af40e8331cde378d",
              "IPY_MODEL_d89d003254ce42f9a44572c7e26940ec"
            ],
            "layout": "IPY_MODEL_2d42bd2c408a4bd3bdebc2865ea42483"
          }
        },
        "d0a9a73a5fe14afb8b15a2addf4f080d": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "DescriptionStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "DescriptionStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "description_width": ""
          }
        },
        "d89d003254ce42f9a44572c7e26940ec": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "HTMLModel",
          "state": {
            "_dom_classes": [],
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "HTMLModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/controls",
            "_view_module_version": "1.5.0",
            "_view_name": "HTMLView",
            "description": "",
            "description_tooltip": null,
            "layout": "IPY_MODEL_b0797e6674144b1b92d2ff70918b917c",
            "placeholder": "​",
            "style": "IPY_MODEL_d0a9a73a5fe14afb8b15a2addf4f080d",
            "value": " 4.27G/4.27G [03:38&lt;00:00, 21.7MB/s]"
          }
        },
        "eec53708d80f41e8b402060c9a1352b1": {
          "model_module": "@jupyter-widgets/base",
          "model_module_version": "1.2.0",
          "model_name": "LayoutModel",
          "state": {
            "_model_module": "@jupyter-widgets/base",
            "_model_module_version": "1.2.0",
            "_model_name": "LayoutModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "LayoutView",
            "align_content": null,
            "align_items": null,
            "align_self": null,
            "border": null,
            "bottom": null,
            "display": null,
            "flex": null,
            "flex_flow": null,
            "grid_area": null,
            "grid_auto_columns": null,
            "grid_auto_flow": null,
            "grid_auto_rows": null,
            "grid_column": null,
            "grid_gap": null,
            "grid_row": null,
            "grid_template_areas": null,
            "grid_template_columns": null,
            "grid_template_rows": null,
            "height": null,
            "justify_content": null,
            "justify_items": null,
            "left": null,
            "margin": null,
            "max_height": null,
            "max_width": null,
            "min_height": null,
            "min_width": null,
            "object_fit": null,
            "object_position": null,
            "order": null,
            "overflow": null,
            "overflow_x": null,
            "overflow_y": null,
            "padding": null,
            "right": null,
            "top": null,
            "visibility": null,
            "width": "20px"
          }
        },
        "f87f682db17a4e6b837b9c14c85ec36f": {
          "model_module": "@jupyter-widgets/controls",
          "model_module_version": "1.5.0",
          "model_name": "ProgressStyleModel",
          "state": {
            "_model_module": "@jupyter-widgets/controls",
            "_model_module_version": "1.5.0",
            "_model_name": "ProgressStyleModel",
            "_view_count": null,
            "_view_module": "@jupyter-widgets/base",
            "_view_module_version": "1.2.0",
            "_view_name": "StyleView",
            "bar_color": null,
            "description_width": ""
          }
        }
      }
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
