{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "DnQGl_D4786T"
   },
   "source": [
    "AUTHORS: Alejandro and Selina primary, contributions listed by section\n",
    "1) Selina primary with significant modularization by Alejandro, one chunk to save a dataset by Alex\n",
    "2) Alejandro and Selina\n",
    "3) Alejandro sole\n",
    "4) Selina sole"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "PV4N5XSIKfTa"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os \n",
    "import torch_geometric.transforms as T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SIMP_LLM.DRKG_loading   import  get_triplets, read_tsv,filter_drkg,map_drkg_relationships,filter_interaction_subset,print_head\n",
    "from SIMP_LLM.DRKG_translate import  load_lookups\n",
    "from SIMP_LLM.DRKG_entity_processing import get_unique_entities, get_entity_lookup, convert_entitynames, flip_headtail\n",
    "from SIMP_LLM.raredisease_loading import get_orphan_data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "VsLJKwgRxArZ"
   },
   "source": [
    "# 1) Load Data\n",
    "\n",
    "AUTHORS: First pass by Selina, first 2 cells in this section modularized by Alejandro, rest of section by Selina except one cell by Alex to save the data (drkg_translated) to a CSV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Triplets:\n",
      "\n",
      "[['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2157'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5264'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2158'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::3309'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::28912'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::811'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2159'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::821'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5627'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5624']]\n",
      "\n",
      " data/drkg.tsv  Dataframe:\n",
      "\n",
      "+----+------------+--------------------------------+-------------+\n",
      "|    | 0          | 1                              | 2           |\n",
      "|----+------------+--------------------------------+-------------|\n",
      "|  0 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2157  |\n",
      "|  1 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::5264  |\n",
      "|  2 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2158  |\n",
      "|  3 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::3309  |\n",
      "|  4 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::28912 |\n",
      "+----+------------+--------------------------------+-------------+\n",
      "\n",
      " data/relation_glossary.tsv  Dataframe:\n",
      "\n",
      "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
      "|    | Relation-name                              | Data-source   | Connected entity-types   | Interaction-type      | Description                                                                                                                                                                     | Reference for the description        |\n",
      "|----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------|\n",
      "|  0 | DGIDB::ACTIVATOR::Gene:Compound            | DGIDB         | Compound:Gene            | activation            | An activator interaction is when a drug activates a biological response from a target, although the mechanism by which it does so may not be understood.                        | http://www.dgidb.org/getting_started |\n",
      "|  1 | DGIDB::AGONIST::Gene:Compound              | DGIDB         | Compound:Gene            | agonism               | An agonist interaction occurs when a drug binds to a target receptor and activates the receptor to produce a biological response.                                               | http://www.dgidb.org/getting_started |\n",
      "|  2 | DGIDB::ALLOSTERIC MODULATOR::Gene:Compound | DGIDB         | Compound:Gene            | allosteric modulation | An allosteric modulator interaction occurs when drugs exert their effects on their protein targets via a different binding site than the natural (orthosteric) ligand site.     | http://www.dgidb.org/getting_started |\n",
      "|  3 | DGIDB::ANTAGONIST::Gene:Compound           | DGIDB         | Compound:Gene            | antagonism            | An antagonist interaction occurs when a drug blocks or dampens agonist-mediated responses rather than provoking a biological response itself upon binding to a target receptor. | http://www.dgidb.org/getting_started |\n",
      "|  4 | DGIDB::ANTIBODY::Gene:Compound             | DGIDB         | Compound:Gene            | antibody              | An antibody interaction occurs when an antibody drug specifically binds the target molecule.                                                                                    | http://www.dgidb.org/getting_started |\n",
      "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
      "Number of Rows Before Filtering: 107\n",
      "Number of Rows After Filtering: 10\n",
      "\n",
      "\n",
      "Filtered:\n",
      "24    DRUGBANK::treats::Compound:Disease\n",
      "35             GNBR::T::Compound:Disease\n",
      "36             GNBR::C::Compound:Disease\n",
      "37            GNBR::Sa::Compound:Disease\n",
      "38            GNBR::Pa::Compound:Disease\n",
      "39            GNBR::Mp::Compound:Disease\n",
      "40            GNBR::Pr::Compound:Disease\n",
      "41             GNBR::J::Compound:Disease\n",
      "68       Hetionet::CtD::Compound:Disease\n",
      "73       Hetionet::CpD::Compound:Disease\n",
      "dtype: object\n",
      "\n",
      "Relationships Mapped:\n",
      "0                    Compound treats the disease\n",
      "1    treatment/therapy (including investigatory)\n",
      "2            inhibits cell growth (esp. cancers)\n",
      "3                      side effect/adverse event\n",
      "4                            alleviates, reduces\n",
      "5            biomarkers (of disease progression)\n",
      "6                           prevents, suppresses\n",
      "7                   role in disease pathogenesis\n",
      "8                                      treatment\n",
      "9                                     palliation\n",
      "Name: Interaction-type, dtype: object\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n",
      "|         | 0                 | 1                                  | 2                     |\n",
      "|---------+-------------------+------------------------------------+-----------------------|\n",
      "| 1518268 | Compound::DB00003 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D003550 |\n",
      "| 1518269 | Compound::DB00004 | DRUGBANK::treats::Compound:Disease | Disease::MESH:C063419 |\n",
      "| 1518270 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D013167 |\n",
      "| 1518271 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D006086 |\n",
      "| 1518272 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D017497 |\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n"
     ]
    }
   ],
   "source": [
    "### 1) Read: This section reads DRKG and a glossary (used to map entities from codes to words)\n",
    "DATA_DIR           = os.path.join(\"data\")\n",
    "verbose            =  True \n",
    "triplets,drkg_df   =  get_triplets(drkg_file = os.path.join(DATA_DIR  ,'drkg.tsv'),             verbose=verbose)  # Read triplets (head,relationship,tail)\n",
    "relation_glossary  =  read_tsv(relation_file = os.path.join(DATA_DIR  ,'relation_glossary.tsv'),verbose=verbose)  # Read relationship mapping  \n",
    "\n",
    "\n",
    "### 2) Filter & Map Interactions: This section returns a list of interactions e.g. DRUGBANK::treats::Compound:Disease )\n",
    "# 2.1: First  we filter the interactions to only Compound-Disease\n",
    "# 2.2: Then   we map the codes -> text  (this will be use to further filter interactions based on text) e.g.  Hetionet::CpD::Compound:Disease -> palliation\n",
    "# 2.3: We use natural text to fitler  interactions based on terms such as \"treat\" (but we return the orignal interaction name )\n",
    "\n",
    "\n",
    "\n",
    "# modularize this in create_dataframe\n",
    "drkg_rx_dx_relations        = filter_drkg(data_frame = drkg_df ,  filter_column = 1 ,  filter_term = r'.*?Compound:Disease', verbose = verbose) # 2.1 Filter only Compound-Disease Interactions\n",
    "drkg_rx_dx_relations_mapped = map_drkg_relationships(drkg_rx_dx_relations,relation_glossary,verbose=verbose)                                    # 2.2 Map codes to text \n",
    "\n",
    "### 2.3 Filter Drug interactions Interaction types to only include: treat inhibit or alleviate interactions  ###\n",
    "drkg_rx_dx_relation_subset =  filter_interaction_subset(df                  = drkg_rx_dx_relations_mapped,\n",
    "                                                        filter_colunm_name = 'Interaction-type' ,\n",
    "                                                        regex_string       =  'treat|inhibit|alleviate',\n",
    "                                                        return_colunm_name =  'Relation-name')\n",
    "\n",
    "# 3) Use Filter Interactions to get Gilter DRKG \n",
    "drkg_df_filtered = drkg_df[drkg_df[1].isin(drkg_rx_dx_relation_subset)] # 3.1 Filter DRKG  to only  Compund-Disease \n",
    "print_head(df=drkg_df_filtered)\n",
    "\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "rx_dx_triplets   = drkg_df_filtered.values.tolist()                     # 3.2 Convert filtered DRKG to list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "KYBnEGL4Nixb",
    "outputId": "5bc6b100-d023-43d7-cf3c-488dc73b6475"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " data/hetionet-v1.0-nodes.tsv  Dataframe:\n",
      "\n",
      "+----+-------------------------+---------------------------+---------+\n",
      "|    | id                      | name                      | kind    |\n",
      "|----+-------------------------+---------------------------+---------|\n",
      "|  0 | Anatomy::UBERON:0000002 | uterine cervix            | Anatomy |\n",
      "|  1 | Anatomy::UBERON:0000004 | nose                      | Anatomy |\n",
      "|  2 | Anatomy::UBERON:0000006 | islet of Langerhans       | Anatomy |\n",
      "|  3 | Anatomy::UBERON:0000007 | pituitary gland           | Anatomy |\n",
      "|  4 | Anatomy::UBERON:0000010 | peripheral nervous system | Anatomy |\n",
      "+----+-------------------------+---------------------------+---------+\n",
      "\n",
      " Sample of Hetionet Data Types (Before processing):\n",
      "\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "|       | id                               | name                                      | kind                |\n",
      "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
      "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
      "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
      "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
      "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
      "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
      "| 14863 | Gene::1                          | A1BG                                      | Gene                |\n",
      "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
      "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
      "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
      "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
      "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "\n",
      " Sample of Hetionet Data Types (After processing):\n",
      "\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "|       | id                               | name                                      | kind                |\n",
      "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
      "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
      "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
      "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
      "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
      "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
      "| 14863 | Gene::1                          | A1BG gene                                 | Gene                |\n",
      "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
      "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
      "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
      "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
      "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "\n",
      " Gene ID Dataframe (After processing):\n",
      "\n",
      "+----+----------+--------------------------------------------+\n",
      "|    | GeneID   | description                                |\n",
      "|----+----------+--------------------------------------------|\n",
      "|  0 | Gene::1  | alpha-1-B glycoprotein (A1BG)              |\n",
      "|  1 | Gene::2  | alpha-2-macroglobulin (A2M)                |\n",
      "|  2 | Gene::3  | alpha-2-macroglobulin pseudogene 1 (A2MP1) |\n",
      "|  3 | Gene::9  | N-acetyltransferase 1 (NAT1)               |\n",
      "|  4 | Gene::10 | N-acetyltransferase 2 (NAT2)               |\n",
      "+----+----------+--------------------------------------------+\n",
      "\n",
      " DrugBank Dataframe (Before processing):\n",
      "\n",
      "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
      "|    | DrugBank ID   | Accession Numbers    | Common name         | CAS         | UNII       | Synonyms                                                                                                                                                              |   Standard InChI Key |\n",
      "|----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------|\n",
      "|  0 | DB00001       | BIOD00024 | BTD00024 | Lepirudin           | 138068-37-8 | Y43GF64R34 | [Leu1, Thr2]-63-desulfohirudin | Desulfatohirudin | Hirudin variant-1 | Lepirudin | Lepirudin recombinant | R-hirudin                                                 |                  nan |\n",
      "|  1 | DB00002       | BIOD00071 | BTD00071 | Cetuximab           | 205923-56-4 | PQX0D8J21J | Cetuximab | Cétuximab | Cetuximabum                                                                                                                                   |                  nan |\n",
      "|  2 | DB00003       | BIOD00001 | BTD00001 | Dornase alfa        | 143831-71-4 | 953A26OA1Y | Deoxyribonuclease (human clone 18-1 protein moiety) | Dornasa alfa | Dornase alfa | Dornase alfa, recombinant | Dornase alpha | Recombinant deoxyribonuclease (DNAse) |                  nan |\n",
      "|  3 | DB00004       | BIOD00084 | BTD00084 | Denileukin diftitox | 173146-27-5 | 25E79B5CTM | Denileukin | Denileukin diftitox | Interleukin-2/diptheria toxin fusion protein                                                                                       |                  nan |\n",
      "|  4 | DB00005       | BIOD00052 | BTD00052 | Etanercept          | 185243-69-0 | OP401G7OJC | Etanercept | etanercept-szzs | etanercept-ykro | Recombinant human TNF | rhu TNFR:Fc | rhu-TNFR:Fc | TNFR-Immunoadhesin                                               |                  nan |\n",
      "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
      "\n",
      " DrugBank Dataframe (After processing):\n",
      "\n",
      "+----+-------------------+---------------------+\n",
      "|    | DrugBank ID       | Common name         |\n",
      "|----+-------------------+---------------------|\n",
      "|  0 | Compound::DB00001 | Lepirudin           |\n",
      "|  1 | Compound::DB00002 | Cetuximab           |\n",
      "|  2 | Compound::DB00003 | Dornase alfa        |\n",
      "|  3 | Compound::DB00004 | Denileukin diftitox |\n",
      "|  4 | Compound::DB00005 | Etanercept          |\n",
      "+----+-------------------+---------------------+\n",
      "\n",
      " data/mimTitles.txt  Dataframe (Before processing):\n",
      "\n",
      "+----+----------------------+----------------------------------------+\n",
      "|    | MIM Number           | name                                   |\n",
      "|----+----------------------+----------------------------------------|\n",
      "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
      "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
      "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
      "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
      "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
      "+----+----------------------+----------------------------------------+\n",
      "\n",
      " data/mimTitles.txt  Dataframe (After processing):\n",
      "\n",
      "+----+----------------------+----------------------------------------+\n",
      "|    | MIM Number           | name                                   |\n",
      "|----+----------------------+----------------------------------------|\n",
      "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
      "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
      "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
      "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
      "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
      "+----+----------------------+----------------------------------------+\n",
      "\n",
      " MeSH Dictionary:\n",
      "\n",
      "+---------------+------------------------+\n",
      "|               | 0                      |\n",
      "|---------------+------------------------|\n",
      "| MESH::D000001 | Calcimycin             |\n",
      "| MESH::D000002 | Temefos                |\n",
      "| MESH::D000003 | Abattoirs              |\n",
      "| MESH::D000004 | Abbreviations as Topic |\n",
      "| MESH::D000005 | Abdomen                |\n",
      "+---------------+------------------------+\n"
     ]
    }
   ],
   "source": [
    "# 4) Load Data frames for translation\n",
    "hetionet_df, gene_df, drugbank_df, omim_df, mesh_dict, chebi_df, chembl_df = load_lookups(data_path=DATA_DIR,verbose=verbose)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Long-form orphan disease data (before processing):\n",
      "\n",
      "+----+-----------+------------------------------------------------------------------------+\n",
      "|    | cols      | data                                                                   |\n",
      "|----+-----------+------------------------------------------------------------------------|\n",
      "|  0 | Orphacode | 166024                                                                 |\n",
      "|  1 | Name      | Multiple epiphyseal dysplasia, Al-Gazali type                          |\n",
      "|  2 | Synonym   | Multiple epiphyseal dysplasia-macrocephaly-distinctive facies syndrome |\n",
      "|  3 | Source    | ICD-10                                                                 |\n",
      "|  4 | Reference | Q77.3                                                                  |\n",
      "+----+-----------+------------------------------------------------------------------------+\n",
      "\n",
      " Long-form orphan disease data (after processing):\n",
      "\n",
      "+----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------+\n",
      "|    | cols                    | data                                                                   |   disease_id | code_source   | code   |\n",
      "|----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------|\n",
      "|  0 | Orphacode               | 166024                                                                 |       166024 |               |        |\n",
      "|  1 | Name                    | Multiple epiphyseal dysplasia, Al-Gazali type                          |       166024 |               |        |\n",
      "|  2 | Synonym                 | Multiple epiphyseal dysplasia-macrocephaly-distinctive facies syndrome |       166024 |               |        |\n",
      "|  4 | Reference               | Q77.3                                                                  |       166024 | ICD-10        | Q77.3  |\n",
      "|  6 | Disordermappingrelation | NTBT (ORPHA code's Narrower Term maps to a Broader Term)               |       166024 | ICD-10        | Q77.3  |\n",
      "+----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------+\n",
      "\n",
      " Orphan disease name/summary data:\n",
      "\n",
      "+--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+\n",
      "|   disease_id |   Orphacode | Name                                                   | Synonym                                                    | Definition                                                                                                                                                                                                                                                                                                                                                                                             |   Info |\n",
      "|--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------|\n",
      "|           10 |          10 | 48,XXYY syndrome                                       | nan                                                        | A rare sex chromosome number anomaly disorder characterized, genetically, by the presence of an extra X and Y chromosome in males and, clinically, by tall stature, dysfunctional testes associated with infertility and insufficient testosterone production, cognitive, affective and social functioning impairments, global developmental delay, and an increased risk of congenital malformations. |    nan |\n",
      "|          100 |         100 | Ataxia-telangiectasia                                  | Louis-Bar syndrome                                         | A rare disorder characterized by the association of severe combined immunodeficiency (affecting mainly the humoral immune response) with progressive cerebellar ataxia. It is characterized by neurological signs, telangiectasia, increased susceptibility to infections and a higher risk of cancer.                                                                                                 |    nan |\n",
      "|         1000 |        1000 | Ocular albinism with late-onset sensorineural deafness | Ocular albinism with late-onset sensorineural hearing loss | Ocular albinism with late-onset sensorineural deafness is a rare, X-linked inherited subtype of ocular albinism characterized by severe visual impairment, translucent pale-blue irises, a reduction in the retinal pigment and moderately severe deafness with onset ranging from adolescence to fourth or fifth decade of life.                                                                      |    nan |\n",
      "|       100000 |      100000 | Reticular perineurioma                                 | nan                                                        | nan                                                                                                                                                                                                                                                                                                                                                                                                    |    nan |\n",
      "|       100001 |      100001 | Sclerosing perineurioma                                | nan                                                        | nan                                                                                                                                                                                                                                                                                                                                                                                                    |    nan |\n",
      "+--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+\n",
      "\n",
      " Orphan disease codes:\n",
      "\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+\n",
      "|    |   Orphacode | Name                                          | code_source   | code     | Disordermappingrelation                                  | Disordermappingicdrelation                          | Disordermappingvalidationstatus   |\n",
      "|----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------|\n",
      "|  0 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | ICD-10        | Q77.3    | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         |\n",
      "|  1 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | OMIM          | 607131   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  2 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | UMLS          | C1846722 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  3 |          58 | Alexander disease                             | OMIM          | 203450   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  4 |          58 | Alexander disease                             | MeSH          | D038261  | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+\n"
     ]
    }
   ],
   "source": [
    "# Load orphan disease names and codes (28 Nov 2022 version)\n",
    "orphan_names, orphan_codes = get_orphan_data(os.path.join(DATA_DIR, 'en_product1-Orphadata.xml'), verbose=verbose)\n",
    "\n",
    "# Get orphan disease MeSH codes\n",
    "orphan_codes_mesh = orphan_codes[orphan_codes['code_source']=='MeSH'].copy()\n",
    "orphan_codes_mesh['id'] = 'MESH::'+orphan_codes_mesh['code']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HoImejcxpFYP",
    "outputId": "904e49c4-8889-4f2d-fcce-26dc99331b86"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+------------------------------+--------------------------------+------------------------------------------------------+\n",
      "|    | head_entity                  | relationshipId                 | tail_entity                                          |\n",
      "|----+------------------------------+--------------------------------+------------------------------------------------------|\n",
      "|  0 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | coagulation factor VIII (F8)                         |\n",
      "|  1 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | phytanoyl-CoA 2-hydroxylase (PHYH)                   |\n",
      "|  2 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | coagulation factor IX (F9)                           |\n",
      "|  3 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | heat shock protein family A (Hsp70) member 5 (HSPA5) |\n",
      "|  4 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | immunoglobulin kappa variable 3-20 (IGKV3-20)        |\n",
      "+----+------------------------------+--------------------------------+------------------------------------------------------+\n",
      "Number of unique DRKG entities:  97238\n",
      "Number of translated entities:  64384\n",
      "Number of untranslated entities:  32854\n",
      "Percentage of entities translated:  66.2 %\n",
      "Total DRKG relationships:  5874261\n",
      "Translated DRKG relationships:  5707119\n",
      "Percentage of relationships fully translated:  97.2 %\n"
     ]
    }
   ],
   "source": [
    "# Make dictionaries for codes\n",
    "code_df   = pd.concat([hetionet_df[['name', 'id']], \n",
    "                       gene_df.rename(columns = {\"description\":\"name\", \"GeneID\":\"id\"}),\n",
    "                       drugbank_df.rename(columns = {\"Common name\":\"name\", \"DrugBank ID\":\"id\"}),\n",
    "                       omim_df.rename(columns = {\"MIM Number\":\"id\"}),\n",
    "                       chebi_df.rename(columns = {\"NAME\":\"name\", \"CHEBI_ACCESSION\":\"id\"}),\n",
    "                       chembl_df.rename(columns = {\"pref_name\":\"name\", \"chembl_id\":\"id\"}),\n",
    "                       orphan_codes_mesh.rename(columns = {\"Name\":\"name\"})\n",
    "                       ], ignore_index=True, axis=0).drop_duplicates() \n",
    "code_dict = pd.Series(code_df['name'].values, index=code_df['id']).to_dict() | mesh_dict # Convert node df to dict and merge with MeSH dictionary\n",
    "\n",
    "# Get unique DRKG entities\n",
    "drkg_entities = get_unique_entities(drkg_df, [0,2])\n",
    "\n",
    "# Create and use convert_entitynames function\n",
    "drkg_entity_df, drkg_unmatched = get_entity_lookup(drkg_entities, code_dict)\n",
    "\n",
    "# Create final node dictionary\n",
    "node_dict = pd.Series(drkg_entity_df['name'].values, index=drkg_entity_df['drkg_id']).to_dict() \n",
    "\n",
    "# Initialize translated DRKG and manually clean heads/tails for one case where they were flipped\n",
    "drkg_translated    = drkg_df.copy()\n",
    "drkg_translated = flip_headtail(drkg_translated, 'Gene:Compound')\n",
    "\n",
    "# Map DRKG to translated entity names\n",
    "drkg_translated = convert_entitynames(drkg_translated, 0, node_dict)\n",
    "drkg_translated = convert_entitynames(drkg_translated, 2, node_dict)\n",
    "drkg_translated = drkg_translated.dropna()\n",
    "drkg_translated.columns = [\"head_entity\", \"relationshipId\", \"tail_entity\"]\n",
    "print_head(drkg_translated) \n",
    "\n",
    "# Summarize percentage translated\n",
    "print(\"Number of unique DRKG entities: \", len(drkg_entities)) # should be 97238\n",
    "print(\"Number of translated entities: \", drkg_entity_df.shape[0])\n",
    "print(\"Number of untranslated entities: \", drkg_unmatched.shape[0])\n",
    "pct_entity_translated = drkg_entity_df.shape[0]/len(drkg_entities)\n",
    "print('Percentage of entities translated: ', round(pct_entity_translated*100,1), '%')\n",
    "\n",
    "print('Total DRKG relationships: ', drkg_df.shape[0])\n",
    "print('Translated DRKG relationships: ', drkg_translated.shape[0])\n",
    "pct_translated = drkg_translated.shape[0]/drkg_df.shape[0]\n",
    "print('Percentage of relationships fully translated: ', round(pct_translated*100,1), '%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drkg_translated.to_csv(os.path.join(DATA_DIR, \"drkg_translated.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>Data-source</th>\n",
       "      <th>Connected entity-types</th>\n",
       "      <th>Interaction-type</th>\n",
       "      <th>Description</th>\n",
       "      <th>Reference for the description</th>\n",
       "      <th>head_entity</th>\n",
       "      <th>tail_entity</th>\n",
       "      <th>relation_name</th>\n",
       "      <th>relation_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DGIDB::ACTIVATOR::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>activation</td>\n",
       "      <td>An activator interaction is when a drug activa...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>activation</td>\n",
       "      <td>activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DGIDB::AGONIST::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>agonism</td>\n",
       "      <td>An agonist interaction occurs when a drug bind...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>activation</td>\n",
       "      <td>activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DGIDB::ALLOSTERIC MODULATOR::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>allosteric modulation</td>\n",
       "      <td>An allosteric modulator interaction occurs whe...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>allosteric modulation</td>\n",
       "      <td>allosteric_modulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DGIDB::ANTAGONIST::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>antagonism</td>\n",
       "      <td>An antagonist interaction occurs when a drug b...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>antagonism</td>\n",
       "      <td>antagonism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DGIDB::ANTIBODY::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>antibody</td>\n",
       "      <td>An antibody interaction occurs when an antibod...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>antibody</td>\n",
       "      <td>antibody</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>bioarx::Covid2_acc_host_gene::Disease:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Disease:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Interactions between 27 viral  proteins,  and ...</td>\n",
       "      <td></td>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>bioarx::DrugHumGen:Compound:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>bioarx::DrugVirGen:Compound:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>bioarx::HumGenHumGen:Gene:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Gene:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Protein-protein interaction</td>\n",
       "      <td></td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>bioarx::VirGenHumGen:Gene:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Gene:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Protein-protein interaction</td>\n",
       "      <td></td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        drkg_id   Data-source   \n",
       "0               DGIDB::ACTIVATOR::Gene:Compound         DGIDB  \\\n",
       "1                 DGIDB::AGONIST::Gene:Compound         DGIDB   \n",
       "2    DGIDB::ALLOSTERIC MODULATOR::Gene:Compound         DGIDB   \n",
       "3              DGIDB::ANTAGONIST::Gene:Compound         DGIDB   \n",
       "4                DGIDB::ANTIBODY::Gene:Compound         DGIDB   \n",
       "..                                          ...           ...   \n",
       "102  bioarx::Covid2_acc_host_gene::Disease:Gene  BIBLIOGRAPHY   \n",
       "103            bioarx::DrugHumGen:Compound:Gene  BIBLIOGRAPHY   \n",
       "104            bioarx::DrugVirGen:Compound:Gene  BIBLIOGRAPHY   \n",
       "105              bioarx::HumGenHumGen:Gene:Gene  BIBLIOGRAPHY   \n",
       "106              bioarx::VirGenHumGen:Gene:Gene  BIBLIOGRAPHY   \n",
       "\n",
       "    Connected entity-types       Interaction-type   \n",
       "0            Compound:Gene             activation  \\\n",
       "1            Compound:Gene                agonism   \n",
       "2            Compound:Gene  allosteric modulation   \n",
       "3            Compound:Gene             antagonism   \n",
       "4            Compound:Gene               antibody   \n",
       "..                     ...                    ...   \n",
       "102           Disease:Gene            interaction   \n",
       "103          Compound:Gene            interaction   \n",
       "104          Compound:Gene            interaction   \n",
       "105              Gene:Gene            interaction   \n",
       "106              Gene:Gene            interaction   \n",
       "\n",
       "                                           Description   \n",
       "0    An activator interaction is when a drug activa...  \\\n",
       "1    An agonist interaction occurs when a drug bind...   \n",
       "2    An allosteric modulator interaction occurs whe...   \n",
       "3    An antagonist interaction occurs when a drug b...   \n",
       "4    An antibody interaction occurs when an antibod...   \n",
       "..                                                 ...   \n",
       "102  Interactions between 27 viral  proteins,  and ...   \n",
       "103                                                      \n",
       "104                                                      \n",
       "105                        Protein-protein interaction   \n",
       "106                        Protein-protein interaction   \n",
       "\n",
       "            Reference for the description head_entity tail_entity   \n",
       "0    http://www.dgidb.org/getting_started    Compound        Gene  \\\n",
       "1    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "2    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "3    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "4    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "..                                    ...         ...         ...   \n",
       "102                                           Disease        Gene   \n",
       "103                                          Compound        Gene   \n",
       "104                                          Compound        Gene   \n",
       "105                                              Gene        Gene   \n",
       "106                                              Gene        Gene   \n",
       "\n",
       "             relation_name         relation_label  \n",
       "0               activation             activation  \n",
       "1               activation             activation  \n",
       "2    allosteric modulation  allosteric_modulation  \n",
       "3               antagonism             antagonism  \n",
       "4                 antibody               antibody  \n",
       "..                     ...                    ...  \n",
       "102            interaction            interaction  \n",
       "103            interaction            interaction  \n",
       "104            interaction            interaction  \n",
       "105            interaction            interaction  \n",
       "106            interaction            interaction  \n",
       "\n",
       "[107 rows x 10 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Update relation glossary \n",
    "relation_df = relation_glossary.copy().rename(columns={'Relation-name':'drkg_id'})\n",
    "relation_df[['head_entity','tail_entity']] = relation_df['drkg_id'].str.split('::', expand=True)[2].str.split(':', expand=True) # Set head and tail nodes\n",
    "\n",
    "# Manually fix head and tail nodes for DGIDB relations, which reverse compound-gene interactions\n",
    "relation_df.loc[relation_df['drkg_id'].str.contains('Gene:Compound'),'head_entity'] = 'Compound'\n",
    "relation_df.loc[relation_df['drkg_id'].str.contains('Gene:Compound'),'tail_entity'] = 'Gene'\n",
    "\n",
    "# Fix bioarx entries without the second \"::\" delimiter\n",
    "bioarx_ht = relation_df['drkg_id'].str.split(':', expand=True)[[3,4]]\n",
    "relation_df['head_entity'] = np.where(relation_df['head_entity'].isna(), bioarx_ht[3], relation_df['head_entity'])\n",
    "relation_df['tail_entity'] = np.where(relation_df['tail_entity'].isna(), bioarx_ht[4], relation_df['tail_entity'])\n",
    "\n",
    "# Add mapped relation group labels\n",
    "relation_groups = [['activation', 'agonism', 'agonism, activation', 'activates, stimulates'],\n",
    "    ['antagonism', 'blocking', 'antagonism, blocking'],\n",
    "    ['binding', 'binding, ligand (esp. receptors)'],\n",
    "    ['blocking', 'channel blocking'],\n",
    "    ['inhibition', 'inhibits cell growth (esp. cancers)', 'inhibits'],\n",
    "    ['enzyme', 'enzyme activity'],\n",
    "    ['upregulation', 'increases expression/production'],\n",
    "    ['downregulation', 'decreases expression/production'],\n",
    "    ['Compound treats the disease', 'treatment/therapy (including investigatory)', 'treatment']]\n",
    "\n",
    "relation_df['relation_name'] = relation_df['Interaction-type']\n",
    "\n",
    "for grp in relation_groups:\n",
    "    relation_df_subset = relation_df[relation_df['Interaction-type'].isin(grp)].copy()\n",
    "    for entities in relation_df_subset['Connected entity-types'].unique():\n",
    "        subgrp = relation_df_subset[relation_df_subset['Connected entity-types'] == entities]['Interaction-type'].unique()\n",
    "        relation_df.loc[(relation_df_subset['Connected entity-types'] == entities) & (relation_df['Interaction-type'].isin(subgrp)), 'relation_name'] = subgrp[0]\n",
    "\n",
    "# Remove special characters from relation names (for embedding) and create relation label with underscores instead of spaces (for edge name in heterograph)\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace(',|/', ' or', regex=True)\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace('esp.','especially')\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace('\\(|\\)|-|\\.', '', regex=True)\n",
    "relation_df['relation_label'] = relation_df['relation_name'].str.replace(' ', '_')\n",
    "\n",
    "# Check if any relationshp names still have non alpha numeric values except space\n",
    "error_relation_names = relation_df['relation_name'][relation_df['relation_name'].str.replace(' ', '').str.contains(r\"[^a-zA-Z0-9]+\", regex=True)].drop_duplicates()\n",
    "if len(error_relation_names):\n",
    "    print('Warning: The following relation names contain special characters, which can interfere with PyG/GraphSage')\n",
    "    print(error_relation_names)\n",
    "    \n",
    "relation_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+------------------------------+------------------------------------------------------+--------------------------------+------------------+--------------------+--------------------+\n",
      "|    | head_entity                  | tail_entity                                          | drkg_id                        | relation_label   | head_entity_type   | tail_entity_type   |\n",
      "|----+------------------------------+------------------------------------------------------+--------------------------------+------------------+--------------------+--------------------|\n",
      "|  0 | coagulation factor VIII (F8) | coagulation factor VIII (F8)                         | bioarx::HumGenHumGen:Gene:Gene | INTERACTION      | GENE               | GENE               |\n",
      "|  1 | coagulation factor VIII (F8) | phytanoyl-CoA 2-hydroxylase (PHYH)                   | bioarx::HumGenHumGen:Gene:Gene | INTERACTION      | GENE               | GENE               |\n",
      "|  2 | coagulation factor VIII (F8) | coagulation factor IX (F9)                           | bioarx::HumGenHumGen:Gene:Gene | INTERACTION      | GENE               | GENE               |\n",
      "|  3 | coagulation factor VIII (F8) | heat shock protein family A (Hsp70) member 5 (HSPA5) | bioarx::HumGenHumGen:Gene:Gene | INTERACTION      | GENE               | GENE               |\n",
      "|  4 | coagulation factor VIII (F8) | immunoglobulin kappa variable 3-20 (IGKV3-20)        | bioarx::HumGenHumGen:Gene:Gene | INTERACTION      | GENE               | GENE               |\n",
      "+----+------------------------------+------------------------------------------------------+--------------------------------+------------------+--------------------+--------------------+\n"
     ]
    }
   ],
   "source": [
    "relation_data_to_merge = relation_df[\n",
    "    [\"drkg_id\", \"relation_label\", \"head_entity\", \"tail_entity\"]\n",
    "]\n",
    "relation_data_to_merge = relation_data_to_merge.rename(\n",
    "    columns={\"head_entity\": \"head_entity_type\", \"tail_entity\": \"tail_entity_type\"}\n",
    ")\n",
    "for col_name in [\"relation_label\", \"head_entity_type\", \"tail_entity_type\"]:\n",
    "    relation_data_to_merge[col_name] = relation_data_to_merge[col_name].str.upper()\n",
    "drkg_translated_with_relation_labels = drkg_translated.merge(\n",
    "    relation_data_to_merge,\n",
    "    left_on=\"relationshipId\",\n",
    "    right_on=\"drkg_id\",\n",
    "    how=\"left\",\n",
    ")\n",
    "drkg_translated_with_relation_labels.drop([\"relationshipId\"], axis=1, inplace=True)\n",
    "print_head(drkg_translated_with_relation_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "drkg_translated_with_relation_labels.to_csv(os.path.join(DATA_DIR, \"drkg_translated_with_relation_labels.csv\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "drkg_entity_df.to_pickle(os.path.join(DATA_DIR, 'drkg_entity_df.pkl'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save h-r-t triples\n",
    "triplet_cats = relation_df[relation_df['drkg_id'].isin(drkg_translated['tail_entity'])].copy()\n",
    "triplet_cats = triplet_cats[['head_entity', 'relation_label', 'tail_entity']].drop_duplicates()\n",
    "# triplet_cats.to_csv(\"triplets.csv\")\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "oTAJOqDVc_Hg"
   },
   "source": [
    "# 2) BioLinkBERT embedding\n",
    "\n",
    "AUTHORS: Alejandro and Selina (equal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import HeteroData\n",
    "from SIMP_LLM.llm_encode import EntityEncoder\n",
    "from SIMP_LLM.dataloader_mappings import create_mapping, create_edges, embed_entities, embed_edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set variables and load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "id": "QkZKp6CJiw6f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-----------------------------+-----------------------------+\n",
      "|    | drkg_id                            | Data-source   | Connected entity-types   | Interaction-type            | Description                                                                                                                                                                                            | Reference for the description        | head_entity   | tail_entity   | relation_name               | relation_label              |\n",
      "|----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-----------------------------+-----------------------------|\n",
      "|  0 | DGIDB::ACTIVATOR::Gene:Compound    | DGIDB         | Compound:Gene            | activation                  | An activator interaction is when a drug activates a biological response from a target, although the mechanism by which it does so may not be understood.                                               | http://www.dgidb.org/getting_started | Compound      | Gene          | activation                  | activation                  |\n",
      "|  1 | DGIDB::AGONIST::Gene:Compound      | DGIDB         | Compound:Gene            | agonism                     | An agonist interaction occurs when a drug binds to a target receptor and activates the receptor to produce a biological response.                                                                      | http://www.dgidb.org/getting_started | Compound      | Gene          | activation                  | activation                  |\n",
      "|  8 | DGIDB::INHIBITOR::Gene:Compound    | DGIDB         | Compound:Gene            | inhibition                  | In inhibitor interactions, the drug binds to a target and decreases its expression or activity. Most interactions of this class are enzyme inhibitors, which bind an enzyme to reduce enzyme activity. | http://www.dgidb.org/getting_started | Compound      | Gene          | inhibition                  | inhibition                  |\n",
      "| 18 | DRUGBANK::treats::Compound:Disease | DRUGBANK      | Compound:Disease         | Compound treats the disease | We extract  the  associated  conditions  related  to  compound entities in the drugbank.                                                                                                               |                                      | Compound      | Disease       | Compound treats the disease | Compound_treats_the_disease |\n",
      "| 19 | GNBR::A+::Compound:Gene            | GNBR          | Compound:Gene            | agonism, activation         |                                                                                                                                                                                                        |                                      | Compound      | Gene          | activation                  | activation                  |\n",
      "+----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-----------------------------+-----------------------------+\n",
      "+----+---------------+---------------------------------+---------------------------------------------------------+\n",
      "|    | head_entity   | relationshipId                  | tail_entity                                             |\n",
      "|----+---------------+---------------------------------+---------------------------------------------------------|\n",
      "|  0 | SNS-032       | DGIDB::INHIBITOR::Gene:Compound | cyclin dependent kinase 7 (CDK7)                        |\n",
      "|  1 | Quizartinib   | DGIDB::INHIBITOR::Gene:Compound | platelet derived growth factor receptor beta (PDGFRB)   |\n",
      "|  2 | Sufentanil    | DGIDB::AGONIST::Gene:Compound   | opioid receptor mu 1 (OPRM1)                            |\n",
      "|  3 | PF-04691502   | DGIDB::INHIBITOR::Gene:Compound | phosphoinositide-3-kinase regulatory subunit 2 (PIK3R2) |\n",
      "|  4 | Erdafitinib   | DGIDB::INHIBITOR::Gene:Compound | fibroblast growth factor receptor 1 (FGFR1)             |\n",
      "+----+---------------+---------------------------------+---------------------------------------------------------+\n",
      "+-------+-------------------------+-------------------------+-------------------------+---------------+-----------------+-----------------+---------------+\n",
      "|       | drkg_id                 | drkg_dict_id            | name                    | entity_type   | ontology_code   | ontology_name   | code          |\n",
      "|-------+-------------------------+-------------------------+-------------------------+---------------+-----------------+-----------------+---------------|\n",
      "| 19899 | Compound::CHEMBL2108557 | Compound::CHEMBL2108557 | TELBERMIN               | Compound      | CHEMBL2108557   | CHEMBL          | CHEMBL2108557 |\n",
      "| 21697 | Compound::CHEMBL551064  | Compound::CHEMBL551064  | AEW-541                 | Compound      | CHEMBL551064    | CHEMBL          | CHEMBL551064  |\n",
      "| 21962 | Compound::DB00003       | Compound::DB00003       | Dornase alfa            | Compound      | DB00003         | drugbank        | DB00003       |\n",
      "| 21964 | Compound::DB00005       | Compound::DB00005       | Etanercept              | Compound      | DB00005         | drugbank        | DB00005       |\n",
      "| 22049 | Compound::DB00097       | Compound::DB00097       | Choriogonadotropin alfa | Compound      | DB00097         | drugbank        | DB00097       |\n",
      "+-------+-------------------------+-------------------------+-------------------------+---------------+-----------------+-----------------+---------------+\n"
     ]
    }
   ],
   "source": [
    "## Set variables\n",
    "device   = \"cpu\"\n",
    "Encoder  = EntityEncoder(device = device )\n",
    "run_full_sample = 0\n",
    "Sample = 5\n",
    "\n",
    "if run_full_sample:\n",
    "    # Run full DRKG\n",
    "    entity_df = drkg_entity_df.copy()\n",
    "    hrt_data = drkg_translated.copy()\n",
    "    relation_lookup = relation_df.copy()\n",
    "else:\n",
    "    # Change interaction list inputs as needed\n",
    "    interaction_list = ['activation', 'agonism', 'agonism, activation', 'Compound treats the disease', 'treats', 'inhibition'] \n",
    "\n",
    "    # Create relationship subset for testing based on interaction and umbrella relation names from DRKG relations\n",
    "    relation_name_list = relation_df.loc[relation_df['Interaction-type'].isin(interaction_list), 'relation_name']\n",
    "    test_relation_df = relation_df[relation_df['relation_name'].isin(relation_name_list)].copy()\n",
    "    relation_entity_list = pd.concat([relation_df['head_entity'], relation_df['tail_entity']], ignore_index=True, axis=0).unique()\n",
    "    print_head(test_relation_df)\n",
    "\n",
    "    # Create test sample of DRKG relationships filtering to these relations\n",
    "    test_hrt_df = drkg_translated[drkg_translated['relationshipId'].isin(test_relation_df['drkg_id'])]\n",
    "    test_hrt_df = test_hrt_df.groupby('relationshipId').head(Sample).reset_index(drop=True)\n",
    "\n",
    "    # Get dataframe of entities, careful to only use types represented in test_relation_df\n",
    "    test_unique_entities = get_unique_entities(test_hrt_df, columns=['head_entity','tail_entity'])\n",
    "    test_entity_df = drkg_entity_df[drkg_entity_df['name'].isin(test_unique_entities)]\n",
    "    test_entity_df = test_entity_df[test_entity_df['entity_type'].isin(relation_entity_list)]\n",
    "    print_head(test_hrt_df)\n",
    "    print_head(test_entity_df)\n",
    "\n",
    "\n",
    "    entity_df = test_entity_df.copy()\n",
    "    hrt_data = test_hrt_df.copy()\n",
    "    relation_lookup = relation_df.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>drkg_dict_id</th>\n",
       "      <th>name</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_code</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>43514</th>\n",
       "      <td>Disease::MESH:D003550</td>\n",
       "      <td>MESH::D003550</td>\n",
       "      <td>Cystic Fibrosis</td>\n",
       "      <td>Disease</td>\n",
       "      <td>MESH:D003550</td>\n",
       "      <td>MESH</td>\n",
       "      <td>D003550</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     drkg_id   drkg_dict_id             name entity_type   \n",
       "43514  Disease::MESH:D003550  MESH::D003550  Cystic Fibrosis     Disease  \\\n",
       "\n",
       "      ontology_code ontology_name     code  \n",
       "43514  MESH:D003550          MESH  D003550  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hrt_data[hrt_data['relationshipId'].str.contains('Disease')]\n",
    "entity_df[entity_df['name']=='Cystic Fibrosis']"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "ywHcEVMiAdOv"
   },
   "source": [
    "### Build HeteroData Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Parent directory data2/ckpts/entity does not exist.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 5\u001b[0m\n\u001b[1;32m      2\u001b[0m data \u001b[39m=\u001b[39m HeteroData()\n\u001b[1;32m      4\u001b[0m \u001b[39m# Embed entities, add to graph, and save embedding mapping dictionary of dictionaries\u001b[39;00m\n\u001b[0;32m----> 5\u001b[0m mapping_dict \u001b[39m=\u001b[39m embed_entities(entity_df, data, Encoder, device) \n\u001b[1;32m      7\u001b[0m \u001b[39m# Embed relationships, add to graph, and save relation embeddings/mapping dictionary\u001b[39;00m\n\u001b[1;32m      8\u001b[0m relation_X, relation_mapping \u001b[39m=\u001b[39m embed_edges(hrt_data, relation_lookup, data, mapping_dict, Encoder, device)\n",
      "File \u001b[0;32m~/SIMPLLM/SIMP_LLM/dataloader_mappings.py:98\u001b[0m, in \u001b[0;36membed_entities\u001b[0;34m(entity_df, graph_obj, encoder, device)\u001b[0m\n\u001b[1;32m     96\u001b[0m     entity_X, entity_mapping \u001b[39m=\u001b[39m create_mapping(entity_names, encoder\u001b[39m=\u001b[39mencoder, device\u001b[39m=\u001b[39mdevice) \u001b[39m# Maps entities to indices\u001b[39;00m\n\u001b[1;32m     97\u001b[0m     graph_obj[entity]\u001b[39m.\u001b[39mx \u001b[39m=\u001b[39m entity_X                                                          \u001b[39m# Assign entity type embeddings to graph object\u001b[39;00m\n\u001b[0;32m---> 98\u001b[0m     torch\u001b[39m.\u001b[39;49msave(entity_X, \u001b[39mf\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mdata2/ckpts/entity/\u001b[39;49m\u001b[39m{\u001b[39;49;00mentity\u001b[39m}\u001b[39;49;00m\u001b[39m.pt\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n\u001b[1;32m    100\u001b[0m     mapping_dict[entity] \u001b[39m=\u001b[39m entity_mapping                                                   \u001b[39m# Add entity type mapping to overall mapping dictionary\u001b[39;00m\n\u001b[1;32m    102\u001b[0m \u001b[39mreturn\u001b[39;00m mapping_dict\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch/serialization.py:440\u001b[0m, in \u001b[0;36msave\u001b[0;34m(obj, f, pickle_module, pickle_protocol, _use_new_zipfile_serialization)\u001b[0m\n\u001b[1;32m    437\u001b[0m _check_save_filelike(f)\n\u001b[1;32m    439\u001b[0m \u001b[39mif\u001b[39;00m _use_new_zipfile_serialization:\n\u001b[0;32m--> 440\u001b[0m     \u001b[39mwith\u001b[39;00m _open_zipfile_writer(f) \u001b[39mas\u001b[39;00m opened_zipfile:\n\u001b[1;32m    441\u001b[0m         _save(obj, opened_zipfile, pickle_module, pickle_protocol)\n\u001b[1;32m    442\u001b[0m         \u001b[39mreturn\u001b[39;00m\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch/serialization.py:315\u001b[0m, in \u001b[0;36m_open_zipfile_writer\u001b[0;34m(name_or_buffer)\u001b[0m\n\u001b[1;32m    313\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    314\u001b[0m     container \u001b[39m=\u001b[39m _open_zipfile_writer_buffer\n\u001b[0;32m--> 315\u001b[0m \u001b[39mreturn\u001b[39;00m container(name_or_buffer)\n",
      "File \u001b[0;32m~/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/torch/serialization.py:288\u001b[0m, in \u001b[0;36m_open_zipfile_writer_file.__init__\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m    287\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, name) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m--> 288\u001b[0m     \u001b[39msuper\u001b[39m()\u001b[39m.\u001b[39m\u001b[39m__init__\u001b[39m(torch\u001b[39m.\u001b[39;49m_C\u001b[39m.\u001b[39;49mPyTorchFileWriter(\u001b[39mstr\u001b[39;49m(name)))\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Parent directory data2/ckpts/entity does not exist."
     ]
    }
   ],
   "source": [
    "# Initialize heterograph object\n",
    "data = HeteroData()\n",
    "\n",
    "# Embed entities, add to graph, and save embedding mapping dictionary of dictionaries\n",
    "mapping_dict = embed_entities(entity_df, data, Encoder, device) \n",
    "\n",
    "# Embed relationships, add to graph, and save relation embeddings/mapping dictionary\n",
    "relation_X, relation_mapping = embed_edges(hrt_data, relation_lookup, data, mapping_dict, Encoder, device)\n",
    "\n",
    "# Print summary\n",
    "#data = T.ToUndirected()(data)\n",
    "\n",
    "print(data)\n",
    "for ent_type in entity_df['entity_type'].unique():\n",
    "    print(f\"Unique {ent_type}s: {len(mapping_dict[ent_type])} \\t Matrix shape: {data[ent_type].x.shape }\")\n",
    "    # print(mapping_dict[ent_type]) # Prints whole dictionary so delete/uncomment if using all entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HeteroData(\n",
       "  \u001b[1mCompound\u001b[0m={ x=[29, 768] },\n",
       "  \u001b[1mDisease\u001b[0m={ x=[14, 768] },\n",
       "  \u001b[1mGene\u001b[0m={ x=[51, 768] },\n",
       "  \u001b[1m(Compound, activation, Gene)\u001b[0m={\n",
       "    edge_index=[2, 15],\n",
       "    edge_label=[15, 768]\n",
       "  },\n",
       "  \u001b[1m(Compound, inhibition, Gene)\u001b[0m={\n",
       "    edge_index=[2, 10],\n",
       "    edge_label=[10, 768]\n",
       "  },\n",
       "  \u001b[1m(Compound, Compound_treats_the_disease, Disease)\u001b[0m={\n",
       "    edge_index=[2, 15],\n",
       "    edge_label=[15, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, activates_or_stimulates, Gene)\u001b[0m={\n",
       "    edge_index=[2, 20],\n",
       "    edge_label=[20, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, inhibition, Gene)\u001b[0m={\n",
       "    edge_index=[2, 10],\n",
       "    edge_label=[10, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, rev_activation, Compound)\u001b[0m={\n",
       "    edge_index=[2, 15],\n",
       "    edge_label=[15, 768]\n",
       "  },\n",
       "  \u001b[1m(Gene, rev_inhibition, Compound)\u001b[0m={\n",
       "    edge_index=[2, 10],\n",
       "    edge_label=[10, 768]\n",
       "  },\n",
       "  \u001b[1m(Disease, rev_Compound_treats_the_disease, Compound)\u001b[0m={\n",
       "    edge_index=[2, 15],\n",
       "    edge_label=[15, 768]\n",
       "  }\n",
       ")"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = T.ToUndirected()(data)\n",
    "data\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3) GRAPH SAGE\n",
    "AUTHORS: Alejandro"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch_geometric as pyg\n",
    "import torch_geometric\n",
    "from torch_geometric.nn import SAGEConv, to_hetero\n",
    "from   torch.utils.data      import Dataset, DataLoader\n",
    "from   torch_geometric.data  import Data\n",
    "from   torch_geometric.utils import negative_sampling\n",
    "\n",
    "from torch_geometric.nn import SAGEConv, to_hetero\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class GNNStack(torch.nn.Module):\n",
    "    def __init__(self, input_dim:int, hidden_dim:int, output_dim:int, layers:int, dropout:float=0.3, return_embedding=False):\n",
    "        \"\"\"\n",
    "            A stack of GraphSAGE Module \n",
    "            input_dim        <int>:   Input dimension\n",
    "            hidden_dim       <int>:   Hidden dimension\n",
    "            output_dim       <int>:   Output dimension\n",
    "            layers           <int>:   Number of layers\n",
    "            dropout          <float>: Dropout rate\n",
    "            return_embedding <bool>:  Whether to return the return_embeddingedding of the input graph\n",
    "        \"\"\"\n",
    "        \n",
    "        super(GNNStack, self).__init__()\n",
    "        graphSage_conv               = pyg.nn.SAGEConv\n",
    "        self.dropout                 = dropout\n",
    "        self.layers                  = layers\n",
    "        self.return_embedding        = return_embedding\n",
    "\n",
    "        ### Initalize the layers ###\n",
    "        self.convs                   = nn.ModuleList()                      # ModuleList to hold the layers\n",
    "        for l in range(self.layers):\n",
    "            if l == 0:\n",
    "                ### First layer  maps from input_dim to hidden_dim ###\n",
    "                self.convs.append(graphSage_conv(input_dim, hidden_dim))\n",
    "            else:\n",
    "                ### All other layers map from hidden_dim to hidden_dim ###\n",
    "                self.convs.append(graphSage_conv(hidden_dim, hidden_dim))\n",
    "\n",
    "        # post-message-passing processing MLP\n",
    "        self.post_mp = nn.Sequential(\n",
    "                                     nn.Linear(hidden_dim, hidden_dim), \n",
    "                                     nn.Dropout(self.dropout),\n",
    "                                     nn.Linear(hidden_dim, output_dim))\n",
    "\n",
    "    def forward(self, x, edge_index):\n",
    "        for i in range(self.layers):\n",
    "            x = self.convs[i](x, edge_index)\n",
    "            x = F.relu(x)\n",
    "            x = F.dropout(x, p=self.dropout, training=self.training)\n",
    "\n",
    "        x = self.post_mp(x)\n",
    "\n",
    "        # Return final layer of return_embeddingeddings if specified\n",
    "        if self.return_embedding:\n",
    "            return x\n",
    "\n",
    "        # Else return class probabilities\n",
    "        return F.log_softmax(x, dim=1)\n",
    "\n",
    "    def loss(self, pred, label):\n",
    "        return F.nll_loss(pred, label)\n",
    "    \n",
    "\n",
    "\n",
    "class LinkPredictorMLP(nn.Module):\n",
    "    def __init__(self, in_channels:int, hidden_channels:int, out_channels:int, n_layers:int,dropout_probabilty:float=0.3):\n",
    "        \"\"\"\n",
    "        Args:\n",
    "            in_channels (int):     Number of input features.\n",
    "            hidden_channels (int): Number of hidden features.\n",
    "            out_channels (int):    Number of output features.\n",
    "            n_layers (int):        Number of MLP layers.\n",
    "            dropout (float):       Dropout probability.\n",
    "            \"\"\"\n",
    "        super(LinkPredictorMLP, self).__init__()\n",
    "        self.dropout_probabilty    = dropout_probabilty  # dropout probability\n",
    "        self.mlp_layers            = nn.ModuleList()     # ModuleList: is a list of modules\n",
    "        self.non_linearity         = F.relu              # non-linearity\n",
    "        \n",
    "        for i in range(n_layers - 1):                                 \n",
    "            if i == 0:\n",
    "                self.mlp_layers.append(nn.Linear(in_channels, hidden_channels))          # input layer (in_channels, hidden_channels)\n",
    "            else:\n",
    "                self.mlp_layers.append(nn.Linear(hidden_channels, hidden_channels))      # hidden layers (hidden_channels, hidden_channels)\n",
    "\n",
    "        self.mlp_layers.append(nn.Linear(hidden_channels, out_channels))                 # output layer (hidden_channels, out_channels)\n",
    "\n",
    "\n",
    "    def reset_parameters(self):\n",
    "        for mlp_layer in self.mlp_layers:\n",
    "            mlp_layer.reset_parameters()\n",
    "\n",
    "    def forward(self, x_i, x_j):\n",
    "        x = x_i * x_j                                                     # element-wise multiplication\n",
    "        for mlp_layer in self.mlp_layers[:-1]:                            # iterate over all layers except the last one\n",
    "            x = mlp_layer(x)                                              # apply linear transformation\n",
    "            x = self.non_linearity(x)                                     # Apply non linear activation function\n",
    "            x = F.dropout(x, p=self.dropout_probabilty,training=self.training)      # Apply dropout\n",
    "        x = self.mlp_layers[-1](x)                                        # apply linear transformation to the last layer\n",
    "        x = torch.sigmoid(x)                                              # apply sigmoid activation function to get the probability\n",
    "        return x\n",
    "    \n",
    "### We will use This function to save our best model during trainnig ###\n",
    "def save_torch_model(model,epoch,PATH:str,optimizer):\n",
    "    print(f\"Saving Model in Path {PATH}\")\n",
    "    torch.save({'epoch': epoch,\n",
    "                'model_state_dict': model.state_dict(),\n",
    "                'optimizer':optimizer,      \n",
    "                }, PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GraphModule(\n",
      "  (convs): ModuleList(\n",
      "    (0): ModuleDict(\n",
      "      (Compound__activation__Gene): SAGEConv(768, 524, aggr=mean)\n",
      "      (Compound__inhibition__Gene): SAGEConv(768, 524, aggr=mean)\n",
      "      (Compound__Compound_treats_the_disease__Disease): SAGEConv(768, 524, aggr=mean)\n",
      "      (Gene__activates_or_stimulates__Gene): SAGEConv(768, 524, aggr=mean)\n",
      "      (Gene__inhibition__Gene): SAGEConv(768, 524, aggr=mean)\n",
      "      (Gene__rev_activation__Compound): SAGEConv(768, 524, aggr=mean)\n",
      "      (Gene__rev_inhibition__Compound): SAGEConv(768, 524, aggr=mean)\n",
      "      (Disease__rev_Compound_treats_the_disease__Compound): SAGEConv(768, 524, aggr=mean)\n",
      "    )\n",
      "    (1-2): 2 x ModuleDict(\n",
      "      (Compound__activation__Gene): SAGEConv(524, 524, aggr=mean)\n",
      "      (Compound__inhibition__Gene): SAGEConv(524, 524, aggr=mean)\n",
      "      (Compound__Compound_treats_the_disease__Disease): SAGEConv(524, 524, aggr=mean)\n",
      "      (Gene__activates_or_stimulates__Gene): SAGEConv(524, 524, aggr=mean)\n",
      "      (Gene__inhibition__Gene): SAGEConv(524, 524, aggr=mean)\n",
      "      (Gene__rev_activation__Compound): SAGEConv(524, 524, aggr=mean)\n",
      "      (Gene__rev_inhibition__Compound): SAGEConv(524, 524, aggr=mean)\n",
      "      (Disease__rev_Compound_treats_the_disease__Compound): SAGEConv(524, 524, aggr=mean)\n",
      "    )\n",
      "  )\n",
      "  (post_mp): ModuleList(\n",
      "    (0): ModuleDict(\n",
      "      (Compound): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Disease): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Gene): Linear(in_features=524, out_features=524, bias=True)\n",
      "    )\n",
      "    (1): ModuleDict(\n",
      "      (Compound): Dropout(p=0.7, inplace=False)\n",
      "      (Disease): Dropout(p=0.7, inplace=False)\n",
      "      (Gene): Dropout(p=0.7, inplace=False)\n",
      "    )\n",
      "    (2): ModuleDict(\n",
      "      (Compound): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Disease): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Gene): Linear(in_features=524, out_features=524, bias=True)\n",
      "    )\n",
      "  )\n",
      ")\n",
      "\n",
      "\n",
      "\n",
      "def forward(self, x, edge_index):\n",
      "    x_dict = torch_geometric_nn_to_hetero_transformer_get_dict(x);  x = None\n",
      "    x__Compound = x_dict.get('Compound', None)\n",
      "    x__Disease = x_dict.get('Disease', None)\n",
      "    x__Gene = x_dict.get('Gene', None);  x_dict = None\n",
      "    edge_index_dict = torch_geometric_nn_to_hetero_transformer_get_dict(edge_index);  edge_index = None\n",
      "    edge_index__Compound__activation__Gene = edge_index_dict.get(('Compound', 'activation', 'Gene'), None)\n",
      "    edge_index__Compound__inhibition__Gene = edge_index_dict.get(('Compound', 'inhibition', 'Gene'), None)\n",
      "    edge_index__Compound__Compound_treats_the_disease__Disease = edge_index_dict.get(('Compound', 'Compound_treats_the_disease', 'Disease'), None)\n",
      "    edge_index__Gene__activates_or_stimulates__Gene = edge_index_dict.get(('Gene', 'activates_or_stimulates', 'Gene'), None)\n",
      "    edge_index__Gene__inhibition__Gene = edge_index_dict.get(('Gene', 'inhibition', 'Gene'), None)\n",
      "    edge_index__Gene__rev_activation__Compound = edge_index_dict.get(('Gene', 'rev_activation', 'Compound'), None)\n",
      "    edge_index__Gene__rev_inhibition__Compound = edge_index_dict.get(('Gene', 'rev_inhibition', 'Compound'), None)\n",
      "    edge_index__Disease__rev_Compound_treats_the_disease__Compound = edge_index_dict.get(('Disease', 'rev_Compound_treats_the_disease', 'Compound'), None);  edge_index_dict = None\n",
      "    convs_0__Gene1 = getattr(self.convs, \"0\").Compound__activation__Gene((x__Compound, x__Gene), edge_index__Compound__activation__Gene)\n",
      "    convs_0__Gene2 = getattr(self.convs, \"0\").Compound__inhibition__Gene((x__Compound, x__Gene), edge_index__Compound__inhibition__Gene)\n",
      "    convs_0__Disease = getattr(self.convs, \"0\").Compound__Compound_treats_the_disease__Disease((x__Compound, x__Disease), edge_index__Compound__Compound_treats_the_disease__Disease)\n",
      "    convs_0__Gene3 = getattr(self.convs, \"0\").Gene__activates_or_stimulates__Gene(x__Gene, edge_index__Gene__activates_or_stimulates__Gene)\n",
      "    convs_0__Gene4 = getattr(self.convs, \"0\").Gene__inhibition__Gene(x__Gene, edge_index__Gene__inhibition__Gene)\n",
      "    convs_0__Compound1 = getattr(self.convs, \"0\").Gene__rev_activation__Compound((x__Gene, x__Compound), edge_index__Gene__rev_activation__Compound)\n",
      "    convs_0__Compound2 = getattr(self.convs, \"0\").Gene__rev_inhibition__Compound((x__Gene, x__Compound), edge_index__Gene__rev_inhibition__Compound);  x__Gene = None\n",
      "    convs_0__Compound3 = getattr(self.convs, \"0\").Disease__rev_Compound_treats_the_disease__Compound((x__Disease, x__Compound), edge_index__Disease__rev_Compound_treats_the_disease__Compound);  x__Disease = x__Compound = None\n",
      "    convs_0__Gene_1 = torch.add(convs_0__Gene1, convs_0__Gene2);  convs_0__Gene1 = convs_0__Gene2 = None\n",
      "    convs_0__Gene_2 = torch.add(convs_0__Gene3, convs_0__Gene4);  convs_0__Gene3 = convs_0__Gene4 = None\n",
      "    convs_0__Gene = torch.add(convs_0__Gene_1, convs_0__Gene_2);  convs_0__Gene_1 = convs_0__Gene_2 = None\n",
      "    convs_0__Compound_1 = torch.add(convs_0__Compound1, convs_0__Compound2);  convs_0__Compound1 = convs_0__Compound2 = None\n",
      "    convs_0__Compound = torch.add(convs_0__Compound3, convs_0__Compound_1);  convs_0__Compound3 = convs_0__Compound_1 = None\n",
      "    relu__Compound = torch.nn.functional.relu(convs_0__Compound, inplace = False);  convs_0__Compound = None\n",
      "    relu__Disease = torch.nn.functional.relu(convs_0__Disease, inplace = False);  convs_0__Disease = None\n",
      "    relu__Gene = torch.nn.functional.relu(convs_0__Gene, inplace = False);  convs_0__Gene = None\n",
      "    dropout__Compound = torch.nn.functional.dropout(relu__Compound, p = 0.7, training = True, inplace = False);  relu__Compound = None\n",
      "    dropout__Disease = torch.nn.functional.dropout(relu__Disease, p = 0.7, training = True, inplace = False);  relu__Disease = None\n",
      "    dropout__Gene = torch.nn.functional.dropout(relu__Gene, p = 0.7, training = True, inplace = False);  relu__Gene = None\n",
      "    convs_1__Gene1 = getattr(self.convs, \"1\").Compound__activation__Gene((dropout__Compound, dropout__Gene), edge_index__Compound__activation__Gene)\n",
      "    convs_1__Gene2 = getattr(self.convs, \"1\").Compound__inhibition__Gene((dropout__Compound, dropout__Gene), edge_index__Compound__inhibition__Gene)\n",
      "    convs_1__Disease = getattr(self.convs, \"1\").Compound__Compound_treats_the_disease__Disease((dropout__Compound, dropout__Disease), edge_index__Compound__Compound_treats_the_disease__Disease)\n",
      "    convs_1__Gene3 = getattr(self.convs, \"1\").Gene__activates_or_stimulates__Gene(dropout__Gene, edge_index__Gene__activates_or_stimulates__Gene)\n",
      "    convs_1__Gene4 = getattr(self.convs, \"1\").Gene__inhibition__Gene(dropout__Gene, edge_index__Gene__inhibition__Gene)\n",
      "    convs_1__Compound1 = getattr(self.convs, \"1\").Gene__rev_activation__Compound((dropout__Gene, dropout__Compound), edge_index__Gene__rev_activation__Compound)\n",
      "    convs_1__Compound2 = getattr(self.convs, \"1\").Gene__rev_inhibition__Compound((dropout__Gene, dropout__Compound), edge_index__Gene__rev_inhibition__Compound);  dropout__Gene = None\n",
      "    convs_1__Compound3 = getattr(self.convs, \"1\").Disease__rev_Compound_treats_the_disease__Compound((dropout__Disease, dropout__Compound), edge_index__Disease__rev_Compound_treats_the_disease__Compound);  dropout__Disease = dropout__Compound = None\n",
      "    convs_1__Gene_1 = torch.add(convs_1__Gene1, convs_1__Gene2);  convs_1__Gene1 = convs_1__Gene2 = None\n",
      "    convs_1__Gene_2 = torch.add(convs_1__Gene3, convs_1__Gene4);  convs_1__Gene3 = convs_1__Gene4 = None\n",
      "    convs_1__Gene = torch.add(convs_1__Gene_1, convs_1__Gene_2);  convs_1__Gene_1 = convs_1__Gene_2 = None\n",
      "    convs_1__Compound_1 = torch.add(convs_1__Compound1, convs_1__Compound2);  convs_1__Compound1 = convs_1__Compound2 = None\n",
      "    convs_1__Compound = torch.add(convs_1__Compound3, convs_1__Compound_1);  convs_1__Compound3 = convs_1__Compound_1 = None\n",
      "    relu_1__Compound = torch.nn.functional.relu(convs_1__Compound, inplace = False);  convs_1__Compound = None\n",
      "    relu_1__Disease = torch.nn.functional.relu(convs_1__Disease, inplace = False);  convs_1__Disease = None\n",
      "    relu_1__Gene = torch.nn.functional.relu(convs_1__Gene, inplace = False);  convs_1__Gene = None\n",
      "    dropout_1__Compound = torch.nn.functional.dropout(relu_1__Compound, p = 0.7, training = True, inplace = False);  relu_1__Compound = None\n",
      "    dropout_1__Disease = torch.nn.functional.dropout(relu_1__Disease, p = 0.7, training = True, inplace = False);  relu_1__Disease = None\n",
      "    dropout_1__Gene = torch.nn.functional.dropout(relu_1__Gene, p = 0.7, training = True, inplace = False);  relu_1__Gene = None\n",
      "    convs_2__Gene1 = getattr(self.convs, \"2\").Compound__activation__Gene((dropout_1__Compound, dropout_1__Gene), edge_index__Compound__activation__Gene);  edge_index__Compound__activation__Gene = None\n",
      "    convs_2__Gene2 = getattr(self.convs, \"2\").Compound__inhibition__Gene((dropout_1__Compound, dropout_1__Gene), edge_index__Compound__inhibition__Gene);  edge_index__Compound__inhibition__Gene = None\n",
      "    convs_2__Disease = getattr(self.convs, \"2\").Compound__Compound_treats_the_disease__Disease((dropout_1__Compound, dropout_1__Disease), edge_index__Compound__Compound_treats_the_disease__Disease);  edge_index__Compound__Compound_treats_the_disease__Disease = None\n",
      "    convs_2__Gene3 = getattr(self.convs, \"2\").Gene__activates_or_stimulates__Gene(dropout_1__Gene, edge_index__Gene__activates_or_stimulates__Gene);  edge_index__Gene__activates_or_stimulates__Gene = None\n",
      "    convs_2__Gene4 = getattr(self.convs, \"2\").Gene__inhibition__Gene(dropout_1__Gene, edge_index__Gene__inhibition__Gene);  edge_index__Gene__inhibition__Gene = None\n",
      "    convs_2__Compound1 = getattr(self.convs, \"2\").Gene__rev_activation__Compound((dropout_1__Gene, dropout_1__Compound), edge_index__Gene__rev_activation__Compound);  edge_index__Gene__rev_activation__Compound = None\n",
      "    convs_2__Compound2 = getattr(self.convs, \"2\").Gene__rev_inhibition__Compound((dropout_1__Gene, dropout_1__Compound), edge_index__Gene__rev_inhibition__Compound);  dropout_1__Gene = edge_index__Gene__rev_inhibition__Compound = None\n",
      "    convs_2__Compound3 = getattr(self.convs, \"2\").Disease__rev_Compound_treats_the_disease__Compound((dropout_1__Disease, dropout_1__Compound), edge_index__Disease__rev_Compound_treats_the_disease__Compound);  dropout_1__Disease = dropout_1__Compound = edge_index__Disease__rev_Compound_treats_the_disease__Compound = None\n",
      "    convs_2__Gene_1 = torch.add(convs_2__Gene1, convs_2__Gene2);  convs_2__Gene1 = convs_2__Gene2 = None\n",
      "    convs_2__Gene_2 = torch.add(convs_2__Gene3, convs_2__Gene4);  convs_2__Gene3 = convs_2__Gene4 = None\n",
      "    convs_2__Gene = torch.add(convs_2__Gene_1, convs_2__Gene_2);  convs_2__Gene_1 = convs_2__Gene_2 = None\n",
      "    convs_2__Compound_1 = torch.add(convs_2__Compound1, convs_2__Compound2);  convs_2__Compound1 = convs_2__Compound2 = None\n",
      "    convs_2__Compound = torch.add(convs_2__Compound3, convs_2__Compound_1);  convs_2__Compound3 = convs_2__Compound_1 = None\n",
      "    relu_2__Compound = torch.nn.functional.relu(convs_2__Compound, inplace = False);  convs_2__Compound = None\n",
      "    relu_2__Disease = torch.nn.functional.relu(convs_2__Disease, inplace = False);  convs_2__Disease = None\n",
      "    relu_2__Gene = torch.nn.functional.relu(convs_2__Gene, inplace = False);  convs_2__Gene = None\n",
      "    dropout_2__Compound = torch.nn.functional.dropout(relu_2__Compound, p = 0.7, training = True, inplace = False);  relu_2__Compound = None\n",
      "    dropout_2__Disease = torch.nn.functional.dropout(relu_2__Disease, p = 0.7, training = True, inplace = False);  relu_2__Disease = None\n",
      "    dropout_2__Gene = torch.nn.functional.dropout(relu_2__Gene, p = 0.7, training = True, inplace = False);  relu_2__Gene = None\n",
      "    post_mp_0__Compound = getattr(self.post_mp, \"0\").Compound(dropout_2__Compound);  dropout_2__Compound = None\n",
      "    post_mp_0__Disease = getattr(self.post_mp, \"0\").Disease(dropout_2__Disease);  dropout_2__Disease = None\n",
      "    post_mp_0__Gene = getattr(self.post_mp, \"0\").Gene(dropout_2__Gene);  dropout_2__Gene = None\n",
      "    post_mp_1__Compound = getattr(self.post_mp, \"1\").Compound(post_mp_0__Compound);  post_mp_0__Compound = None\n",
      "    post_mp_1__Disease = getattr(self.post_mp, \"1\").Disease(post_mp_0__Disease);  post_mp_0__Disease = None\n",
      "    post_mp_1__Gene = getattr(self.post_mp, \"1\").Gene(post_mp_0__Gene);  post_mp_0__Gene = None\n",
      "    post_mp_2__Compound = getattr(self.post_mp, \"2\").Compound(post_mp_1__Compound);  post_mp_1__Compound = None\n",
      "    post_mp_2__Disease = getattr(self.post_mp, \"2\").Disease(post_mp_1__Disease);  post_mp_1__Disease = None\n",
      "    post_mp_2__Gene = getattr(self.post_mp, \"2\").Gene(post_mp_1__Gene);  post_mp_1__Gene = None\n",
      "    return {'Compound': post_mp_2__Compound, 'Disease': post_mp_2__Disease, 'Gene': post_mp_2__Gene}\n",
      "    \n",
      "# To see more debug info, please use `graph_module.print_readable()`\n",
      "LinkPredictorMLP(\n",
      "  (mlp_layers): ModuleList(\n",
      "    (0-1): 2 x Linear(in_features=524, out_features=524, bias=True)\n",
      "    (2): Linear(in_features=524, out_features=1, bias=True)\n",
      "  )\n",
      ")\n",
      "Models Loaded to cpu\n"
     ]
    }
   ],
   "source": [
    "epochs        = 500\n",
    "hidden_dim    = 524      # 256 \n",
    "dropout       = 0.7\n",
    "num_layers    = 3\n",
    "learning_rate = 1e-4\n",
    "node_emb_dim  = 768\n",
    "\n",
    "\n",
    "\n",
    "HomoGNN         = GNNStack(node_emb_dim, hidden_dim, hidden_dim, num_layers, dropout, return_embedding=True).to(device) # the graph neural network that takes all the node embeddings as inputs to message pass and agregate\n",
    "HeteroGNN       = to_hetero(HomoGNN   , data.metadata(), aggr='sum')\n",
    "link_predictor  = LinkPredictorMLP(hidden_dim, hidden_dim, 1, num_layers , dropout).to(device) # the MLP that takes embeddings of a pair of nodes and predicts the existence of an edge between them\n",
    "#optimizer      = torch.optim.AdamW(list(model.parameters()) + list(link_predictor.parameters() ), lr=learning_rate, weight_decay=1e-4)\n",
    "optimizer       = torch.optim.Adam(list(HeteroGNN.parameters()) + list(link_predictor.parameters() ), lr=learning_rate)\n",
    "\n",
    "print(HeteroGNN )\n",
    "print(link_predictor)\n",
    "print(f\"Models Loaded to {device}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HeteroData(\n",
      "  \u001b[1mCompound\u001b[0m={ x=[29, 768] },\n",
      "  \u001b[1mDisease\u001b[0m={ x=[14, 768] },\n",
      "  \u001b[1mGene\u001b[0m={ x=[51, 768] },\n",
      "  \u001b[1m(Compound, activation, Gene)\u001b[0m={\n",
      "    edge_index=[2, 15],\n",
      "    edge_label=[15, 768]\n",
      "  },\n",
      "  \u001b[1m(Compound, inhibition, Gene)\u001b[0m={\n",
      "    edge_index=[2, 10],\n",
      "    edge_label=[10, 768]\n",
      "  },\n",
      "  \u001b[1m(Compound, Compound_treats_the_disease, Disease)\u001b[0m={\n",
      "    edge_index=[2, 15],\n",
      "    edge_label=[15, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, activates_or_stimulates, Gene)\u001b[0m={\n",
      "    edge_index=[2, 20],\n",
      "    edge_label=[20, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, inhibition, Gene)\u001b[0m={\n",
      "    edge_index=[2, 10],\n",
      "    edge_label=[10, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, rev_activation, Compound)\u001b[0m={\n",
      "    edge_index=[2, 15],\n",
      "    edge_label=[15, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, rev_inhibition, Compound)\u001b[0m={\n",
      "    edge_index=[2, 10],\n",
      "    edge_label=[10, 768]\n",
      "  },\n",
      "  \u001b[1m(Disease, rev_Compound_treats_the_disease, Compound)\u001b[0m={\n",
      "    edge_index=[2, 15],\n",
      "    edge_label=[15, 768]\n",
      "  }\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "print(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_emb   = HeteroGNN(data.x_dict, data.edge_index_dict)\n",
    "edge_index = data['Compound', 'Compound_treats_the_disease', 'Disease'].edge_index \n",
    "pos_pred   = link_predictor(node_emb[\"Compound\"][edge_index[0]], node_emb[\"Disease\"][edge_index[1]])   # (B, )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Compound': tensor([[ 0.5875, -1.0616, -0.7926,  ..., -0.0278, -0.0979,  0.5913],\n",
       "         [ 0.2394,  0.2247,  1.2243,  ...,  0.8381,  1.1482, -0.6650],\n",
       "         [-0.6849, -1.8316, -0.0687,  ...,  0.7773, -0.1265,  1.0705],\n",
       "         ...,\n",
       "         [ 0.4461,  0.2052, -0.9088,  ...,  0.0704, -0.5500,  0.4950],\n",
       "         [-0.8308,  0.2126,  1.3543,  ..., -0.1609,  0.9807,  0.3113],\n",
       "         [-0.0263,  0.0214,  0.8781,  ...,  0.2766, -0.5748,  0.3060]],\n",
       "        grad_fn=<AddmmBackward0>),\n",
       " 'Disease': tensor([[ 1.1654,  1.1531, -0.0988,  ..., -0.4830, -0.7136, -0.3199],\n",
       "         [-0.0638,  0.4340, -1.4309,  ..., -0.4483, -1.2654,  0.1086],\n",
       "         [ 0.6284, -0.2473,  0.5736,  ..., -0.1155, -0.5782, -1.0763],\n",
       "         ...,\n",
       "         [ 0.0311,  0.0726,  1.3918,  ..., -0.5468, -0.1218, -0.0914],\n",
       "         [-0.1779,  0.6497,  0.6867,  ..., -0.5519,  0.2511, -0.0758],\n",
       "         [ 0.7770,  0.5642,  0.9319,  ...,  0.3275,  0.5305,  0.4870]],\n",
       "        grad_fn=<AddmmBackward0>),\n",
       " 'Gene': tensor([[-0.1290, -0.7859,  1.0558,  ..., -0.1844,  0.5041,  0.5360],\n",
       "         [-0.4826,  0.8699,  0.0759,  ..., -1.5977,  1.6642, -0.6225],\n",
       "         [-3.8885,  2.7971, -1.1174,  ..., -0.9936, -0.0363, -3.0334],\n",
       "         ...,\n",
       "         [ 0.8810,  0.8204,  0.3814,  ..., -1.1545,  0.0333, -0.5118],\n",
       "         [ 0.6700,  0.6417, -0.1798,  ...,  2.3699, -0.3231, -0.0218],\n",
       "         [-1.2696,  1.1270,  0.6545,  ...,  1.2150,  0.0812, -1.7771]],\n",
       "        grad_fn=<AddmmBackward0>)}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "node_emb"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4) Identify rare diseases in DRKG\n",
    "\n",
    "AUTHORS: Selina"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SIMP_LLM.raredisease_loading import get_drkg_entity_ontologies, read_and_process_doid, create_orphanet_regex, merge_regex, find_drkg_rarediseases, check_raredisease_multiple_codes, read_and_process_rep_drugs"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download and process data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "DRKG code counts:\n",
      "   matched ontology_name  count\n",
      "0        0          MESH   1002\n",
      "1        1          DOID    127\n",
      "2        1          MESH   4284\n",
      "3        1          OMIM     78\n",
      "4        1      UMLS CUI   5701\n",
      "\n",
      "Orphanet code types:  ['ICD-10' 'OMIM' 'UMLS' 'MeSH' 'ICD-11' 'GARD' 'MedDRA']\n",
      "\n",
      " DOID Dataframe (After processing):\n",
      "\n",
      "+-----+--------------+--------------------------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------------------------------------------------------------------------------------------------------------+----------------------+------------------------------------------------+---------------------------------------------+\n",
      "|     | id           | Preferred Label          | Synonyms                                       | Definitions                                                                                                                                                                     |   CUI | database_cross_reference                                                                                         | has_alternative_id   | has_exact_synonym                              | Parents                                     |\n",
      "|-----+--------------+--------------------------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------------------------------------------------------------------------------------------------------------+----------------------+------------------------------------------------+---------------------------------------------|\n",
      "| 572 | DOID:0001816 | angiosarcoma             | hemangiosarcoma                                | A vascular cancer that derives_from the cells that line the walls of blood vessels or lymphatic vessels.                                                                        |   nan | SNOMEDCT_US_2022_09_01:39000009|NCI:C3088|UMLS_CUI:C0018923|MESH:D006394|ICDO:9120/3|NCI:C9275|UMLS_CUI:C0854893 | DOID:4508|DOID:267   | hemangiosarcoma                                | http://purl.obolibrary.org/obo/DOID_175     |\n",
      "| 573 | DOID:0002116 | pterygium                | surfer's eye                                   | A corneal disease that is characterized by a triangular tissue growth located_in cornea of the eye that is the result of collagen degeneration and fibrovascular proliferation. |   nan | UMLS_CUI:C0033999                                                                                                | nan                  | surfer's eye                                   | http://purl.obolibrary.org/obo/DOID_10124   |\n",
      "| 574 | DOID:0014667 | disease of metabolism    | metabolic disease                              | A disease that involving errors in metabolic processes of building or degradation of molecules.                                                                                 |   nan | ICD10CM:E88.9|ICD9CM:277.9|NCI:C3235|SNOMEDCT_US_2022_09_01:75934005|MESH:D008659|UMLS_CUI:C0025517              | nan                  | metabolic disease                              | http://purl.obolibrary.org/obo/DOID_4       |\n",
      "| 576 | DOID:0040002 | aspirin allergy          | ASA allergy|acetylsalicylic acid allergy       | A drug allergy that has_allergic_trigger acetylsalicylic acid.                                                                                                                  |   nan | UMLS_CUI:C0004058|SNOMEDCT_US_2022_09_01:293586001                                                               | nan                  | ASA allergy|acetylsalicylic acid allergy       | http://purl.obolibrary.org/obo/DOID_0060500 |\n",
      "| 577 | DOID:0040003 | benzylpenicillin allergy | penicillin G allergy|benzyl penicillin allergy | A beta-lactam allergy that has_allergic_trigger benzylpenicillin.                                                                                                               |   nan | SNOMEDCT_US_2022_09_01:294499007|UMLS_CUI:C0571411                                                               | nan                  | penicillin G allergy|benzyl penicillin allergy | http://purl.obolibrary.org/obo/DOID_0060519 |\n",
      "+-----+--------------+--------------------------+------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+-------+------------------------------------------------------------------------------------------------------------------+----------------------+------------------------------------------------+---------------------------------------------+\n",
      "\n",
      " Orphanet Dataframe with regex (After processing):\n",
      "\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+-------------------------------------------+\n",
      "|    |   Orphacode | Name                                          | code_source   | code     | Disordermappingrelation                                  | Disordermappingicdrelation                          | Disordermappingvalidationstatus   | regex                                     |\n",
      "|----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+-------------------------------------------|\n",
      "|  0 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | ICD-10        | Q77\\.3   | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         | (?:^|\\|)ICD10(?:(?!\\|).)*:Q77.3.*         |\n",
      "|  1 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | OMIM          | 607131   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | (?:^|\\|)OMIM(?:(?!\\|).)*:607131(?:$|\\|)   |\n",
      "|  2 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | UMLS          | C1846722 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | (?:^|\\|)UMLS(?:(?!\\|).)*:C1846722(?:$|\\|) |\n",
      "|  3 |          58 | Alexander disease                             | OMIM          | 203450   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | (?:^|\\|)OMIM(?:(?!\\|).)*:203450(?:$|\\|)   |\n",
      "|  4 |          58 | Alexander disease                             | MeSH          | D038261  | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | (?:^|\\|)MESH(?:(?!\\|).)*:D038261(?:$|\\|)  |\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+-------------------------------------------+\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/selinapi/SIMPLLM/SIMP_LLM/raredisease_loading.py:204: DtypeWarning: Columns (8,16,17,18,26,31,32,33,47,52,53,55,56,58,62,63,65,69,71,73,77,83,84,92,104,108) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df_raw = pd.read_csv(relation_file)\n"
     ]
    }
   ],
   "source": [
    "# All drkg entities \n",
    "drkg_all_entities = pd.concat([drkg_entity_df, drkg_unmatched], ignore_index=True, axis=0) \n",
    "\n",
    " # Check disease code ontologies in DRKG - extra code in DKRG called DOID - and Orphanet\n",
    "if verbose:\n",
    "    condition_list = ['Disease', 'Symptom', 'Side Effect']\n",
    "    drkg_ontology_counts = get_drkg_entity_ontologies(drkg_all_entities, condition_list)\n",
    "    print('DRKG code counts:')\n",
    "    print(drkg_ontology_counts)\n",
    "    print('\\nOrphanet code types: ', orphan_codes['code_source'].unique())\n",
    "\n",
    "# Download and process DOID (disease ontology) codes\n",
    "doid_df  =  read_and_process_doid(relation_file = os.path.join(DATA_DIR,'DOID.csv'), verbose=verbose) \n",
    "\n",
    "# Create regex for match Orphanet mapped codes to cross reference codes in DOID\n",
    "orphan_codes_match = create_orphanet_regex(orphan_codes, verbose=verbose)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Code and regex matching between Orphacodes and DRKG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " DRKG-Orphacode matches:\n",
      "\n",
      "+----+-----------------------+----------------+------------------------------------------------------------------------+---------------+-----------------+-----------------+---------+-----------+-------------+------------------------------------------------------------------------+---------------+----------------------------------------------------+------------------------------+-----------------------------------+---------------------+--------------+----------+------+----------+---------+\n",
      "|    | drkg_id               | drkg_dict_id   | name                                                                   | entity_type   | ontology_code   | ontology_name   | code    |   matched |   Orphacode | Name                                                                   | code_source   | Disordermappingrelation                            |   Disordermappingicdrelation | Disordermappingvalidationstatus   | code_source_upper   | match_type   |   code_x |   id |   code_y |   key_0 |\n",
      "|----+-----------------------+----------------+------------------------------------------------------------------------+---------------+-----------------+-----------------+---------+-----------+-------------+------------------------------------------------------------------------+---------------+----------------------------------------------------+------------------------------+-----------------------------------+---------------------+--------------+----------+------+----------+---------|\n",
      "|  0 | Disease::MESH:C531609 | MESH::C531609  | Pseudopelade of Brocq                                                  | Disease       | MESH:C531609    | MESH            | C531609 |         1 |         129 | Pseudopelade of Brocq                                                  | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "|  1 | Disease::MESH:C531616 | MESH::C531616  | AL amyloidosis                                                         | Disease       | MESH:C531616    | MESH            | C531616 |         1 |       85443 | AL amyloidosis                                                         | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "|  2 | Disease::MESH:C531629 | MESH::C531629  | Balantidiasis                                                          | Disease       | MESH:C531629    | MESH            | C531629 |         1 |        1223 | Balantidiasis                                                          | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "|  3 | Disease::MESH:C531816 | MESH::C531816  | Severe combined immunodeficiency due to adenosine deaminase deficiency | Disease       | MESH:C531816    | MESH            | C531816 |         1 |         277 | Severe combined immunodeficiency due to adenosine deaminase deficiency | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "|  4 | Disease::MESH:C535276 | MESH::C535276  | Primary angiitis of the central nervous system                         | Disease       | MESH:C535276    | MESH            | C535276 |         1 |      140989 | Primary angiitis of the central nervous system                         | MeSH          | E (Exact mapping: the two concepts are equivalent) |                          nan | Validated                         | MESH                | MeSH/OMIM    |      nan |  nan |      nan |     nan |\n",
      "+----+-----------------------+----------------+------------------------------------------------------------------------+---------------+-----------------+-----------------+---------+-----------+-------------+------------------------------------------------------------------------+---------------+----------------------------------------------------+------------------------------+-----------------------------------+---------------------+--------------+----------+------+----------+---------+\n",
      "                                        count\n",
      "match_type   entity_type ontology_name       \n",
      "DOID regex   Disease     DOID              21\n",
      "Disease name Disease     DOID               7\n",
      "                         MESH              54\n",
      "             Side Effect UMLS CUI         134\n",
      "             Symptom     MESH               3\n",
      "MeSH/OMIM    Disease     MESH            1033\n",
      "                         OMIM              67\n",
      "             Symptom     MESH              41\n",
      "\n",
      "Unique matched rare disease DRKG IDs:  1321\n",
      "Unique matched Orphacodes:  1197\n",
      "DRKG IDs associated with multiple Orphacodes:  31\n",
      "Orphacodes associated with multiple DRKG IDs:  44\n",
      "+-----+-----------------------+--------------------------------------+--------------+----------------+\n",
      "|     | drkg_id               | name                                 | match_type   |   ct_orphacode |\n",
      "|-----+-----------------------+--------------------------------------+--------------+----------------|\n",
      "|   1 | Disease::DOID:10021   | duodenum cancer                      | DOID regex   |              2 |\n",
      "|   2 | Disease::DOID:10153   | ileum cancer                         | DOID regex   |              3 |\n",
      "|  10 | Disease::DOID:14268   | sclerosing cholangitis               | DOID regex   |              4 |\n",
      "| 153 | Disease::MESH:C536385 | Focal facial dermal dysplasia type I | MeSH/OMIM    |              2 |\n",
      "| 184 | Disease::MESH:C536611 | Brooke-Spiegler syndrome             | MeSH/OMIM    |              2 |\n",
      "+-----+-----------------------+--------------------------------------+--------------+----------------+\n",
      "+----+-------------+----------------------------------+--------------+--------------+\n",
      "|    |   Orphacode | name                             | match_type   |   ct_drkg_id |\n",
      "|----+-------------+----------------------------------+--------------+--------------|\n",
      "|  1 |         100 | Ataxia Telangiectasia            | MeSH/OMIM    |            2 |\n",
      "|  4 |      100067 | Waterhouse-Friderichsen Syndrome | MeSH/OMIM    |            2 |\n",
      "| 21 |        1041 | Hydrops Fetalis                  | MeSH/OMIM    |            2 |\n",
      "| 43 |        1183 | Opsoclonus-Myoclonus Syndrome    | MeSH/OMIM    |            2 |\n",
      "| 49 |        1223 | Balantidiasis                    | MeSH/OMIM    |            2 |\n",
      "+----+-------------+----------------------------------+--------------+--------------+\n",
      "Number of unmatched Orphacodes:  7132\n",
      "+----+-------------+---------------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+---------------------+\n",
      "|    |   Orphacode | Name                                              | code_source   | code     | Disordermappingrelation                                  | Disordermappingicdrelation                          | Disordermappingvalidationstatus   | code_source_upper   |\n",
      "|----+-------------+---------------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+---------------------|\n",
      "|  0 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type     | ICD-10        | Q77.3    | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         | ICD-10              |\n",
      "|  1 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type     | OMIM          | 607131   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | OMIM                |\n",
      "|  2 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type     | UMLS          | C1846722 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | UMLS                |\n",
      "|  9 |      166032 | Multiple epiphyseal dysplasia, with miniepiphyses | ICD-10        | Q77.3    | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         | ICD-10              |\n",
      "| 10 |      166032 | Multiple epiphyseal dysplasia, with miniepiphyses | UMLS          | C1836307 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         | UMLS                |\n",
      "+----+-------------+---------------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+---------------------+\n",
      "Rare disease heads in untranslated DRKG:  3610\n",
      "+-------+-----------------------+-------------------------------------------------+------------+\n",
      "|       | 0                     | 1                                               | 2          |\n",
      "|-------+-----------------------+-------------------------------------------------+------------|\n",
      "| 84697 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::6502 |\n",
      "| 84698 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::3840 |\n",
      "| 84699 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::8575 |\n",
      "| 84700 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::928  |\n",
      "| 84701 | Disease::MESH:D065207 | bioarx::Coronavirus_ass_host_gene::Disease:Gene | Gene::7113 |\n",
      "+-------+-----------------------+-------------------------------------------------+------------+\n",
      "\n",
      " Rare disease tails in untranslated DRKG:  31854\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n",
      "|         | 0                 | 1                                  | 2                     |\n",
      "|---------+-------------------+------------------------------------+-----------------------|\n",
      "| 1518268 | Compound::DB00003 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D003550 |\n",
      "| 1518271 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D006086 |\n",
      "| 1518273 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D001171 |\n",
      "| 1518276 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D017511 |\n",
      "| 1518278 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D013262 |\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n",
      "+-------+----------------------------------------------+-------------------------------------------------+------------------------------------------------------------------------+\n",
      "|       | head_entity                                  | relationshipId                                  | tail_entity                                                            |\n",
      "|-------+----------------------------------------------+-------------------------------------------------+------------------------------------------------------------------------|\n",
      "| 84697 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | S-phase kinase associated protein 2 (SKP2)                             |\n",
      "| 84698 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | karyopherin subunit alpha 4 (KPNA4)                                    |\n",
      "| 84699 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | protein activator of interferon induced protein kinase EIF2AK2 (PRKRA) |\n",
      "| 84700 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | CD9 molecule (CD9)                                                     |\n",
      "| 84701 | Middle East Respiratory Syndrome Coronavirus | bioarx::Coronavirus_ass_host_gene::Disease:Gene | transmembrane serine protease 2 (TMPRSS2)                              |\n",
      "+-------+----------------------------------------------+-------------------------------------------------+------------------------------------------------------------------------+\n"
     ]
    }
   ],
   "source": [
    "# Match Orphacodes and DOIDs - Takes ~3min to run\n",
    "orphacode_doid_regex = merge_regex(orphan_codes_match, 'regex', doid_df, 'database_cross_reference')\n",
    "\n",
    "# Match rare diseases in DRKG\n",
    "matched_rarediseases = find_drkg_rarediseases(drkg_all_entities, orphan_codes, orphacode_doid_regex, verbose=verbose)\n",
    "\n",
    "# Check for rare diseases with multiple codes\n",
    "if verbose:\n",
    "    multiple_orphacode, multiple_drkg = check_raredisease_multiple_codes(matched_rarediseases)\n",
    "    print_head(multiple_orphacode)\n",
    "    print_head(multiple_drkg)\n",
    "\n",
    "# Get non-matched rare diseases (could still be in DRKG but under different name or code - need to check that embeddings are separate)\n",
    "unmatched_rarediseases = orphan_codes[~orphan_codes['Orphacode'].isin(matched_rarediseases['Orphacode'])]\n",
    "if verbose:\n",
    "    print('Number of unmatched Orphacodes: ', len(unmatched_rarediseases['Orphacode'].unique()))\n",
    "    print_head(unmatched_rarediseases)\n",
    "\n",
    "# Get all DRKG relationships involving matched rare diseases\n",
    "raredisease_heads = drkg_df[drkg_df[0].isin(matched_rarediseases['drkg_id'])]\n",
    "raredisease_tails = drkg_df[drkg_df[2].isin(matched_rarediseases['drkg_id'])]\n",
    "\n",
    "if verbose:\n",
    "    print('Rare disease heads in untranslated DRKG: ', raredisease_heads.shape[0])\n",
    "    print_head(raredisease_heads)\n",
    "    print('\\n Rare disease tails in untranslated DRKG: ',raredisease_tails.shape[0])\n",
    "    print_head(raredisease_tails)\n",
    "\n",
    "# Get DRKG rare disease entries\n",
    "raredisease_index = raredisease_heads.index.tolist() + raredisease_tails.index.tolist()\n",
    "drkg_raredisease = drkg_translated.loc[drkg_translated.index.isin(raredisease_index)]\n",
    "if verbose:\n",
    "    print_head(drkg_raredisease)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Check drug repurposing hub (DRH) database for examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " data/repurposing_drugs_20200324.txt  Drug Repurposing Dataframe:\n",
      "\n",
      "+----+------------------------------+------------------+---------------------------------+----------------------------------------------------------------------------------------------+----------------------+---------------------+\n",
      "|    | pert_iname                   | clinical_phase   | moa                             | target                                                                                       | disease_area         | indication          |\n",
      "|----+------------------------------+------------------+---------------------------------+----------------------------------------------------------------------------------------------+----------------------+---------------------|\n",
      "|  0 | (R)-(-)-apomorphine          | Launched         | dopamine receptor agonist       | ADRA2A|ADRA2B|ADRA2C|CALY|DRD1|DRD2|DRD3|DRD4|DRD5|HTR1A|HTR1B|HTR1D|HTR2A|HTR2B|HTR2C|HTR5A | neurology/psychiatry | Parkinson's Disease |\n",
      "|  1 | (R)-(-)-rolipram             | Phase 1          | phosphodiesterase inhibitor     | PDE4A|PDE4B|PDE4C|PDE4D|PDE5A                                                                | nan                  | nan                 |\n",
      "|  2 | (R)-baclofen                 | Phase 3          | benzodiazepine receptor agonist | GABBR1|GABBR2                                                                                | nan                  | nan                 |\n",
      "|  3 | (S)-(+)-rolipram             | Phase 1          | phosphodiesterase inhibitor     | PDE4B|PDE4D                                                                                  | nan                  | nan                 |\n",
      "|  4 | [sar9,met(o2)11]-substance-p | Preclinical      | tachykinin antagonist           | TACR1                                                                                        | nan                  | nan                 |\n",
      "+----+------------------------------+------------------+---------------------------------+----------------------------------------------------------------------------------------------+----------------------+---------------------+\n"
     ]
    }
   ],
   "source": [
    "rep_drugs_df       =  read_and_process_rep_drugs(os.path.join(DATA_DIR, 'repurposing_drugs_20200324.txt'), verbose=verbose)     # Process entity names for clarity (e.g., F8 -> Gene F8) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get repurposed drugs with indications\n",
    "drug_with_indication = rep_drugs_df[~rep_drugs_df['indication'].isna()]\n",
    "\n",
    "# Find entries of DRH data with drug in DRKG and indication in DRKG rare diseases\n",
    "drh_in_drkg = drug_with_indication[(drug_with_indication['indication'].str.lower().isin(matched_rarediseases['name'].str.lower())) & \n",
    "                                   (drug_with_indication['pert_iname'].str.lower().isin(drkg_entity_df['name'].str.lower())) ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(114, 11)\n",
      "(31, 6)\n",
      "(35, 11)\n",
      "(20, 6)\n"
     ]
    }
   ],
   "source": [
    "# Find DRKG rare disease relationships whose drug-treatment relationship in DRH is not represented in DRKG,\n",
    "# first by matching drug-disease pairs in both\n",
    "drh_matches = drh_in_drkg.merge(drkg_raredisease, how='left', left_on=[drh_in_drkg['pert_iname'].str.lower(), drh_in_drkg['indication'].str.lower()], right_on=[drkg_raredisease['head_entity'].str.lower(), drkg_raredisease['tail_entity'].str.lower()])\n",
    "print(drh_matches.shape)\n",
    "\n",
    "# Then dropping matches\n",
    "drh_unmatched_try1 = drh_matches[drh_matches['head_entity'].isna()].drop(columns=['head_entity', 'relationshipId', 'tail_entity','key_0', 'key_1'])\n",
    "print(drh_unmatched_try1.shape)\n",
    "\n",
    "# Try replacing 'cancer' with 'neoplasms' and dropping further matches\n",
    "drh_unmatched_try1['indication'] = drh_unmatched_try1['indication'].str.replace('cancer', 'neoplasms')\n",
    "\n",
    "drh_matches_try2 = drh_unmatched_try1.merge(drkg_translated, how='left', left_on=[drh_unmatched_try1['pert_iname'].str.lower(), drh_unmatched_try1['indication'].str.lower()], right_on=[drkg_translated['head_entity'].str.lower(), drkg_translated['tail_entity'].str.lower()])\n",
    "drh_matches_try2\n",
    "print(drh_matches_try2.shape)\n",
    "\n",
    "# # Then dropping matches again\n",
    "drh_unmatched = drh_matches_try2[drh_matches_try2['head_entity'].isna()].drop(columns=['head_entity', 'relationshipId', 'tail_entity', 'key_0', 'key_1']).drop_duplicates()\n",
    "print(drh_unmatched.shape)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>drkg_dict_id</th>\n",
       "      <th>name</th>\n",
       "      <th>entity_type</th>\n",
       "      <th>ontology_code</th>\n",
       "      <th>ontology_name</th>\n",
       "      <th>code</th>\n",
       "      <th>matched</th>\n",
       "      <th>Orphacode</th>\n",
       "      <th>Name</th>\n",
       "      <th>code_source</th>\n",
       "      <th>Disordermappingrelation</th>\n",
       "      <th>Disordermappingicdrelation</th>\n",
       "      <th>Disordermappingvalidationstatus</th>\n",
       "      <th>code_source_upper</th>\n",
       "      <th>match_type</th>\n",
       "      <th>code_x</th>\n",
       "      <th>id</th>\n",
       "      <th>code_y</th>\n",
       "      <th>key_0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1187</th>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease</td>\n",
       "      <td>DOID:10283</td>\n",
       "      <td>DOID</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>1331</td>\n",
       "      <td>Familial prostate cancer</td>\n",
       "      <td>MeSH</td>\n",
       "      <td>E (Exact mapping: the two concepts are equival...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Validated</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DOID regex</td>\n",
       "      <td>10283</td>\n",
       "      <td>DOID:10283</td>\n",
       "      <td>C537243</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  drkg_id         drkg_dict_id             name entity_type   \n",
       "1187  Disease::DOID:10283  Disease::DOID:10283  prostate cancer     Disease  \\\n",
       "\n",
       "     ontology_code ontology_name code  matched Orphacode   \n",
       "1187    DOID:10283          DOID  NaN        1      1331  \\\n",
       "\n",
       "                          Name code_source   \n",
       "1187  Familial prostate cancer        MeSH  \\\n",
       "\n",
       "                                Disordermappingrelation   \n",
       "1187  E (Exact mapping: the two concepts are equival...  \\\n",
       "\n",
       "     Disordermappingicdrelation Disordermappingvalidationstatus   \n",
       "1187                        NaN                       Validated  \\\n",
       "\n",
       "     code_source_upper  match_type code_x          id   code_y key_0  \n",
       "1187               NaN  DOID regex  10283  DOID:10283  C537243   NaN  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check if prostate cancer included in rare diseases\n",
    "matched_rarediseases[matched_rarediseases['Name'].str.lower().str.contains('prostate cancer')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>compound</th>\n",
       "      <th>compound_drkg_id</th>\n",
       "      <th>rare_disease</th>\n",
       "      <th>disease_drkg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Alpelisib</td>\n",
       "      <td>Compound::DB12015</td>\n",
       "      <td>breast cancer</td>\n",
       "      <td>Disease::DOID:1612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Ambenonium</td>\n",
       "      <td>Compound::DB01122</td>\n",
       "      <td>Myasthenia Gravis</td>\n",
       "      <td>Disease::MESH:D009157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Apalutamide</td>\n",
       "      <td>Compound::DB11901</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Artesunate</td>\n",
       "      <td>Compound::DB09274</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>Disease::MESH:D008288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Artesunate</td>\n",
       "      <td>Compound::DB09274</td>\n",
       "      <td>malaria</td>\n",
       "      <td>Disease::DOID:12365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Darolutamide</td>\n",
       "      <td>Compound::DB12941</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Delamanid</td>\n",
       "      <td>Compound::DB11637</td>\n",
       "      <td>Tuberculosis</td>\n",
       "      <td>Disease::MESH:D014376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Didox</td>\n",
       "      <td>Compound::DB12948</td>\n",
       "      <td>breast cancer</td>\n",
       "      <td>Disease::DOID:1612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Enzalutamide</td>\n",
       "      <td>Compound::DB08899</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Leucovorin</td>\n",
       "      <td>Compound::DB00650</td>\n",
       "      <td>Osteosarcoma</td>\n",
       "      <td>Disease::MESH:D012516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Leuprolide</td>\n",
       "      <td>Compound::DB00007</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>Lumefantrine</td>\n",
       "      <td>Compound::DB06708</td>\n",
       "      <td>Malaria</td>\n",
       "      <td>Disease::MESH:D008288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Lumefantrine</td>\n",
       "      <td>Compound::DB06708</td>\n",
       "      <td>malaria</td>\n",
       "      <td>Disease::DOID:12365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Panobinostat</td>\n",
       "      <td>Compound::DB06603</td>\n",
       "      <td>Multiple Myeloma</td>\n",
       "      <td>Disease::MESH:D009101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Pomalidomide</td>\n",
       "      <td>Compound::DB08910</td>\n",
       "      <td>Multiple Myeloma</td>\n",
       "      <td>Disease::MESH:D009101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Raltitrexed</td>\n",
       "      <td>Compound::DB00293</td>\n",
       "      <td>Mesothelioma</td>\n",
       "      <td>Disease::MESH:D008654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Selinexor</td>\n",
       "      <td>Compound::DB11942</td>\n",
       "      <td>Multiple Myeloma</td>\n",
       "      <td>Disease::MESH:D009101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Solriamfetol</td>\n",
       "      <td>Compound::DB14754</td>\n",
       "      <td>Narcolepsy</td>\n",
       "      <td>Disease::MESH:D009290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Trofosfamide</td>\n",
       "      <td>Compound::DB12902</td>\n",
       "      <td>Medulloblastoma</td>\n",
       "      <td>Disease::MESH:D008527</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        compound   compound_drkg_id       rare_disease        disease_drkg_id\n",
       "0      Alpelisib  Compound::DB12015      breast cancer     Disease::DOID:1612\n",
       "1     Ambenonium  Compound::DB01122  Myasthenia Gravis  Disease::MESH:D009157\n",
       "4    Apalutamide  Compound::DB11901    prostate cancer    Disease::DOID:10283\n",
       "5     Artesunate  Compound::DB09274            Malaria  Disease::MESH:D008288\n",
       "6     Artesunate  Compound::DB09274            malaria    Disease::DOID:12365\n",
       "8   Darolutamide  Compound::DB12941    prostate cancer    Disease::DOID:10283\n",
       "9      Delamanid  Compound::DB11637       Tuberculosis  Disease::MESH:D014376\n",
       "11         Didox  Compound::DB12948      breast cancer     Disease::DOID:1612\n",
       "12  Enzalutamide  Compound::DB08899    prostate cancer    Disease::DOID:10283\n",
       "15    Leucovorin  Compound::DB00650       Osteosarcoma  Disease::MESH:D012516\n",
       "16    Leuprolide  Compound::DB00007    prostate cancer    Disease::DOID:10283\n",
       "17  Lumefantrine  Compound::DB06708            Malaria  Disease::MESH:D008288\n",
       "18  Lumefantrine  Compound::DB06708            malaria    Disease::DOID:12365\n",
       "20  Panobinostat  Compound::DB06603   Multiple Myeloma  Disease::MESH:D009101\n",
       "21  Pomalidomide  Compound::DB08910   Multiple Myeloma  Disease::MESH:D009101\n",
       "22   Raltitrexed  Compound::DB00293       Mesothelioma  Disease::MESH:D008654\n",
       "24     Selinexor  Compound::DB11942   Multiple Myeloma  Disease::MESH:D009101\n",
       "25  Solriamfetol  Compound::DB14754         Narcolepsy  Disease::MESH:D009290\n",
       "27  Trofosfamide  Compound::DB12902    Medulloblastoma  Disease::MESH:D008527"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Merge DRKG name of diseases (matching cancer/neoplasms too)\n",
    "drkg_test = drh_unmatched.merge(matched_rarediseases[['name', 'drkg_id']], how='left', left_on=[drh_unmatched['indication'].str.lower()], right_on=[matched_rarediseases['name'].str.lower().str.replace('cancer', 'neoplasms')])\n",
    "drkg_test = drkg_test.drop(columns=['key_0']).drop_duplicates().rename(columns={'name':'rare_disease', 'drkg_id':'disease_drkg_id'})\n",
    "\n",
    "# Merge DRKG name of drugs\n",
    "drkg_test = drkg_test.merge(drkg_entity_df[['name', 'drkg_id']], how='left', left_on=[drkg_test['pert_iname'].str.lower()], right_on=[drkg_entity_df['name'].str.lower()])\n",
    "drkg_test = drkg_test[['name', 'drkg_id', 'rare_disease', 'disease_drkg_id']].rename(columns={'name':'compound', 'drkg_id':'compound_drkg_id'})\n",
    "drkg_test = drkg_test[drkg_test['disease_drkg_id'].str.contains('Disease')]\n",
    "print(len(drkg_test))\n",
    "drkg_test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>compound</th>\n",
       "      <th>compound_drkg_id</th>\n",
       "      <th>rare_disease</th>\n",
       "      <th>disease_drkg_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>alpelisib</td>\n",
       "      <td>Compound::DB12015</td>\n",
       "      <td>breast cancer</td>\n",
       "      <td>Disease::DOID:1612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ambenonium</td>\n",
       "      <td>Compound::DB01122</td>\n",
       "      <td>myasthenia gravis</td>\n",
       "      <td>Disease::MESH:D009157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>apalutamide</td>\n",
       "      <td>Compound::DB11901</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>artesunate</td>\n",
       "      <td>Compound::DB09274</td>\n",
       "      <td>malaria</td>\n",
       "      <td>Disease::MESH:D008288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>darolutamide</td>\n",
       "      <td>Compound::DB12941</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>delamanid</td>\n",
       "      <td>Compound::DB11637</td>\n",
       "      <td>tuberculosis</td>\n",
       "      <td>Disease::MESH:D014376</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>didox</td>\n",
       "      <td>Compound::DB12948</td>\n",
       "      <td>breast cancer</td>\n",
       "      <td>Disease::DOID:1612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>enzalutamide</td>\n",
       "      <td>Compound::DB08899</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>leucovorin</td>\n",
       "      <td>Compound::DB00650</td>\n",
       "      <td>osteosarcoma</td>\n",
       "      <td>Disease::MESH:D012516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>leuprolide</td>\n",
       "      <td>Compound::DB00007</td>\n",
       "      <td>prostate cancer</td>\n",
       "      <td>Disease::DOID:10283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>lumefantrine</td>\n",
       "      <td>Compound::DB06708</td>\n",
       "      <td>malaria</td>\n",
       "      <td>Disease::MESH:D008288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>panobinostat</td>\n",
       "      <td>Compound::DB06603</td>\n",
       "      <td>multiple myeloma</td>\n",
       "      <td>Disease::MESH:D009101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>pomalidomide</td>\n",
       "      <td>Compound::DB08910</td>\n",
       "      <td>multiple myeloma</td>\n",
       "      <td>Disease::MESH:D009101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>raltitrexed</td>\n",
       "      <td>Compound::DB00293</td>\n",
       "      <td>mesothelioma</td>\n",
       "      <td>Disease::MESH:D008654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>selinexor</td>\n",
       "      <td>Compound::DB11942</td>\n",
       "      <td>multiple myeloma</td>\n",
       "      <td>Disease::MESH:D009101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>solriamfetol</td>\n",
       "      <td>Compound::DB14754</td>\n",
       "      <td>narcolepsy</td>\n",
       "      <td>Disease::MESH:D009290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>trofosfamide</td>\n",
       "      <td>Compound::DB12902</td>\n",
       "      <td>medulloblastoma</td>\n",
       "      <td>Disease::MESH:D008527</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        compound   compound_drkg_id       rare_disease        disease_drkg_id\n",
       "0      alpelisib  Compound::DB12015      breast cancer     Disease::DOID:1612\n",
       "1     ambenonium  Compound::DB01122  myasthenia gravis  Disease::MESH:D009157\n",
       "4    apalutamide  Compound::DB11901    prostate cancer    Disease::DOID:10283\n",
       "5     artesunate  Compound::DB09274            malaria  Disease::MESH:D008288\n",
       "8   darolutamide  Compound::DB12941    prostate cancer    Disease::DOID:10283\n",
       "9      delamanid  Compound::DB11637       tuberculosis  Disease::MESH:D014376\n",
       "11         didox  Compound::DB12948      breast cancer     Disease::DOID:1612\n",
       "12  enzalutamide  Compound::DB08899    prostate cancer    Disease::DOID:10283\n",
       "15    leucovorin  Compound::DB00650       osteosarcoma  Disease::MESH:D012516\n",
       "16    leuprolide  Compound::DB00007    prostate cancer    Disease::DOID:10283\n",
       "17  lumefantrine  Compound::DB06708            malaria  Disease::MESH:D008288\n",
       "20  panobinostat  Compound::DB06603   multiple myeloma  Disease::MESH:D009101\n",
       "21  pomalidomide  Compound::DB08910   multiple myeloma  Disease::MESH:D009101\n",
       "22   raltitrexed  Compound::DB00293       mesothelioma  Disease::MESH:D008654\n",
       "24     selinexor  Compound::DB11942   multiple myeloma  Disease::MESH:D009101\n",
       "25  solriamfetol  Compound::DB14754         narcolepsy  Disease::MESH:D009290\n",
       "27  trofosfamide  Compound::DB12902    medulloblastoma  Disease::MESH:D008527"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check for rare disease links in DRKG\n",
    "test_rd_drkg = drkg_translated[drkg_translated['relationshipId'].str.contains('Compound:Disease')]\n",
    "test_rd_drkg[test_rd_drkg['head_entity'].str.lower() == 'apalutamide']\n",
    "\n",
    "# Count distinct links\n",
    "drkg_test_ct = drkg_test.copy()\n",
    "drkg_test_ct['compound'] = drkg_test_ct['compound'].str.lower()\n",
    "drkg_test_ct['rare_disease'] = drkg_test_ct['rare_disease'].str.lower()\n",
    "drkg_test_ct = drkg_test_ct.drop_duplicates(subset=['compound', 'rare_disease'])\n",
    "print(len(drkg_test_ct))\n",
    "drkg_test_ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "drkg_test.to_csv(\"test_raredisease_links.csv\")\n",
    "drkg_test_ct.to_csv(\"test_raredisease_links_unique.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "2d42bd2c408a4bd3bdebc2865ea42483": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": "hidden",
      "width": null
     }
    },
    "6e79954f9c514f099b336e3a11ad635e": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_6f2e0878218c4b38974a74a673b70751",
      "placeholder": "​",
      "style": "IPY_MODEL_98ccadd9e4a34d8282282a332fad6e0b",
      "value": "Downloading chembl_32_sqlite.tar.gz: 100%"
     }
    },
    "6f2e0878218c4b38974a74a673b70751": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "98ccadd9e4a34d8282282a332fad6e0b": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "a95662c4dc0949f4af40e8331cde378d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "info",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_eec53708d80f41e8b402060c9a1352b1",
      "max": 1,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_f87f682db17a4e6b837b9c14c85ec36f",
      "value": 1
     }
    },
    "b0797e6674144b1b92d2ff70918b917c": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "b355679609b749d698029d4f39d45503": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_6e79954f9c514f099b336e3a11ad635e",
       "IPY_MODEL_a95662c4dc0949f4af40e8331cde378d",
       "IPY_MODEL_d89d003254ce42f9a44572c7e26940ec"
      ],
      "layout": "IPY_MODEL_2d42bd2c408a4bd3bdebc2865ea42483"
     }
    },
    "d0a9a73a5fe14afb8b15a2addf4f080d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "d89d003254ce42f9a44572c7e26940ec": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_b0797e6674144b1b92d2ff70918b917c",
      "placeholder": "​",
      "style": "IPY_MODEL_d0a9a73a5fe14afb8b15a2addf4f080d",
      "value": " 4.27G/4.27G [03:38&lt;00:00, 21.7MB/s]"
     }
    },
    "eec53708d80f41e8b402060c9a1352b1": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": "20px"
     }
    },
    "f87f682db17a4e6b837b9c14c85ec36f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
