{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "DnQGl_D4786T"
   },
   "source": [
    "DRKG\n",
    "\n",
    "Adapted from: https://github.com/gnn4dr/DRKG/blob/master/drkg_with_dgl/loading_drkg_in_dgl.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "id": "PV4N5XSIKfTa"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/selinapi/opt/anaconda3/envs/simp_llm/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os \n",
    "import torch_geometric.transforms as T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from SIMP_LLM.DRKG_loading   import  get_triplets, read_tsv,filter_drkg,map_drkg_relationships,filter_interaction_subset,print_head\n",
    "from SIMP_LLM.DRKG_translate import  load_lookups\n",
    "from SIMP_LLM.DRKG_entity_processing import get_unique_entities, get_entity_lookup, convert_entitynames, flip_headtail\n",
    "from SIMP_LLM.raredisease_loading import get_orphan_data"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "VsLJKwgRxArZ"
   },
   "source": [
    "# 1) Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Triplets:\n",
      "\n",
      "[['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2157'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5264'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2158'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::3309'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::28912'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::811'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2159'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::821'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5627'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5624']]\n",
      "\n",
      " data/drkg.tsv  Dataframe:\n",
      "\n",
      "+----+------------+--------------------------------+-------------+\n",
      "|    | 0          | 1                              | 2           |\n",
      "|----+------------+--------------------------------+-------------|\n",
      "|  0 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2157  |\n",
      "|  1 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::5264  |\n",
      "|  2 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2158  |\n",
      "|  3 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::3309  |\n",
      "|  4 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::28912 |\n",
      "+----+------------+--------------------------------+-------------+\n",
      "\n",
      " data/relation_glossary.tsv  Dataframe:\n",
      "\n",
      "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
      "|    | Relation-name                              | Data-source   | Connected entity-types   | Interaction-type      | Description                                                                                                                                                                     | Reference for the description        |\n",
      "|----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------|\n",
      "|  0 | DGIDB::ACTIVATOR::Gene:Compound            | DGIDB         | Compound:Gene            | activation            | An activator interaction is when a drug activates a biological response from a target, although the mechanism by which it does so may not be understood.                        | http://www.dgidb.org/getting_started |\n",
      "|  1 | DGIDB::AGONIST::Gene:Compound              | DGIDB         | Compound:Gene            | agonism               | An agonist interaction occurs when a drug binds to a target receptor and activates the receptor to produce a biological response.                                               | http://www.dgidb.org/getting_started |\n",
      "|  2 | DGIDB::ALLOSTERIC MODULATOR::Gene:Compound | DGIDB         | Compound:Gene            | allosteric modulation | An allosteric modulator interaction occurs when drugs exert their effects on their protein targets via a different binding site than the natural (orthosteric) ligand site.     | http://www.dgidb.org/getting_started |\n",
      "|  3 | DGIDB::ANTAGONIST::Gene:Compound           | DGIDB         | Compound:Gene            | antagonism            | An antagonist interaction occurs when a drug blocks or dampens agonist-mediated responses rather than provoking a biological response itself upon binding to a target receptor. | http://www.dgidb.org/getting_started |\n",
      "|  4 | DGIDB::ANTIBODY::Gene:Compound             | DGIDB         | Compound:Gene            | antibody              | An antibody interaction occurs when an antibody drug specifically binds the target molecule.                                                                                    | http://www.dgidb.org/getting_started |\n",
      "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
      "Number of Rows Before Filtering: 107\n",
      "Number of Rows After Filtering: 10\n",
      "\n",
      "\n",
      "Filtered:\n",
      "24    DRUGBANK::treats::Compound:Disease\n",
      "35             GNBR::T::Compound:Disease\n",
      "36             GNBR::C::Compound:Disease\n",
      "37            GNBR::Sa::Compound:Disease\n",
      "38            GNBR::Pa::Compound:Disease\n",
      "39            GNBR::Mp::Compound:Disease\n",
      "40            GNBR::Pr::Compound:Disease\n",
      "41             GNBR::J::Compound:Disease\n",
      "68       Hetionet::CtD::Compound:Disease\n",
      "73       Hetionet::CpD::Compound:Disease\n",
      "dtype: object\n",
      "\n",
      "Relationships Mapped:\n",
      "0                    Compound treats the disease\n",
      "1    treatment/therapy (including investigatory)\n",
      "2            inhibits cell growth (esp. cancers)\n",
      "3                      side effect/adverse event\n",
      "4                            alleviates, reduces\n",
      "5            biomarkers (of disease progression)\n",
      "6                           prevents, suppresses\n",
      "7                   role in disease pathogenesis\n",
      "8                                      treatment\n",
      "9                                     palliation\n",
      "Name: Interaction-type, dtype: object\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n",
      "|         | 0                 | 1                                  | 2                     |\n",
      "|---------+-------------------+------------------------------------+-----------------------|\n",
      "| 1518268 | Compound::DB00003 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D003550 |\n",
      "| 1518269 | Compound::DB00004 | DRUGBANK::treats::Compound:Disease | Disease::MESH:C063419 |\n",
      "| 1518270 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D013167 |\n",
      "| 1518271 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D006086 |\n",
      "| 1518272 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D017497 |\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n"
     ]
    }
   ],
   "source": [
    "### 1) Read: This section reads DRKG and a glossary (used to map entities from codes to words)\n",
    "DATA_DIR           = os.path.join(\"data\")\n",
    "verbose            =  True \n",
    "triplets,drkg_df   =  get_triplets(drkg_file = os.path.join(DATA_DIR  ,'drkg.tsv'),             verbose=verbose)  # Read triplets (head,relationship,tail)\n",
    "relation_glossary  =  read_tsv(relation_file = os.path.join(DATA_DIR  ,'relation_glossary.tsv'),verbose=verbose)  # Read relationship mapping  \n",
    "\n",
    "\n",
    "### 2) Filter & Map Interactions: This section returns a list of interactions e.g. DRUGBANK::treats::Compound:Disease )\n",
    "# 2.1: First  we filter the interactions to only Compound-Disease\n",
    "# 2.2: Then   we map the codes -> text  (this will be use to further filter interactions based on text) e.g.  Hetionet::CpD::Compound:Disease -> palliation\n",
    "# 2.3: We use natural text to fitler  interactions based on terms such as \"treat\" (but we return the orignal interaction name )\n",
    "\n",
    "\n",
    "\n",
    "# modularize this in create_dataframe\n",
    "drkg_rx_dx_relations        = filter_drkg(data_frame = drkg_df ,  filter_column = 1 ,  filter_term = r'.*?Compound:Disease', verbose = verbose) # 2.1 Filter only Compound-Disease Interactions\n",
    "drkg_rx_dx_relations_mapped = map_drkg_relationships(drkg_rx_dx_relations,relation_glossary,verbose=verbose)                                    # 2.2 Map codes to text \n",
    "\n",
    "### 2.3 Filter Drug interactions Interaction types to only include: treat inhibit or alleviate interactions  ###\n",
    "drkg_rx_dx_relation_subset =  filter_interaction_subset(df                  = drkg_rx_dx_relations_mapped,\n",
    "                                                        filter_colunm_name = 'Interaction-type' ,\n",
    "                                                        regex_string       =  'treat|inhibit|alleviate',\n",
    "                                                        return_colunm_name =  'Relation-name')\n",
    "\n",
    "# 3) Use Filter Interactions to get Gilter DRKG \n",
    "drkg_df_filtered = drkg_df[drkg_df[1].isin(drkg_rx_dx_relation_subset)] # 3.1 Filter DRKG  to only  Compund-Disease \n",
    "print_head(df=drkg_df_filtered)\n",
    "\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "rx_dx_triplets   = drkg_df_filtered.values.tolist()                     # 3.2 Convert filtered DRKG to list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "KYBnEGL4Nixb",
    "outputId": "5bc6b100-d023-43d7-cf3c-488dc73b6475"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " data/hetionet-v1.0-nodes.tsv  Dataframe:\n",
      "\n",
      "+----+-------------------------+---------------------------+---------+\n",
      "|    | id                      | name                      | kind    |\n",
      "|----+-------------------------+---------------------------+---------|\n",
      "|  0 | Anatomy::UBERON:0000002 | uterine cervix            | Anatomy |\n",
      "|  1 | Anatomy::UBERON:0000004 | nose                      | Anatomy |\n",
      "|  2 | Anatomy::UBERON:0000006 | islet of Langerhans       | Anatomy |\n",
      "|  3 | Anatomy::UBERON:0000007 | pituitary gland           | Anatomy |\n",
      "|  4 | Anatomy::UBERON:0000010 | peripheral nervous system | Anatomy |\n",
      "+----+-------------------------+---------------------------+---------+\n",
      "\n",
      " Sample of Hetionet Data Types (Before processing):\n",
      "\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "|       | id                               | name                                      | kind                |\n",
      "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
      "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
      "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
      "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
      "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
      "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
      "| 14863 | Gene::1                          | A1BG                                      | Gene                |\n",
      "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
      "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
      "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
      "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
      "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "\n",
      " Sample of Hetionet Data Types (After processing):\n",
      "\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "|       | id                               | name                                      | kind                |\n",
      "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
      "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
      "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
      "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
      "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
      "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
      "| 14863 | Gene::1                          | A1BG gene                                 | Gene                |\n",
      "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
      "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
      "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
      "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
      "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "\n",
      " Gene ID Dataframe (After processing):\n",
      "\n",
      "+----+----------+--------------------------------------------+\n",
      "|    | GeneID   | description                                |\n",
      "|----+----------+--------------------------------------------|\n",
      "|  0 | Gene::1  | alpha-1-B glycoprotein (A1BG)              |\n",
      "|  1 | Gene::2  | alpha-2-macroglobulin (A2M)                |\n",
      "|  2 | Gene::3  | alpha-2-macroglobulin pseudogene 1 (A2MP1) |\n",
      "|  3 | Gene::9  | N-acetyltransferase 1 (NAT1)               |\n",
      "|  4 | Gene::10 | N-acetyltransferase 2 (NAT2)               |\n",
      "+----+----------+--------------------------------------------+\n",
      "\n",
      " DrugBank Dataframe (Before processing):\n",
      "\n",
      "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
      "|    | DrugBank ID   | Accession Numbers    | Common name         | CAS         | UNII       | Synonyms                                                                                                                                                              |   Standard InChI Key |\n",
      "|----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------|\n",
      "|  0 | DB00001       | BIOD00024 | BTD00024 | Lepirudin           | 138068-37-8 | Y43GF64R34 | [Leu1, Thr2]-63-desulfohirudin | Desulfatohirudin | Hirudin variant-1 | Lepirudin | Lepirudin recombinant | R-hirudin                                                 |                  nan |\n",
      "|  1 | DB00002       | BIOD00071 | BTD00071 | Cetuximab           | 205923-56-4 | PQX0D8J21J | Cetuximab | Cétuximab | Cetuximabum                                                                                                                                   |                  nan |\n",
      "|  2 | DB00003       | BIOD00001 | BTD00001 | Dornase alfa        | 143831-71-4 | 953A26OA1Y | Deoxyribonuclease (human clone 18-1 protein moiety) | Dornasa alfa | Dornase alfa | Dornase alfa, recombinant | Dornase alpha | Recombinant deoxyribonuclease (DNAse) |                  nan |\n",
      "|  3 | DB00004       | BIOD00084 | BTD00084 | Denileukin diftitox | 173146-27-5 | 25E79B5CTM | Denileukin | Denileukin diftitox | Interleukin-2/diptheria toxin fusion protein                                                                                       |                  nan |\n",
      "|  4 | DB00005       | BIOD00052 | BTD00052 | Etanercept          | 185243-69-0 | OP401G7OJC | Etanercept | etanercept-szzs | etanercept-ykro | Recombinant human TNF | rhu TNFR:Fc | rhu-TNFR:Fc | TNFR-Immunoadhesin                                               |                  nan |\n",
      "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
      "\n",
      " DrugBank Dataframe (After processing):\n",
      "\n",
      "+----+-------------------+---------------------+\n",
      "|    | DrugBank ID       | Common name         |\n",
      "|----+-------------------+---------------------|\n",
      "|  0 | Compound::DB00001 | Lepirudin           |\n",
      "|  1 | Compound::DB00002 | Cetuximab           |\n",
      "|  2 | Compound::DB00003 | Dornase alfa        |\n",
      "|  3 | Compound::DB00004 | Denileukin diftitox |\n",
      "|  4 | Compound::DB00005 | Etanercept          |\n",
      "+----+-------------------+---------------------+\n",
      "\n",
      " data/mimTitles.txt  Dataframe (Before processing):\n",
      "\n",
      "+----+----------------------+----------------------------------------+\n",
      "|    | MIM Number           | name                                   |\n",
      "|----+----------------------+----------------------------------------|\n",
      "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
      "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
      "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
      "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
      "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
      "+----+----------------------+----------------------------------------+\n",
      "\n",
      " data/mimTitles.txt  Dataframe (After processing):\n",
      "\n",
      "+----+----------------------+----------------------------------------+\n",
      "|    | MIM Number           | name                                   |\n",
      "|----+----------------------+----------------------------------------|\n",
      "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
      "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
      "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
      "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
      "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
      "+----+----------------------+----------------------------------------+\n",
      "\n",
      " MeSH Dictionary:\n",
      "\n",
      "+---------------+------------------------+\n",
      "|               | 0                      |\n",
      "|---------------+------------------------|\n",
      "| MESH::D000001 | Calcimycin             |\n",
      "| MESH::D000002 | Temefos                |\n",
      "| MESH::D000003 | Abattoirs              |\n",
      "| MESH::D000004 | Abbreviations as Topic |\n",
      "| MESH::D000005 | Abdomen                |\n",
      "+---------------+------------------------+\n"
     ]
    }
   ],
   "source": [
    "# 4) Load Data frames for translation\n",
    "hetionet_df, gene_df, drugbank_df, omim_df, mesh_dict, chebi_df, chembl_df = load_lookups(data_path=DATA_DIR,verbose=verbose)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Long-form orphan disease data (before processing):\n",
      "\n",
      "+----+-----------+------------------------------------------------------------------------+\n",
      "|    | cols      | data                                                                   |\n",
      "|----+-----------+------------------------------------------------------------------------|\n",
      "|  0 | Orphacode | 166024                                                                 |\n",
      "|  1 | Name      | Multiple epiphyseal dysplasia, Al-Gazali type                          |\n",
      "|  2 | Synonym   | Multiple epiphyseal dysplasia-macrocephaly-distinctive facies syndrome |\n",
      "|  3 | Source    | ICD-10                                                                 |\n",
      "|  4 | Reference | Q77.3                                                                  |\n",
      "+----+-----------+------------------------------------------------------------------------+\n",
      "\n",
      " Long-form orphan disease data (after processing):\n",
      "\n",
      "+----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------+\n",
      "|    | cols                    | data                                                                   |   disease_id | code_source   | code   |\n",
      "|----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------|\n",
      "|  0 | Orphacode               | 166024                                                                 |       166024 |               |        |\n",
      "|  1 | Name                    | Multiple epiphyseal dysplasia, Al-Gazali type                          |       166024 |               |        |\n",
      "|  2 | Synonym                 | Multiple epiphyseal dysplasia-macrocephaly-distinctive facies syndrome |       166024 |               |        |\n",
      "|  4 | Reference               | Q77.3                                                                  |       166024 | ICD-10        | Q77.3  |\n",
      "|  6 | Disordermappingrelation | NTBT (ORPHA code's Narrower Term maps to a Broader Term)               |       166024 | ICD-10        | Q77.3  |\n",
      "+----+-------------------------+------------------------------------------------------------------------+--------------+---------------+--------+\n",
      "\n",
      " Orphan disease name/summary data:\n",
      "\n",
      "+--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+\n",
      "|   disease_id |   Orphacode | Name                                                   | Synonym                                                    | Definition                                                                                                                                                                                                                                                                                                                                                                                             |   Info |\n",
      "|--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------|\n",
      "|           10 |          10 | 48,XXYY syndrome                                       | nan                                                        | A rare sex chromosome number anomaly disorder characterized, genetically, by the presence of an extra X and Y chromosome in males and, clinically, by tall stature, dysfunctional testes associated with infertility and insufficient testosterone production, cognitive, affective and social functioning impairments, global developmental delay, and an increased risk of congenital malformations. |    nan |\n",
      "|          100 |         100 | Ataxia-telangiectasia                                  | Louis-Bar syndrome                                         | A rare disorder characterized by the association of severe combined immunodeficiency (affecting mainly the humoral immune response) with progressive cerebellar ataxia. It is characterized by neurological signs, telangiectasia, increased susceptibility to infections and a higher risk of cancer.                                                                                                 |    nan |\n",
      "|         1000 |        1000 | Ocular albinism with late-onset sensorineural deafness | Ocular albinism with late-onset sensorineural hearing loss | Ocular albinism with late-onset sensorineural deafness is a rare, X-linked inherited subtype of ocular albinism characterized by severe visual impairment, translucent pale-blue irises, a reduction in the retinal pigment and moderately severe deafness with onset ranging from adolescence to fourth or fifth decade of life.                                                                      |    nan |\n",
      "|       100000 |      100000 | Reticular perineurioma                                 | nan                                                        | nan                                                                                                                                                                                                                                                                                                                                                                                                    |    nan |\n",
      "|       100001 |      100001 | Sclerosing perineurioma                                | nan                                                        | nan                                                                                                                                                                                                                                                                                                                                                                                                    |    nan |\n",
      "+--------------+-------------+--------------------------------------------------------+------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------+\n",
      "\n",
      " Orphan disease codes:\n",
      "\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+\n",
      "|    |   Orphacode | Name                                          | code_source   | code     | Disordermappingrelation                                  | Disordermappingicdrelation                          | Disordermappingvalidationstatus   |\n",
      "|----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------|\n",
      "|  0 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | ICD-10        | Q77.3    | NTBT (ORPHA code's Narrower Term maps to a Broader Term) | Attributed (The ICD code is attributed by Orphanet) | Validated                         |\n",
      "|  1 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | OMIM          | 607131   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  2 |      166024 | Multiple epiphyseal dysplasia, Al-Gazali type | UMLS          | C1846722 | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  3 |          58 | Alexander disease                             | OMIM          | 203450   | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "|  4 |          58 | Alexander disease                             | MeSH          | D038261  | E (Exact mapping: the two concepts are equivalent)       | nan                                                 | Validated                         |\n",
      "+----+-------------+-----------------------------------------------+---------------+----------+----------------------------------------------------------+-----------------------------------------------------+-----------------------------------+\n"
     ]
    }
   ],
   "source": [
    "# Load orphan disease names and codes (28 Nov 2022 version)\n",
    "orphan_names, orphan_codes = get_orphan_data(os.path.join(DATA_DIR, 'en_product1-Orphadata.xml'), verbose=verbose)\n",
    "\n",
    "# Get orphan disease MeSH codes\n",
    "orphan_codes_mesh = orphan_codes[orphan_codes['code_source']=='MeSH'].copy()\n",
    "orphan_codes_mesh['id'] = 'MESH::'+orphan_codes_mesh['code']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HoImejcxpFYP",
    "outputId": "904e49c4-8889-4f2d-fcce-26dc99331b86"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+------------------------------+--------------------------------+------------------------------------------------------+\n",
      "|    | 0                            | 1                              | 2                                                    |\n",
      "|----+------------------------------+--------------------------------+------------------------------------------------------|\n",
      "|  0 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | coagulation factor VIII (F8)                         |\n",
      "|  1 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | phytanoyl-CoA 2-hydroxylase (PHYH)                   |\n",
      "|  2 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | coagulation factor IX (F9)                           |\n",
      "|  3 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | heat shock protein family A (Hsp70) member 5 (HSPA5) |\n",
      "|  4 | coagulation factor VIII (F8) | bioarx::HumGenHumGen:Gene:Gene | immunoglobulin kappa variable 3-20 (IGKV3-20)        |\n",
      "+----+------------------------------+--------------------------------+------------------------------------------------------+\n",
      "Number of unique DRKG entities:  97238\n",
      "Number of translated entities:  64384\n",
      "Number of untranslated entities:  32854\n",
      "Percentage of entities translated:  66.2 %\n",
      "Total DRKG relationships:  5874261\n",
      "Translated DRKG relationships:  5707119\n",
      "Percentage of relationships fully translated:  97.2 %\n"
     ]
    }
   ],
   "source": [
    "# Make dictionaries for codes\n",
    "code_df   = pd.concat([hetionet_df[['name', 'id']], \n",
    "                       gene_df.rename(columns = {\"description\":\"name\", \"GeneID\":\"id\"}),\n",
    "                       drugbank_df.rename(columns = {\"Common name\":\"name\", \"DrugBank ID\":\"id\"}),\n",
    "                       omim_df.rename(columns = {\"MIM Number\":\"id\"}),\n",
    "                       chebi_df.rename(columns = {\"NAME\":\"name\", \"CHEBI_ACCESSION\":\"id\"}),\n",
    "                       chembl_df.rename(columns = {\"pref_name\":\"name\", \"chembl_id\":\"id\"}),\n",
    "                       orphan_codes_mesh.rename(columns = {\"Name\":\"name\"}) # SP 05/24/23 added orphan disease MeSH terms\n",
    "                       ], ignore_index=True, axis=0).drop_duplicates() \n",
    "code_dict = pd.Series(code_df['name'].values, index=code_df['id']).to_dict() | mesh_dict # Convert node df to dict and merge with MeSH dictionary\n",
    "\n",
    "# Get unique DRKG entities\n",
    "drkg_entities = get_unique_entities(drkg_df, [0,2])\n",
    "\n",
    "# Create and use convert_entitynames function\n",
    "drkg_entity_df, drkg_unmatched = get_entity_lookup(drkg_entities, code_dict)\n",
    "\n",
    "# Create final node dictionary\n",
    "node_dict = pd.Series(drkg_entity_df['name'].values, index=drkg_entity_df['drkg_id']).to_dict() \n",
    "\n",
    "# Initialize translated DRKG and manually clean heads/tails for one case where they were flipped\n",
    "drkg_translated    = drkg_df.copy()\n",
    "drkg_translated = flip_headtail(drkg_translated, 'Gene:Compound')\n",
    "\n",
    "# Map DRKG to translated entity names\n",
    "drkg_translated = convert_entitynames(drkg_translated, 0, node_dict)\n",
    "drkg_translated = convert_entitynames(drkg_translated, 2, node_dict)\n",
    "drkg_translated = drkg_translated.dropna()\n",
    "print_head(drkg_translated) \n",
    "\n",
    "# Summarize percentage translated\n",
    "print(\"Number of unique DRKG entities: \", len(drkg_entities)) # should be 97238\n",
    "print(\"Number of translated entities: \", drkg_entity_df.shape[0])\n",
    "print(\"Number of untranslated entities: \", drkg_unmatched.shape[0])\n",
    "pct_entity_translated = drkg_entity_df.shape[0]/len(drkg_entities)\n",
    "print('Percentage of entities translated: ', round(pct_entity_translated*100,1), '%')\n",
    "\n",
    "print('Total DRKG relationships: ', drkg_df.shape[0])\n",
    "print('Translated DRKG relationships: ', drkg_translated.shape[0])\n",
    "pct_translated = drkg_translated.shape[0]/drkg_df.shape[0]\n",
    "print('Percentage of relationships fully translated: ', round(pct_translated*100,1), '%')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drkg_id</th>\n",
       "      <th>Data-source</th>\n",
       "      <th>Connected entity-types</th>\n",
       "      <th>Interaction-type</th>\n",
       "      <th>Description</th>\n",
       "      <th>Reference for the description</th>\n",
       "      <th>head_entity</th>\n",
       "      <th>tail_entity</th>\n",
       "      <th>relation_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DGIDB::ACTIVATOR::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>activation</td>\n",
       "      <td>An activator interaction is when a drug activa...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DGIDB::AGONIST::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>agonism</td>\n",
       "      <td>An agonist interaction occurs when a drug bind...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>activation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DGIDB::ALLOSTERIC MODULATOR::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>allosteric modulation</td>\n",
       "      <td>An allosteric modulator interaction occurs whe...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>allosteric modulation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DGIDB::ANTAGONIST::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>antagonism</td>\n",
       "      <td>An antagonist interaction occurs when a drug b...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>antagonism</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DGIDB::ANTIBODY::Gene:Compound</td>\n",
       "      <td>DGIDB</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>antibody</td>\n",
       "      <td>An antibody interaction occurs when an antibod...</td>\n",
       "      <td>http://www.dgidb.org/getting_started</td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>antibody</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>bioarx::Covid2_acc_host_gene::Disease:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Disease:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Interactions between 27 viral  proteins,  and ...</td>\n",
       "      <td></td>\n",
       "      <td>Disease</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>bioarx::DrugHumGen:Compound:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>bioarx::DrugVirGen:Compound:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Compound:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Compound</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>bioarx::HumGenHumGen:Gene:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Gene:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Protein-protein interaction</td>\n",
       "      <td></td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>bioarx::VirGenHumGen:Gene:Gene</td>\n",
       "      <td>BIBLIOGRAPHY</td>\n",
       "      <td>Gene:Gene</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Protein-protein interaction</td>\n",
       "      <td></td>\n",
       "      <td>Gene</td>\n",
       "      <td>Gene</td>\n",
       "      <td>interaction</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>107 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        drkg_id   Data-source   \n",
       "0               DGIDB::ACTIVATOR::Gene:Compound         DGIDB  \\\n",
       "1                 DGIDB::AGONIST::Gene:Compound         DGIDB   \n",
       "2    DGIDB::ALLOSTERIC MODULATOR::Gene:Compound         DGIDB   \n",
       "3              DGIDB::ANTAGONIST::Gene:Compound         DGIDB   \n",
       "4                DGIDB::ANTIBODY::Gene:Compound         DGIDB   \n",
       "..                                          ...           ...   \n",
       "102  bioarx::Covid2_acc_host_gene::Disease:Gene  BIBLIOGRAPHY   \n",
       "103            bioarx::DrugHumGen:Compound:Gene  BIBLIOGRAPHY   \n",
       "104            bioarx::DrugVirGen:Compound:Gene  BIBLIOGRAPHY   \n",
       "105              bioarx::HumGenHumGen:Gene:Gene  BIBLIOGRAPHY   \n",
       "106              bioarx::VirGenHumGen:Gene:Gene  BIBLIOGRAPHY   \n",
       "\n",
       "    Connected entity-types       Interaction-type   \n",
       "0            Compound:Gene             activation  \\\n",
       "1            Compound:Gene                agonism   \n",
       "2            Compound:Gene  allosteric modulation   \n",
       "3            Compound:Gene             antagonism   \n",
       "4            Compound:Gene               antibody   \n",
       "..                     ...                    ...   \n",
       "102           Disease:Gene            interaction   \n",
       "103          Compound:Gene            interaction   \n",
       "104          Compound:Gene            interaction   \n",
       "105              Gene:Gene            interaction   \n",
       "106              Gene:Gene            interaction   \n",
       "\n",
       "                                           Description   \n",
       "0    An activator interaction is when a drug activa...  \\\n",
       "1    An agonist interaction occurs when a drug bind...   \n",
       "2    An allosteric modulator interaction occurs whe...   \n",
       "3    An antagonist interaction occurs when a drug b...   \n",
       "4    An antibody interaction occurs when an antibod...   \n",
       "..                                                 ...   \n",
       "102  Interactions between 27 viral  proteins,  and ...   \n",
       "103                                                      \n",
       "104                                                      \n",
       "105                        Protein-protein interaction   \n",
       "106                        Protein-protein interaction   \n",
       "\n",
       "            Reference for the description head_entity tail_entity   \n",
       "0    http://www.dgidb.org/getting_started    Compound        Gene  \\\n",
       "1    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "2    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "3    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "4    http://www.dgidb.org/getting_started    Compound        Gene   \n",
       "..                                    ...         ...         ...   \n",
       "102                                           Disease        Gene   \n",
       "103                                          Compound        Gene   \n",
       "104                                          Compound        Gene   \n",
       "105                                              Gene        Gene   \n",
       "106                                              Gene        Gene   \n",
       "\n",
       "             relation_name  \n",
       "0               activation  \n",
       "1               activation  \n",
       "2    allosteric modulation  \n",
       "3               antagonism  \n",
       "4                 antibody  \n",
       "..                     ...  \n",
       "102            interaction  \n",
       "103            interaction  \n",
       "104            interaction  \n",
       "105            interaction  \n",
       "106            interaction  \n",
       "\n",
       "[107 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "\n",
    "\n",
    "# Update relation glossary \n",
    "relation_df = relation_glossary.copy().rename(columns={'Relation-name':'drkg_id'})\n",
    "relation_df[['head_entity','tail_entity']] = relation_df['drkg_id'].str.split('::', expand=True)[2].str.split(':', expand=True) # Set head and tail nodes\n",
    "\n",
    "# Manually fix head and tail nodes for DGIDB relations, which reverse compound-gene interactions\n",
    "relation_df.loc[relation_df['drkg_id'].str.contains('Gene:Compound'),'head_entity'] = 'Compound'\n",
    "relation_df.loc[relation_df['drkg_id'].str.contains('Gene:Compound'),'tail_entity'] = 'Gene'\n",
    "\n",
    "# Fix bioarx entries without the second \"::\" delimiter\n",
    "bioarx_ht = relation_df['drkg_id'].str.split(':', expand=True)[[3,4]]\n",
    "relation_df['head_entity'] = np.where(relation_df['head_entity'].isna(), bioarx_ht[3], relation_df['head_entity'])\n",
    "relation_df['tail_entity'] = np.where(relation_df['tail_entity'].isna(), bioarx_ht[4], relation_df['tail_entity'])\n",
    "\n",
    "# Add mapped relation group labels\n",
    "relation_groups = [['activation', 'agonism', 'agonism, activation', 'activates, stimulates'],\n",
    "    ['antagonism', 'blocking', 'antagonism, blocking'],\n",
    "    ['binding', 'binding, ligand (esp. receptors)'],\n",
    "    ['blocking', 'channel blocking'],\n",
    "    ['inhibition', 'inhibits cell growth (esp. cancers)', 'inhibits'],\n",
    "    ['enzyme', 'enzyme activity'],\n",
    "    ['upregulation', 'increases expression/production'],\n",
    "    ['downregulation', 'decreases expression/production'],\n",
    "    ['Compound treats the disease', 'treatment/therapy (including investigatory)', 'treatment']]\n",
    "\n",
    "relation_df['relation_name'] = relation_df['Interaction-type']\n",
    "\n",
    "for grp in relation_groups:\n",
    "    relation_df_subset = relation_df[relation_df['Interaction-type'].isin(grp)].copy()\n",
    "    for entities in relation_df_subset['Connected entity-types'].unique():\n",
    "        subgrp = relation_df_subset[relation_df_subset['Connected entity-types'] == entities]['Interaction-type'].unique()\n",
    "        relation_df.loc[(relation_df_subset['Connected entity-types'] == entities) & (relation_df['Interaction-type'].isin(subgrp)), 'relation_name'] = subgrp[0]\n",
    "\n",
    "# Remove special characters from relation names\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace(',|/', ' or ', regex=True)\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace('esp.','especially')\n",
    "relation_df['relation_name'] = relation_df['relation_name'].str.replace('\\(|\\)|-|\\.', '', regex=True)\n",
    "\n",
    "# Check if any relationshp names still have non alpha numeric values except space\n",
    "error_relation_names = relation_df['relation_name'][relation_df['relation_name'].str.replace(' ', '').str.contains(r\"[^a-zA-Z0-9]+\", regex=True)].drop_duplicates()\n",
    "if len(error_relation_names):\n",
    "    print('Warning: The following relation names contain special characters, which can interfere with PyG/GraphSage')\n",
    "    print(error_relation_names)\n",
    "    \n",
    "relation_df\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "oTAJOqDVc_Hg"
   },
   "source": [
    "# 3) BioLinkBERT embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from torch_geometric.data import HeteroData\n",
    "from SIMP_LLM.llm_encode import EntityEncoder\n",
    "from SIMP_LLM.dataloader_mappings import create_mapping, create_edges, embed_entities, embed_edges"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Set variables and load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "id": "QkZKp6CJiw6f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-------------------------+\n",
      "|    | drkg_id                            | Data-source   | Connected entity-types   | Interaction-type            | Description                                                                                                                                                                                            | Reference for the description        | head_entity   | tail_entity   | relation_name           |\n",
      "|----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-------------------------|\n",
      "|  0 | DGIDB::ACTIVATOR::Gene:Compound    | DGIDB         | Compound:Gene            | activation                  | An activator interaction is when a drug activates a biological response from a target, although the mechanism by which it does so may not be understood.                                               | http://www.dgidb.org/getting_started | Compound      | Gene          | Compound activates gene |\n",
      "|  1 | DGIDB::AGONIST::Gene:Compound      | DGIDB         | Compound:Gene            | agonism                     | An agonist interaction occurs when a drug binds to a target receptor and activates the receptor to produce a biological response.                                                                      | http://www.dgidb.org/getting_started | Compound      | Gene          | Compound activates gene |\n",
      "|  8 | DGIDB::INHIBITOR::Gene:Compound    | DGIDB         | Compound:Gene            | inhibition                  | In inhibitor interactions, the drug binds to a target and decreases its expression or activity. Most interactions of this class are enzyme inhibitors, which bind an enzyme to reduce enzyme activity. | http://www.dgidb.org/getting_started | Compound      | Gene          | Inhibition              |\n",
      "| 18 | DRUGBANK::treats::Compound:Disease | DRUGBANK      | Compound:Disease         | Compound treats the disease | We extract  the  associated  conditions  related  to  compound entities in the drugbank.                                                                                                               |                                      | Compound      | Disease       | Compound treats disease |\n",
      "| 19 | GNBR::A+::Compound:Gene            | GNBR          | Compound:Gene            | agonism, activation         |                                                                                                                                                                                                        |                                      | Compound      | Gene          | Compound activates gene |\n",
      "+----+------------------------------------+---------------+--------------------------+-----------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+---------------+---------------+-------------------------+\n",
      "+----+-------------+---------------------------------+---------------------------------------------------------+\n",
      "|    | 0           | 1                               | 2                                                       |\n",
      "|----+-------------+---------------------------------+---------------------------------------------------------|\n",
      "|  0 | SNS-032     | DGIDB::INHIBITOR::Gene:Compound | cyclin dependent kinase 7 (CDK7)                        |\n",
      "|  1 | Quizartinib | DGIDB::INHIBITOR::Gene:Compound | platelet derived growth factor receptor beta (PDGFRB)   |\n",
      "|  2 | Sufentanil  | DGIDB::AGONIST::Gene:Compound   | opioid receptor mu 1 (OPRM1)                            |\n",
      "|  3 | PF-04691502 | DGIDB::INHIBITOR::Gene:Compound | phosphoinositide-3-kinase regulatory subunit 2 (PIK3R2) |\n",
      "|  4 | Ramelteon   | DGIDB::AGONIST::Gene:Compound   | melatonin receptor 1A (MTNR1A)                          |\n",
      "+----+-------------+---------------------------------+---------------------------------------------------------+\n",
      "+-------+-------------------+-------------------+----------------+---------------+-----------------+-----------------+---------+\n",
      "|       | drkg_id           | drkg_dict_id      | name           | entity_type   | ontology_code   | ontology_name   | code    |\n",
      "|-------+-------------------+-------------------+----------------+---------------+-----------------+-----------------+---------|\n",
      "| 21962 | Compound::DB00003 | Compound::DB00003 | Dornase alfa   | Compound      | DB00003         | drugbank        | DB00003 |\n",
      "| 21964 | Compound::DB00005 | Compound::DB00005 | Etanercept     | Compound      | DB00005         | drugbank        | DB00005 |\n",
      "| 22319 | Compound::DB00371 | Compound::DB00371 | Meprobamate    | Compound      | DB00371         | drugbank        | DB00371 |\n",
      "| 22653 | Compound::DB00708 | Compound::DB00708 | Sufentanil     | Compound      | DB00708         | drugbank        | DB00708 |\n",
      "| 22688 | Compound::DB00743 | Compound::DB00743 | Gadobenic acid | Compound      | DB00743         | drugbank        | DB00743 |\n",
      "+-------+-------------------+-------------------+----------------+---------------+-----------------+-----------------+---------+\n"
     ]
    }
   ],
   "source": [
    "## Set variables\n",
    "device   = \"cpu\"\n",
    "Encoder  = EntityEncoder(device = device )\n",
    "run_full_sample = 0\n",
    "\n",
    "if run_full_sample:\n",
    "    # Run full DRKG\n",
    "    entity_df = drkg_entity_df.copy()\n",
    "    hrt_data = drkg_translated.copy()\n",
    "    relation_lookup = relation_df.copy()\n",
    "else:\n",
    "    # Create relationship subset for testing\n",
    "    test_relation_df = relation_df[relation_df['Connected entity-types'].isin(['Compound:Gene', 'Disease:Gene', 'Compound:Disease', 'Gene:Gene'])].copy()\n",
    "    test_relation_df['relation_name'] = None\n",
    "\n",
    "    activation_list = ['activation', 'agonism', 'agonism, activation'] \n",
    "    treat_list = ['Compound treats the disease', 'treats']\n",
    "    gene_drug_list = ['inhibition']\n",
    "\n",
    "    test_relation_df['relation_name'][test_relation_df['Interaction-type'].isin(activation_list)] = 'Compound activates gene'\n",
    "    test_relation_df['relation_name'][test_relation_df['Interaction-type'].isin(treat_list)] = 'Compound treats disease'\n",
    "    test_relation_df['relation_name'][test_relation_df['Interaction-type'].isin(gene_drug_list)] = 'Inhibition'\n",
    "    test_relation_df = test_relation_df[~test_relation_df['relation_name'].isna()]\n",
    "    print_head(test_relation_df)\n",
    "\n",
    "    # Create test sample of DRKG relationships filtering to these relations (for full sample: delete and use drkg_entity_df)\n",
    "    test_hrt_df = drkg_translated[drkg_translated[1].isin(test_relation_df['drkg_id'])]\n",
    "    test_hrt_df = test_hrt_df.groupby(1).head(3).reset_index(drop=True)\n",
    "    test_unique_entities = get_unique_entities(test_hrt_df, columns=[0,2])\n",
    "    test_entity_df = drkg_entity_df[drkg_entity_df['name'].isin(test_unique_entities)]\n",
    "    print_head(test_hrt_df)\n",
    "    print_head(test_entity_df)\n",
    "\n",
    "\n",
    "    entity_df = test_entity_df.copy()\n",
    "    hrt_data = test_hrt_df.copy()\n",
    "    relation_lookup = relation_df.copy()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {
    "id": "ywHcEVMiAdOv"
   },
   "source": [
    "### Build HeteroData Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HeteroData(\n",
      "  \u001b[1mCompound\u001b[0m={ x=[13, 768] },\n",
      "  \u001b[1mDisease\u001b[0m={ x=[3, 768] },\n",
      "  \u001b[1mGene\u001b[0m={ x=[24, 768] },\n",
      "  \u001b[1m(Compound, activation, Gene)\u001b[0m={\n",
      "    edge_index=[2, 9],\n",
      "    edge_label=[9, 768]\n",
      "  },\n",
      "  \u001b[1m(Compound, inhibition, Gene)\u001b[0m={\n",
      "    edge_index=[2, 3],\n",
      "    edge_label=[3, 768]\n",
      "  },\n",
      "  \u001b[1m(Compound, Compound treats the disease, Disease)\u001b[0m={\n",
      "    edge_index=[2, 3],\n",
      "    edge_label=[3, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, activates or  stimulates, Gene)\u001b[0m={\n",
      "    edge_index=[2, 3],\n",
      "    edge_label=[3, 768]\n",
      "  },\n",
      "  \u001b[1m(Gene, inhibition, Gene)\u001b[0m={\n",
      "    edge_index=[2, 3],\n",
      "    edge_label=[3, 768]\n",
      "  }\n",
      ")\n",
      "Unique Compounds: 13 \t Matrix shape: torch.Size([13, 768])\n",
      "Unique Diseases: 3 \t Matrix shape: torch.Size([3, 768])\n",
      "Unique Genes: 24 \t Matrix shape: torch.Size([24, 768])\n"
     ]
    }
   ],
   "source": [
    "# Initialize heterograph object\n",
    "data = HeteroData()\n",
    "\n",
    "# Embed entities, add to graph, and save embedding mapping dictionary of dictionaries\n",
    "mapping_dict = embed_entities(entity_df, data, Encoder, device) \n",
    "\n",
    "# Embed relationships, add to graph, and save relation embeddings/mapping dictionary\n",
    "relation_X, relation_mapping = embed_edges(hrt_data, relation_lookup, data, mapping_dict, Encoder, device)\n",
    "\n",
    "# Print summary\n",
    "#data = T.ToUndirected()(data)\n",
    "\n",
    "print(data)\n",
    "for ent_type in entity_df['entity_type'].unique():\n",
    "    print(f\"Unique {ent_type}s: {len(mapping_dict[ent_type])} \\t Matrix shape: {data[ent_type].x.shape }\")\n",
    "    # print(mapping_dict[ent_type]) # Prints whole dictionary so delete/uncomment if using all entities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "torch.Size([2, 3])\n",
      "HeteroData(\n",
      "  \u001b[1mCompound\u001b[0m={ x=[13, 768] },\n",
      "  \u001b[1mDisease\u001b[0m={ x=[3, 768] },\n",
      "  \u001b[1m(Compound, treats, Disease)\u001b[0m={ edge_index=[2, 3] },\n",
      "  \u001b[1m(Disease, rev_treats, Compound)\u001b[0m={ edge_index=[2, 3] }\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "# Experimental: Ignore ##\n",
    "data2 = HeteroData()\n",
    "data2['Compound'].x  = data[\"Compound\"][\"x\"]\n",
    "data2['Disease'].x   = data[\"Disease\"][\"x\"]\n",
    "ctd                  = data[(\"Compound\", \"Compound treats the disease\", \"Disease\")][\"edge_index\"] #.to_sparse()\n",
    "data2['Compound', 'treats', 'Disease'].edge_index = ctd\n",
    "target_label                                     = data2['Compound', 'treats', 'Disease'].edge_index\n",
    "#data['Compound', 'treats', 'Disease'].edge_label =  torch.ones(target_label.shape[1],)\n",
    "data2 = T.ToUndirected()(data2)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "print(target_label.shape)\n",
    "\n",
    "print(data2)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GRAPH SAGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch_geometric as pyg\n",
    "import torch_geometric\n",
    "from torch_geometric.nn import SAGEConv, to_hetero\n",
    "from   torch.utils.data      import Dataset, DataLoader\n",
    "from   torch_geometric.data  import Data\n",
    "from   torch_geometric.utils import negative_sampling\n",
    "\n",
    "from torch_geometric.nn import SAGEConv, to_hetero\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "class GNNStack(torch.nn.Module):\n",
    "    def __init__(self, input_dim:int, hidden_dim:int, output_dim:int, layers:int, dropout:float=0.3, return_embedding=False):\n",
    "        \"\"\"\n",
    "            A stack of GraphSAGE Module \n",
    "            input_dim        <int>:   Input dimension\n",
    "            hidden_dim       <int>:   Hidden dimension\n",
    "            output_dim       <int>:   Output dimension\n",
    "            layers           <int>:   Number of layers\n",
    "            dropout          <float>: Dropout rate\n",
    "            return_embedding <bool>:  Whether to return the return_embeddingedding of the input graph\n",
    "        \"\"\"\n",
    "        \n",
    "        super(GNNStack, self).__init__()\n",
    "        graphSage_conv               = pyg.nn.SAGEConv\n",
    "        self.dropout                 = dropout\n",
    "        self.layers                  = layers\n",
    "        self.return_embedding        = return_embedding\n",
    "\n",
    "        ### Initalize the layers ###\n",
    "        self.convs                   = nn.ModuleList()                      # ModuleList to hold the layers\n",
    "        for l in range(self.layers):\n",
    "            if l == 0:\n",
    "                ### First layer  maps from input_dim to hidden_dim ###\n",
    "                self.convs.append(graphSage_conv(input_dim, hidden_dim))\n",
    "            else:\n",
    "                ### All other layers map from hidden_dim to hidden_dim ###\n",
    "                self.convs.append(graphSage_conv(hidden_dim, hidden_dim))\n",
    "\n",
    "        # post-message-passing processing MLP\n",
    "        self.post_mp = nn.Sequential(\n",
    "                                     nn.Linear(hidden_dim, hidden_dim), \n",
    "                                     nn.Dropout(self.dropout),\n",
    "                                     nn.Linear(hidden_dim, output_dim))\n",
    "\n",
    "    def forward(self, x, edge_index):\n",
    "        for i in range(self.layers):\n",
    "            x = self.convs[i](x, edge_index)\n",
    "            x = F.relu(x)\n",
    "            x = F.dropout(x, p=self.dropout, training=self.training)\n",
    "\n",
    "        x = self.post_mp(x)\n",
    "\n",
    "        # Return final layer of return_embeddingeddings if specified\n",
    "        if self.return_embedding:\n",
    "            return x\n",
    "\n",
    "        # Else return class probabilities\n",
    "        return F.log_softmax(x, dim=1)\n",
    "\n",
    "    def loss(self, pred, label):\n",
    "        return F.nll_loss(pred, label)\n",
    "    \n",
    "\n",
    "\n",
    "class LinkPredictorMLP(nn.Module):\n",
    "    def __init__(self, in_channels:int, hidden_channels:int, out_channels:int, n_layers:int,dropout_probabilty:float=0.3):\n",
    "        \"\"\"\n",
    "        Args:\n",
    "            in_channels (int):     Number of input features.\n",
    "            hidden_channels (int): Number of hidden features.\n",
    "            out_channels (int):    Number of output features.\n",
    "            n_layers (int):        Number of MLP layers.\n",
    "            dropout (float):       Dropout probability.\n",
    "            \"\"\"\n",
    "        super(LinkPredictorMLP, self).__init__()\n",
    "        self.dropout_probabilty    = dropout_probabilty  # dropout probability\n",
    "        self.mlp_layers            = nn.ModuleList()     # ModuleList: is a list of modules\n",
    "        self.non_linearity         = F.relu              # non-linearity\n",
    "        \n",
    "        for i in range(n_layers - 1):                                 \n",
    "            if i == 0:\n",
    "                self.mlp_layers.append(nn.Linear(in_channels, hidden_channels))          # input layer (in_channels, hidden_channels)\n",
    "            else:\n",
    "                self.mlp_layers.append(nn.Linear(hidden_channels, hidden_channels))      # hidden layers (hidden_channels, hidden_channels)\n",
    "\n",
    "        self.mlp_layers.append(nn.Linear(hidden_channels, out_channels))                 # output layer (hidden_channels, out_channels)\n",
    "\n",
    "\n",
    "    def reset_parameters(self):\n",
    "        for mlp_layer in self.mlp_layers:\n",
    "            mlp_layer.reset_parameters()\n",
    "\n",
    "    def forward(self, x_i, x_j):\n",
    "        x = x_i * x_j                                                     # element-wise multiplication\n",
    "        for mlp_layer in self.mlp_layers[:-1]:                            # iterate over all layers except the last one\n",
    "            x = mlp_layer(x)                                              # apply linear transformation\n",
    "            x = self.non_linearity(x)                                     # Apply non linear activation function\n",
    "            x = F.dropout(x, p=self.dropout_probabilty,training=self.training)      # Apply dropout\n",
    "        x = self.mlp_layers[-1](x)                                        # apply linear transformation to the last layer\n",
    "        x = torch.sigmoid(x)                                              # apply sigmoid activation function to get the probability\n",
    "        return x\n",
    "    \n",
    "### We will use This function to save our best model during trainnig ###\n",
    "def save_torch_model(model,epoch,PATH:str,optimizer):\n",
    "    print(f\"Saving Model in Path {PATH}\")\n",
    "    torch.save({'epoch': epoch,\n",
    "                'model_state_dict': model.state_dict(),\n",
    "                'optimizer':optimizer,      \n",
    "                }, PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GraphModule(\n",
      "  (convs): ModuleList(\n",
      "    (0): ModuleDict(\n",
      "      (Compound__treats__Disease): SAGEConv(768, 524, aggr=mean)\n",
      "      (Disease__rev_treats__Compound): SAGEConv(768, 524, aggr=mean)\n",
      "    )\n",
      "    (1-2): 2 x ModuleDict(\n",
      "      (Compound__treats__Disease): SAGEConv(524, 524, aggr=mean)\n",
      "      (Disease__rev_treats__Compound): SAGEConv(524, 524, aggr=mean)\n",
      "    )\n",
      "  )\n",
      "  (post_mp): ModuleList(\n",
      "    (0): ModuleDict(\n",
      "      (Compound): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Disease): Linear(in_features=524, out_features=524, bias=True)\n",
      "    )\n",
      "    (1): ModuleDict(\n",
      "      (Compound): Dropout(p=0.7, inplace=False)\n",
      "      (Disease): Dropout(p=0.7, inplace=False)\n",
      "    )\n",
      "    (2): ModuleDict(\n",
      "      (Compound): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (Disease): Linear(in_features=524, out_features=524, bias=True)\n",
      "    )\n",
      "  )\n",
      ")\n",
      "\n",
      "\n",
      "\n",
      "def forward(self, x, edge_index):\n",
      "    x_dict = torch_geometric_nn_to_hetero_transformer_get_dict(x);  x = None\n",
      "    x__Compound = x_dict.get('Compound', None)\n",
      "    x__Disease = x_dict.get('Disease', None);  x_dict = None\n",
      "    edge_index_dict = torch_geometric_nn_to_hetero_transformer_get_dict(edge_index);  edge_index = None\n",
      "    edge_index__Compound__treats__Disease = edge_index_dict.get(('Compound', 'treats', 'Disease'), None)\n",
      "    edge_index__Disease__rev_treats__Compound = edge_index_dict.get(('Disease', 'rev_treats', 'Compound'), None);  edge_index_dict = None\n",
      "    convs_0__Disease = getattr(self.convs, \"0\").Compound__treats__Disease((x__Compound, x__Disease), edge_index__Compound__treats__Disease)\n",
      "    convs_0__Compound = getattr(self.convs, \"0\").Disease__rev_treats__Compound((x__Disease, x__Compound), edge_index__Disease__rev_treats__Compound);  x__Disease = x__Compound = None\n",
      "    relu__Compound = torch.nn.functional.relu(convs_0__Compound, inplace = False);  convs_0__Compound = None\n",
      "    relu__Disease = torch.nn.functional.relu(convs_0__Disease, inplace = False);  convs_0__Disease = None\n",
      "    dropout__Compound = torch.nn.functional.dropout(relu__Compound, p = 0.7, training = True, inplace = False);  relu__Compound = None\n",
      "    dropout__Disease = torch.nn.functional.dropout(relu__Disease, p = 0.7, training = True, inplace = False);  relu__Disease = None\n",
      "    convs_1__Disease = getattr(self.convs, \"1\").Compound__treats__Disease((dropout__Compound, dropout__Disease), edge_index__Compound__treats__Disease)\n",
      "    convs_1__Compound = getattr(self.convs, \"1\").Disease__rev_treats__Compound((dropout__Disease, dropout__Compound), edge_index__Disease__rev_treats__Compound);  dropout__Disease = dropout__Compound = None\n",
      "    relu_1__Compound = torch.nn.functional.relu(convs_1__Compound, inplace = False);  convs_1__Compound = None\n",
      "    relu_1__Disease = torch.nn.functional.relu(convs_1__Disease, inplace = False);  convs_1__Disease = None\n",
      "    dropout_1__Compound = torch.nn.functional.dropout(relu_1__Compound, p = 0.7, training = True, inplace = False);  relu_1__Compound = None\n",
      "    dropout_1__Disease = torch.nn.functional.dropout(relu_1__Disease, p = 0.7, training = True, inplace = False);  relu_1__Disease = None\n",
      "    convs_2__Disease = getattr(self.convs, \"2\").Compound__treats__Disease((dropout_1__Compound, dropout_1__Disease), edge_index__Compound__treats__Disease);  edge_index__Compound__treats__Disease = None\n",
      "    convs_2__Compound = getattr(self.convs, \"2\").Disease__rev_treats__Compound((dropout_1__Disease, dropout_1__Compound), edge_index__Disease__rev_treats__Compound);  dropout_1__Disease = dropout_1__Compound = edge_index__Disease__rev_treats__Compound = None\n",
      "    relu_2__Compound = torch.nn.functional.relu(convs_2__Compound, inplace = False);  convs_2__Compound = None\n",
      "    relu_2__Disease = torch.nn.functional.relu(convs_2__Disease, inplace = False);  convs_2__Disease = None\n",
      "    dropout_2__Compound = torch.nn.functional.dropout(relu_2__Compound, p = 0.7, training = True, inplace = False);  relu_2__Compound = None\n",
      "    dropout_2__Disease = torch.nn.functional.dropout(relu_2__Disease, p = 0.7, training = True, inplace = False);  relu_2__Disease = None\n",
      "    post_mp_0__Compound = getattr(self.post_mp, \"0\").Compound(dropout_2__Compound);  dropout_2__Compound = None\n",
      "    post_mp_0__Disease = getattr(self.post_mp, \"0\").Disease(dropout_2__Disease);  dropout_2__Disease = None\n",
      "    post_mp_1__Compound = getattr(self.post_mp, \"1\").Compound(post_mp_0__Compound);  post_mp_0__Compound = None\n",
      "    post_mp_1__Disease = getattr(self.post_mp, \"1\").Disease(post_mp_0__Disease);  post_mp_0__Disease = None\n",
      "    post_mp_2__Compound = getattr(self.post_mp, \"2\").Compound(post_mp_1__Compound);  post_mp_1__Compound = None\n",
      "    post_mp_2__Disease = getattr(self.post_mp, \"2\").Disease(post_mp_1__Disease);  post_mp_1__Disease = None\n",
      "    return {'Compound': post_mp_2__Compound, 'Disease': post_mp_2__Disease}\n",
      "    \n",
      "# To see more debug info, please use `graph_module.print_readable()`\n",
      "LinkPredictorMLP(\n",
      "  (mlp_layers): ModuleList(\n",
      "    (0-1): 2 x Linear(in_features=524, out_features=524, bias=True)\n",
      "    (2): Linear(in_features=524, out_features=1, bias=True)\n",
      "  )\n",
      ")\n",
      "Models Loaded to cpu\n"
     ]
    }
   ],
   "source": [
    "epochs        = 500\n",
    "hidden_dim    = 524      # 256 \n",
    "dropout       = 0.7\n",
    "num_layers    = 3\n",
    "learning_rate = 1e-4\n",
    "node_emb_dim  = 768\n",
    "\n",
    "HomoGNN         = GNNStack(node_emb_dim, hidden_dim, hidden_dim, num_layers, dropout, return_embedding=True).to(device) # the graph neural network that takes all the node embeddings as inputs to message pass and agregate\n",
    "HeteroGNN       = to_hetero(HomoGNN   , data2.metadata(), aggr='sum')\n",
    "link_predictor  = LinkPredictorMLP(hidden_dim, hidden_dim, 1, num_layers , dropout).to(device) # the MLP that takes embeddings of a pair of nodes and predicts the existence of an edge between them\n",
    "#optimizer      = torch.optim.AdamW(list(model.parameters()) + list(link_predictor.parameters() ), lr=learning_rate, weight_decay=1e-4)\n",
    "optimizer       = torch.optim.Adam(list(HeteroGNN.parameters()) + list(link_predictor.parameters() ), lr=learning_rate)\n",
    "\n",
    "print(HeteroGNN )\n",
    "print(link_predictor)\n",
    "print(f\"Models Loaded to {device}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HeteroData(\n",
      "  \u001b[1mCompound\u001b[0m={ x=[13, 768] },\n",
      "  \u001b[1mDisease\u001b[0m={ x=[3, 768] },\n",
      "  \u001b[1m(Compound, treats, Disease)\u001b[0m={ edge_index=[2, 3] },\n",
      "  \u001b[1m(Disease, rev_treats, Compound)\u001b[0m={ edge_index=[2, 3] }\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "print(data2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_emb   = HeteroGNN(data2.x_dict, data2.edge_index_dict)\n",
    "edge_index = data2['Compound', 'treats', 'Disease'].edge_index \n",
    "pos_pred    = link_predictor(node_emb[\"Compound\"][edge_index[0]], node_emb[\"Disease\"][edge_index[1]])   # (B, )\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "2d42bd2c408a4bd3bdebc2865ea42483": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": "hidden",
      "width": null
     }
    },
    "6e79954f9c514f099b336e3a11ad635e": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_6f2e0878218c4b38974a74a673b70751",
      "placeholder": "​",
      "style": "IPY_MODEL_98ccadd9e4a34d8282282a332fad6e0b",
      "value": "Downloading chembl_32_sqlite.tar.gz: 100%"
     }
    },
    "6f2e0878218c4b38974a74a673b70751": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "98ccadd9e4a34d8282282a332fad6e0b": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "a95662c4dc0949f4af40e8331cde378d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "info",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_eec53708d80f41e8b402060c9a1352b1",
      "max": 1,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_f87f682db17a4e6b837b9c14c85ec36f",
      "value": 1
     }
    },
    "b0797e6674144b1b92d2ff70918b917c": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "b355679609b749d698029d4f39d45503": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_6e79954f9c514f099b336e3a11ad635e",
       "IPY_MODEL_a95662c4dc0949f4af40e8331cde378d",
       "IPY_MODEL_d89d003254ce42f9a44572c7e26940ec"
      ],
      "layout": "IPY_MODEL_2d42bd2c408a4bd3bdebc2865ea42483"
     }
    },
    "d0a9a73a5fe14afb8b15a2addf4f080d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "d89d003254ce42f9a44572c7e26940ec": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_b0797e6674144b1b92d2ff70918b917c",
      "placeholder": "​",
      "style": "IPY_MODEL_d0a9a73a5fe14afb8b15a2addf4f080d",
      "value": " 4.27G/4.27G [03:38&lt;00:00, 21.7MB/s]"
     }
    },
    "eec53708d80f41e8b402060c9a1352b1": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": "20px"
     }
    },
    "f87f682db17a4e6b837b9c14c85ec36f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
