{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "DnQGl_D4786T"
   },
   "source": [
    "DRKG\n",
    "\n",
    "Adapted from: https://github.com/gnn4dr/DRKG/blob/master/drkg_with_dgl/loading_drkg_in_dgl.ipynb\n",
    "\n",
    "https://medium.com/@pytorch_geometric/link-prediction-on-heterogeneous-graphs-with-pyg-6d5c29677c70"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "id": "PV4N5XSIKfTa"
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/simp/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "from SIMP_LLM.DRKG_loading   import  get_triplets, read_tsv,filter_drkg,map_drkg_relationships,filter_interaction_subset,print_head\n",
    "from SIMP_LLM.DRKG_translate import  load_lookups"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "VsLJKwgRxArZ"
   },
   "source": [
    "# 1) Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " Triplets:\n",
      "\n",
      "[['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2157'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5264'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2158'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::3309'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::28912'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::811'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::2159'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::821'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5627'], ['Gene::2157', 'bioarx::HumGenHumGen:Gene:Gene', 'Gene::5624']]\n",
      "\n",
      " data/drkg.tsv  Dataframe:\n",
      "\n",
      "+----+------------+--------------------------------+-------------+\n",
      "|    | 0          | 1                              | 2           |\n",
      "|----+------------+--------------------------------+-------------|\n",
      "|  0 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2157  |\n",
      "|  1 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::5264  |\n",
      "|  2 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::2158  |\n",
      "|  3 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::3309  |\n",
      "|  4 | Gene::2157 | bioarx::HumGenHumGen:Gene:Gene | Gene::28912 |\n",
      "+----+------------+--------------------------------+-------------+\n",
      "\n",
      " data/relation_glossary.tsv  Dataframe:\n",
      "\n",
      "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
      "|    | Relation-name                              | Data-source   | Connected entity-types   | Interaction-type      | Description                                                                                                                                                                     | Reference for the description        |\n",
      "|----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------|\n",
      "|  0 | DGIDB::ACTIVATOR::Gene:Compound            | DGIDB         | Compound:Gene            | activation            | An activator interaction is when a drug activates a biological response from a target, although the mechanism by which it does so may not be understood.                        | http://www.dgidb.org/getting_started |\n",
      "|  1 | DGIDB::AGONIST::Gene:Compound              | DGIDB         | Compound:Gene            | agonism               | An agonist interaction occurs when a drug binds to a target receptor and activates the receptor to produce a biological response.                                               | http://www.dgidb.org/getting_started |\n",
      "|  2 | DGIDB::ALLOSTERIC MODULATOR::Gene:Compound | DGIDB         | Compound:Gene            | allosteric modulation | An allosteric modulator interaction occurs when drugs exert their effects on their protein targets via a different binding site than the natural (orthosteric) ligand site.     | http://www.dgidb.org/getting_started |\n",
      "|  3 | DGIDB::ANTAGONIST::Gene:Compound           | DGIDB         | Compound:Gene            | antagonism            | An antagonist interaction occurs when a drug blocks or dampens agonist-mediated responses rather than provoking a biological response itself upon binding to a target receptor. | http://www.dgidb.org/getting_started |\n",
      "|  4 | DGIDB::ANTIBODY::Gene:Compound             | DGIDB         | Compound:Gene            | antibody              | An antibody interaction occurs when an antibody drug specifically binds the target molecule.                                                                                    | http://www.dgidb.org/getting_started |\n",
      "+----+--------------------------------------------+---------------+--------------------------+-----------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------+\n",
      "Number of Rows Before Filtering: 107\n",
      "Number of Rows After Filtering: 10\n",
      "\n",
      "\n",
      "Filtered:\n",
      "24    DRUGBANK::treats::Compound:Disease\n",
      "35             GNBR::T::Compound:Disease\n",
      "36             GNBR::C::Compound:Disease\n",
      "37            GNBR::Sa::Compound:Disease\n",
      "38            GNBR::Pa::Compound:Disease\n",
      "39            GNBR::Mp::Compound:Disease\n",
      "40            GNBR::Pr::Compound:Disease\n",
      "41             GNBR::J::Compound:Disease\n",
      "68       Hetionet::CtD::Compound:Disease\n",
      "73       Hetionet::CpD::Compound:Disease\n",
      "dtype: object\n",
      "\n",
      "Relationships Mapped:\n",
      "0                    Compound treats the disease\n",
      "1    treatment/therapy (including investigatory)\n",
      "2            inhibits cell growth (esp. cancers)\n",
      "3                      side effect/adverse event\n",
      "4                            alleviates, reduces\n",
      "5            biomarkers (of disease progression)\n",
      "6                           prevents, suppresses\n",
      "7                   role in disease pathogenesis\n",
      "8                                      treatment\n",
      "9                                     palliation\n",
      "Name: Interaction-type, dtype: object\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n",
      "|         | 0                 | 1                                  | 2                     |\n",
      "|---------+-------------------+------------------------------------+-----------------------|\n",
      "| 1518268 | Compound::DB00003 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D003550 |\n",
      "| 1518269 | Compound::DB00004 | DRUGBANK::treats::Compound:Disease | Disease::MESH:C063419 |\n",
      "| 1518270 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D013167 |\n",
      "| 1518271 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D006086 |\n",
      "| 1518272 | Compound::DB00005 | DRUGBANK::treats::Compound:Disease | Disease::MESH:D017497 |\n",
      "+---------+-------------------+------------------------------------+-----------------------+\n"
     ]
    }
   ],
   "source": [
    "### 1) Read: This section reads DRKG and a glossary (used to map entities from codes to words)\n",
    "DATA_DIR           = os.path.join(\"data\")\n",
    "verbose            =  True \n",
    "triplets,drkg_df   =  get_triplets(drkg_file = os.path.join(DATA_DIR  ,'drkg.tsv'),             verbose=verbose)  # Read triplets (head,relationship,tail)\n",
    "relation_glossary  =  read_tsv(relation_file = os.path.join(DATA_DIR  ,'relation_glossary.tsv'),verbose=verbose)  # Read relationship mapping  \n",
    "\n",
    "\n",
    "### 2) Filter & Map Interactions: This section returns a list of interactions e.g. DRUGBANK::treats::Compound:Disease )\n",
    "# 2.1: First  we filter the interactions to only Compound-Disease\n",
    "# 2.2: Then   we map the codes -> text  (this will be use to further filter interactions based on text) e.g.  Hetionet::CpD::Compound:Disease -> palliation\n",
    "# 2.3: We use natural text to fitler  interactions based on terms such as \"treat\" (but we return the orignal interaction name )\n",
    "\n",
    "\n",
    "\n",
    "# modularize this in create_dataframe\n",
    "drkg_rx_dx_relations        = filter_drkg(data_frame = drkg_df ,  filter_column = 1 ,  filter_term = r'.*?Compound:Disease', verbose = verbose) # 2.1 Filter only Compound-Disease Interactions\n",
    "drkg_rx_dx_relations_mapped = map_drkg_relationships(drkg_rx_dx_relations,relation_glossary,verbose=verbose)                                    # 2.2 Map codes to text \n",
    "\n",
    "### 2.3 Filter Drug interactions Interaction types to only include: treat inhibit or alleviate interactions  ###\n",
    "drkg_rx_dx_relation_subset =  filter_interaction_subset(df                  = drkg_rx_dx_relations_mapped,\n",
    "                                                        filter_colunm_name = 'Interaction-type' ,\n",
    "                                                        regex_string       =  'treat|inhibit|alleviate',\n",
    "                                                        return_colunm_name =  'Relation-name')\n",
    "\n",
    "# 3) Use Filter Interactions to get Gilter DRKG \n",
    "drkg_df_filtered = drkg_df[drkg_df[1].isin(drkg_rx_dx_relation_subset)] # 3.1 Filter DRKG  to only  Compund-Disease \n",
    "print_head(df=drkg_df_filtered)\n",
    "\n",
    "\n",
    "\n",
    "###\n",
    "\n",
    "rx_dx_triplets   = drkg_df_filtered.values.tolist()                     # 3.2 Convert filtered DRKG to list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "KYBnEGL4Nixb",
    "outputId": "5bc6b100-d023-43d7-cf3c-488dc73b6475"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      " data/hetionet-v1.0-nodes.tsv  Dataframe:\n",
      "\n",
      "+----+-------------------------+---------------------------+---------+\n",
      "|    | id                      | name                      | kind    |\n",
      "|----+-------------------------+---------------------------+---------|\n",
      "|  0 | Anatomy::UBERON:0000002 | uterine cervix            | Anatomy |\n",
      "|  1 | Anatomy::UBERON:0000004 | nose                      | Anatomy |\n",
      "|  2 | Anatomy::UBERON:0000006 | islet of Langerhans       | Anatomy |\n",
      "|  3 | Anatomy::UBERON:0000007 | pituitary gland           | Anatomy |\n",
      "|  4 | Anatomy::UBERON:0000010 | peripheral nervous system | Anatomy |\n",
      "+----+-------------------------+---------------------------+---------+\n",
      "\n",
      " Sample of Hetionet Data Types (Before processing):\n",
      "\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "|       | id                               | name                                      | kind                |\n",
      "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
      "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
      "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
      "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
      "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
      "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
      "| 14863 | Gene::1                          | A1BG                                      | Gene                |\n",
      "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
      "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
      "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
      "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
      "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "\n",
      " Sample of Hetionet Data Types (After processing):\n",
      "\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "|       | id                               | name                                      | kind                |\n",
      "|-------+----------------------------------+-------------------------------------------+---------------------|\n",
      "|     0 | Anatomy::UBERON:0000002          | uterine cervix                            | Anatomy             |\n",
      "|   402 | Biological Process::GO:0000002   | mitochondrial genome maintenance          | Biological Process  |\n",
      "| 11783 | Cellular Component::GO:0000015   | phosphopyruvate hydratase complex         | Cellular Component  |\n",
      "| 13174 | Compound::DB00014                | Goserelin                                 | Compound            |\n",
      "| 14726 | Disease::DOID:0050156            | idiopathic pulmonary fibrosis             | Disease             |\n",
      "| 14863 | Gene::1                          | A1BG gene                                 | Gene                |\n",
      "| 35808 | Molecular Function::GO:0000010   | trans-hexaprenyltranstransferase activity | Molecular Function  |\n",
      "| 38692 | Pathway::PC7_10399               | a4b7 Integrin signaling                   | Pathway             |\n",
      "| 40514 | Pharmacologic Class::N0000000069 | Calcium Channel Antagonists               | Pharmacologic Class |\n",
      "| 40859 | Side Effect::C0000727            | Acute abdomen                             | Side Effect         |\n",
      "| 46593 | Symptom::D000006                 | Abdomen, Acute                            | Symptom             |\n",
      "+-------+----------------------------------+-------------------------------------------+---------------------+\n",
      "\n",
      " Gene ID Dataframe (After processing):\n",
      "\n",
      "+----+----------+--------------------------------------------+\n",
      "|    | GeneID   | description                                |\n",
      "|----+----------+--------------------------------------------|\n",
      "|  0 | Gene::1  | alpha-1-B glycoprotein (A1BG)              |\n",
      "|  1 | Gene::2  | alpha-2-macroglobulin (A2M)                |\n",
      "|  2 | Gene::3  | alpha-2-macroglobulin pseudogene 1 (A2MP1) |\n",
      "|  3 | Gene::9  | N-acetyltransferase 1 (NAT1)               |\n",
      "|  4 | Gene::10 | N-acetyltransferase 2 (NAT2)               |\n",
      "+----+----------+--------------------------------------------+\n",
      "\n",
      " DrugBank Dataframe (Before processing):\n",
      "\n",
      "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
      "|    | DrugBank ID   | Accession Numbers    | Common name         | CAS         | UNII       | Synonyms                                                                                                                                                              |   Standard InChI Key |\n",
      "|----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------|\n",
      "|  0 | DB00001       | BIOD00024 | BTD00024 | Lepirudin           | 138068-37-8 | Y43GF64R34 | [Leu1, Thr2]-63-desulfohirudin | Desulfatohirudin | Hirudin variant-1 | Lepirudin | Lepirudin recombinant | R-hirudin                                                 |                  nan |\n",
      "|  1 | DB00002       | BIOD00071 | BTD00071 | Cetuximab           | 205923-56-4 | PQX0D8J21J | Cetuximab | Cétuximab | Cetuximabum                                                                                                                                   |                  nan |\n",
      "|  2 | DB00003       | BIOD00001 | BTD00001 | Dornase alfa        | 143831-71-4 | 953A26OA1Y | Deoxyribonuclease (human clone 18-1 protein moiety) | Dornasa alfa | Dornase alfa | Dornase alfa, recombinant | Dornase alpha | Recombinant deoxyribonuclease (DNAse) |                  nan |\n",
      "|  3 | DB00004       | BIOD00084 | BTD00084 | Denileukin diftitox | 173146-27-5 | 25E79B5CTM | Denileukin | Denileukin diftitox | Interleukin-2/diptheria toxin fusion protein                                                                                       |                  nan |\n",
      "|  4 | DB00005       | BIOD00052 | BTD00052 | Etanercept          | 185243-69-0 | OP401G7OJC | Etanercept | etanercept-szzs | etanercept-ykro | Recombinant human TNF | rhu TNFR:Fc | rhu-TNFR:Fc | TNFR-Immunoadhesin                                               |                  nan |\n",
      "+----+---------------+----------------------+---------------------+-------------+------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------+----------------------+\n",
      "\n",
      " DrugBank Dataframe (After processing):\n",
      "\n",
      "+----+-------------------+---------------------+\n",
      "|    | DrugBank ID       | Common name         |\n",
      "|----+-------------------+---------------------|\n",
      "|  0 | Compound::DB00001 | Lepirudin           |\n",
      "|  1 | Compound::DB00002 | Cetuximab           |\n",
      "|  2 | Compound::DB00003 | Dornase alfa        |\n",
      "|  3 | Compound::DB00004 | Denileukin diftitox |\n",
      "|  4 | Compound::DB00005 | Etanercept          |\n",
      "+----+-------------------+---------------------+\n",
      "\n",
      " data/mimTitles.txt  Dataframe (Before processing):\n",
      "\n",
      "+----+----------------------+----------------------------------------+\n",
      "|    | MIM Number           | name                                   |\n",
      "|----+----------------------+----------------------------------------|\n",
      "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
      "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
      "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
      "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
      "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
      "+----+----------------------+----------------------------------------+\n",
      "\n",
      " data/mimTitles.txt  Dataframe (After processing):\n",
      "\n",
      "+----+----------------------+----------------------------------------+\n",
      "|    | MIM Number           | name                                   |\n",
      "|----+----------------------+----------------------------------------|\n",
      "|  0 | Disease::OMIM:100050 | AARSKOG SYNDROME, AUTOSOMAL DOMINANT   |\n",
      "|  1 | Disease::OMIM:100070 | AORTIC ANEURYSM, FAMILIAL ABDOMINAL, 1 |\n",
      "|  2 | Disease::OMIM:100100 | PRUNE BELLY SYNDROME                   |\n",
      "|  3 | Disease::OMIM:100200 | ABDUCENS PALSY                         |\n",
      "|  4 | Disease::OMIM:100300 | ADAMS-OLIVER SYNDROME 1                |\n",
      "+----+----------------------+----------------------------------------+\n",
      "\n",
      " MeSH Dictionary:\n",
      "\n",
      "+---------------+------------------------+\n",
      "|               | 0                      |\n",
      "|---------------+------------------------|\n",
      "| MESH::D000001 | Calcimycin             |\n",
      "| MESH::D000002 | Temefos                |\n",
      "| MESH::D000003 | Abattoirs              |\n",
      "| MESH::D000004 | Abbreviations as Topic |\n",
      "| MESH::D000005 | Abdomen                |\n",
      "+---------------+------------------------+\n"
     ]
    }
   ],
   "source": [
    "# 4) Load Data frames for translation\n",
    "hetionet_df, gene_df, drugbank_df, omim_df, mesh_dict, chebi_df, chembl_df = load_lookups(data_path=DATA_DIR,verbose=verbose)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "HoImejcxpFYP",
    "outputId": "904e49c4-8889-4f2d-fcce-26dc99331b86"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "+----+--------------+-------------+---------------+\n",
      "|    | 0            | 1           | 2             |\n",
      "|----+--------------+-------------+---------------|\n",
      "|  0 | Gene ID 2157 | interaction | Gene ID 2157  |\n",
      "|  1 | Gene ID 2157 | interaction | Gene ID 5264  |\n",
      "|  2 | Gene ID 2157 | interaction | Gene ID 2158  |\n",
      "|  3 | Gene ID 2157 | interaction | Gene ID 3309  |\n",
      "|  4 | Gene ID 2157 | interaction | Gene ID 28912 |\n",
      "+----+--------------+-------------+---------------+\n"
     ]
    }
   ],
   "source": [
    "# Make dictionaries\n",
    "relation_glossary_relation_dict = pd.Series(relation_glossary['Interaction-type'].values, index=relation_glossary['Relation-name']).to_dict()\n",
    "\n",
    "node_df   = pd.concat([hetionet_df[['name', 'id']], \n",
    "                       gene_df.rename(columns = {\"symbol\":\"name\", \"GeneID\":\"id\"}),\n",
    "                       drugbank_df.rename(columns = {\"Common name\":\"name\", \"DrugBank ID\":\"id\"}),\n",
    "                       omim_df.rename(columns = {\"MIM Number\":\"id\"}),\n",
    "                       chebi_df.rename(columns = {\"NAME\":\"name\", \"CHEBI_ACCESSION\":\"id\"}),\n",
    "                       chembl_df.rename(columns = {\"pref_name\":\"name\", \"chembl_id\":\"id\"})\n",
    "                       ], ignore_index=True, axis=0).drop_duplicates() \n",
    "node_dict = pd.Series(node_df['name'].values, index=node_df['id']).to_dict() | mesh_dict # Convert node df to dict and merge with MeSH dictionary\n",
    "\n",
    "\n",
    "# Create and use convert_entitynames function\n",
    "def convert_entitynames(df, col, node_dict):\n",
    "  df_update = df.copy()\n",
    "  df_update[col] = df_update[col].str.replace(r'.*?MESH:', \"MESH::\", regex=True) # Remove MeSH labeling\n",
    "  df_update[col] = df_update[col].map(node_dict).fillna(df_update[col])    # Translate dictionary\n",
    "  df_update[col] = df_update[col].str.replace(\"Gene::\", \"Gene ID \") # For remaining uncoverted Gene IDs, remove \"::\"\n",
    "  df_update[col] = df_update[col].str.replace(\"Disease::\", \"\") # For remaining diseases (appears to be just SARS-COVID related names), remove label\n",
    "  return df_update\n",
    "\n",
    "df_med    = drkg_df.copy()\n",
    "df_med[1] = df_med[1].map(relation_glossary_relation_dict).fillna(df_med[1])\n",
    "\n",
    "df_med = convert_entitynames(df_med, 0, node_dict)\n",
    "df_med = convert_entitynames(df_med, 2, node_dict)\n",
    "\n",
    "print_head(df_med) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 419
    },
    "id": "IcZZ_cEn-hIF",
    "outputId": "83b65c42-cf6f-41e2-9f30-f1c3618d957a"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1518268</th>\n",
       "      <td>Dornase alfa</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Cystic Fibrosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518269</th>\n",
       "      <td>Denileukin diftitox</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>MESH::C063419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518270</th>\n",
       "      <td>Etanercept</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Spondylitis, Ankylosing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518271</th>\n",
       "      <td>Etanercept</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Graft vs Host Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518272</th>\n",
       "      <td>Etanercept</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Hidradenitis Suppurativa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520263</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Choroiditis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520264</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Adrenal Hyperplasia, Congenital</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520265</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Fanconi Anemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520266</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Keratitis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520267</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Crohn Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0                            1   \n",
       "1518268         Dornase alfa  Compound treats the disease  \\\n",
       "1518269  Denileukin diftitox  Compound treats the disease   \n",
       "1518270           Etanercept  Compound treats the disease   \n",
       "1518271           Etanercept  Compound treats the disease   \n",
       "1518272           Etanercept  Compound treats the disease   \n",
       "...                      ...                          ...   \n",
       "1520263       Hydrocortisone  Compound treats the disease   \n",
       "1520264       Hydrocortisone  Compound treats the disease   \n",
       "1520265       Hydrocortisone  Compound treats the disease   \n",
       "1520266       Hydrocortisone  Compound treats the disease   \n",
       "1520267       Hydrocortisone  Compound treats the disease   \n",
       "\n",
       "                                       2  \n",
       "1518268                  Cystic Fibrosis  \n",
       "1518269                    MESH::C063419  \n",
       "1518270          Spondylitis, Ankylosing  \n",
       "1518271            Graft vs Host Disease  \n",
       "1518272         Hidradenitis Suppurativa  \n",
       "...                                  ...  \n",
       "1520263                      Choroiditis  \n",
       "1520264  Adrenal Hyperplasia, Congenital  \n",
       "1520265                   Fanconi Anemia  \n",
       "1520266                        Keratitis  \n",
       "1520267                    Crohn Disease  \n",
       "\n",
       "[2000 rows x 3 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter DRKG in natural language to drug-treats-disease relationships\n",
    "# rx_dx        = df_med[df_med.iloc[:,1] ==   'Compound treats the disease']\n",
    "rx_dx        =  df_med.loc[drkg_df_filtered.index]\n",
    "rx_dx_subset =  rx_dx[0:2000]\n",
    "rx_dx_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 470
    },
    "id": "ZvGYU-hIFDqM",
    "outputId": "07e87c76-114e-4e9d-8873-b287f8a53950"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of pairs 5874261\n",
      "Dropped taxonomy pairs 14663\n",
      "Dropped untranslated pairs 62779\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gene ID 2157</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Gene ID 2157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene ID 2157</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Gene ID 5264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gene ID 2157</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Gene ID 2158</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gene ID 2157</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Gene ID 3309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gene ID 2157</td>\n",
       "      <td>interaction</td>\n",
       "      <td>Gene ID 28912</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5874256</th>\n",
       "      <td>Gene ID 29099</td>\n",
       "      <td>reaction</td>\n",
       "      <td>Gene ID 1643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5874257</th>\n",
       "      <td>Gene ID 51645</td>\n",
       "      <td>reaction</td>\n",
       "      <td>Gene ID 3183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5874258</th>\n",
       "      <td>Gene ID 865</td>\n",
       "      <td>catalysis</td>\n",
       "      <td>Gene ID 983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5874259</th>\n",
       "      <td>Gene ID 1066</td>\n",
       "      <td>binding</td>\n",
       "      <td>Gene ID 7365</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5874260</th>\n",
       "      <td>Gene ID 6118</td>\n",
       "      <td>binding</td>\n",
       "      <td>Gene ID 1111</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5796819 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0            1              2\n",
       "0         Gene ID 2157  interaction   Gene ID 2157\n",
       "1         Gene ID 2157  interaction   Gene ID 5264\n",
       "2         Gene ID 2157  interaction   Gene ID 2158\n",
       "3         Gene ID 2157  interaction   Gene ID 3309\n",
       "4         Gene ID 2157  interaction  Gene ID 28912\n",
       "...                ...          ...            ...\n",
       "5874256  Gene ID 29099     reaction   Gene ID 1643\n",
       "5874257  Gene ID 51645     reaction   Gene ID 3183\n",
       "5874258    Gene ID 865    catalysis    Gene ID 983\n",
       "5874259   Gene ID 1066      binding   Gene ID 7365\n",
       "5874260   Gene ID 6118      binding   Gene ID 1111\n",
       "\n",
       "[5796819 rows x 3 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Remove codes that are not relevant or have no conversion\n",
    "drkg_translated = df_med.copy()\n",
    "\n",
    "# Remove taxonomy\n",
    "remove_tax = drkg_translated[drkg_translated[2].str.contains('Tax::')]\n",
    "drkg_translated = drkg_translated.drop(remove_tax.index)\n",
    "\n",
    "# Check for any untranslated terms\n",
    "remove_untranslated = drkg_translated[(drkg_translated[0].str.contains('::'))|(drkg_translated[2].str.contains('::'))]\n",
    "drkg_translated = drkg_translated.drop(remove_untranslated.index)\n",
    "\n",
    "# Summarize \n",
    "print('Total number of pairs ' + str(drkg_df.shape[0]))\n",
    "print('Dropped taxonomy pairs ' + str(len(remove_tax.index)))\n",
    "print('Dropped untranslated pairs ' + str(len(remove_untranslated.index)))\n",
    "drkg_translated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 419
    },
    "id": "o2sd24zgkGjo",
    "outputId": "f912827f-45a0-493b-8506-3936d10ea4e2"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Compound::Bioarxivdrug:0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Compound::Bioarxivdrug:1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Compound::Bioarxivdrug:10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Compound::Bioarxivdrug:11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Compound::Bioarxivdrug:2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14416</th>\n",
       "      <td>MESH::C580539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14417</th>\n",
       "      <td>MESH::C585640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14418</th>\n",
       "      <td>MESH::D000071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14419</th>\n",
       "      <td>MESH::D018290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14420</th>\n",
       "      <td>MESH::D019952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>14421 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               0\n",
       "0       Compound::Bioarxivdrug:0\n",
       "1       Compound::Bioarxivdrug:1\n",
       "2      Compound::Bioarxivdrug:10\n",
       "3      Compound::Bioarxivdrug:11\n",
       "4       Compound::Bioarxivdrug:2\n",
       "...                          ...\n",
       "14416              MESH::C580539\n",
       "14417              MESH::C585640\n",
       "14418              MESH::D000071\n",
       "14419              MESH::D018290\n",
       "14420              MESH::D019952\n",
       "\n",
       "[14421 rows x 1 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Check unique values of untranslated\n",
    "drkg_test1 = np.unique(remove_untranslated[0][remove_untranslated[0].str.contains('::')])\n",
    "drkg_test2 = np.unique(remove_untranslated[2][remove_untranslated[2].str.contains('::')])\n",
    "\n",
    "drkg_mesh_list = drkg_test1.tolist() +  drkg_test2.tolist()\n",
    "drkg_mesh_unique = pd.DataFrame(pd.unique(drkg_mesh_list))\n",
    "drkg_mesh_unique"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 324
    },
    "id": "ibNECdMLoCjA",
    "outputId": "8c8b6a0e-16db-4e37-9a38-a14c8967bb21"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7751\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1429</th>\n",
       "      <td>MESH::C000020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>MESH::C000050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1431</th>\n",
       "      <td>MESH::C000121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1432</th>\n",
       "      <td>MESH::C000154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1433</th>\n",
       "      <td>MESH::C000188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14416</th>\n",
       "      <td>MESH::C580539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14417</th>\n",
       "      <td>MESH::C585640</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14418</th>\n",
       "      <td>MESH::D000071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14419</th>\n",
       "      <td>MESH::D018290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14420</th>\n",
       "      <td>MESH::D019952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7751 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   0\n",
       "1429   MESH::C000020\n",
       "1430   MESH::C000050\n",
       "1431   MESH::C000121\n",
       "1432   MESH::C000154\n",
       "1433   MESH::C000188\n",
       "...              ...\n",
       "14416  MESH::C580539\n",
       "14417  MESH::C585640\n",
       "14418  MESH::D000071\n",
       "14419  MESH::D018290\n",
       "14420  MESH::D019952\n",
       "\n",
       "[7751 rows x 1 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drkg_untranslated = drkg_mesh_unique.copy()\n",
    "drkg_untranslated = drkg_untranslated[drkg_untranslated[0].str.startswith('MESH::')]\n",
    "print(len(drkg_untranslated))\n",
    "drkg_untranslated"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "oTAJOqDVc_Hg"
   },
   "source": [
    "# 3) BioLinkBERT embedding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1518268</th>\n",
       "      <td>Dornase alfa</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Cystic Fibrosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518269</th>\n",
       "      <td>Denileukin diftitox</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>MESH::C063419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518270</th>\n",
       "      <td>Etanercept</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Spondylitis, Ankylosing</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518271</th>\n",
       "      <td>Etanercept</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Graft vs Host Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1518272</th>\n",
       "      <td>Etanercept</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Hidradenitis Suppurativa</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520263</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Choroiditis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520264</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Adrenal Hyperplasia, Congenital</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520265</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Fanconi Anemia</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520266</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Keratitis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520267</th>\n",
       "      <td>Hydrocortisone</td>\n",
       "      <td>Compound treats the disease</td>\n",
       "      <td>Crohn Disease</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2000 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           0                            1   \n",
       "1518268         Dornase alfa  Compound treats the disease  \\\n",
       "1518269  Denileukin diftitox  Compound treats the disease   \n",
       "1518270           Etanercept  Compound treats the disease   \n",
       "1518271           Etanercept  Compound treats the disease   \n",
       "1518272           Etanercept  Compound treats the disease   \n",
       "...                      ...                          ...   \n",
       "1520263       Hydrocortisone  Compound treats the disease   \n",
       "1520264       Hydrocortisone  Compound treats the disease   \n",
       "1520265       Hydrocortisone  Compound treats the disease   \n",
       "1520266       Hydrocortisone  Compound treats the disease   \n",
       "1520267       Hydrocortisone  Compound treats the disease   \n",
       "\n",
       "                                       2  \n",
       "1518268                  Cystic Fibrosis  \n",
       "1518269                    MESH::C063419  \n",
       "1518270          Spondylitis, Ankylosing  \n",
       "1518271            Graft vs Host Disease  \n",
       "1518272         Hidradenitis Suppurativa  \n",
       "...                                  ...  \n",
       "1520263                      Choroiditis  \n",
       "1520264  Adrenal Hyperplasia, Congenital  \n",
       "1520265                   Fanconi Anemia  \n",
       "1520266                        Keratitis  \n",
       "1520267                    Crohn Disease  \n",
       "\n",
       "[2000 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rx_dx_subset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "id": "QkZKp6CJiw6f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique Drugs:   499 Matrix shape: torch.Size([499, 768])\n",
      "Unique Disases: 810 Matrix shape: torch.Size([810, 768])\n"
     ]
    }
   ],
   "source": [
    "from torch_geometric.data import HeteroData\n",
    "from SIMP_LLM.llm_encode import EntityEncoder\n",
    "from SIMP_LLM.dataloader_mappings import create_mapping,create_edges\n",
    "\n",
    "\n",
    "## Example of loading data without anything to encode\n",
    "device    = \"cpu\"\n",
    "Encoder  = EntityEncoder(device = device )\n",
    "\n",
    "\n",
    "### DX RX Relationship ###\n",
    "rx_X,rx_mapping = create_mapping(rx_dx_subset[0].to_list(),encoder= Encoder ,device=device) # Maps drugs to indices\n",
    "dx_X,dx_mapping = create_mapping(rx_dx_subset[2].to_list(),encoder= Encoder ,device=device) # Maps drugs to indices\n",
    "## As of now this only encodes 'Compound treats the disease', but  the idea is that this is used to encode every entity \n",
    "relationship_X,relationship_mapping = create_mapping(rx_dx_subset[1].to_list(),encoder= Encoder ,device=device)  \n",
    "\n",
    "print(f\"Unique Drugs:   {len(rx_mapping)} Matrix shape: {rx_X.shape}\")\n",
    "print(f\"Unique Disases: {len(dx_mapping)} Matrix shape: {dx_X.shape }\")\n",
    "relationship_feature = relationship_X[relationship_mapping['Compound treats the disease'],:].reshape(1,-1)\n",
    "\n",
    "\n",
    "## NEED To add other reationships"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "ywHcEVMiAdOv"
   },
   "source": [
    "### Build HeteroData Object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "SGIzkUOYAe3q",
    "outputId": "09482160-f318-435e-d52f-298f68ff2997"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HeteroData(\n",
      "  \u001b[1mcompounds\u001b[0m={ x=[499, 768] },\n",
      "  \u001b[1mdisease\u001b[0m={ x=[810, 768] },\n",
      "  \u001b[1m(compounds, treats, disease)\u001b[0m={ edge_index=[2, 2000] },\n",
      "  \u001b[1m(disease, rev_treats, compounds)\u001b[0m={ edge_index=[2, 2000] }\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "### Create PyG Hetero Graph:\n",
    "import torch_geometric.transforms as T\n",
    "\n",
    "data = HeteroData()\n",
    "data['compounds'].x = rx_X\n",
    "data['disease'].x   = dx_X\n",
    "\n",
    "#data['compounds2'].x = rx_X\n",
    "#data['disease2'].x   = dx_X\n",
    "#print(data)\n",
    "\n",
    "Edge_index,edge_attribute = create_edges(df             =  rx_dx_subset,\n",
    "                                          src_index_col  = 0, \n",
    "                                          src_mapping    = rx_mapping , \n",
    "                                          dst_index_col  = 2, \n",
    "                                          dst_mapping    = dx_mapping ,\n",
    "                                          edge_attr      = relationship_feature)\n",
    "\n",
    "data['compounds', 'treats', 'disease'].edge_index = Edge_index\n",
    "data = T.ToUndirected()(data)\n",
    "\n",
    "\n",
    "print(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.nn.functional as F\n",
    "import torch_geometric as pyg\n",
    "import torch_geometric\n",
    "from torch_geometric.nn import SAGEConv, to_hetero\n",
    "from   torch.utils.data      import Dataset, DataLoader\n",
    "from   torch_geometric.data  import Data\n",
    "from   torch_geometric.utils import negative_sampling\n",
    "\n",
    "from torch_geometric.nn import SAGEConv, to_hetero\n",
    "\n",
    "import torch_geometric.transforms as T\n",
    "\n",
    "\n",
    "class GNNStack(torch.nn.Module):\n",
    "    def __init__(self, input_dim:int, hidden_dim:int, output_dim:int, layers:int, dropout:float=0.3, return_embedding=False):\n",
    "        \"\"\"\n",
    "            A stack of GraphSAGE Module \n",
    "            input_dim        <int>:   Input dimension\n",
    "            hidden_dim       <int>:   Hidden dimension\n",
    "            output_dim       <int>:   Output dimension\n",
    "            layers           <int>:   Number of layers\n",
    "            dropout          <float>: Dropout rate\n",
    "            return_embedding <bool>:  Whether to return the return_embeddingedding of the input graph\n",
    "        \"\"\"\n",
    "        \n",
    "        super(GNNStack, self).__init__()\n",
    "        graphSage_conv               = pyg.nn.SAGEConv\n",
    "        self.dropout                 = dropout\n",
    "        self.layers                  = layers\n",
    "        self.return_embedding        = return_embedding\n",
    "\n",
    "        ### Initalize the layers ###\n",
    "        self.convs                   = nn.ModuleList()                      # ModuleList to hold the layers\n",
    "        for l in range(self.layers):\n",
    "            if l == 0:\n",
    "                ### First layer  maps from input_dim to hidden_dim ###\n",
    "                self.convs.append(graphSage_conv(input_dim, hidden_dim))\n",
    "            else:\n",
    "                ### All other layers map from hidden_dim to hidden_dim ###\n",
    "                self.convs.append(graphSage_conv(hidden_dim, hidden_dim))\n",
    "\n",
    "        # post-message-passing processing MLP\n",
    "        self.post_mp = nn.Sequential(\n",
    "                                     nn.Linear(hidden_dim, hidden_dim), \n",
    "                                     nn.Dropout(self.dropout),\n",
    "                                     nn.Linear(hidden_dim, output_dim))\n",
    "\n",
    "    def forward(self, x, edge_index,device=\"cpu\"):\n",
    "        for i in range(self.layers):\n",
    "            x = self.convs[i](x, edge_index)\n",
    "            x = F.relu(x)\n",
    "            x = F.dropout(x, p=self.dropout, training=self.training)\n",
    "\n",
    "        x = self.post_mp(x)\n",
    "\n",
    "        # Return final layer of return_embeddingeddings if specified\n",
    "        if self.return_embedding:\n",
    "            return x\n",
    "\n",
    "        # Else return class probabilities\n",
    "        return F.log_softmax(x, dim=1)\n",
    "\n",
    "    def loss(self, pred, label):\n",
    "        return F.nll_loss(pred, label)\n",
    "    \n",
    "\n",
    "\n",
    "class LinkPredictorMLP(nn.Module):\n",
    "    def __init__(self, in_channels:int, hidden_channels:int, out_channels:int, n_layers:int,dropout_probabilty:float=0.3):\n",
    "        \"\"\"\n",
    "        Args:\n",
    "            in_channels (int):     Number of input features.\n",
    "            hidden_channels (int): Number of hidden features.\n",
    "            out_channels (int):    Number of output features.\n",
    "            n_layers (int):        Number of MLP layers.\n",
    "            dropout (float):       Dropout probability.\n",
    "            \"\"\"\n",
    "        super(LinkPredictorMLP, self).__init__()\n",
    "        self.dropout_probabilty    = dropout_probabilty  # dropout probability\n",
    "        self.mlp_layers            = nn.ModuleList()     # ModuleList: is a list of modules\n",
    "        self.non_linearity         = F.relu              # non-linearity\n",
    "        \n",
    "        for i in range(n_layers - 1):                                 \n",
    "            if i == 0:\n",
    "                self.mlp_layers.append(nn.Linear(in_channels, hidden_channels))          # input layer (in_channels, hidden_channels)\n",
    "            else:\n",
    "                self.mlp_layers.append(nn.Linear(hidden_channels, hidden_channels))      # hidden layers (hidden_channels, hidden_channels)\n",
    "\n",
    "        self.mlp_layers.append(nn.Linear(hidden_channels, out_channels))                 # output layer (hidden_channels, out_channels)\n",
    "\n",
    "\n",
    "    def reset_parameters(self):\n",
    "        for mlp_layer in self.mlp_layers:\n",
    "            mlp_layer.reset_parameters()\n",
    "\n",
    "    def forward(self, x_i, x_j):\n",
    "        x = x_i * x_j                                                     # element-wise multiplication\n",
    "        for mlp_layer in self.mlp_layers[:-1]:                            # iterate over all layers except the last one\n",
    "            x = mlp_layer(x)                                              # apply linear transformation\n",
    "            x = self.non_linearity(x)                                     # Apply non linear activation function\n",
    "            x = F.dropout(x, p=self.dropout_probabilty,training=self.training)      # Apply dropout\n",
    "        x = self.mlp_layers[-1](x)                                        # apply linear transformation to the last layer\n",
    "        x = torch.sigmoid(x)                                              # apply sigmoid activation function to get the probability\n",
    "        return x\n",
    "    \n",
    "### We will use This function to save our best model during trainnig ###\n",
    "def save_torch_model(model,epoch,PATH:str,optimizer):\n",
    "    print(f\"Saving Model in Path {PATH}\")\n",
    "    torch.save({'epoch': epoch,\n",
    "                'model_state_dict': model.state_dict(),\n",
    "                'optimizer':optimizer,      \n",
    "                }, PATH)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "def train(model, link_predictor, data, optimizer,device:str=\"cuda\"):\n",
    "    \"\"\"\n",
    "    Runs offline training for model, link_predictor and node embeddings given the message\n",
    "    edges and supervision edges.\n",
    "    :param model: Torch Graph model used for updating node embeddings based on message passing \n",
    "        (If None, no embbeding is performed) \n",
    "    :param link_predictor: Torch model used for predicting whether edge exists or not\n",
    "    :param emb: (N, d) Initial node embeddings for all N nodes in graph\n",
    "    :param edge_index: (2, E) Edge index for all edges in the graph\n",
    "\n",
    "    :param optimizer: Torch Optimizer to update model parameters\n",
    "    :return: Average supervision loss over all positive (and correspondingly sampled negative) edges\n",
    "    \"\"\"\n",
    "    if model != None:\n",
    "        model.train()\n",
    "    link_predictor.train()\n",
    "    train_losses = []\n",
    "\n",
    "    optimizer.zero_grad()                                 # Reset Gradients\n",
    "    #edge_index     = torch.tensor(edge_index).T           # Reshape edge index     (2,|E|)\n",
    "    #x              = x.squeeze(dim=1)                     # Reshape Feature matrix (|N|,D)\n",
    "    #x , edge_index = x.to(device) , edge_index.to(device) # Move data to devices\n",
    "\n",
    "\n",
    "    ### Step 1: Get Embeddings:\n",
    "    # Run message passing on the inital node embeddings to get updated embeddings\n",
    "\n",
    "    ### This model has the option of only running link predictor without graphsage, for that case the node embedding\n",
    "    ### is equal to the original embedding (X)\n",
    "    if model !=  None:\n",
    "        node_emb   = model(data.x_dict, data.edge_index_dict) # Embed Bert Embeddigns with graphsage (N, d) \n",
    "\n",
    "    else:\n",
    "        node_emb = x                     # Else (None) use Bert Embedddings\n",
    "    # Predict the class probabilities on the batch of positive edges using link_predictor\n",
    "    #print(node_emb[edge_index[0]].shape)\n",
    "    \n",
    "  \n",
    "    edge_index     = data['compounds', 'treats', 'disease'].edge_label_index\n",
    "    pred           = link_predictor(node_emb[\"compounds\"][edge_index[0]], node_emb[\"disease\"][edge_index[0]])   \n",
    "    ground_truth   = data['compounds', 'treats', 'disease'].edge_label.to(device)\n",
    "\n",
    "    loss = F.binary_cross_entropy_with_logits(pred, ground_truth.unsqueeze(1))\n",
    "    # Backpropagate and update parameters\n",
    "    loss.backward()\n",
    "    optimizer.step()\n",
    "\n",
    "    train_losses.append(loss.item())\n",
    "    return sum(train_losses) / len(train_losses)\n",
    "\n",
    "\n",
    "def evaluate(model, predictor,dataset,device=\"cpu\",threshold=0.5,ppi_:int=0.9,verbose:bool=False,best_accuracy=0,show_extra_metrics:bool=False):\n",
    "    if model != None:\n",
    "        model.eval()\n",
    "    possitive_acc  = 0 \n",
    "    negative_acc   = 0\n",
    "    batches        = 0\n",
    "    if show_extra_metrics:\n",
    "        yhat_total     = []\n",
    "        y_total        = []\n",
    "        \n",
    "    for x, edge_index in dataset:                              # Get X and Index from Dataste\n",
    "\n",
    "        edge_index      = torch.tensor(edge_index).T           # Reshape edge index     (2,|E|)\n",
    "        number_of_edges =  edge_index.shape[1]                 # Retrive number of edges \n",
    "        permutations    =  torch.randperm(number_of_edges)     # Create Permutations for edge index\n",
    "        edge_index      = edge_index[:,permutations]           # Run permutation\n",
    "        limit           = int(ppi_*number_of_edges)            # get limit  (based on ppis to embed)\n",
    "        ppi_index_embed = edge_index[:,0:limit]                # PPI to embed with GraphSage \n",
    "        ppi_index_infer = edge_index[:,limit:]                 # PPI to make inference\n",
    "        \n",
    "        x                    = x.squeeze(dim=1)                          # Reshape Feature matrix (|N|,D)\n",
    "        x ,ppi_index_embed   = x.to(device) , ppi_index_embed.to(device) # Move data to devices\n",
    "        if model !=  None:\n",
    "            node_emb             = model(x,ppi_index_embed)              # Get all node embeddings\n",
    "        else:\n",
    "            node_emb = x                                                 # Else (None) use Bert Embedddings\n",
    "            \n",
    "        if verbose:\n",
    "            print(f\" {limit} Positive Protein Interactions were used to Embed a graph with {number_of_edges} ppi's\")\n",
    "        \n",
    "        del ppi_index_embed \n",
    "        with torch.no_grad():\n",
    "            ### Positive PPI ###\n",
    "            positive_pairs_embeddings = node_emb[ppi_index_infer[0]].to(device), node_emb[ppi_index_infer[1]].to(device)\n",
    "            predictions               = predictor(positive_pairs_embeddings[0], positive_pairs_embeddings[1]) \n",
    "            y                         = torch.ones_like(input=predictions)\n",
    "            predictions,y             = predictions.cpu(),y.cpu()\n",
    "            possitive_acc            += accuracy_score(predictions > threshold  ,y)\n",
    "            if show_extra_metrics:\n",
    "                yhat_total.extend(predictions.tolist())\n",
    "                y_total.extend(y.tolist())\n",
    "                \n",
    "            else:\n",
    "                del y, predictions , positive_pairs_embeddings,ppi_index_infer\n",
    "\n",
    "            ### Negative PPI ##\n",
    "            edge_index  =  edge_index.to(device)\n",
    "            neg_edge    = negative_sampling(edge_index       = edge_index,        # Possitve PPI's\n",
    "                                         num_nodes        = x.shape[0],           # Total number of nodes in graph\n",
    "                                         num_neg_samples  = edge_index.shape[1],  # Same Number of edges as in positive example\n",
    "                                         method           = 'dense',              # Method for edge generation\n",
    "                                         force_undirected = True)                 # Our graph is undirected\n",
    "\n",
    "            negative_pairs_embeddings = node_emb[neg_edge[0]].to(device), node_emb[neg_edge[1]].to(device)\n",
    "            predictions               = predictor(negative_pairs_embeddings[0], negative_pairs_embeddings[1])   \n",
    "            y                         = torch.zeros_like(input=predictions)\n",
    "            predictions,y             = predictions.cpu(),y.cpu()\n",
    "            negative_acc             += accuracy_score(predictions > threshold,y)\n",
    "            if show_extra_metrics:\n",
    "                yhat_total.extend(predictions.tolist())\n",
    "                y_total.extend(y.tolist())\n",
    "                \n",
    "            else:\n",
    "                del y,  predictions  ,negative_pairs_embeddings \n",
    "            batches +=1\n",
    "\n",
    "    negative_acc  = negative_acc/batches\n",
    "    possitive_acc = possitive_acc/batches\n",
    "    total_acc     = 0.5*possitive_acc  + 0.5*negative_acc\n",
    "    if show_extra_metrics == False:\n",
    "        print(f\"Sensitivity (poss_acc):{possitive_acc:.4f} Specificity (negative_acc):{negative_acc:.4f} accuracy:{total_acc:.4f}\")\n",
    "    \n",
    "    elif show_extra_metrics == True:\n",
    "        \n",
    "        fig, ax = plt.subplots(1, 2,figsize=(10,2))\n",
    "        fpr, tpr, thresholds = metrics.roc_curve( y_total, yhat_total)\n",
    "        \n",
    "        sens      =  tpr\n",
    "        spec      =  1 - fpr\n",
    "        j         = sens + spec -1\n",
    "        opt_index = np.where(j == np.max(j))[0][0]\n",
    "        op_point  = thresholds[opt_index]\n",
    "        \n",
    "        print(f\"Youdens  index: {op_point:.4f} Sensitivity: {round(sens[opt_index],4)} Specificity: {round(spec[opt_index],4)}\")\n",
    "       \n",
    "        ax[0].set_title(\"ROC Curve\")\n",
    "        ax[1].set_title(\"Confussion Matrix\")\n",
    "        if model == None:\n",
    "            ax[0].plot(fpr,tpr,label=\"MLP\") \n",
    "        else:\n",
    "            ax[0].plot(fpr,tpr,label=\"GraphSage+MLP\") \n",
    "        ax[0].plot([0, 1], [0, 1], 'k--')\n",
    "        ax[0].set_ylabel('True Positive Rate')\n",
    "        ax[0].set_xlabel('False Positive Rate')\n",
    "        ax[0].legend()\n",
    "       \n",
    "    \n",
    "        cfm = metrics.confusion_matrix(y_total, np.array(yhat_total)> op_point)\n",
    "        \n",
    "        cmn = cfm.astype('float') / cfm.sum(axis=1)[:, np.newaxis] # Normalise\n",
    "        disp = ConfusionMatrixDisplay(cmn)\n",
    "        disp.plot(ax=ax[1])\n",
    "        \n",
    "        plt.show()\n",
    "     \n",
    "    return total_acc "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<bound method HeteroData.metadata of HeteroData(\n",
       "  \u001b[1mcompounds\u001b[0m={ x=[499, 768] },\n",
       "  \u001b[1mdisease\u001b[0m={ x=[810, 768] },\n",
       "  \u001b[1m(compounds, treats, disease)\u001b[0m={ edge_index=[2, 2000] },\n",
       "  \u001b[1m(disease, rev_treats, compounds)\u001b[0m={ edge_index=[2, 2000] }\n",
       ")>"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GraphModule(\n",
      "  (convs): ModuleList(\n",
      "    (0): ModuleDict(\n",
      "      (compounds__treats__disease): SAGEConv(768, 524, aggr=mean)\n",
      "      (disease__rev_treats__compounds): SAGEConv(768, 524, aggr=mean)\n",
      "    )\n",
      "    (1-2): 2 x ModuleDict(\n",
      "      (compounds__treats__disease): SAGEConv(524, 524, aggr=mean)\n",
      "      (disease__rev_treats__compounds): SAGEConv(524, 524, aggr=mean)\n",
      "    )\n",
      "  )\n",
      "  (post_mp): ModuleList(\n",
      "    (0): ModuleDict(\n",
      "      (compounds): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (disease): Linear(in_features=524, out_features=524, bias=True)\n",
      "    )\n",
      "    (1): ModuleDict(\n",
      "      (compounds): Dropout(p=0.7, inplace=False)\n",
      "      (disease): Dropout(p=0.7, inplace=False)\n",
      "    )\n",
      "    (2): ModuleDict(\n",
      "      (compounds): Linear(in_features=524, out_features=524, bias=True)\n",
      "      (disease): Linear(in_features=524, out_features=524, bias=True)\n",
      "    )\n",
      "  )\n",
      ")\n",
      "\n",
      "\n",
      "\n",
      "def forward(self, x, edge_index, device = 'cpu'):\n",
      "    x_dict = torch_geometric_nn_to_hetero_transformer_get_dict(x);  x = None\n",
      "    x__compounds = x_dict.get('compounds', None)\n",
      "    x__disease = x_dict.get('disease', None);  x_dict = None\n",
      "    edge_index_dict = torch_geometric_nn_to_hetero_transformer_get_dict(edge_index);  edge_index = None\n",
      "    edge_index__compounds__treats__disease = edge_index_dict.get(('compounds', 'treats', 'disease'), None)\n",
      "    edge_index__disease__rev_treats__compounds = edge_index_dict.get(('disease', 'rev_treats', 'compounds'), None);  edge_index_dict = None\n",
      "    device_1 = device\n",
      "    convs_0__disease = getattr(self.convs, \"0\").compounds__treats__disease((x__compounds, x__disease), edge_index__compounds__treats__disease)\n",
      "    convs_0__compounds = getattr(self.convs, \"0\").disease__rev_treats__compounds((x__disease, x__compounds), edge_index__disease__rev_treats__compounds);  x__disease = x__compounds = None\n",
      "    relu__compounds = torch.nn.functional.relu(convs_0__compounds, inplace = False);  convs_0__compounds = None\n",
      "    relu__disease = torch.nn.functional.relu(convs_0__disease, inplace = False);  convs_0__disease = None\n",
      "    dropout__compounds = torch.nn.functional.dropout(relu__compounds, p = 0.7, training = True, inplace = False);  relu__compounds = None\n",
      "    dropout__disease = torch.nn.functional.dropout(relu__disease, p = 0.7, training = True, inplace = False);  relu__disease = None\n",
      "    convs_1__disease = getattr(self.convs, \"1\").compounds__treats__disease((dropout__compounds, dropout__disease), edge_index__compounds__treats__disease)\n",
      "    convs_1__compounds = getattr(self.convs, \"1\").disease__rev_treats__compounds((dropout__disease, dropout__compounds), edge_index__disease__rev_treats__compounds);  dropout__disease = dropout__compounds = None\n",
      "    relu_1__compounds = torch.nn.functional.relu(convs_1__compounds, inplace = False);  convs_1__compounds = None\n",
      "    relu_1__disease = torch.nn.functional.relu(convs_1__disease, inplace = False);  convs_1__disease = None\n",
      "    dropout_1__compounds = torch.nn.functional.dropout(relu_1__compounds, p = 0.7, training = True, inplace = False);  relu_1__compounds = None\n",
      "    dropout_1__disease = torch.nn.functional.dropout(relu_1__disease, p = 0.7, training = True, inplace = False);  relu_1__disease = None\n",
      "    convs_2__disease = getattr(self.convs, \"2\").compounds__treats__disease((dropout_1__compounds, dropout_1__disease), edge_index__compounds__treats__disease);  edge_index__compounds__treats__disease = None\n",
      "    convs_2__compounds = getattr(self.convs, \"2\").disease__rev_treats__compounds((dropout_1__disease, dropout_1__compounds), edge_index__disease__rev_treats__compounds);  dropout_1__disease = dropout_1__compounds = edge_index__disease__rev_treats__compounds = None\n",
      "    relu_2__compounds = torch.nn.functional.relu(convs_2__compounds, inplace = False);  convs_2__compounds = None\n",
      "    relu_2__disease = torch.nn.functional.relu(convs_2__disease, inplace = False);  convs_2__disease = None\n",
      "    dropout_2__compounds = torch.nn.functional.dropout(relu_2__compounds, p = 0.7, training = True, inplace = False);  relu_2__compounds = None\n",
      "    dropout_2__disease = torch.nn.functional.dropout(relu_2__disease, p = 0.7, training = True, inplace = False);  relu_2__disease = None\n",
      "    post_mp_0__compounds = getattr(self.post_mp, \"0\").compounds(dropout_2__compounds);  dropout_2__compounds = None\n",
      "    post_mp_0__disease = getattr(self.post_mp, \"0\").disease(dropout_2__disease);  dropout_2__disease = None\n",
      "    post_mp_1__compounds = getattr(self.post_mp, \"1\").compounds(post_mp_0__compounds);  post_mp_0__compounds = None\n",
      "    post_mp_1__disease = getattr(self.post_mp, \"1\").disease(post_mp_0__disease);  post_mp_0__disease = None\n",
      "    post_mp_2__compounds = getattr(self.post_mp, \"2\").compounds(post_mp_1__compounds);  post_mp_1__compounds = None\n",
      "    post_mp_2__disease = getattr(self.post_mp, \"2\").disease(post_mp_1__disease);  post_mp_1__disease = None\n",
      "    return {'compounds': post_mp_2__compounds, 'disease': post_mp_2__disease}\n",
      "    \n",
      "# To see more debug info, please use `graph_module.print_readable()`\n",
      "LinkPredictorMLP(\n",
      "  (mlp_layers): ModuleList(\n",
      "    (0-1): 2 x Linear(in_features=524, out_features=524, bias=True)\n",
      "    (2): Linear(in_features=524, out_features=1, bias=True)\n",
      "  )\n",
      ")\n",
      "Models Loaded to cpu\n"
     ]
    }
   ],
   "source": [
    "epochs        = 500\n",
    "hidden_dim    = 524      # 256 \n",
    "dropout       = 0.7\n",
    "num_layers    = 3\n",
    "learning_rate = 1e-4\n",
    "node_emb_dim  = 768\n",
    "device        = \"cpu\"\n",
    "\n",
    "HomoGNN         = GNNStack(node_emb_dim, hidden_dim, hidden_dim, num_layers, dropout, return_embedding=True).to(device) # the graph neural network that takes all the node embeddings as inputs to message pass and agregate\n",
    "HeteroGNN       = to_hetero(HomoGNN   , data.metadata(), aggr='sum')\n",
    "link_predictor  = LinkPredictorMLP(hidden_dim, hidden_dim, 1, num_layers , dropout).to(device) # the MLP that takes embeddings of a pair of nodes and predicts the existence of an edge between them\n",
    "#optimizer      = torch.optim.AdamW(list(model.parameters()) + list(link_predictor.parameters() ), lr=learning_rate, weight_decay=1e-4)\n",
    "optimizer       = torch.optim.Adam(list(HeteroGNN.parameters()) + list(link_predictor.parameters() ), lr=learning_rate)\n",
    "\n",
    "print(HeteroGNN )\n",
    "print(link_predictor)\n",
    "print(f\"Models Loaded to {device}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Train Data:\n",
      "HeteroData(\n",
      "  \u001b[1mcompounds\u001b[0m={ x=[499, 768] },\n",
      "  \u001b[1mdisease\u001b[0m={ x=[810, 768] },\n",
      "  \u001b[1m(compounds, treats, disease)\u001b[0m={\n",
      "    edge_index=[2, 800],\n",
      "    edge_label=[3200],\n",
      "    edge_label_index=[2, 3200]\n",
      "  },\n",
      "  \u001b[1m(disease, rev_treats, compounds)\u001b[0m={ edge_index=[2, 2000] },\n",
      "  \u001b[1m(compounds, rev_treats, disease)\u001b[0m={}\n",
      ")\n",
      "Validation Data:\n",
      "HeteroData(\n",
      "  \u001b[1mcompounds\u001b[0m={ x=[499, 768] },\n",
      "  \u001b[1mdisease\u001b[0m={ x=[810, 768] },\n",
      "  \u001b[1m(compounds, treats, disease)\u001b[0m={\n",
      "    edge_index=[2, 1600],\n",
      "    edge_label=[800],\n",
      "    edge_label_index=[2, 800]\n",
      "  },\n",
      "  \u001b[1m(disease, rev_treats, compounds)\u001b[0m={ edge_index=[2, 2000] },\n",
      "  \u001b[1m(compounds, rev_treats, disease)\u001b[0m={}\n",
      ")\n",
      "Test Data:\n",
      "HeteroData(\n",
      "  \u001b[1mcompounds\u001b[0m={ x=[499, 768] },\n",
      "  \u001b[1mdisease\u001b[0m={ x=[810, 768] },\n",
      "  \u001b[1m(compounds, treats, disease)\u001b[0m={\n",
      "    edge_index=[2, 1800],\n",
      "    edge_label=[800],\n",
      "    edge_label_index=[2, 800]\n",
      "  },\n",
      "  \u001b[1m(disease, rev_treats, compounds)\u001b[0m={ edge_index=[2, 2000] },\n",
      "  \u001b[1m(compounds, rev_treats, disease)\u001b[0m={}\n",
      ")\n"
     ]
    }
   ],
   "source": [
    "transform = T.RandomLinkSplit(\n",
    "    num_val=0.1,\n",
    "    num_test=0.1,\n",
    "    disjoint_train_ratio=0.5,\n",
    "    neg_sampling_ratio=3,\n",
    "    add_negative_train_samples=True,\n",
    "    edge_types=('compounds', 'treats', 'disease'),\n",
    "    rev_edge_types=('compounds', 'rev_treats', 'disease'), \n",
    ")\n",
    "\n",
    "train_data, val_data, test_data = transform(data)\n",
    "\n",
    "\n",
    "print(f\"Train Data:\\n{train_data}\")\n",
    "print(f\"Validation Data:\\n{val_data}\")\n",
    "print(f\"Test Data:\\n{test_data}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "#val = val_data.to(\"cuda\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "node_emb    = HeteroGNN( val_data.x_dict,  val_data.edge_index_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch 1: loss: 0.84755\n",
      "Epoch 2: loss: 0.84638\n",
      "Epoch 3: loss: 0.84612\n",
      "Epoch 4: loss: 0.84475\n",
      "Epoch 5: loss: 0.84357\n",
      "Epoch 6: loss: 0.84228\n",
      "Epoch 7: loss: 0.84044\n",
      "Epoch 8: loss: 0.83859\n",
      "Epoch 9: loss: 0.83694\n",
      "Epoch 10: loss: 0.83306\n",
      "Epoch 11: loss: 0.82913\n",
      "Epoch 12: loss: 0.82392\n",
      "Epoch 13: loss: 0.81768\n",
      "Epoch 14: loss: 0.80855\n",
      "Epoch 15: loss: 0.7965\n",
      "Epoch 16: loss: 0.78448\n",
      "Epoch 17: loss: 0.76982\n",
      "Epoch 18: loss: 0.75282\n",
      "Epoch 19: loss: 0.73616\n",
      "Epoch 20: loss: 0.71981\n",
      "Epoch 21: loss: 0.70881\n",
      "Epoch 22: loss: 0.70251\n",
      "Epoch 23: loss: 0.69938\n",
      "Epoch 24: loss: 0.69692\n",
      "Epoch 25: loss: 0.69536\n",
      "Epoch 26: loss: 0.69475\n",
      "Epoch 27: loss: 0.69412\n",
      "Epoch 28: loss: 0.69354\n",
      "Epoch 29: loss: 0.69334\n",
      "Epoch 30: loss: 0.6932\n",
      "Epoch 31: loss: 0.69324\n",
      "Epoch 32: loss: 0.69322\n",
      "Epoch 33: loss: 0.69316\n",
      "Epoch 34: loss: 0.69314\n",
      "Epoch 35: loss: 0.69316\n",
      "Epoch 36: loss: 0.69315\n",
      "Epoch 37: loss: 0.69315\n",
      "Epoch 38: loss: 0.69313\n",
      "Epoch 39: loss: 0.69315\n",
      "Epoch 40: loss: 0.69315\n",
      "Epoch 41: loss: 0.69315\n",
      "Epoch 42: loss: 0.69315\n",
      "Epoch 43: loss: 0.69315\n",
      "Epoch 44: loss: 0.69315\n",
      "Epoch 45: loss: 0.69315\n",
      "Epoch 46: loss: 0.69315\n",
      "Epoch 47: loss: 0.69315\n",
      "Epoch 48: loss: 0.69315\n",
      "Epoch 49: loss: 0.69315\n",
      "Epoch 50: loss: 0.69315\n",
      "Epoch 51: loss: 0.69315\n",
      "Epoch 52: loss: 0.69315\n",
      "Epoch 53: loss: 0.69315\n",
      "Epoch 54: loss: 0.69315\n",
      "Epoch 55: loss: 0.69315\n",
      "Epoch 56: loss: 0.69315\n",
      "Epoch 57: loss: 0.69315\n",
      "Epoch 58: loss: 0.69315\n",
      "Epoch 59: loss: 0.69315\n",
      "Epoch 60: loss: 0.69315\n",
      "Epoch 61: loss: 0.69315\n",
      "Epoch 62: loss: 0.69315\n",
      "Epoch 63: loss: 0.69315\n",
      "Epoch 64: loss: 0.69315\n",
      "Epoch 65: loss: 0.69315\n",
      "Epoch 66: loss: 0.69315\n",
      "Epoch 67: loss: 0.69315\n",
      "Epoch 68: loss: 0.69315\n",
      "Epoch 69: loss: 0.69315\n",
      "Epoch 70: loss: 0.69315\n",
      "Epoch 71: loss: 0.69315\n",
      "Epoch 72: loss: 0.69315\n",
      "Epoch 73: loss: 0.69315\n",
      "Epoch 74: loss: 0.69315\n",
      "Epoch 75: loss: 0.69315\n",
      "Epoch 76: loss: 0.69315\n",
      "Epoch 77: loss: 0.69315\n",
      "Epoch 78: loss: 0.69315\n",
      "Epoch 79: loss: 0.69315\n",
      "Epoch 80: loss: 0.69315\n",
      "Epoch 81: loss: 0.69315\n",
      "Epoch 82: loss: 0.69315\n",
      "Epoch 83: loss: 0.69315\n",
      "Epoch 84: loss: 0.69315\n",
      "Epoch 85: loss: 0.69315\n",
      "Epoch 86: loss: 0.69315\n",
      "Epoch 87: loss: 0.69315\n",
      "Epoch 88: loss: 0.69315\n",
      "Epoch 89: loss: 0.69315\n",
      "Epoch 90: loss: 0.69315\n",
      "Epoch 91: loss: 0.69315\n",
      "Epoch 92: loss: 0.69315\n",
      "Epoch 93: loss: 0.69315\n",
      "Epoch 94: loss: 0.69315\n",
      "Epoch 95: loss: 0.69315\n",
      "Epoch 96: loss: 0.69315\n",
      "Epoch 97: loss: 0.69315\n",
      "Epoch 98: loss: 0.69315\n",
      "Epoch 99: loss: 0.69315\n",
      "Epoch 100: loss: 0.69315\n",
      "Epoch 101: loss: 0.69315\n",
      "Epoch 102: loss: 0.69315\n",
      "Epoch 103: loss: 0.69315\n",
      "Epoch 104: loss: 0.69315\n",
      "Epoch 105: loss: 0.69315\n",
      "Epoch 106: loss: 0.69315\n",
      "Epoch 107: loss: 0.69315\n",
      "Epoch 108: loss: 0.69315\n",
      "Epoch 109: loss: 0.69315\n",
      "Epoch 110: loss: 0.69315\n",
      "Epoch 111: loss: 0.69315\n",
      "Epoch 112: loss: 0.69315\n",
      "Epoch 113: loss: 0.69315\n",
      "Epoch 114: loss: 0.69315\n",
      "Epoch 115: loss: 0.69315\n",
      "Epoch 116: loss: 0.69315\n",
      "Epoch 117: loss: 0.69315\n",
      "Epoch 118: loss: 0.69315\n",
      "Epoch 119: loss: 0.69315\n",
      "Epoch 120: loss: 0.69315\n",
      "Epoch 121: loss: 0.69315\n",
      "Epoch 122: loss: 0.69315\n",
      "Epoch 123: loss: 0.69315\n",
      "Epoch 124: loss: 0.69315\n",
      "Epoch 125: loss: 0.69315\n",
      "Epoch 126: loss: 0.69315\n",
      "Epoch 127: loss: 0.69315\n",
      "Epoch 128: loss: 0.69315\n",
      "Epoch 129: loss: 0.69315\n",
      "Epoch 130: loss: 0.69315\n",
      "Epoch 131: loss: 0.69315\n",
      "Epoch 132: loss: 0.69315\n",
      "Epoch 133: loss: 0.69315\n",
      "Epoch 134: loss: 0.69315\n",
      "Epoch 135: loss: 0.69315\n",
      "Epoch 136: loss: 0.69315\n",
      "Epoch 137: loss: 0.69315\n",
      "Epoch 138: loss: 0.69315\n",
      "Epoch 139: loss: 0.69315\n",
      "Epoch 140: loss: 0.69315\n",
      "Epoch 141: loss: 0.69315\n",
      "Epoch 142: loss: 0.69315\n",
      "Epoch 143: loss: 0.69315\n",
      "Epoch 144: loss: 0.69315\n",
      "Epoch 145: loss: 0.69315\n",
      "Epoch 146: loss: 0.69315\n",
      "Epoch 147: loss: 0.69315\n",
      "Epoch 148: loss: 0.69315\n",
      "Epoch 149: loss: 0.69315\n",
      "Epoch 150: loss: 0.69315\n",
      "Epoch 151: loss: 0.69315\n",
      "Epoch 152: loss: 0.69315\n",
      "Epoch 153: loss: 0.69315\n",
      "Epoch 154: loss: 0.69315\n",
      "Epoch 155: loss: 0.69315\n",
      "Epoch 156: loss: 0.69315\n",
      "Epoch 157: loss: 0.69315\n",
      "Epoch 158: loss: 0.69315\n",
      "Epoch 159: loss: 0.69315\n",
      "Epoch 160: loss: 0.69315\n",
      "Epoch 161: loss: 0.69315\n",
      "Epoch 162: loss: 0.69315\n",
      "Epoch 163: loss: 0.69315\n",
      "Epoch 164: loss: 0.69315\n",
      "Epoch 165: loss: 0.69315\n",
      "Epoch 166: loss: 0.69315\n",
      "Epoch 167: loss: 0.69315\n",
      "Epoch 168: loss: 0.69315\n",
      "Epoch 169: loss: 0.69315\n",
      "Epoch 170: loss: 0.69315\n",
      "Epoch 171: loss: 0.69315\n",
      "Epoch 172: loss: 0.69315\n",
      "Epoch 173: loss: 0.69315\n",
      "Epoch 174: loss: 0.69315\n",
      "Epoch 175: loss: 0.69315\n",
      "Epoch 176: loss: 0.69315\n",
      "Epoch 177: loss: 0.69315\n",
      "Epoch 178: loss: 0.69315\n",
      "Epoch 179: loss: 0.69315\n",
      "Epoch 180: loss: 0.69315\n",
      "Epoch 181: loss: 0.69315\n",
      "Epoch 182: loss: 0.69315\n",
      "Epoch 183: loss: 0.69315\n",
      "Epoch 184: loss: 0.69315\n",
      "Epoch 185: loss: 0.69315\n",
      "Epoch 186: loss: 0.69315\n",
      "Epoch 187: loss: 0.69315\n",
      "Epoch 188: loss: 0.69315\n",
      "Epoch 189: loss: 0.69315\n",
      "Epoch 190: loss: 0.69315\n",
      "Epoch 191: loss: 0.69315\n",
      "Epoch 192: loss: 0.69315\n",
      "Epoch 193: loss: 0.69315\n",
      "Epoch 194: loss: 0.69315\n",
      "Epoch 195: loss: 0.69315\n",
      "Epoch 196: loss: 0.69315\n",
      "Epoch 197: loss: 0.69315\n",
      "Epoch 198: loss: 0.69315\n",
      "Epoch 199: loss: 0.69315\n",
      "Epoch 200: loss: 0.69315\n",
      "Epoch 201: loss: 0.69315\n",
      "Epoch 202: loss: 0.69315\n",
      "Epoch 203: loss: 0.69315\n",
      "Epoch 204: loss: 0.69315\n",
      "Epoch 205: loss: 0.69315\n",
      "Epoch 206: loss: 0.69315\n",
      "Epoch 207: loss: 0.69315\n",
      "Epoch 208: loss: 0.69315\n",
      "Epoch 209: loss: 0.69315\n",
      "Epoch 210: loss: 0.69315\n",
      "Epoch 211: loss: 0.69315\n",
      "Epoch 212: loss: 0.69315\n",
      "Epoch 213: loss: 0.69315\n",
      "Epoch 214: loss: 0.69315\n",
      "Epoch 215: loss: 0.69315\n",
      "Epoch 216: loss: 0.69315\n",
      "Epoch 217: loss: 0.69315\n",
      "Epoch 218: loss: 0.69315\n",
      "Epoch 219: loss: 0.69315\n",
      "Epoch 220: loss: 0.69315\n",
      "Epoch 221: loss: 0.69315\n",
      "Epoch 222: loss: 0.69315\n",
      "Epoch 223: loss: 0.69315\n",
      "Epoch 224: loss: 0.69315\n",
      "Epoch 225: loss: 0.69315\n",
      "Epoch 226: loss: 0.69315\n",
      "Epoch 227: loss: 0.69315\n",
      "Epoch 228: loss: 0.69315\n",
      "Epoch 229: loss: 0.69315\n",
      "Epoch 230: loss: 0.69315\n",
      "Epoch 231: loss: 0.69315\n",
      "Epoch 232: loss: 0.69315\n",
      "Epoch 233: loss: 0.69315\n",
      "Epoch 234: loss: 0.69315\n",
      "Epoch 235: loss: 0.69315\n",
      "Epoch 236: loss: 0.69315\n",
      "Epoch 237: loss: 0.69315\n",
      "Epoch 238: loss: 0.69315\n",
      "Epoch 239: loss: 0.69315\n",
      "Epoch 240: loss: 0.69315\n",
      "Epoch 241: loss: 0.69315\n",
      "Epoch 242: loss: 0.69315\n",
      "Epoch 243: loss: 0.69315\n",
      "Epoch 244: loss: 0.69315\n",
      "Epoch 245: loss: 0.69315\n",
      "Epoch 246: loss: 0.69315\n",
      "Epoch 247: loss: 0.69315\n",
      "Epoch 248: loss: 0.69315\n",
      "Epoch 249: loss: 0.69315\n",
      "Epoch 250: loss: 0.69315\n",
      "Epoch 251: loss: 0.69315\n",
      "Epoch 252: loss: 0.69315\n",
      "Epoch 253: loss: 0.69315\n",
      "Epoch 254: loss: 0.69315\n",
      "Epoch 255: loss: 0.69315\n",
      "Epoch 256: loss: 0.69315\n",
      "Epoch 257: loss: 0.69315\n",
      "Epoch 258: loss: 0.69315\n",
      "Epoch 259: loss: 0.69315\n",
      "Epoch 260: loss: 0.69315\n",
      "Epoch 261: loss: 0.69315\n",
      "Epoch 262: loss: 0.69315\n",
      "Epoch 263: loss: 0.69315\n",
      "Epoch 264: loss: 0.69315\n",
      "Epoch 265: loss: 0.69315\n",
      "Epoch 266: loss: 0.69315\n",
      "Epoch 267: loss: 0.69315\n",
      "Epoch 268: loss: 0.69315\n",
      "Epoch 269: loss: 0.69315\n",
      "Epoch 270: loss: 0.69315\n",
      "Epoch 271: loss: 0.69315\n",
      "Epoch 272: loss: 0.69315\n",
      "Epoch 273: loss: 0.69315\n",
      "Epoch 274: loss: 0.69315\n",
      "Epoch 275: loss: 0.69315\n",
      "Epoch 276: loss: 0.69315\n",
      "Epoch 277: loss: 0.69315\n",
      "Epoch 278: loss: 0.69315\n",
      "Epoch 279: loss: 0.69315\n",
      "Epoch 280: loss: 0.69315\n",
      "Epoch 281: loss: 0.69315\n",
      "Epoch 282: loss: 0.69315\n",
      "Epoch 283: loss: 0.69315\n",
      "Epoch 284: loss: 0.69315\n",
      "Epoch 285: loss: 0.69315\n",
      "Epoch 286: loss: 0.69315\n",
      "Epoch 287: loss: 0.69315\n",
      "Epoch 288: loss: 0.69315\n",
      "Epoch 289: loss: 0.69315\n",
      "Epoch 290: loss: 0.69315\n",
      "Epoch 291: loss: 0.69315\n",
      "Epoch 292: loss: 0.69315\n",
      "Epoch 293: loss: 0.69315\n",
      "Epoch 294: loss: 0.69315\n",
      "Epoch 295: loss: 0.69315\n",
      "Epoch 296: loss: 0.69315\n",
      "Epoch 297: loss: 0.69315\n",
      "Epoch 298: loss: 0.69315\n",
      "Epoch 299: loss: 0.69315\n",
      "Epoch 300: loss: 0.69315\n",
      "Epoch 301: loss: 0.69315\n",
      "Epoch 302: loss: 0.69315\n",
      "Epoch 303: loss: 0.69315\n",
      "Epoch 304: loss: 0.69315\n",
      "Epoch 305: loss: 0.69315\n",
      "Epoch 306: loss: 0.69315\n",
      "Epoch 307: loss: 0.69315\n",
      "Epoch 308: loss: 0.69315\n",
      "Epoch 309: loss: 0.69315\n",
      "Epoch 310: loss: 0.69315\n",
      "Epoch 311: loss: 0.69315\n",
      "Epoch 312: loss: 0.69315\n",
      "Epoch 313: loss: 0.69315\n",
      "Epoch 314: loss: 0.69315\n",
      "Epoch 315: loss: 0.69315\n",
      "Epoch 316: loss: 0.69315\n",
      "Epoch 317: loss: 0.69315\n",
      "Epoch 318: loss: 0.69315\n",
      "Epoch 319: loss: 0.69315\n",
      "Epoch 320: loss: 0.69315\n",
      "Epoch 321: loss: 0.69315\n",
      "Epoch 322: loss: 0.69315\n",
      "Epoch 323: loss: 0.69315\n",
      "Epoch 324: loss: 0.69315\n",
      "Epoch 325: loss: 0.69315\n",
      "Epoch 326: loss: 0.69315\n",
      "Epoch 327: loss: 0.69315\n",
      "Epoch 328: loss: 0.69315\n",
      "Epoch 329: loss: 0.69315\n",
      "Epoch 330: loss: 0.69315\n",
      "Epoch 331: loss: 0.69315\n",
      "Epoch 332: loss: 0.69315\n",
      "Epoch 333: loss: 0.69315\n",
      "Epoch 334: loss: 0.69315\n",
      "Epoch 335: loss: 0.69315\n",
      "Epoch 336: loss: 0.69315\n",
      "Epoch 337: loss: 0.69315\n",
      "Epoch 338: loss: 0.69315\n",
      "Epoch 339: loss: 0.69315\n",
      "Epoch 340: loss: 0.69315\n",
      "Epoch 341: loss: 0.69315\n",
      "Epoch 342: loss: 0.69315\n",
      "Epoch 343: loss: 0.69315\n",
      "Epoch 344: loss: 0.69315\n",
      "Epoch 345: loss: 0.69315\n",
      "Epoch 346: loss: 0.69315\n",
      "Epoch 347: loss: 0.69315\n",
      "Epoch 348: loss: 0.69315\n",
      "Epoch 349: loss: 0.69315\n",
      "Epoch 350: loss: 0.69315\n",
      "Epoch 351: loss: 0.69315\n",
      "Epoch 352: loss: 0.69315\n",
      "Epoch 353: loss: 0.69315\n",
      "Epoch 354: loss: 0.69315\n",
      "Epoch 355: loss: 0.69315\n",
      "Epoch 356: loss: 0.69315\n",
      "Epoch 357: loss: 0.69315\n",
      "Epoch 358: loss: 0.69315\n",
      "Epoch 359: loss: 0.69315\n",
      "Epoch 360: loss: 0.69315\n",
      "Epoch 361: loss: 0.69315\n",
      "Epoch 362: loss: 0.69315\n",
      "Epoch 363: loss: 0.69315\n",
      "Epoch 364: loss: 0.69315\n",
      "Epoch 365: loss: 0.69315\n",
      "Epoch 366: loss: 0.69315\n",
      "Epoch 367: loss: 0.69315\n",
      "Epoch 368: loss: 0.69315\n",
      "Epoch 369: loss: 0.69315\n",
      "Epoch 370: loss: 0.69315\n",
      "Epoch 371: loss: 0.69315\n",
      "Epoch 372: loss: 0.69315\n",
      "Epoch 373: loss: 0.69315\n",
      "Epoch 374: loss: 0.69315\n",
      "Epoch 375: loss: 0.69315\n",
      "Epoch 376: loss: 0.69315\n",
      "Epoch 377: loss: 0.69315\n",
      "Epoch 378: loss: 0.69315\n",
      "Epoch 379: loss: 0.69315\n",
      "Epoch 380: loss: 0.69315\n",
      "Epoch 381: loss: 0.69315\n",
      "Epoch 382: loss: 0.69315\n",
      "Epoch 383: loss: 0.69315\n",
      "Epoch 384: loss: 0.69315\n",
      "Epoch 385: loss: 0.69315\n",
      "Epoch 386: loss: 0.69315\n",
      "Epoch 387: loss: 0.69315\n",
      "Epoch 388: loss: 0.69315\n",
      "Epoch 389: loss: 0.69315\n",
      "Epoch 390: loss: 0.69315\n",
      "Epoch 391: loss: 0.69315\n",
      "Epoch 392: loss: 0.69315\n",
      "Epoch 393: loss: 0.69315\n",
      "Epoch 394: loss: 0.69315\n",
      "Epoch 395: loss: 0.69315\n",
      "Epoch 396: loss: 0.69315\n",
      "Epoch 397: loss: 0.69315\n",
      "Epoch 398: loss: 0.69315\n",
      "Epoch 399: loss: 0.69315\n",
      "Epoch 400: loss: 0.69315\n",
      "Epoch 401: loss: 0.69315\n",
      "Epoch 402: loss: 0.69315\n",
      "Epoch 403: loss: 0.69315\n",
      "Epoch 404: loss: 0.69315\n",
      "Epoch 405: loss: 0.69315\n",
      "Epoch 406: loss: 0.69315\n",
      "Epoch 407: loss: 0.69315\n",
      "Epoch 408: loss: 0.69315\n",
      "Epoch 409: loss: 0.69315\n",
      "Epoch 410: loss: 0.69315\n",
      "Epoch 411: loss: 0.69315\n",
      "Epoch 412: loss: 0.69315\n",
      "Epoch 413: loss: 0.69315\n",
      "Epoch 414: loss: 0.69315\n",
      "Epoch 415: loss: 0.69315\n",
      "Epoch 416: loss: 0.69315\n",
      "Epoch 417: loss: 0.69315\n",
      "Epoch 418: loss: 0.69315\n",
      "Epoch 419: loss: 0.69315\n",
      "Epoch 420: loss: 0.69315\n",
      "Epoch 421: loss: 0.69315\n",
      "Epoch 422: loss: 0.69315\n",
      "Epoch 423: loss: 0.69315\n",
      "Epoch 424: loss: 0.69315\n",
      "Epoch 425: loss: 0.69315\n",
      "Epoch 426: loss: 0.69315\n",
      "Epoch 427: loss: 0.69315\n",
      "Epoch 428: loss: 0.69315\n",
      "Epoch 429: loss: 0.69315\n"
     ]
    }
   ],
   "source": [
    "train_loss                      = []\n",
    "train_accuracy                  = []\n",
    "show_metrics_every              = 20\n",
    "best_accuracy                   = 0 \n",
    "best_graphsage_model_path       = \"\"\n",
    "best_link_predictor_model_path  = \"\"\n",
    "\n",
    "for epoch in range(1,epochs):\n",
    "    \n",
    "    ### TRAIN ####\n",
    "    loss = train(model =  HeteroGNN, link_predictor = link_predictor, data = val_data, optimizer =  optimizer,device=\"cpu\")\n",
    "    train_loss.append(loss)\n",
    "    print(f\"Epoch {epoch}: loss: {round(loss, 5)}\")\n",
    "    \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_loss                      = []\n",
    "train_accuracy                  = []\n",
    "show_metrics_every              = 20\n",
    "best_accuracy                   = 0 \n",
    "best_graphsage_model_path       = \"\"\n",
    "best_link_predictor_model_path  = \"\"\n",
    "\n",
    "for epoch in range(1,epochs):\n",
    "    \n",
    "    ### TRAIN ####\n",
    "    loss = train(model, link_predictor,train_dataset, optimizer,device) # Get Loss\n",
    "    train_loss.append(loss)\n",
    "    print(f\"Epoch {epoch}: loss: {round(loss, 5)}\")\n",
    "    \n",
    "    ### EVALUATE ###\n",
    "    if (epoch % 20 == 0) or (epoch ==1):\n",
    "        accuracy = evaluate(model, link_predictor ,test_dataset,device=device,best_accuracy=best_accuracy,show_extra_metrics=True)\n",
    "        \n",
    "    else:\n",
    "        accuracy = evaluate(model, link_predictor ,test_dataset,device=device,best_accuracy=best_accuracy)\n",
    "    \n",
    "    train_accuracy.append(accuracy)\n",
    "    ### SAVE ###\n",
    "    if best_accuracy < accuracy:\n",
    "        if os.path.exists(best_graphsage_model_path):\n",
    "            \n",
    "            os.remove(best_graphsage_model_path)\n",
    "            \n",
    "        if os.path.exists(best_link_predictor_model_path):\n",
    "            os.remove(best_link_predictor_model_path)\n",
    "        print(f\"Replacing models: {best_graphsage_model_path }  {best_link_predictor_model_path}\")\n",
    "            \n",
    "        best_accuracy  = accuracy\n",
    "        best_graphsage_model_path      = f\"GraphSage_epoch_{epoch}.pt\"\n",
    "        best_link_predictor_model_path =  f\"link_predictor_epoch_{epoch}.pt\"\n",
    "        print(f\"with: Best models at {best_graphsage_model_path }  {best_link_predictor_model_path}\")\n",
    "        save_torch_model(model,         epoch=epoch,PATH=best_graphsage_model_path ,     optimizer=optimizer)\n",
    "        save_torch_model(link_predictor,epoch=epoch,PATH=best_link_predictor_model_path, optimizer=optimizer)\n",
    "\n",
    "        \n",
    "#### Load Best Models ####\n",
    "\n",
    "print(f\"Loading best models:  {best_graphsage_model_path }  {best_link_predictor_model_path}\")\n",
    "checkpoint = torch.load(best_graphsage_model_path)\n",
    "model.load_state_dict(checkpoint['model_state_dict'])\n",
    "\n",
    "checkpoint = torch.load(best_link_predictor_model_path)\n",
    "link_predictor.load_state_dict(checkpoint['model_state_dict'])\n",
    "\n",
    "del checkpoint\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "provenance": []
  },
  "gpuClass": "standard",
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "2d42bd2c408a4bd3bdebc2865ea42483": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": "hidden",
      "width": null
     }
    },
    "6e79954f9c514f099b336e3a11ad635e": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_6f2e0878218c4b38974a74a673b70751",
      "placeholder": "​",
      "style": "IPY_MODEL_98ccadd9e4a34d8282282a332fad6e0b",
      "value": "Downloading chembl_32_sqlite.tar.gz: 100%"
     }
    },
    "6f2e0878218c4b38974a74a673b70751": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "98ccadd9e4a34d8282282a332fad6e0b": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "a95662c4dc0949f4af40e8331cde378d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "info",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_eec53708d80f41e8b402060c9a1352b1",
      "max": 1,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_f87f682db17a4e6b837b9c14c85ec36f",
      "value": 1
     }
    },
    "b0797e6674144b1b92d2ff70918b917c": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "b355679609b749d698029d4f39d45503": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_6e79954f9c514f099b336e3a11ad635e",
       "IPY_MODEL_a95662c4dc0949f4af40e8331cde378d",
       "IPY_MODEL_d89d003254ce42f9a44572c7e26940ec"
      ],
      "layout": "IPY_MODEL_2d42bd2c408a4bd3bdebc2865ea42483"
     }
    },
    "d0a9a73a5fe14afb8b15a2addf4f080d": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "d89d003254ce42f9a44572c7e26940ec": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_b0797e6674144b1b92d2ff70918b917c",
      "placeholder": "​",
      "style": "IPY_MODEL_d0a9a73a5fe14afb8b15a2addf4f080d",
      "value": " 4.27G/4.27G [03:38&lt;00:00, 21.7MB/s]"
     }
    },
    "eec53708d80f41e8b402060c9a1352b1": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": "20px"
     }
    },
    "f87f682db17a4e6b837b9c14c85ec36f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
